{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NetColoc analysis of rare variants in Autism spectrum disorder (ASD) and Congenital Heart Disease (CHD)\n",
    "\n",
    "Example of NetColoc workflow on genes associated with rare exome variants in ASD and CHD\n",
    "\n",
    "**Some background:**\n",
    "\n",
    "Here we introduce NetColoc, a tool which evaluates the extent to which two gene sets are related in network space, i.e. the extent to which they are colocalized in a molecular interaction network, and interrogates the underlying biological pathways and processes using multiscale community detection. This framework may be applied to any number of scenarios in which gene sets have been associated with a phenotype or condition, including rare and common variants within the same disease, genes associated with two comorbid diseases, genetically correlated GWAS phenotypes, GWAS across two different species, or gene expression changes after treatment with two different drugs, to name a few. NetColoc relies on a dual network propagation approach to identify the region of network space which is significantly proximal to both input gene sets, and as such is highly effective for small to medium input gene sets\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Obtain input gene sets and interactome_\n",
    "\n",
    "### 1. Load required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DISCLAIMIER: cdapsutil is experimental and may contain errors and interfaces may change\n"
     ]
    }
   ],
   "source": [
    "# load required packages\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import re\n",
    "import random\n",
    "\n",
    "from IPython.display import display\n",
    "\n",
    "import getpass\n",
    "import ndex2\n",
    "\n",
    "import json\n",
    "import cdapsutil\n",
    "\n",
    "from gprofiler import GProfiler\n",
    "gp = GProfiler(\"MyToolName/0.1\")\n",
    "\n",
    "from scipy.stats import hypergeom\n",
    "from scipy.stats import norm\n",
    "\n",
    "# latex rendering of text in graphs\n",
    "import matplotlib as mpl\n",
    "mpl.rc('text', usetex = False)\n",
    "mpl.rc('font', family = 'serif')\n",
    "\n",
    "from matplotlib import rcParams\n",
    "rcParams['font.family'] = 'sans-serif'\n",
    "rcParams['font.sans-serif'] = ['Arial']\n",
    "\n",
    "sns.set(font_scale=1.4)\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "sns.set_style(\"ticks\", {\"xtick.major.size\": 15, \"ytick.major.size\": 15})\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "import sys\n",
    "\n",
    "# % matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "\n",
    "from netcoloc import netprop_zscore\n",
    "from netcoloc import netprop\n",
    "from netcoloc import network_colocalization\n",
    "from netcoloc import validation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Load two gene sets of interest\n",
    "\n",
    "\n",
    "Identify two gene sets of interest. Gene sets should come from experimental data (not manual curation) to avoid bias. \n",
    "\n",
    "**Usage Note**: gene sets should be < 500 genes (propagation algorithm breaks down if seeded with larger sets). If your gene set is larger, only use the top 500 as seeds to the network propagation.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set names of geneset 1 and geneset 2\n",
    "# ------ customize this section based on your gene sets and how they should be labeled -------\n",
    "d1_name='ASD'\n",
    "d2_name='CHD'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number ASD genes:\n",
      "102\n",
      "['CHD8', 'SCN2A', 'SYNGAP1', 'ADNP', 'FOXP1']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# ------ customize this section based your input genesets -------\n",
    "\n",
    "# load rare variants Satterstrom et al 2020; PMID: 31981491\n",
    "\n",
    "D1_df = pd.read_csv('data/Satterstrom--Top-102-ASD-genes--May2019.csv')\n",
    "D1_df.index = D1_df['gene']\n",
    "print('number '+d1_name+' genes:')\n",
    "print(len(D1_df))\n",
    "D1_genes = D1_df.index.tolist() # define rare variant genes to seed network propagation\n",
    "print(D1_genes[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number CHD genes:\n",
      "66\n",
      "['KMT2D', 'CHD7', 'PTPN11', 'NSD1', 'NOTCH1']\n"
     ]
    }
   ],
   "source": [
    "# ------ customize this section based your input genesets -------\n",
    "\n",
    "# load D2 rare variant genes from Jin et al 2017; PMID: 28991257\n",
    "D2_df = pd.read_csv('data/CHD_HC.tsv',sep='\\t')\n",
    "\n",
    "D2_genes = D2_df['0'].tolist()\n",
    "print('number '+d2_name+' genes:')\n",
    "print(len(D2_genes))\n",
    "print(D2_genes[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of ASD and CHD genes overlapping:\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "# how much overlap between gene sets?\n",
    "print('number of '+d1_name+' and '+d2_name+' genes overlapping:')\n",
    "print(len(np.intersect1d(D1_genes,D2_genes)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Select molecular interaction network to use for the analysis. \n",
    "\n",
    "Identify network UUID on  Ndex (Pratt et al. 2015) and use this to import to jupyter notebook using. \n",
    "\n",
    "**Coverage**. Larger, denser interactomes will be more inclusive and be amenable to creating more granular models. Human curated interactomes are smaller, sparser and are biased towards known biology. Many, however, have richer descriptions of the relationships. Data derived interactomes based on specific projects have the advantage that the experimental context is well-defined and consistent.  \n",
    "\n",
    "\n",
    "**Interaction Types**. The edges that were useful in computing the coloc may not be useful for interpretation. For example, the edges in PCNet are not typed. For purposes of interpretation we need to know how the genes relate to each other. Further, we are best able to understand physical interactions, and so it may be most useful to review the nodes in a community or other subnetwork using a protein-protein interactome, or at least one in which the edges can be filtered when needed.\n",
    "\n",
    "**Net recommendation**: use an inclusive interactome for generating the model but then annotate subsystem networks with relationships derived from richer, if less comprehensive, sources. Or from sources specifically relevant to the experimental context.\n",
    "\n",
    "\n",
    "**Usage note**: PCnet is a general purpose interactome, a good starting place https://www.sciencedirect.com/science/article/pii/S2405471218300954"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nodes:\n",
      "17185\n",
      "\n",
      "number of edges:\n",
      "420534\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# interactome_uuid='4de852d9-9908-11e9-bcaf-0ac135e8bacf' # for PCNet\n",
    "interactome_uuid='275bd84e-3d18-11e8-a935-0ac135e8bacf' # for STRING high confidence\n",
    "ndex_server='public.ndexbio.org'\n",
    "ndex_user=None\n",
    "ndex_password=None\n",
    "G_int = ndex2.create_nice_cx_from_server(\n",
    "            ndex_server, \n",
    "            username=ndex_user, \n",
    "            password=ndex_password, \n",
    "            uuid=interactome_uuid\n",
    "        ).to_networkx()\n",
    "nodes = list(G_int.nodes)\n",
    "\n",
    "# pcnet appears to have some self edges... should remove them. \n",
    "G_int.remove_edges_from(nx.selfloop_edges(G_int))\n",
    "\n",
    "# print out interactome num nodes and edges for diagnostic purposes\n",
    "print('number of nodes:')\n",
    "print(len(G_int.nodes))\n",
    "print('\\nnumber of edges:')\n",
    "print(len(G_int.edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "int_nodes = list(G_int.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Identify network colocalized gene network_\n",
    "\n",
    "### 4. Precalculate matrices needed for propagation. This step should take a few minutes \n",
    "\n",
    "Network propagation from genes on selected interactome\n",
    "- Control for degree of input genes\n",
    "- Generate a proximity z-score, which defines genes which are closer to input set than expected by chance.\n",
    "- Repeat for rare and common variant genes, defined above\n",
    "\n",
    "Background on network propagation: https://www.nature.com/articles/nrg.2017.38.pdf?origin=ppub\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "calculating w_prime\n",
      "\n",
      "calculating w_double_prime\n"
     ]
    }
   ],
   "source": [
    "# pre calculate mats used for netprop... this step takes a few minutes, more for denser interactomes\n",
    "print('\\ncalculating w_prime')\n",
    "w_prime = netprop.get_normalized_adjacency_matrix(G_int, conserve_heat=True)\n",
    "\n",
    "print('\\ncalculating w_double_prime')\n",
    "w_double_prime = netprop.get_individual_heats_matrix(w_prime, .5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Subset seed genes to those found in the selected interactome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "102\n",
      "98\n",
      "66\n",
      "65\n"
     ]
    }
   ],
   "source": [
    "# subset seed genes to those found in interactome\n",
    "print(len(D1_genes))\n",
    "D1_genes = list(np.intersect1d(D1_genes,int_nodes))\n",
    "print(len(D1_genes))\n",
    "\n",
    "print(len(D2_genes))\n",
    "D2_genes = list(np.intersect1d(D2_genes,int_nodes))\n",
    "print(len(D2_genes))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6. Run network propagation from both seed gene sets independently."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calculating D1 variant z-scores: \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5343d167290647979e6df267b03479d6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=1000.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KDM6B</th>\n",
       "      <td>10.991122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASXL3</th>\n",
       "      <td>10.905475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRAF7</th>\n",
       "      <td>10.895410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2E</th>\n",
       "      <td>10.783263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DIP2A</th>\n",
       "      <td>10.639522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               z\n",
       "KDM6B  10.991122\n",
       "ASXL3  10.905475\n",
       "TRAF7  10.895410\n",
       "KMT2E  10.783263\n",
       "DIP2A  10.639522"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# D1 netprop\n",
    "print('\\nCalculating D1 variant z-scores: ')\n",
    "z_D1, Fnew_D1, Fnew_rand_D1 = netprop_zscore.calculate_heat_zscores(w_double_prime, int_nodes, \n",
    "                                                                    dict(G_int.degree), \n",
    "                                                                    D1_genes, num_reps=1000,\n",
    "                                                                    minimum_bin_size=100)\n",
    "\n",
    "z_D1 = pd.DataFrame({'z':z_D1})\n",
    "\n",
    "z_D1.sort_values('z',ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calculating D2 variant z-scores: \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6f98c8bc89cb46ceb32915bbe59a8762",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=1000.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CLUH</th>\n",
       "      <td>13.015273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZEB2</th>\n",
       "      <td>10.239921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NR6A1</th>\n",
       "      <td>9.672462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHD7</th>\n",
       "      <td>9.607006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD11</th>\n",
       "      <td>9.116133</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                z\n",
       "CLUH    13.015273\n",
       "ZEB2    10.239921\n",
       "NR6A1    9.672462\n",
       "CHD7     9.607006\n",
       "SAMD11   9.116133"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# D2 variant netprop\n",
    "print('\\nCalculating D2 variant z-scores: ')\n",
    "z_D2, Fnew_D2, Fnew_rand_D2 = netprop_zscore.calculate_heat_zscores(w_double_prime, int_nodes, \n",
    "                                                                    dict(G_int.degree), \n",
    "                                                                    D2_genes, num_reps=1000,\n",
    "                                                                    minimum_bin_size=100)\n",
    "\n",
    "z_D2 = pd.DataFrame({'z':z_D2})\n",
    "z_D2.sort_values('z',ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7. Compute the network colocalization enrichment for gene set pair. If gene sets are significantly colocalized, proceed with analysis. \n",
    "\n",
    "\n",
    "Size of network co-localization subgraph compared to null model created by permuting individual propagation z-scores. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ASD + CHD\n",
      "size of network intersection = 350\n",
      "observed size/ expected size = 2.1875\n",
      "p = 1.97776641905858e-46\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "seed_dict = {d1_name:D1_genes,d2_name:D2_genes}\n",
    "# --------------------------------------------------------------------------------------------\n",
    "\n",
    "zthresh=3\n",
    "\n",
    "z_d1d2_size, high_z_rand = network_colocalization.calculate_expected_overlap(\n",
    "    z_D1['z'],\n",
    "    z_D2['z'],\n",
    "    plot=False,\n",
    "    num_reps=100,\n",
    "    z_score_threshold=zthresh,\n",
    "    z1_threshold=1.5,\n",
    "    z2_threshold=1.5\n",
    ")\n",
    "\n",
    "ztemp = (z_d1d2_size - np.mean(high_z_rand)) / np.std(high_z_rand)\n",
    "ptemp = norm.sf(ztemp)\n",
    "print(d1_name + ' + ' + d2_name)\n",
    "print('size of network intersection = ' + str(z_d1d2_size))\n",
    "obs_exp_temp = float(z_d1d2_size) / np.mean(high_z_rand)\n",
    "print('observed size/ expected size = ' + str(obs_exp_temp))\n",
    "print('p = ' + str(ptemp))\n",
    "\n",
    "# save the num overlap and overlap p-val\n",
    "network_num_overlap = z_d1d2_size\n",
    "network_pval_overlap = ptemp\n",
    "network_obs_exp = obs_exp_temp\n",
    "network_exp_mean_overlap = np.mean(high_z_rand)\n",
    "network_exp_std_overlap = np.std(high_z_rand)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8. Plot the sizes of the observed and expected NetColoc subgraphs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-1, 2)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAALQAAAFnCAYAAAAPEGMHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nO3deVyNef8/8NdpE6Vs0aSIUkmLtYwtFcYwhixlCxWSupFtcBtjbmVuo0RykylburNLhvtmLPkyQ5SlwWkhtGhDqVROy/X7o1/nduZ06oo61+nyfj4e5/HQ5zrnXO94nct1ruXzFjAMw4AQnlDiugBCmhIFmvAKBZrwCgWa8AoFmvCKCtcFyFt5eTkePnwIHR0dKCsrc10OaaSqqirk5+fDwsIC6urqUss/u0A/fPgQs2bN4roM8okiIyMxcOBAqfHPLtA6OjoAav5CdHV1Oa6GNFZOTg5mzZol/nf8q88u0LW7Gbq6utDX1+e4GvKxZO0u0pdCwisUaMIrFGjCKxRowisUaMIrcg90bm4uli9fDltbW/Tr1w8LFy5EamqqePny5cthamoq8RgxYoR4eXV1NYKDgzF8+HBYW1vD3d0dL168kPevQRSUXAPNMAwWLFiAnJwchIeH48SJE1BXV8e8efPw7t07AEBycjKWLFmCGzduiB/R0dHi99i1axeioqLg5+eHo0ePQllZGR4eHnj//r08fxWioOQa6FevXsHIyAj+/v6wsLCAkZERFi9ejFevXiElJQUikQjPnz+HpaUldHR0xI8OHToAAEQiEfbt2wcfHx/Y2dnBzMwMQUFBePXqFf7zn//I81chCkqugdbR0UFQUBB69OgBoCbg4eHh6Ny5M0xMTPD06VNUVlbC2Ni4ztcLhUKUlpZi8ODB4jFNTU2Ym5sjPj5eLr8DUWycnSlcs2YNTp8+DTU1NezevRsaGhpITk6GiooKQkNDcf36dSgrK8POzg5Lly5F27ZtkZubCwDo0qWLxHt17twZ2dnZUusoKipCUVGRxFhOTk7z/VKEc5wF2sPDA7NmzcK///1veHt7IzIyUvzlUF9fH3v27MGLFy+wZcsWJCUl4dChQygrKwMAqKmpSbyXmpoaRCKR1DoOHjyIkJCQ5v9lFEx5eXmdV6JxSV41cRboXr16AQD8/f3x4MEDRERE4KeffoKnpye0tLQAACYmJujUqROmT5+O+/fvi/9CRCKRRKhFIhHatGkjtY65c+fCyclJYqz24hY+U1dXh0Ag4LoMCfK6F1uugc7Ly0NcXBy++eYb8V+4kpISjI2NkZubCyUlJXGYa5mamgIAsrOzYWBgIH4fTU1Nifeta79bS0tL6v0Iv8n1S2F2djZWrlyJhIQE8VhFRQUeP34MIyMjeHt7w8vLS+I1iYmJAABjY2OYmZlBU1MTt2/fFi8vKSnB48ePYWNjI59fgig0uQba0tIStra22LBhA+Lj45GSkoLvvvsOhYWFmDdvHsaNG4crV65g7969SE9PR2xsLNatW4evvvoKpqamUFNTw+zZsxEUFIRLly4hKSkJvr6+6NKlC8aMGSPPX4UoKLnucigpKWHnzp0ICAjAsmXLUFxcjIEDByIyMhIGBgYwMDBAdXU1wsLCsGvXLrRt2xbjx4+Hr6+v+D2WLFmCqqoqbNiwAWVlZRgwYADCwsKkviiSz5Pgc5s5KTMzE46Ojrh8+TKvL/Dn65fChv796OIkwisUaMIrFGjCKxRowisUaMIrFGjCKxRowisUaMIrFGjCKxRowisUaMIrFGjCKxRowisUaMIrFGjCKxRowisUaMIrFGjCKxRowisUaMIrFGjCKxRowisUaMIrFGjCKxRowisK1zRIKBTC1dUVffv2xciRIxEeHi7xemoaROqjUE2D3rx5g3nz5qF79+44efIkli5diuDgYBw7dkz8HtQ0iNSLkaO8vDxm2bJlTFpamnhMKBQyJiYmzN27d5ndu3czQ4cOZSoqKsTLg4KCGEdHR4ZhGOb9+/dM3759mcOHD4uXFxcXM9bW1szp06dZ1ZCRkcGYmJgwGRkZTfRbKSYACvVoKg39+ylU06D4+HgMHDgQKir/mxTV1tYWGRkZyM3NpaZBpEEK1TQoNzdXaib+zp07A6iZLD0vLw8ANQ0isilU06Dy8vI6GwIBwPv376lpEGmQQjUNUldXlwpm7c9t2rShpkGkQQrVNEhXV1e8W/HhawBAV1dXPGk2NQ0isihU06BBgwYhISEBlZWV4uW3bt2CoaEhdHR0qGkQaZBCNQ2aMmUKysrKsG7dOjx58gTR0dE4cOAAPD09AYCaBpGGNdkBQpYKCwuZ9evXM0OHDmWsrKwYd3d3JikpSbw8MTGRcXFxYSwsLBh7e3vm4MGDEq+vrKxktm7dynz55ZdM3759GQ8PDyY9PZ31+uk4NL+PQ1PTIJ6ipkGE8AAFmvAKBZrwCgWa8AoFmvAKBZrwCgWa8AoFmvAKBZrwCgWa8AoFmvAKBZrwCgWa8AoFmvAK61uwsrKyUF5eDiMjI5SUlGD79u3IzMzE+PHjMWHChOaskRDWWG2h//jjD4wdOxYnTpwAAGzcuBFRUVHIysrC6tWrER0d3axFEsIWq0Dv2rULAwYMwMKFC/Hu3TtcvHgR8+fPx9mzZzF37lwcOHCgmcskhB1WgX78+DHc3NzQvn17xMXFoaKiAl9//TUAYMSIEXj27FmzFkkIW6wCraqqKr6l58aNG+jYsSPMzMwAAG/fvoWGhkbzVUhII7D6UmhpaYljx45BS0sL586dw1dffQWgZm66sLAwWFpaNmuRhLDFagu9atUqJCQkYPr06VBVVcXChQsBABMmTEBWVhaWLVvWrEUSwharLbSZmRl+++03pKWloVevXmjdujWAmqMd/fv3h46OTrMWSQhbrE+saGpqwsrKCoWFhbh//z5KS0thZ2dHYSYKhfWJldjYWGzZsgXPnz+HQCDA8ePHsXPnTnTu3BkbN26EkhKddCTcY5XCa9euYfHixdDT08OGDRvEk4bY2tri5MmTCAsLa9YiCWGLVaCDg4Px1VdfITw8HM7OzuJAu7m5Yf78+Th16hTrFZaUlGDz5s1wcHBAv379MHnyZFy+fFm8PDAwEKamplKPDydwjIyMhKOjI6ysrODi4oLExETW6yf8xirQqampmDhxYp3LBg8eXOfs+bKsXbsWsbGx8PPzQ3R0NMaMGQMfHx/cvHkTAJCcnAxnZ2fcuHFD4lHbpuLUqVPYunUrli1bhlOnTqFHjx6YP38+Xr9+zboGwl+sAq2lpYWXL1/WuSwzMxNt27ZltbL8/HxcvHgR69atw5AhQ9C9e3csWrQINjY24utEUlJSYG5uDh0dHYlHrdDQUMycORMTJkyAsbEx/P39oampiSNHjrCqgfAbq0A7Ojpi586dEo15BAIBsrKysHfvXjg4OLBaWevWrfHLL79g4MCBEuMCgQBv375FUVERsrOz65y8HKg5kfP8+XPY2tqKx5SVlTFgwABqGkQAsDzKsWLFCiQmJsLV1RXt27cHACxbtgw5OTkwMDDA8uXLWa1MU1MTI0aMkBi7f/8+bt26hfXr1yMlJQUAcPbsWfz9739HRUUFbGxssGLFCnTu3Bm5ubkAambz/1Dnzp3x559/Sq2PmgZ9flgFWktLC0ePHsWZM2dw69YtFBQUoG3btpg7dy4mT54sPtHSWE+fPoWPjw+sra3h4uKC48ePA6gJfnBwMPLz8xEUFARXV1dER0dT0yDSINbHodXU1DBt2jRMmzatSVZ8584d+Pj4QE9PD6GhoVBVVcWMGTMwfvx4aGtrA6g5Q2liYgI7OztcunRJ3N+wrsZC1DSIAI0I9IsXL3D16lWUlpaiurpaYplAIIC3tzfrlcbExGDdunWwsbFBcHCwuAGQQCAQh7lWly5d0K5dO2RnZ2Po0KEAapoEmZqaip+Tl5cn1bsQoKZBnyNWgT579iy+++47qSDXakygz549i9WrV2PChAnYvHkzVFVVxcv8/PwQHx8vcQdMRkYGCgoKYGxsjA4dOqBHjx64ffs2hg8fDgCoqqpCQkICXFxcWK2f8BybvhZjx45lZs+ezaSnpzMikYiprKyUerCRnZ3NWFtbM3PmzGFyc3OZvLw88aOgoIC5c+cOY25uzmzatIl59uwZExcXx3z77bfM1KlTmaqqKoZhGObo0aOMpaUlc/LkSSY1NZVZs2YNY2Njw7x+/ZpVDdRjhd89VlhtoTMzM7F27VoYGBh8ymcHFy9eRFlZGW7duiXewtbq378/oqKisGfPHoSEhMDJyQlqampwdHTEqlWrxNeKODs7o6SkBDt27EBhYSH69OmDffv2oUOHDp9UG+EJNp+Kr7/+mjl9+nSTfcq4RFtofm+hWZ1Y8fLywu7du5GWlvYpnx1Cmh2rXY5Tp06hoKAA48ePh5aWltRxZ4FAgKtXrzZLgYQ0BqtA6+rqSp2dI0QRsQr0Tz/91Nx1ENIkWJ9YAYDc3FzcvHkTeXl5cHJyQl5eHkxMTCSOJRPCJdaBDggIwIEDB1BZWQmBQIChQ4di69atePPmDQ4cOECHzYhCYHWUY9++fdi3bx/+9re/4dy5c+I7Vry8vJCfn48dO3Y0a5GEsMUq0FFRUVi4cCE8PT1haGgoHre1tcWSJUsQGxvbTOUR0jisAp2Tk4P+/fvXuax79+548+ZNkxZFyMdiFWg9PT0kJCTUuezBgwfQ09Nr0qII+VisvhROmzYN27dvh7q6OhwdHQEAxcXFOHfuHMLCwuDp6dmsRRLCFqtAe3h4ICsrC8HBwQgODgZQM4UBAEycOBELFixovgoJaQRWgRYIBPjhhx8wd+5c3Lp1C4WFhWjbti1sbGzQq1ev5q6RENYadWLF0NBQ4igHIYqGVaBdXV3FE57/lZKSEtq0aQNDQ0O4uLige/fuTVogIY3B6iiHgYEBHjx4gHv37gEAOnXqBIFAgAcPHiA+Ph5v3rxBTEwMJk2ahKSkpGYtmJD6NOpqu3379qFr167i8by8PHh4eMDOzg4LFiyAt7c3goKCEBoa2mwFE1IfVlvoY8eOYcmSJRJhBmomePHy8kJkZCRUVFQwbdo03L9/v1kKJYQNVoEuKyurd/7nd+/eAQA0NDTqnPCFEHlhFej+/fsjJCQEeXl5EuO5ubn417/+hX79+gEA4uLi0K1bt6avkhCWWO1Dr127FrNmzcLo0aNhbW2NDh064PXr13jw4AE0NTWxY8cOxMbGIjQ0FBs3bmzmkgmRjdUWumfPnjh//jzc3d1RWVmJpKQkKCsrw9PTExcuXICRkRG0tbWxfft2mvCFcIr1iZX27dtj6dKlMpfX7nYQwiXq9EN4hQJNeEXugW6oaVBmZiY8PT3Rv39/DBkyBFu3bpVoGARQ0yAim9wDXV/TIJFIBA8PDwgEAhw5cgSbNm3CiRMnsHPnTvHrqWkQqRfbOcWEQiETHh4u8fPatWuZJ0+esJ6XLC8vjzExMWGuXr0qMT5nzhxm+fLlTExMDNOnTx+msLBQvOzYsWNM3759mbKyMoZhGGbMmDHMli1bxMsrKysZe3t7JiQkhFUNNLcdzW2HuLg4ODs748yZM+IxkUiEuLg4TJs2DQ8fPmT14WmoaVB8fDx69+4tMem5ra0tSktL8ejRI2oaRBrEKtBBQUEYOnQoTp48KR6zsrLChQsXYGtri61bt7JaWW3ToNoZ+4H/NQ0aOXIkcnNz62wIBNTcqFtf06C6eiUWFRUhMzNT4kFNg/iN1XHo5ORk7Ny5U9z8UvxiFRVMnz6ddResv/pr06BLly5BQ0ND4jm1DYLev39PTYNIg1gFuk2bNjIbb7569eqjpgKrq2mQurp6nQ2BamtQV1eXGPvwOdQ0iAAsdzlGjBiBnTt3SvUCFAqFCAkJkeo92JCYmBi4ubmhT58+iIiIQLt27QDU7Er89QKo2p91dXXF0yXU9RxZTYP09fUlHjSLKr+xCvTKlSvRunVrODs7w8HBAdOmTYOjoyMmT54MVVVVrFq1ivUKa5sGff311wgNDZXYnx40aBCEQqFEs8y4uDhoaGjA3NxcomlQrdqmQTY2NqxrIPzFapejY8eOiImJwalTp5CQkICCggLo6elh3rx5mDx5stR+ryw5OTn4/vvvYWtri1WrVqGwsFC8TFVVFaNGjcL27dvh6+uLVatW4eXLlwgMDISbm5t4v9nd3R1+fn7o0aMHrKysEB4ejnfv3jVZ/0TSsrG+OEldXR0zZ87EzJkzP3plbJoGhYWF4R//+AecnZ2hpaUFFxcXiZZx1DSI1Efw/w/CN+jp06fYsWMH4uLiUFJSgnbt2mHAgAHw8fGBiYlJc9fZZDIzM+Ho6IjLly9DX1+f63Kajay79LnCMmYNaujfj9UWOiUlBdOnT4eqqiocHBzQqVMn5OXlITY2FtevX8exY8dowhmiEFgFOjAwEF27dkVkZKREq+GioiK4urpi+/bt2LVrV7MVSQhbrI5yxMfHw8vLS6pvtpaWFhYuXIg7d+40S3GENBarQKuqqso8eaKmpoaKioomLepzUfs/n6xHYGAg1yW2OKx2OSwtLcXXIH84nUF1dTUiIiJgYWHRbAXy2YoVK7BixQoAwNSpUwEAJ06c4LKkFo9VoJcuXYrp06dj/PjxGDt2LDp16oRXr17hv//9LzIyMrB///7mrpMQVlgF2sLCAmFhYQgMDMSePXvAMAwEAgEsLCzwyy+/YNCgQc1dJyGssD6xMnjwYBw/fhxlZWUoKiqqs0UyIVyTGeiMjIx6X/jq1SuJnw0MDJqmIkI+gcxAjx49ulFnm4RCYZMURMinkBlo6u9NWiKZgf7rhfGEtASsvxS+fv0aYWFhiIuLQ3FxMdq3b48BAwbA3d0dOjo6zVkjIayxOlP48uVLTJo0CREREdDQ0IClpSXU1NQQEREBJyenOm9QJYQLrLbQAQEBUFFRwfnz5yXmf05PT4ebmxuCgoLw888/N1uRhLDFagv9+++/429/+5vUZObdunWDt7c3bty40SzFEdJYrAJdVVWF9u3b17msXbt24pYUhHCNVaDNzMwQHR1d57Lo6Gi6uJ8oDFb70IsXL4a7uzvc3Nwwfvx46OjoID8/H7/++ivi4uLE/b8J4RqrQA8ZMgRbtmzB1q1bsX79evF4p06dsHnzZowePbrZCiSkMVgfh544cSK+/fZbpKWl4e3bt9DW1kbPnj0V7mZM8nlrcB+6pKREPA2YQCCAkZER+vfvjzt37qC4uLjZCySkMeoN9MWLF2Fvb4/Dhw9LjGdlZWHjxo2wt7fHtWvXmrVARffX7gKEWzJ3OR4+fIjly5fD0tISY8eOlVjWtWtXREREICAgAN7e3jh9+vRne6RDRUUFP/744ye/z/PnzwGgSd7rhx9++OT3aKlkbqH37t0LCwsLREREwMrKSmr5oEGDEBERgR49emDPnj3NWiQhbMkMdGJiImbOnCk1J/SH1NTUMHv2bNy7d++jVh4aGooZM2ZIjAUGBsLU1FTq8eF/7dQ0iMgiM9AFBQWsrqLr1q3bRzXsiYyMRFBQkNR4cnIynJ2dcePGDYlH7QeLmgaR+sgMdJcuXRq8DQuo+YLYsWNH1ivMzc3FokWLEBAQgB49ekgtT0lJgbm5OXR0dCQetUJDQzFz5kxMmDABxsbG8Pf3h6amJo4cOcK6BsJfMgM9bNgwHDlyBNXV1TJfXFVVhaNHjzaqLfKjR4+goaGBmJgYWFtbSywrKipCdnY2jI2N63wtNQ0iDZEZ6Llz5+LZs2dYunQp8vPzpZa/evUKvr6+EAqFmDt3LusVOjg4IDAwsM6balNSUgDUTIo+ZswY2Nvb47vvvhPP2E9Ng0hDZH7j6969OwICArB69WrY29ujT58+0NfXR1VVFbKysiAUCqGiogJ/f/86j4J8jNpAa2pqIjg4GPn5+QgKCoKrqyuio6OpaRBpUL2nvh0dHXHmzBlERETg//7v/3DlyhUoKytDT08Pc+bMwaxZs9C1a9cmK2bGjBkYP368uE+hmZkZTExMYGdnh0uXLon3ualpEJGlwWs59PX1sXbtWqxdu7bZixEIBBJNN4GaL6ft2rVDdnY2hg4dCqCmSZCpqan4OfU1DfrrjKmE31hdDz1nzhykpqbWuezRo0eYMGFCkxTj5+eHSZMmSYxlZGSgoKAAxsbG1DSINEjmFjo6Olp8MuP27ds4e/as1C1YQE2XqszMzCYpZuzYsYiKioKfnx9mz56NvLw88T76yJEjAVDTIFI/mYFOTk4WzyoqEAiwd+9emW+ycOHCJilm4MCB2LNnD0JCQuDk5AQ1NTU4Ojpi1apV4ml8qWkQqY/MQPv6+mL27NlgGAajRo3Cjh07pOaBVlZWhqampkSvwcb45z//KTU2fPhwqQ5Zf+Xu7g53d/ePWifhN5mBVlNTEx/BOHToEMzNzT86uITIC6s7VmxsbFBdXY3z58/jxo0byMvLw/r163H//n1YWVmhZ8+ezV0nIaywCnRJSQkWLFiAe/fuQUdHB69evcK7d+9w+vRp+Pn54fDhwzAzM2vuWglpEKvDdgEBAUhLS0NUVBSuXr0qbqK4fft2GBgYYPv27c1aJCFssQr0xYsX4evri379+kncFNu+fXssWLAA9+/fb7YCCWkMVrsc7969k7ogqJaGhob4GgvSOPHx8bh7967E2IeHR/v374+BAwfKu6wWjVWgTUxMEB0dLT658aHLly9LnIYm7A0cOJAC28RYBdrLywve3t5wc3PDqFGjIBAIcPPmTRw9ehQnTpzAjh07mrtOQlhhtQ9dew3z8+fPsWnTJjAMg4CAAFy5cgV+fn40cxJRGKxnTho3bhzGjRuHZ8+eoaCgAFpaWujZs6dEZ1lCuMY60LV69OhR572AhCgCmYGeM2cO6zcRCAQ4ePBgkxREyKeQGeiKiooGJ2JMSUlBSUkJlJWVm7wwQj6GzEBHRUXJfFFJSQn++c9/4u7duzA0NKzzqjlCuNDofehbt25h3bp1yMnJwZw5c7BixQq0atWqOWojpNFYB7q8vBxbt25FVFQUunbtikOHDtFJAaJwWAU6Pj4e69atQ0ZGBmbMmIFVq1ahdevWzV0bIY1Wb6BFIhG2bduGgwcP4osvvsD+/fsxePBgedVGSKPJDHRiYiK+++47PH/+HNOmTcOaNWvqnPuCEEUiM9AzZsxAdXU12rZti6dPn2LBggUy30QgEEjN8k8IF2QGun///vKsg5AmITPQERER8qyDkCZBVxYRXqFAE16hQBNe4TTQdTUNyszMhKenJ/r3748hQ4Zg69atUr0AqWkQkYVVoF++fImKigqJnz+14WRdTYNEIhE8PDwgEAhw5MgRbNq0CSdOnMDOnTvFz6GmQaQ+MgN97do1lJSUAKiZ+PzPP/8EUDN9raOjI5KTkz9qhfU1Dbpw4QKysrKwZcsWmJiYwNHREStXrsShQ4dQXl4OgJoGkfrJDPTKlSthY2ODiRMngmEY3LhxA2lpaWAYRjzRzMeor2lQfHw8evfuLTHpua2tLUpLS/Ho0SNqGkQaJPM49O3bt5GcnIw7d+7A398fR48exb/+9S9oa2tDIBAgKioKQ4YMQe/evRt1S5aDgwMcHBzqXJabm1tnQyCgppWEuro6gLqbBtX+D/KhoqIiFBUVSYxR0yB+kxlogUAAMzMzmJmZwd/fHzt37kSPHj2QkJAAHx8fpKSk4Nq1a8jPz0fr1q0/upvsh8rLy6GhoSExVtsg6P3799Q0iDRIZqD379+PgQMHwtzcHEBNwNu3bw97e3sANQ3S+/Tpgzdv3nz0/vRfqaur19kQCADatGkj3kJT0yAii8xAX7hwAYGBgVBVVYVAIMDRo0fx+vVr9OnTBwDE9xt26NABX375ZZMUo6urC6FQKDFW26NQV1cXenp64jFqGkTqIvNL4ZEjR5CQkIDdu3eDYRg8f/4cGzduhL29PQQCAbZu3YqQkBBcvnwZL1++bJJiBg0aBKFQKLHfGxcXBw0NDZibm1PTINKgeo9Dt2rVSnxB/+rVq3Hjxg1cvHgRDMOgQ4cOuHv3LtavXw9HR8cmKWbUqFHo0qULfH19kZSUhCtXriAwMBBubm7i/WZ3d3ccPHgQp06dwpMnT7B+/XpqGkTEWN2CpaenJ74RVl9fH3p6eli0aBF69eoF4H8tiz9Vq1atEBYWhn/84x9wdnaGlpYWXFxc4O3tLX4ONQ0i9WEV6CtXroj/rKSkJPEzgDr3X9moa/qD7t27Izw8vN7XUdMgIgtdnER4hQJNeIUCTXiFAk14hQJNeIUCTXiFAk14hQJNeIUCTXiFAk14hQJNeIUCTXiFAk14hQJNeIUCTXiFAk14hQJNeIUCTXiFAk14hQJNeIUCTXiFAk14hQJNeIUCTXiFAk14RSEDnZaWBlNTU6nH8ePHAQBCoRCurq7o27cvRo4c2eBMS+TzwWoqMHlLTk6GpqYm/vvf/0qMt23bFm/evMG8efMwevRobNy4EYmJidi4cSPatm0LZ2dnjiomikIhA52SkgIjIyPo6OhILTtw4ABUVVWxceNGqKiowMjICC9evMDevXsp0EQxdzmSk5NhZGRU57L4+HgMHDgQKir/+yza2toiIyOjyWZBJS2Xwm6hu3fvjunTpyM9PR2GhoZYvHgxhg0bhtzcXBgbG0s8v7axUHZ2tsRMqNQ06POjcIEuLS1FZmYmOnTogBUrVohbwM2fPx/79u1DeXl5nU2DgJrGQh+ipkGfH4ULdJs2bZCQkABVVVVxUC0sLPD06VOEhYU12FjoQ9Q06POjcIEGINXaDQBMTExw9epVGBgYiBsJ1fqwsdCHqGnQ50fhvhTeu3cP/fr1k2pI//DhQ/Tq1QuDBg1CQkKCRK/xW7duwdDQsM6jIuTzonCBtrCwgL6+Pr7//nskJCTg6dOn8PPzw7179+Dl5YUpU6agrKwM69atw5MnTxAdHY0DBw7A09OT69KJAlC4XQ5VVVWEhYUhMDAQS5YsQVFRkbgxUG0T0PDwcPj7+8PJyQk6OjpYsWIFJk+ezHHlRBEoXKCBmiZEP//8s8zllpaWOHLkiBwrIi2FwkzGfWIAAA4vSURBVO1yEPIpKNCEVyjQhFco0IRXKNCEVyjQhFco0IRXKNCEVyjQhFco0IRXKNCEVyjQhFco0IRXKNCEVyjQhFco0IRXKNCEVyjQhFco0IRXKNCEVyjQhFco0IRXKNCEVyjQhFdaZKCrq6sRHByM4cOHw9raGu7u7njx4gXXZREF0CIDvWvXLkRFRcHPzw9Hjx6FsrIyPDw8pOaHJp+fFhdokUiEffv2wcfHB3Z2djAzM0NQUBBevXqF//znP1yXRzjW4gItFApRWlqKwYMHi8c0NTVhbm6O+Ph4DisjiqDFBbq2MdCHvVSAmj4r2dnZXJREFIhCzj5an7KyMgCos8/KX1tV1NU0KCsrC0DTNg8qKSlpsvdqCpmZmRJdwhRBZmZmk7xP7b9bVVVVncsV67dmQV1dHUDNvvSHoRaJRFI9VuprGsTnPisxMTHo2bMn12VIcHR0bNL3y8/PR/fu3aXGW1ygv/jiCwA1fVU0NTXF43l5eVLt3upqGiQSiZCRkQFDQ0MoKys3f8Es1TYzioyMlOoVwxVFrKmqqgr5+fmwsLCoc3mLC7SZmRk0NTVx+/Zt8VaopKQEjx8/xsyZMyWeK6tpkKJtvT6kq6sLfX19rsuQoGg11bVlrtXiAq2mpobZs2cjKCgInTp1gr6+PgIDA9GlSxeMGTOG6/IIx1pcoAFgyZIlqKqqwoYNG1BWVoYBAwYgLCxM6osi+fy0yEArKytj5cqVWLlyJdelEAXT4o5D85WWlhZ8fHwUqlGoItbUEAHDMAzXRRDSVGgLTXiFAk14hQJNeIUCTXilRR62a6lkXVdSFx8fn2as5H+io6NZP3fSpEnNWEnToKMccmRnZyfxc15eHpSVldG1a1eoqqoiIyMDVVVVMDMzw4kTJ+RSk5mZmcTPAoEADMNAVVUVKioqKCsrg7KyMrS0tHDz5k251PRJGMKJgwcPMq6urkxubq54rKCggFm4cCETFBQktzoqKyvFj5iYGGbChAnMn3/+KV6emprKTJs2jTl8+LDcavoUFGiODBkyhBEKhVLjycnJjI2NDQcVMczIkSOZe/fuSY0nJiYyQ4cO5aCixqMvhRwRiUR49+6d1Pjr1685qKbG27dv0apVK6nx6upqlJeXc1BR41GgOeLg4ID169fjjz/+QFFREd6+fYvY2Fh8//33+OabbzipydbWFj/++KPElBCpqan48ccfMXLkSE5qaiz6UsiRkpISLF26FL///jsEAgEAgGEYjB07Fj///DMnVw7m5ubCw8MDT58+hYaGBhiGQWlpKaysrLB3715oa2vLvabGokBz7NmzZ0hJSQEAmJubw8DAgNN6qqqq8McffyA1NRUA0Lt3bwwePFj8oVN0tMvBoaqqKqSlpeHFixf48ssvUVhYKL4JmCvKysrQ0dGBrq4upk+fjm7durWYMAN0YoUzhYWFcHV1xZMnTwAAX3/9NUJCQvD06VMcPHgQXbt2lXtN5eXl8Pb2xu+//w4lJSVcuHABP/30E16/fo3Q0FC0a9dO7jU1Fm2hORIYGIh27drhxo0b4jvZf/zxR3Ts2BE///wzJzUFBwcjPz8fx48fFx/t8PLywtu3bxEQEMBJTY3G5THDz5mdnR3z4MEDhmEYpm/fvkx6ejrDMAzz4MEDxtbWlpOaHB0dmZs3b0rVFBcXR8ehSf0KCgrq/C9cQ0ODs2O+eXl5dd7d3blzZ6kJexQVBZojlpaWOH/+vNT4oUOHZM450dyMjY3x+++/S42fPXsWvXr14qCixqMvhRzx9fWFu7s77t+/j8rKSuzatQupqalISUnBvn37OKnJ29sbvr6+SE1NRVVVFU6cOIFnz57h0qVL2LFjByc1NRYdh+aQUCjEvn37IBQKUV1dDRMTE7i7u8PKyoqzmq5du4bQ0FCJmjw9PTFq1CjOamoMCjRH7ty5g379+klNqigSiXDlyhWMHTtW7jVVVlYq3CSPjUX70ByZM2dOnV+0srKysHr1ag4qAoYNG4bNmzcjKSmJk/U3BdpCy1FkZKR4/zgrKwtffPEFlJQktylFRUXQ1dXF2bNn5V7fwYMHcfbsWTx8+BBmZmaYPHkyvvnmG3To0EHutXwsCrQclZaW4pdffkF1dTVCQ0Mxd+5ctG7dWrxcIBBAQ0MD48aNg56eHmd1pqWl4cyZMzh37hxyc3MxcuRIODk5wcHBgbOa2KJAcyQkJAQeHh4SgVZEUVFRCAgIQGlpKYRCIdflNIgCzZGKigoEBATgiy++wLx58wAA48aNg4ODA5YvXy61KyJP6enpiImJwdmzZ5GVlYUhQ4Zg8uTJnHxRbayW/ZW2Bdu2bRvOnz8Pf39/8di8efOwc+dOqKmpYcmSJXKvKTIyEjExMUhMTIShoSGmTJmCSZMmoXPnznKv5aNxc8adDB8+XHzdxIeuX7/OjBw5koOKGGbAgAHM999/z9y9e5eT9TcF2kJzpLi4WKqTFwDo6+tzdl/h6NGj4e7uDkNDQ07W3xToODRHevfujWPHjkmNnz59GiYmJhxUBPz2228t/sQKfSnkyM2bNzF//nz06dMH1tbWAIA///wTDx8+xK5du6QmpZGHlStXQltbG0uXLm1Rc0J/iALNoQcPHuDgwYNITU2Fqqoqevbsifnz50vNZiQvM2fOxN27dyEQCKClpSV1SDE2NpaTuhqDAk3EGpp7T17z7X0KCjSH4uPjceDAAaSlpSE8PBznzp1Dr169ONnd4Av6UsiRhIQEzJs3DxUVFcjIyEBlZSVycnLg5eWF3377jbO64uPj4ePjg3HjxiE7OxthYWG4du0aZ/U0FgWaI9u3b8fixYsRGhoqPrKwfv16LFiwALt37+akJkX9kDUGBZojjx8/xrhx46TGp06dirS0NA4qUswPWWNRoDmipqZW5/XQmZmZ0NDQ4KAixfyQNRYFmiOjRo3Ctm3bUFBQIB5LSUnB5s2bYW9vz0lNivghaywKNEdWr16NsrIyDBkyBGVlZfj2228xceJEqKqqYtWqVZzUpIgfssaiw3Yc++OPPyRuSB0+fDhnl44WFxdj/vz5SExMBMMwaN26NcrLy9G7d2/s37+fZh8lDSspKUFaWhqUlJRgZGSkEBf8K9KHrLEo0BypqKjApk2bcOrUKVRVVQEAWrVqhVmzZmHlypUKMeOnSCTC48ePYWxsDE1NTa7LYYeDS1YJwzCbN29mBg8ezBw5coRJSkpiHj9+zBw+fJixtbVlQkJCOKkpKyuLcXV1Ze7fv8+8f/+ecXJyYkxNTRkbGxvm4cOHnNTUWBRojgwePJi5cuWK1PjFixeZESNGcFARw3h7ezNTp05lMjIymJMnTzIDBgxg7t27x/z9739n5s6dy0lNjdWyL35twUQiEbp16yY1bmRkxNnEiHFxcTh06BD09fUREBCAYcOGoW/fvtDW1sbkyZM5qamxWsaePg85OTlh+/btEIlE4rHq6mrs378fU6ZM4aSmyspKaGtrg2EY3Lx5E19++aW4rpZy4X/LqJKHCgoKEBsbC3t7e1hYWEBFRQVCoRDZ2dmwtLTErFmzxM+NjIyUS03m5uY4ceIEOnbsiKKiItjZ2UEkEiEsLAy9e/eWSw2figLNETU1Nan2bba2thxVU2P16tVYtGgRCgsLsWDBAujq6mLDhg347bffEB4ezmltbNFhOwVUUVEBVVVVTtZdXV2NkpIS8S1YT58+RYcOHdC+fXtO6mks2ofmyNq1a+ucqT81NRVTp07loKIaSkpKePDgAfbv34/IyEi8fv26xYQZoF0Ozly/fh2TJ0/Gtm3bxPcQHjhwANu2bUOfPn04qSk/Px9ubm548uQJtLW1UV1djeLiYgwaNAg7duxoGZM2cnzY8LP15s0bxtPTk7GysmJ++eUXZt68eYy1tTWzf/9+prq6mpOaFi9ezLi4uDDPnz8XjyUlJTETJ05kli9fzklNjUWB5tiaNWsYU1NTpk+fPkx8fDyntVhbWzOPHj2SGr937x7Tr18/DipqPNqH5khxcTHWrFmDM2fOwNnZGebm5li8eDFiYmI4q0lbW1viuHgtgUCANm3acFDRR+D6E/W5GjZsGDN06FAmNjaWYRiGqaysZIKCghhzc3Nm0aJFcqujqqpK/Dhy5Ajz7bffMklJSeLl6enpzJQpU5jDhw/LraZPQYftOLJ48WL4+/tLHUG4d+8eVq9eLbebUs3MzCSu7GMYBgKBAOrq6lBSUkJpaSkEAgG0tbVx8+ZNudT0KSjQHPr1118RFhaGZ8+e4ddff8W///1vGBgYYNKkSXL7L/7UqVOsL1V1cnJq5mo+HR2248jly5exbt06iRtQtbW1sXnzZigrK8PFxUUudfz1oqOrV69i9+7dSElJgaqqKoyNjbFgwYIW0Y4CoC00Z6ZOnQonJyfMmjUL/fr1Q0xMDAwMDHDo0CEcPXoU586dk3tNly9fho+PD+zs7GBjY4Pq6mrcvn0b169fx549e1rGjE4c7r9/1j5sDv/hn9PT0xkrKytOanJycmK2bdsmNb5t2zbGxcWFg4oajw7bcaRt27bIzs6WGk9JSamzqb08PHnypM795EmTJrWY3oUUaI5888032Lx5Mx49egSBQIDi4mJcvXoVmzZtqnOyF3no2LEjcnJypMZzcnJazHFoCjRHli1bJm7MU1paiilTpsDLywvm5ubw9fXlpCYHBwds2rRJYmssFArh7+9PXwoJO+np6Xj8+DGqq6thamoKIyMjzmp5+/Yt3NzcIBQK0bp1awgEApSWltK8HKTlqq6uxvXr15GSkgIAMDU1xbBhw2heDkK40DI+doSwRIEmvEKBJrxCgSa8QoEmvPL/AOWpT8rNxY4qAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 144x360 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot the observed and expected overlaps\n",
    "plt.figure(figsize=(2,5))\n",
    "plt.bar([0,1],[network_exp_mean_overlap,network_num_overlap],color=['gray','black'],width=1)\n",
    "plt.errorbar([0],[network_exp_mean_overlap],[2*network_exp_std_overlap],color='k',capsize=4)\n",
    "plt.ylabel('# NetColoc genes')\n",
    "plt.xticks([0,1],['expected','observed'],rotation='vertical')\n",
    "plt.xlim([-1,2])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 9. Compute network colocalization subgraph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "350\n",
      "1810\n"
     ]
    }
   ],
   "source": [
    "# select genes in network intersection, make a subgraph\n",
    "\n",
    "G_overlap = network_colocalization.calculate_network_overlap_subgraph(G_int,z_D1['z'],z_D2['z'],z_score_threshold=zthresh,\n",
    "                                                                     z1_threshold=1.5,z2_threshold=1.5)\n",
    "print(len(G_overlap.nodes()))\n",
    "print(len(G_overlap.edges()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10. Transform NetColoc subgraph edges to cosine similarities\n",
    "\n",
    "Cosine similarity transformation (and network transformations in general) provide better visualization of clusters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing the adjacency matrix...\n",
      "computing the cosine similarity...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/brinrosenthal/anaconda3/lib/python3.7/site-packages/scipy/spatial/distance.py:714: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  dist = 1.0 - uv / np.sqrt(uu * vv)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rank transforming...\n",
      "number of transformed edges returned = \n",
      "1721\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/brinrosenthal/anaconda3/lib/python3.7/site-packages/netcoloc/network_colocalization.py:241: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.\n",
      "  sim_rank.values[[np.arange(sim_rank.shape[0])]*2] = 0\n"
     ]
    }
   ],
   "source": [
    "G_cosSim=network_colocalization.transform_edges(G_overlap,method='cosine_sim',edge_weight_threshold=0.95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Compute network colocalized systems map_\n",
    "\n",
    "### 11. Convert network colocalization subgraph to form used in community detection module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d1_seeds</th>\n",
       "      <th>d2_seeds</th>\n",
       "      <th>z_d1</th>\n",
       "      <th>z_d2</th>\n",
       "      <th>z_both</th>\n",
       "      <th>d1_name</th>\n",
       "      <th>d2_name</th>\n",
       "      <th>sum_seeds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KDM5B</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8.745623</td>\n",
       "      <td>8.198657</td>\n",
       "      <td>71.702356</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLUH</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5.294906</td>\n",
       "      <td>13.015273</td>\n",
       "      <td>68.914648</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2C</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8.819306</td>\n",
       "      <td>7.349744</td>\n",
       "      <td>64.819635</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTEN</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6.944277</td>\n",
       "      <td>7.072284</td>\n",
       "      <td>49.111900</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF3G</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8.506393</td>\n",
       "      <td>5.452404</td>\n",
       "      <td>46.380288</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNB1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6.026975</td>\n",
       "      <td>6.929012</td>\n",
       "      <td>41.760980</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POGZ</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.462101</td>\n",
       "      <td>7.565479</td>\n",
       "      <td>41.323410</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HDLBP</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6.986262</td>\n",
       "      <td>5.816466</td>\n",
       "      <td>40.635353</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NSD1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6.411852</td>\n",
       "      <td>6.329823</td>\n",
       "      <td>40.585882</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KDM6B</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10.991122</td>\n",
       "      <td>3.405209</td>\n",
       "      <td>37.427066</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADNP</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>7.887236</td>\n",
       "      <td>4.654060</td>\n",
       "      <td>36.707673</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KDM5A</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4.575175</td>\n",
       "      <td>7.306914</td>\n",
       "      <td>33.430408</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.438264</td>\n",
       "      <td>5.951293</td>\n",
       "      <td>32.364702</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2E</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10.783263</td>\n",
       "      <td>2.628955</td>\n",
       "      <td>28.348710</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HDGFRP2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.702384</td>\n",
       "      <td>6.002973</td>\n",
       "      <td>28.228286</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         d1_seeds  d2_seeds       z_d1       z_d2     z_both d1_name d2_name  \\\n",
       "KDM5B           1         1   8.745623   8.198657  71.702356     ASD     CHD   \n",
       "CLUH            0         1   5.294906  13.015273  68.914648     ASD     CHD   \n",
       "KMT2C           1         1   8.819306   7.349744  64.819635     ASD     CHD   \n",
       "PTEN            1         1   6.944277   7.072284  49.111900     ASD     CHD   \n",
       "EIF3G           1         0   8.506393   5.452404  46.380288     ASD     CHD   \n",
       "CTNNB1          1         1   6.026975   6.929012  41.760980     ASD     CHD   \n",
       "POGZ            1         1   5.462101   7.565479  41.323410     ASD     CHD   \n",
       "HDLBP           1         0   6.986262   5.816466  40.635353     ASD     CHD   \n",
       "NSD1            1         1   6.411852   6.329823  40.585882     ASD     CHD   \n",
       "KDM6B           1         0  10.991122   3.405209  37.427066     ASD     CHD   \n",
       "ADNP            1         0   7.887236   4.654060  36.707673     ASD     CHD   \n",
       "KDM5A           0         1   4.575175   7.306914  33.430408     ASD     CHD   \n",
       "KLF10           0         0   5.438264   5.951293  32.364702     ASD     CHD   \n",
       "KMT2E           1         0  10.783263   2.628955  28.348710     ASD     CHD   \n",
       "HDGFRP2         0         0   4.702384   6.002973  28.228286     ASD     CHD   \n",
       "\n",
       "         sum_seeds  \n",
       "KDM5B            2  \n",
       "CLUH             1  \n",
       "KMT2C            2  \n",
       "PTEN             2  \n",
       "EIF3G            1  \n",
       "CTNNB1           2  \n",
       "POGZ             2  \n",
       "HDLBP            1  \n",
       "NSD1             2  \n",
       "KDM6B            1  \n",
       "ADNP             1  \n",
       "KDM5A            1  \n",
       "KLF10            0  \n",
       "KMT2E            1  \n",
       "HDGFRP2          0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compile dataframe of metadata for overlapping nodes\n",
    "node_df = pd.DataFrame(index=list(G_overlap.nodes))\n",
    "node_df['d1_seeds']=0\n",
    "node_df['d2_seeds']=0\n",
    "node_df['d1_seeds'].loc[list(np.intersect1d(seed_dict[d1_name],node_df.index.tolist()))]=1\n",
    "node_df['d2_seeds'].loc[list(np.intersect1d(seed_dict[d2_name],node_df.index.tolist()))]=1\n",
    "node_df['z_d1']=z_D1.loc[list(G_overlap.nodes)]['z']\n",
    "node_df['z_d2']=z_D2.loc[list(G_overlap.nodes)]['z']\n",
    "node_df['z_both']=node_df['z_d1']*node_df['z_d2']\n",
    "node_df['d1_name']=d1_name\n",
    "node_df['d2_name']=d2_name\n",
    "node_df['sum_seeds']=node_df['d1_seeds']+node_df['d2_seeds']\n",
    "\n",
    "node_df = node_df.sort_values('z_both',ascending=False)\n",
    "node_df.head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 12. Run community detection on NetColoc subgraph (recommend HiDef).\n",
    "\n",
    "Documentation for CDAPS utils to build multiscale systems map in notebook\n",
    "\n",
    "https://cdapsutil.readthedocs.io/en/latest/quicktutorial.html#example  \n",
    "https://cdapsutil.readthedocs.io/en/latest/cdapsutil.html#community-detection  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "350\n",
      "1810\n"
     ]
    }
   ],
   "source": [
    "print(len(G_overlap.nodes()))\n",
    "print(len(G_overlap.edges()))\n",
    "G_overlap_cx = ndex2.create_nice_cx_from_networkx(G_overlap)\n",
    "G_overlap_cx.set_name(d1_name+'_'+d2_name+'_NetColoc_subgraph') \n",
    "for node_id, node in G_overlap_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_overlap_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "cd = cdapsutil.CommunityDetection()\n",
    "\n",
    "# Run HiDeF on CDAPS REST service\n",
    "G_hier = cd.run_community_detection(G_overlap_cx, algorithm='hidefv1.1beta',arguments={'--maxres':'20'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hierarchy name: hidefv1.1beta_(none)_ASD_CHD_NetColoc_subgraph\n",
      "# nodes: 96\n",
      "# edges: 114\n"
     ]
    }
   ],
   "source": [
    "# Print information about hierarchy\n",
    "print('Hierarchy name: ' + str(G_hier.get_name()))\n",
    "print('# nodes: ' + str(len(G_hier.get_nodes())))\n",
    "print('# edges: ' + str(len(G_hier.get_edges())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sr = cdapsutil.ServiceRunner()\n",
    "# sr.get_algorithms() # this will print out available options for each algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 13. Convert NetColoc hierarchy to networkx format, and write out features of the hierarchy to a pandas dataframe, for easier access in python.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nodes:\n",
      "96\n",
      "\n",
      "number of edges:\n",
      "114\n"
     ]
    }
   ],
   "source": [
    "G_hier = G_hier.to_networkx(mode='default')\n",
    "G_hier\n",
    "\n",
    "nodes = G_hier.nodes()\n",
    "\n",
    "# print out interactome num nodes and edges for diagnostic purposes\n",
    "print('number of nodes:')\n",
    "print(len(G_hier.nodes()))\n",
    "print('\\nnumber of edges:')\n",
    "print(len(G_hier.edges()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>represents</th>\n",
       "      <th>name</th>\n",
       "      <th>system_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TNRC6C TWIST1 METTL16 HP1BP3 DPY30 ETV2 MYF6 G...</td>\n",
       "      <td>331</td>\n",
       "      <td>False</td>\n",
       "      <td>8.371</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>C350</td>\n",
       "      <td>C350</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NAV1 SPTBN5 SCN3B GLDN CACNA1A SPTBN1 SCN4B SP...</td>\n",
       "      <td>54</td>\n",
       "      <td>False</td>\n",
       "      <td>5.755</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>178</td>\n",
       "      <td>C353</td>\n",
       "      <td>C353</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PCDHA2 SMARCA4 ZBTB14 FHL2 ADNP HP1BP3 TNFRSF2...</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>3.322</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>C386</td>\n",
       "      <td>C386</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PCDHA2 CDCA7L METTL16 HP1BP3 ZBTB14 CLUH GLYR1...</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>5.322</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>41</td>\n",
       "      <td>C356</td>\n",
       "      <td>C356</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PCDHA2 C17orf49 ERICH6 CDCA7L ZBTB14 HP1BP3 MG...</td>\n",
       "      <td>34</td>\n",
       "      <td>False</td>\n",
       "      <td>5.087</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27</td>\n",
       "      <td>C358</td>\n",
       "      <td>C358</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       CD_MemberList  CD_MemberList_Size  \\\n",
       "0  TNRC6C TWIST1 METTL16 HP1BP3 DPY30 ETV2 MYF6 G...                 331   \n",
       "1  NAV1 SPTBN5 SCN3B GLDN CACNA1A SPTBN1 SCN4B SP...                  54   \n",
       "2  PCDHA2 SMARCA4 ZBTB14 FHL2 ADNP HP1BP3 TNFRSF2...                  10   \n",
       "3  PCDHA2 CDCA7L METTL16 HP1BP3 ZBTB14 CLUH GLYR1...                  40   \n",
       "4  PCDHA2 C17orf49 ERICH6 CDCA7L ZBTB14 HP1BP3 MG...                  34   \n",
       "\n",
       "  CD_Labeled CD_MemberList_LogSize CD_CommunityName CD_AnnotatedMembers  \\\n",
       "0      False                 8.371                                        \n",
       "1      False                 5.755                                        \n",
       "2      False                 3.322                                        \n",
       "3      False                 5.322                                        \n",
       "4      False                 5.087                                        \n",
       "\n",
       "  CD_AnnotatedMembers_Size CD_AnnotatedMembers_Overlap  \\\n",
       "0                        0                         0.0   \n",
       "1                        0                         0.0   \n",
       "2                        0                         0.0   \n",
       "3                        0                         0.0   \n",
       "4                        0                         0.0   \n",
       "\n",
       "  CD_AnnotatedMembers_Pvalue  HiDeF_persistence represents  name  system_ID  \n",
       "0                        0.0                  0       C350  C350          0  \n",
       "1                        0.0                178       C353  C353          1  \n",
       "2                        0.0                 14       C386  C386          2  \n",
       "3                        0.0                 41       C356  C356          3  \n",
       "4                        0.0                 27       C358  C358          4  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add node attributes to dataframe for easier access\n",
    "hier_df = pd.DataFrame.from_dict(dict(G_hier.nodes(data=True)), orient='index')\n",
    "hier_df['system_ID']=hier_df.index.tolist()\n",
    "# some columns are not the right type\n",
    "hier_df['CD_MemberList_Size']=[int(x) for x in hier_df['CD_MemberList_Size'].tolist()]\n",
    "hier_df['HiDeF_persistence']=[int(x) for x in hier_df['HiDeF_persistence'].tolist()]\n",
    "hier_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 14. Remove systems with no seed genes (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>represents</th>\n",
       "      <th>name</th>\n",
       "      <th>system_ID</th>\n",
       "      <th>num_d1_seeds</th>\n",
       "      <th>num_d2_seeds</th>\n",
       "      <th>frac_d1_seeds</th>\n",
       "      <th>frac_d2_seeds</th>\n",
       "      <th>frac_no_seeds</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C350</th>\n",
       "      <td>TNRC6C TWIST1 METTL16 HP1BP3 DPY30 ETV2 MYF6 G...</td>\n",
       "      <td>331</td>\n",
       "      <td>False</td>\n",
       "      <td>8.371</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>C350</td>\n",
       "      <td>C350</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>18</td>\n",
       "      <td>0.081571</td>\n",
       "      <td>0.054381</td>\n",
       "      <td>0.864048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C353</th>\n",
       "      <td>NAV1 SPTBN5 SCN3B GLDN CACNA1A SPTBN1 SCN4B SP...</td>\n",
       "      <td>54</td>\n",
       "      <td>False</td>\n",
       "      <td>5.755</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>178</td>\n",
       "      <td>C353</td>\n",
       "      <td>C353</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>0.092593</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.851852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C386</th>\n",
       "      <td>PCDHA2 SMARCA4 ZBTB14 FHL2 ADNP HP1BP3 TNFRSF2...</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>3.322</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>C386</td>\n",
       "      <td>C386</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C356</th>\n",
       "      <td>PCDHA2 CDCA7L METTL16 HP1BP3 ZBTB14 CLUH GLYR1...</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>5.322</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>41</td>\n",
       "      <td>C356</td>\n",
       "      <td>C356</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.075000</td>\n",
       "      <td>0.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C358</th>\n",
       "      <td>PCDHA2 C17orf49 ERICH6 CDCA7L ZBTB14 HP1BP3 MG...</td>\n",
       "      <td>34</td>\n",
       "      <td>False</td>\n",
       "      <td>5.087</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27</td>\n",
       "      <td>C358</td>\n",
       "      <td>C358</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.088235</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.882353</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CD_MemberList  CD_MemberList_Size  \\\n",
       "name                                                                          \n",
       "C350  TNRC6C TWIST1 METTL16 HP1BP3 DPY30 ETV2 MYF6 G...                 331   \n",
       "C353  NAV1 SPTBN5 SCN3B GLDN CACNA1A SPTBN1 SCN4B SP...                  54   \n",
       "C386  PCDHA2 SMARCA4 ZBTB14 FHL2 ADNP HP1BP3 TNFRSF2...                  10   \n",
       "C356  PCDHA2 CDCA7L METTL16 HP1BP3 ZBTB14 CLUH GLYR1...                  40   \n",
       "C358  PCDHA2 C17orf49 ERICH6 CDCA7L ZBTB14 HP1BP3 MG...                  34   \n",
       "\n",
       "     CD_Labeled CD_MemberList_LogSize CD_CommunityName CD_AnnotatedMembers  \\\n",
       "name                                                                         \n",
       "C350      False                 8.371                                        \n",
       "C353      False                 5.755                                        \n",
       "C386      False                 3.322                                        \n",
       "C356      False                 5.322                                        \n",
       "C358      False                 5.087                                        \n",
       "\n",
       "     CD_AnnotatedMembers_Size CD_AnnotatedMembers_Overlap  \\\n",
       "name                                                        \n",
       "C350                        0                         0.0   \n",
       "C353                        0                         0.0   \n",
       "C386                        0                         0.0   \n",
       "C356                        0                         0.0   \n",
       "C358                        0                         0.0   \n",
       "\n",
       "     CD_AnnotatedMembers_Pvalue  HiDeF_persistence represents  name  \\\n",
       "name                                                                  \n",
       "C350                        0.0                  0       C350  C350   \n",
       "C353                        0.0                178       C353  C353   \n",
       "C386                        0.0                 14       C386  C386   \n",
       "C356                        0.0                 41       C356  C356   \n",
       "C358                        0.0                 27       C358  C358   \n",
       "\n",
       "      system_ID  num_d1_seeds  num_d2_seeds  frac_d1_seeds  frac_d2_seeds  \\\n",
       "name                                                                        \n",
       "C350          0            27            18       0.081571       0.054381   \n",
       "C353          1             5             3       0.092593       0.055556   \n",
       "C386          2             1             0       0.100000       0.000000   \n",
       "C356          3             5             3       0.125000       0.075000   \n",
       "C358          4             3             1       0.088235       0.029412   \n",
       "\n",
       "      frac_no_seeds  \n",
       "name                 \n",
       "C350       0.864048  \n",
       "C353       0.851852  \n",
       "C386       0.900000  \n",
       "C356       0.800000  \n",
       "C358       0.882353  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hier_df.index=hier_df['name']\n",
    "hier_df.head()\n",
    "\n",
    "num_d1_seeds, num_d2_seeds = [],[]\n",
    "frac_d1_seeds, frac_d2_seeds=[],[]\n",
    "systems_keep = []\n",
    "for c in hier_df.index.tolist():\n",
    "    system_genes = hier_df['CD_MemberList'].loc[c].split(' ')\n",
    "    num_ASD_temp = len(list(np.intersect1d(system_genes,seed_dict[d1_name])))\n",
    "    num_CHD_temp = len(list(np.intersect1d(system_genes,seed_dict[d2_name])))\n",
    "    if (num_ASD_temp+num_CHD_temp)>0: # keep the system if it has at least 1 seed genes\n",
    "        systems_keep.append(c)\n",
    "        num_d1_seeds.append(num_ASD_temp)\n",
    "        num_d2_seeds.append(num_CHD_temp)\n",
    "        \n",
    "        frac_d1_seeds.append(num_ASD_temp/np.float(len(system_genes)))\n",
    "        frac_d2_seeds.append(num_CHD_temp/np.float(len(system_genes)))\n",
    "\n",
    "        \n",
    "frac_no_seeds = np.subtract(1.0,np.add(frac_d1_seeds,frac_d2_seeds))\n",
    "\n",
    "hier_df = hier_df.loc[systems_keep]\n",
    "hier_df['num_d1_seeds']=num_d1_seeds\n",
    "hier_df['num_d2_seeds']=num_d2_seeds\n",
    "hier_df['frac_d1_seeds']=frac_d1_seeds\n",
    "hier_df['frac_d2_seeds']=frac_d2_seeds\n",
    "hier_df['frac_no_seeds']=frac_no_seeds\n",
    "print(len(hier_df))\n",
    "\n",
    "hier_df.head()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64\n",
      "76\n"
     ]
    }
   ],
   "source": [
    "# prune G_hier--> only keep systems with at least one seed gene\n",
    "\n",
    "nkeep=[]\n",
    "for n in list(G_hier.nodes()):\n",
    "    if G_hier.nodes(data=True)[n]['name'] in systems_keep:\n",
    "        nkeep.append(n)\n",
    "        \n",
    "\n",
    "G_hier = nx.subgraph(G_hier, nkeep)\n",
    "print(len(G_hier.nodes()))\n",
    "print(len(G_hier.edges()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 15. Annotate systems with gprofiler. \n",
    "\n",
    "Annotate systems if they are significantly enriched for a gene ontology biological process term, which has a term size of at least 50 genes, and fewer than 1000 genes. Also require that the GO term is enriched with p<1E-5, and shares at least 3 genes with the focal system to annotate, to increase the stringency of the annotation. Choose the GO term which meets these criteria, and which has the highest sum of precision and recall to annotate. Systems which have no GO terms meeting these criteria are labeled with their unique system ID.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C350\n",
      "331\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>21</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>1.153417e-18</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.109375</td>\n",
       "      <td>query_1</td>\n",
       "      <td>192</td>\n",
       "      <td>0.295775</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>10622</td>\n",
       "      <td>26</td>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>REAC:R-HSA-8936459</td>\n",
       "      <td>6.734109e-22</td>\n",
       "      <td>[REAC:R-HSA-8878171]</td>\n",
       "      <td>0.135417</td>\n",
       "      <td>query_1</td>\n",
       "      <td>192</td>\n",
       "      <td>0.268041</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>30</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>2.076046e-24</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.156250</td>\n",
       "      <td>query_1</td>\n",
       "      <td>192</td>\n",
       "      <td>0.252101</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>10622</td>\n",
       "      <td>12</td>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>REAC:R-HSA-8986944</td>\n",
       "      <td>1.048758e-08</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>query_1</td>\n",
       "      <td>192</td>\n",
       "      <td>0.203390</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Activation of HOX genes during differentiation</td>\n",
       "      <td>10622</td>\n",
       "      <td>21</td>\n",
       "      <td>Activation of HOX genes during differentiation</td>\n",
       "      <td>REAC:R-HSA-5619507</td>\n",
       "      <td>1.033371e-13</td>\n",
       "      <td>[REAC:R-HSA-1266738]</td>\n",
       "      <td>0.109375</td>\n",
       "      <td>query_1</td>\n",
       "      <td>192</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "8                     PKMTs methylate histone lysines                  10622   \n",
       "5   RUNX1 regulates genes involved in megakaryocyt...                  10622   \n",
       "3                                  L1CAM interactions                  10622   \n",
       "21                Transcriptional Regulation by MECP2                  10622   \n",
       "10     Activation of HOX genes during differentiation                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "8                  21                    PKMTs methylate histone lysines   \n",
       "5                  26  RUNX1 regulates genes involved in megakaryocyt...   \n",
       "3                  30                                 L1CAM interactions   \n",
       "21                 12                Transcriptional Regulation by MECP2   \n",
       "10                 21     Activation of HOX genes during differentiation   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "8   REAC:R-HSA-3214841  1.153417e-18  [REAC:R-HSA-3247509]   0.109375   \n",
       "5   REAC:R-HSA-8936459  6.734109e-22  [REAC:R-HSA-8878171]   0.135417   \n",
       "3    REAC:R-HSA-373760  2.076046e-24   [REAC:R-HSA-422475]   0.156250   \n",
       "21  REAC:R-HSA-8986944  1.048758e-08   [REAC:R-HSA-212436]   0.062500   \n",
       "10  REAC:R-HSA-5619507  1.033371e-13  [REAC:R-HSA-1266738]   0.109375   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "8   query_1         192  0.295775         True   REAC         71  \n",
       "5   query_1         192  0.268041         True   REAC         97  \n",
       "3   query_1         192  0.252101         True   REAC        119  \n",
       "21  query_1         192  0.203390         True   REAC         59  \n",
       "10  query_1         192  0.175000         True   REAC        120  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C353\n",
      "54\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>28</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>1.582063e-43</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.636364</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>REAC:R-HSA-375165</td>\n",
       "      <td>3.291188e-13</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.227273</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.169492</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>10622</td>\n",
       "      <td>19</td>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>REAC:R-HSA-5576891</td>\n",
       "      <td>1.487210e-24</td>\n",
       "      <td>[REAC:R-HSA-397014]</td>\n",
       "      <td>0.431818</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.149606</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>REAC:R-HSA-6807878</td>\n",
       "      <td>8.853777e-11</td>\n",
       "      <td>[REAC:R-HSA-199977]</td>\n",
       "      <td>0.227273</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.098039</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>10622</td>\n",
       "      <td>19</td>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>REAC:R-HSA-397014</td>\n",
       "      <td>6.290964e-21</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.431818</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.097436</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             description  effective_domain_size  \\\n",
       "1                     L1CAM interactions                  10622   \n",
       "8  NCAM signaling for neurite out-growth                  10622   \n",
       "5                     Cardiac conduction                  10622   \n",
       "9    COPI-mediated anterograde transport                  10622   \n",
       "6                     Muscle contraction                  10622   \n",
       "\n",
       "   intersection_size                                   name  \\\n",
       "1                 28                     L1CAM interactions   \n",
       "8                 10  NCAM signaling for neurite out-growth   \n",
       "5                 19                     Cardiac conduction   \n",
       "9                 10    COPI-mediated anterograde transport   \n",
       "6                 19                     Muscle contraction   \n",
       "\n",
       "               native       p_value              parents  precision    query  \\\n",
       "1   REAC:R-HSA-373760  1.582063e-43  [REAC:R-HSA-422475]   0.636364  query_1   \n",
       "8   REAC:R-HSA-375165  3.291188e-13  [REAC:R-HSA-422475]   0.227273  query_1   \n",
       "5  REAC:R-HSA-5576891  1.487210e-24  [REAC:R-HSA-397014]   0.431818  query_1   \n",
       "9  REAC:R-HSA-6807878  8.853777e-11  [REAC:R-HSA-199977]   0.227273  query_1   \n",
       "6   REAC:R-HSA-397014  6.290964e-21       [REAC:0000000]   0.431818  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1          44  0.235294         True   REAC        119  \n",
       "8          44  0.169492         True   REAC         59  \n",
       "5          44  0.149606         True   REAC        127  \n",
       "9          44  0.098039         True   REAC        102  \n",
       "6          44  0.097436         True   REAC        195  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C386\n",
      "10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C356\n",
      "40\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C358\n",
      "34\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C394\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C395\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C370\n",
      "16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>4.994361e-09</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>query_1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.084507</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>4.225291e-06</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>query_1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>4.225291e-06</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>query_1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       description  effective_domain_size  intersection_size  \\\n",
       "0  PKMTs methylate histone lysines                  10622                  6   \n",
       "2      Chromatin modifying enzymes                  10622                  6   \n",
       "3           Chromatin organization                  10622                  6   \n",
       "\n",
       "                              name              native       p_value  \\\n",
       "0  PKMTs methylate histone lysines  REAC:R-HSA-3214841  4.994361e-09   \n",
       "2      Chromatin modifying enzymes  REAC:R-HSA-3247509  4.225291e-06   \n",
       "3           Chromatin organization  REAC:R-HSA-4839726  4.225291e-06   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-3247509]   0.461538  query_1          13  0.084507   \n",
       "2  [REAC:R-HSA-4839726]   0.461538  query_1          13  0.021978   \n",
       "3        [REAC:0000000]   0.461538  query_1          13  0.021978   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         71  \n",
       "2         True   REAC        273  \n",
       "3         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C382\n",
      "11\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C351\n",
      "200\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>21</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>1.207683e-22</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.165354</td>\n",
       "      <td>query_1</td>\n",
       "      <td>127</td>\n",
       "      <td>0.295775</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>10622</td>\n",
       "      <td>26</td>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>REAC:R-HSA-8936459</td>\n",
       "      <td>1.257973e-26</td>\n",
       "      <td>[REAC:R-HSA-8878171]</td>\n",
       "      <td>0.204724</td>\n",
       "      <td>query_1</td>\n",
       "      <td>127</td>\n",
       "      <td>0.268041</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>10622</td>\n",
       "      <td>11</td>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>REAC:R-HSA-8986944</td>\n",
       "      <td>1.432223e-09</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.086614</td>\n",
       "      <td>query_1</td>\n",
       "      <td>127</td>\n",
       "      <td>0.186441</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Activation of anterior HOX genes in hindbrain ...</td>\n",
       "      <td>10622</td>\n",
       "      <td>21</td>\n",
       "      <td>Activation of anterior HOX genes in hindbrain ...</td>\n",
       "      <td>REAC:R-HSA-5617472</td>\n",
       "      <td>1.544176e-17</td>\n",
       "      <td>[REAC:R-HSA-5619507]</td>\n",
       "      <td>0.165354</td>\n",
       "      <td>query_1</td>\n",
       "      <td>127</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Activation of HOX genes during differentiation</td>\n",
       "      <td>10622</td>\n",
       "      <td>21</td>\n",
       "      <td>Activation of HOX genes during differentiation</td>\n",
       "      <td>REAC:R-HSA-5619507</td>\n",
       "      <td>1.544176e-17</td>\n",
       "      <td>[REAC:R-HSA-1266738]</td>\n",
       "      <td>0.165354</td>\n",
       "      <td>query_1</td>\n",
       "      <td>127</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "7                     PKMTs methylate histone lysines                  10622   \n",
       "2   RUNX1 regulates genes involved in megakaryocyt...                  10622   \n",
       "19                Transcriptional Regulation by MECP2                  10622   \n",
       "8   Activation of anterior HOX genes in hindbrain ...                  10622   \n",
       "9      Activation of HOX genes during differentiation                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "7                  21                    PKMTs methylate histone lysines   \n",
       "2                  26  RUNX1 regulates genes involved in megakaryocyt...   \n",
       "19                 11                Transcriptional Regulation by MECP2   \n",
       "8                  21  Activation of anterior HOX genes in hindbrain ...   \n",
       "9                  21     Activation of HOX genes during differentiation   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "7   REAC:R-HSA-3214841  1.207683e-22  [REAC:R-HSA-3247509]   0.165354   \n",
       "2   REAC:R-HSA-8936459  1.257973e-26  [REAC:R-HSA-8878171]   0.204724   \n",
       "19  REAC:R-HSA-8986944  1.432223e-09   [REAC:R-HSA-212436]   0.086614   \n",
       "8   REAC:R-HSA-5617472  1.544176e-17  [REAC:R-HSA-5619507]   0.165354   \n",
       "9   REAC:R-HSA-5619507  1.544176e-17  [REAC:R-HSA-1266738]   0.165354   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "7   query_1         127  0.295775         True   REAC         71  \n",
       "2   query_1         127  0.268041         True   REAC         97  \n",
       "19  query_1         127  0.186441         True   REAC         59  \n",
       "8   query_1         127  0.175000         True   REAC        120  \n",
       "9   query_1         127  0.175000         True   REAC        120  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C352\n",
      "82\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>10622</td>\n",
       "      <td>26</td>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>REAC:R-HSA-8936459</td>\n",
       "      <td>1.808700e-35</td>\n",
       "      <td>[REAC:R-HSA-8878171]</td>\n",
       "      <td>0.412698</td>\n",
       "      <td>query_1</td>\n",
       "      <td>63</td>\n",
       "      <td>0.268041</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>14</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>7.039837e-17</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>query_1</td>\n",
       "      <td>63</td>\n",
       "      <td>0.197183</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Activation of HOX genes during differentiation</td>\n",
       "      <td>10622</td>\n",
       "      <td>21</td>\n",
       "      <td>Activation of HOX genes during differentiation</td>\n",
       "      <td>REAC:R-HSA-5619507</td>\n",
       "      <td>1.477214e-24</td>\n",
       "      <td>[REAC:R-HSA-1266738]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>63</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Activation of anterior HOX genes in hindbrain ...</td>\n",
       "      <td>10622</td>\n",
       "      <td>21</td>\n",
       "      <td>Activation of anterior HOX genes in hindbrain ...</td>\n",
       "      <td>REAC:R-HSA-5617472</td>\n",
       "      <td>1.477214e-24</td>\n",
       "      <td>[REAC:R-HSA-5619507]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>63</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>REAC:R-HSA-8986944</td>\n",
       "      <td>2.120412e-11</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.158730</td>\n",
       "      <td>query_1</td>\n",
       "      <td>63</td>\n",
       "      <td>0.169492</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "0   RUNX1 regulates genes involved in megakaryocyt...                  10622   \n",
       "9                     PKMTs methylate histone lysines                  10622   \n",
       "5      Activation of HOX genes during differentiation                  10622   \n",
       "6   Activation of anterior HOX genes in hindbrain ...                  10622   \n",
       "15                Transcriptional Regulation by MECP2                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "0                  26  RUNX1 regulates genes involved in megakaryocyt...   \n",
       "9                  14                    PKMTs methylate histone lysines   \n",
       "5                  21     Activation of HOX genes during differentiation   \n",
       "6                  21  Activation of anterior HOX genes in hindbrain ...   \n",
       "15                 10                Transcriptional Regulation by MECP2   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "0   REAC:R-HSA-8936459  1.808700e-35  [REAC:R-HSA-8878171]   0.412698   \n",
       "9   REAC:R-HSA-3214841  7.039837e-17  [REAC:R-HSA-3247509]   0.222222   \n",
       "5   REAC:R-HSA-5619507  1.477214e-24  [REAC:R-HSA-1266738]   0.333333   \n",
       "6   REAC:R-HSA-5617472  1.477214e-24  [REAC:R-HSA-5619507]   0.333333   \n",
       "15  REAC:R-HSA-8986944  2.120412e-11   [REAC:R-HSA-212436]   0.158730   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "0   query_1          63  0.268041         True   REAC         97  \n",
       "9   query_1          63  0.197183         True   REAC         71  \n",
       "5   query_1          63  0.175000         True   REAC        120  \n",
       "6   query_1          63  0.175000         True   REAC        120  \n",
       "15  query_1          63  0.169492         True   REAC         59  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C355\n",
      "45\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C357\n",
      "38\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>2.541432e-09</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>24</td>\n",
       "      <td>0.098592</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Recruitment and ATM-mediated phosphorylation o...</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Recruitment and ATM-mediated phosphorylation o...</td>\n",
       "      <td>REAC:R-HSA-5693565</td>\n",
       "      <td>1.901221e-07</td>\n",
       "      <td>[REAC:R-HSA-5693606]</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>24</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNA Double Strand Break Response</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>DNA Double Strand Break Response</td>\n",
       "      <td>REAC:R-HSA-5693606</td>\n",
       "      <td>1.901221e-07</td>\n",
       "      <td>[REAC:R-HSA-5693532]</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>24</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>17</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>6.100412e-21</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.708333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>24</td>\n",
       "      <td>0.062271</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>17</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>6.100412e-21</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.708333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>24</td>\n",
       "      <td>0.062271</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "3                    PKMTs methylate histone lysines                  10622   \n",
       "5  Recruitment and ATM-mediated phosphorylation o...                  10622   \n",
       "4                   DNA Double Strand Break Response                  10622   \n",
       "0                        Chromatin modifying enzymes                  10622   \n",
       "1                             Chromatin organization                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "3                  7                    PKMTs methylate histone lysines   \n",
       "5                  6  Recruitment and ATM-mediated phosphorylation o...   \n",
       "4                  6                   DNA Double Strand Break Response   \n",
       "0                 17                        Chromatin modifying enzymes   \n",
       "1                 17                             Chromatin organization   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "3  REAC:R-HSA-3214841  2.541432e-09  [REAC:R-HSA-3247509]   0.291667  query_1   \n",
       "5  REAC:R-HSA-5693565  1.901221e-07  [REAC:R-HSA-5693606]   0.250000  query_1   \n",
       "4  REAC:R-HSA-5693606  1.901221e-07  [REAC:R-HSA-5693532]   0.250000  query_1   \n",
       "0  REAC:R-HSA-3247509  6.100412e-21  [REAC:R-HSA-4839726]   0.708333  query_1   \n",
       "1  REAC:R-HSA-4839726  6.100412e-21        [REAC:0000000]   0.708333  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "3          24  0.098592         True   REAC         71  \n",
       "5          24  0.077922         True   REAC         77  \n",
       "4          24  0.076923         True   REAC         78  \n",
       "0          24  0.062271         True   REAC        273  \n",
       "1          24  0.062271         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C361\n",
      "24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Regulation of PTEN gene transcription</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Regulation of PTEN gene transcription</td>\n",
       "      <td>REAC:R-HSA-8943724</td>\n",
       "      <td>1.641990e-07</td>\n",
       "      <td>[REAC:R-HSA-6807070]</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Potential therapeutics for SARS</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Potential therapeutics for SARS</td>\n",
       "      <td>REAC:R-HSA-9679191</td>\n",
       "      <td>3.044496e-07</td>\n",
       "      <td>[REAC:R-HSA-9679506]</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.063291</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Signaling by TGF-beta Receptor Complex</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Signaling by TGF-beta Receptor Complex</td>\n",
       "      <td>REAC:R-HSA-170834</td>\n",
       "      <td>6.970355e-06</td>\n",
       "      <td>[REAC:R-HSA-9006936]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.056338</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HDACs deacetylate histones</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>HDACs deacetylate histones</td>\n",
       "      <td>REAC:R-HSA-3214815</td>\n",
       "      <td>5.515347e-07</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.053191</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ERCC6 (CSB) and EHMT2 (G9a) positively regulat...</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>ERCC6 (CSB) and EHMT2 (G9a) positively regulat...</td>\n",
       "      <td>REAC:R-HSA-427389</td>\n",
       "      <td>8.202705e-06</td>\n",
       "      <td>[REAC:R-HSA-5250913]</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "2               Regulation of PTEN gene transcription                  10622   \n",
       "5                     Potential therapeutics for SARS                  10622   \n",
       "14             Signaling by TGF-beta Receptor Complex                  10622   \n",
       "7                          HDACs deacetylate histones                  10622   \n",
       "15  ERCC6 (CSB) and EHMT2 (G9a) positively regulat...                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "2                   5              Regulation of PTEN gene transcription   \n",
       "5                   5                    Potential therapeutics for SARS   \n",
       "14                  4             Signaling by TGF-beta Receptor Complex   \n",
       "7                   5                         HDACs deacetylate histones   \n",
       "15                  4  ERCC6 (CSB) and EHMT2 (G9a) positively regulat...   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "2   REAC:R-HSA-8943724  1.641990e-07  [REAC:R-HSA-6807070]   0.416667   \n",
       "5   REAC:R-HSA-9679191  3.044496e-07  [REAC:R-HSA-9679506]   0.416667   \n",
       "14   REAC:R-HSA-170834  6.970355e-06  [REAC:R-HSA-9006936]   0.333333   \n",
       "7   REAC:R-HSA-3214815  5.515347e-07  [REAC:R-HSA-3247509]   0.416667   \n",
       "15   REAC:R-HSA-427389  8.202705e-06  [REAC:R-HSA-5250913]   0.333333   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "2   query_1          12  0.084746         True   REAC         59  \n",
       "5   query_1          12  0.063291         True   REAC         79  \n",
       "14  query_1          12  0.056338         True   REAC         71  \n",
       "7   query_1          12  0.053191         True   REAC         94  \n",
       "15  query_1          12  0.052632         True   REAC         76  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C362\n",
      "21\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.5</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.5</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "0  Chromatin modifying enzymes                  10622                  6   \n",
       "1       Chromatin organization                  10622                  6   \n",
       "\n",
       "                          name              native   p_value  \\\n",
       "0  Chromatin modifying enzymes  REAC:R-HSA-3247509  0.000005   \n",
       "1       Chromatin organization  REAC:R-HSA-4839726  0.000005   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-4839726]        0.5  query_1          12  0.021978   \n",
       "1        [REAC:0000000]        0.5  query_1          12  0.021978   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC        273  \n",
       "1         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C364\n",
      "20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>1.206898e-14</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.727273</td>\n",
       "      <td>query_1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.112676</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>2.478138e-10</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.727273</td>\n",
       "      <td>query_1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.029304</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>2.478138e-10</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.727273</td>\n",
       "      <td>query_1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.029304</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       description  effective_domain_size  intersection_size  \\\n",
       "0  PKMTs methylate histone lysines                  10622                  8   \n",
       "1      Chromatin modifying enzymes                  10622                  8   \n",
       "2           Chromatin organization                  10622                  8   \n",
       "\n",
       "                              name              native       p_value  \\\n",
       "0  PKMTs methylate histone lysines  REAC:R-HSA-3214841  1.206898e-14   \n",
       "1      Chromatin modifying enzymes  REAC:R-HSA-3247509  2.478138e-10   \n",
       "2           Chromatin organization  REAC:R-HSA-4839726  2.478138e-10   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-3247509]   0.727273  query_1          11  0.112676   \n",
       "1  [REAC:R-HSA-4839726]   0.727273  query_1          11  0.029304   \n",
       "2        [REAC:0000000]   0.727273  query_1          11  0.029304   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         71  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C430\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C404\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C416\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C354\n",
      "45\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>10622</td>\n",
       "      <td>20</td>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>REAC:R-HSA-8936459</td>\n",
       "      <td>6.407252e-31</td>\n",
       "      <td>[REAC:R-HSA-8878171]</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>query_1</td>\n",
       "      <td>35</td>\n",
       "      <td>0.206186</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>REAC:R-HSA-8986944</td>\n",
       "      <td>2.097351e-10</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.228571</td>\n",
       "      <td>query_1</td>\n",
       "      <td>35</td>\n",
       "      <td>0.135593</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Ca2+ pathway</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>Ca2+ pathway</td>\n",
       "      <td>REAC:R-HSA-4086398</td>\n",
       "      <td>8.116942e-09</td>\n",
       "      <td>[REAC:R-HSA-3858494]</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>35</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transcriptional regulation by RUNX1</td>\n",
       "      <td>10622</td>\n",
       "      <td>26</td>\n",
       "      <td>Transcriptional regulation by RUNX1</td>\n",
       "      <td>REAC:R-HSA-8878171</td>\n",
       "      <td>3.286705e-34</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.742857</td>\n",
       "      <td>query_1</td>\n",
       "      <td>35</td>\n",
       "      <td>0.110169</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Pre-NOTCH Transcription and Translation</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Pre-NOTCH Transcription and Translation</td>\n",
       "      <td>REAC:R-HSA-1912408</td>\n",
       "      <td>6.293859e-09</td>\n",
       "      <td>[REAC:R-HSA-1912422]</td>\n",
       "      <td>0.228571</td>\n",
       "      <td>query_1</td>\n",
       "      <td>35</td>\n",
       "      <td>0.086022</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "1   RUNX1 regulates genes involved in megakaryocyt...                  10622   \n",
       "11                Transcriptional Regulation by MECP2                  10622   \n",
       "17                                       Ca2+ pathway                  10622   \n",
       "0                 Transcriptional regulation by RUNX1                  10622   \n",
       "16            Pre-NOTCH Transcription and Translation                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "1                  20  RUNX1 regulates genes involved in megakaryocyt...   \n",
       "11                  8                Transcriptional Regulation by MECP2   \n",
       "17                  7                                       Ca2+ pathway   \n",
       "0                  26                Transcriptional regulation by RUNX1   \n",
       "16                  8            Pre-NOTCH Transcription and Translation   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "1   REAC:R-HSA-8936459  6.407252e-31  [REAC:R-HSA-8878171]   0.571429   \n",
       "11  REAC:R-HSA-8986944  2.097351e-10   [REAC:R-HSA-212436]   0.228571   \n",
       "17  REAC:R-HSA-4086398  8.116942e-09  [REAC:R-HSA-3858494]   0.200000   \n",
       "0   REAC:R-HSA-8878171  3.286705e-34   [REAC:R-HSA-212436]   0.742857   \n",
       "16  REAC:R-HSA-1912408  6.293859e-09  [REAC:R-HSA-1912422]   0.228571   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "1   query_1          35  0.206186         True   REAC         97  \n",
       "11  query_1          35  0.135593         True   REAC         59  \n",
       "17  query_1          35  0.118644         True   REAC         59  \n",
       "0   query_1          35  0.110169         True   REAC        236  \n",
       "16  query_1          35  0.086022         True   REAC         93  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C365\n",
      "19\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Activation of HOX genes during differentiation</td>\n",
       "      <td>10622</td>\n",
       "      <td>16</td>\n",
       "      <td>Activation of HOX genes during differentiation</td>\n",
       "      <td>REAC:R-HSA-5619507</td>\n",
       "      <td>1.100674e-30</td>\n",
       "      <td>[REAC:R-HSA-1266738]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Activation of anterior HOX genes in hindbrain ...</td>\n",
       "      <td>10622</td>\n",
       "      <td>16</td>\n",
       "      <td>Activation of anterior HOX genes in hindbrain ...</td>\n",
       "      <td>REAC:R-HSA-5617472</td>\n",
       "      <td>1.100674e-30</td>\n",
       "      <td>[REAC:R-HSA-5619507]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "0     Activation of HOX genes during differentiation                  10622   \n",
       "1  Activation of anterior HOX genes in hindbrain ...                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "0                 16     Activation of HOX genes during differentiation   \n",
       "1                 16  Activation of anterior HOX genes in hindbrain ...   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "0  REAC:R-HSA-5619507  1.100674e-30  [REAC:R-HSA-1266738]        1.0  query_1   \n",
       "1  REAC:R-HSA-5617472  1.100674e-30  [REAC:R-HSA-5619507]        1.0  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "0          16  0.133333         True   REAC        120  \n",
       "1          16  0.133333         True   REAC        120  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C442\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C415\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Regulation of lipid metabolism by PPARalpha</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Regulation of lipid metabolism by PPARalpha</td>\n",
       "      <td>REAC:R-HSA-400206</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   description  effective_domain_size  \\\n",
       "0  Regulation of lipid metabolism by PPARalpha                  10622   \n",
       "\n",
       "   intersection_size                                         name  \\\n",
       "0                  4  Regulation of lipid metabolism by PPARalpha   \n",
       "\n",
       "              native   p_value              parents  precision    query  \\\n",
       "0  REAC:R-HSA-400206  0.000008  [REAC:R-HSA-556833]   0.666667  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "0           6  0.034483         True   REAC        116  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C409\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>4.386404e-08</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.033613</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>REAC:R-HSA-5576891</td>\n",
       "      <td>4.386404e-08</td>\n",
       "      <td>[REAC:R-HSA-397014]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.031496</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>REAC:R-HSA-397014</td>\n",
       "      <td>1.983296e-07</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.020513</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          description  effective_domain_size  intersection_size  \\\n",
       "3  L1CAM interactions                  10622                  4   \n",
       "2  Cardiac conduction                  10622                  4   \n",
       "4  Muscle contraction                  10622                  4   \n",
       "\n",
       "                 name              native       p_value              parents  \\\n",
       "3  L1CAM interactions   REAC:R-HSA-373760  4.386404e-08  [REAC:R-HSA-422475]   \n",
       "2  Cardiac conduction  REAC:R-HSA-5576891  4.386404e-08  [REAC:R-HSA-397014]   \n",
       "4  Muscle contraction   REAC:R-HSA-397014  1.983296e-07       [REAC:0000000]   \n",
       "\n",
       "   precision    query  query_size    recall  significant source  term_size  \n",
       "3        1.0  query_1           4  0.033613         True   REAC        119  \n",
       "2        1.0  query_1           4  0.031496         True   REAC        127  \n",
       "4        1.0  query_1           4  0.020513         True   REAC        195  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C375\n",
      "13\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>REAC:R-HSA-375165</td>\n",
       "      <td>4.705064e-13</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>REAC:R-HSA-6807878</td>\n",
       "      <td>3.953266e-18</td>\n",
       "      <td>[REAC:R-HSA-199977]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.098039</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>11</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>5.567281e-20</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.092437</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ER to Golgi Anterograde Transport</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>ER to Golgi Anterograde Transport</td>\n",
       "      <td>REAC:R-HSA-199977</td>\n",
       "      <td>2.261762e-16</td>\n",
       "      <td>[REAC:R-HSA-199991, REAC:R-HSA-948021]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.064516</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Transport to the Golgi and subsequent modifica...</td>\n",
       "      <td>10622</td>\n",
       "      <td>10</td>\n",
       "      <td>Transport to the Golgi and subsequent modifica...</td>\n",
       "      <td>REAC:R-HSA-948021</td>\n",
       "      <td>1.171782e-15</td>\n",
       "      <td>[REAC:R-HSA-446203]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.053763</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "7              NCAM signaling for neurite out-growth                  10622   \n",
       "2                COPI-mediated anterograde transport                  10622   \n",
       "1                                 L1CAM interactions                  10622   \n",
       "3                  ER to Golgi Anterograde Transport                  10622   \n",
       "4  Transport to the Golgi and subsequent modifica...                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "7                  7              NCAM signaling for neurite out-growth   \n",
       "2                 10                COPI-mediated anterograde transport   \n",
       "1                 11                                 L1CAM interactions   \n",
       "3                 10                  ER to Golgi Anterograde Transport   \n",
       "4                 10  Transport to the Golgi and subsequent modifica...   \n",
       "\n",
       "               native       p_value                                 parents  \\\n",
       "7   REAC:R-HSA-375165  4.705064e-13                     [REAC:R-HSA-422475]   \n",
       "2  REAC:R-HSA-6807878  3.953266e-18                     [REAC:R-HSA-199977]   \n",
       "1   REAC:R-HSA-373760  5.567281e-20                     [REAC:R-HSA-422475]   \n",
       "3   REAC:R-HSA-199977  2.261762e-16  [REAC:R-HSA-199991, REAC:R-HSA-948021]   \n",
       "4   REAC:R-HSA-948021  1.171782e-15                     [REAC:R-HSA-446203]   \n",
       "\n",
       "   precision    query  query_size    recall  significant source  term_size  \n",
       "7   0.583333  query_1          12  0.118644         True   REAC         59  \n",
       "2   0.833333  query_1          12  0.098039         True   REAC        102  \n",
       "1   0.916667  query_1          12  0.092437         True   REAC        119  \n",
       "3   0.833333  query_1          12  0.064516         True   REAC        155  \n",
       "4   0.833333  query_1          12  0.053763         True   REAC        186  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C377\n",
      "12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>1.657207e-14</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>query_1</td>\n",
       "      <td>9</td>\n",
       "      <td>0.067227</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>REAC:R-HSA-5576891</td>\n",
       "      <td>6.364253e-06</td>\n",
       "      <td>[REAC:R-HSA-397014]</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>query_1</td>\n",
       "      <td>9</td>\n",
       "      <td>0.031496</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Axon guidance</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Axon guidance</td>\n",
       "      <td>REAC:R-HSA-422475</td>\n",
       "      <td>2.639737e-09</td>\n",
       "      <td>[REAC:R-HSA-9675108]</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>query_1</td>\n",
       "      <td>9</td>\n",
       "      <td>0.014572</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Nervous system development</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Nervous system development</td>\n",
       "      <td>REAC:R-HSA-9675108</td>\n",
       "      <td>2.827211e-09</td>\n",
       "      <td>[REAC:R-HSA-1266738]</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>query_1</td>\n",
       "      <td>9</td>\n",
       "      <td>0.013937</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  description  effective_domain_size  intersection_size  \\\n",
       "1          L1CAM interactions                  10622                  8   \n",
       "6          Cardiac conduction                  10622                  4   \n",
       "2               Axon guidance                  10622                  8   \n",
       "3  Nervous system development                  10622                  8   \n",
       "\n",
       "                         name              native       p_value  \\\n",
       "1          L1CAM interactions   REAC:R-HSA-373760  1.657207e-14   \n",
       "6          Cardiac conduction  REAC:R-HSA-5576891  6.364253e-06   \n",
       "2               Axon guidance   REAC:R-HSA-422475  2.639737e-09   \n",
       "3  Nervous system development  REAC:R-HSA-9675108  2.827211e-09   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "1   [REAC:R-HSA-422475]   0.888889  query_1           9  0.067227   \n",
       "6   [REAC:R-HSA-397014]   0.444444  query_1           9  0.031496   \n",
       "2  [REAC:R-HSA-9675108]   0.888889  query_1           9  0.014572   \n",
       "3  [REAC:R-HSA-1266738]   0.888889  query_1           9  0.013937   \n",
       "\n",
       "   significant source  term_size  \n",
       "1         True   REAC        119  \n",
       "6         True   REAC        127  \n",
       "2         True   REAC        549  \n",
       "3         True   REAC        574  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C378\n",
      "11\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Transmission across Chemical Synapses</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Transmission across Chemical Synapses</td>\n",
       "      <td>REAC:R-HSA-112315</td>\n",
       "      <td>2.053701e-07</td>\n",
       "      <td>[REAC:R-HSA-112316]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.023166</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neuronal System</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Neuronal System</td>\n",
       "      <td>REAC:R-HSA-112316</td>\n",
       "      <td>1.852806e-06</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             description  effective_domain_size  \\\n",
       "1  Transmission across Chemical Synapses                  10622   \n",
       "2                        Neuronal System                  10622   \n",
       "\n",
       "   intersection_size                                   name  \\\n",
       "1                  6  Transmission across Chemical Synapses   \n",
       "2                  6                        Neuronal System   \n",
       "\n",
       "              native       p_value              parents  precision    query  \\\n",
       "1  REAC:R-HSA-112315  2.053701e-07  [REAC:R-HSA-112316]       0.75  query_1   \n",
       "2  REAC:R-HSA-112316  1.852806e-06       [REAC:0000000]       0.75  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1           8  0.023166         True   REAC        259  \n",
       "2           8  0.015000         True   REAC        400  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C381\n",
      "11\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>Cardiac conduction</td>\n",
       "      <td>REAC:R-HSA-5576891</td>\n",
       "      <td>4.064806e-11</td>\n",
       "      <td>[REAC:R-HSA-397014]</td>\n",
       "      <td>0.636364</td>\n",
       "      <td>query_1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.055118</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>1.998794e-09</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>query_1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.050420</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Stimuli-sensing channels</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Stimuli-sensing channels</td>\n",
       "      <td>REAC:R-HSA-2672351</td>\n",
       "      <td>6.111963e-06</td>\n",
       "      <td>[REAC:R-HSA-983712]</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>query_1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>Muscle contraction</td>\n",
       "      <td>REAC:R-HSA-397014</td>\n",
       "      <td>6.362544e-10</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.636364</td>\n",
       "      <td>query_1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.035897</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                description  effective_domain_size  intersection_size  \\\n",
       "2        Cardiac conduction                  10622                  7   \n",
       "4        L1CAM interactions                  10622                  6   \n",
       "5  Stimuli-sensing channels                  10622                  4   \n",
       "3        Muscle contraction                  10622                  7   \n",
       "\n",
       "                       name              native       p_value  \\\n",
       "2        Cardiac conduction  REAC:R-HSA-5576891  4.064806e-11   \n",
       "4        L1CAM interactions   REAC:R-HSA-373760  1.998794e-09   \n",
       "5  Stimuli-sensing channels  REAC:R-HSA-2672351  6.111963e-06   \n",
       "3        Muscle contraction   REAC:R-HSA-397014  6.362544e-10   \n",
       "\n",
       "               parents  precision    query  query_size    recall  significant  \\\n",
       "2  [REAC:R-HSA-397014]   0.636364  query_1          11  0.055118         True   \n",
       "4  [REAC:R-HSA-422475]   0.545455  query_1          11  0.050420         True   \n",
       "5  [REAC:R-HSA-983712]   0.363636  query_1          11  0.038095         True   \n",
       "3       [REAC:0000000]   0.636364  query_1          11  0.035897         True   \n",
       "\n",
       "  source  term_size  \n",
       "2   REAC        127  \n",
       "4   REAC        119  \n",
       "5   REAC        105  \n",
       "3   REAC        195  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C427\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C372\n",
      "15\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>10622</td>\n",
       "      <td>12</td>\n",
       "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
       "      <td>REAC:R-HSA-8936459</td>\n",
       "      <td>6.197344e-22</td>\n",
       "      <td>[REAC:R-HSA-8878171]</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>query_1</td>\n",
       "      <td>14</td>\n",
       "      <td>0.123711</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transcriptional regulation by RUNX1</td>\n",
       "      <td>10622</td>\n",
       "      <td>14</td>\n",
       "      <td>Transcriptional regulation by RUNX1</td>\n",
       "      <td>REAC:R-HSA-8878171</td>\n",
       "      <td>4.035819e-22</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>14</td>\n",
       "      <td>0.059322</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "1  RUNX1 regulates genes involved in megakaryocyt...                  10622   \n",
       "0                Transcriptional regulation by RUNX1                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "1                 12  RUNX1 regulates genes involved in megakaryocyt...   \n",
       "0                 14                Transcriptional regulation by RUNX1   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "1  REAC:R-HSA-8936459  6.197344e-22  [REAC:R-HSA-8878171]   0.857143  query_1   \n",
       "0  REAC:R-HSA-8878171  4.035819e-22   [REAC:R-HSA-212436]   1.000000  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1          14  0.123711         True   REAC         97  \n",
       "0          14  0.059322         True   REAC        236  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C367\n",
      "18\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ca2+ pathway</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>Ca2+ pathway</td>\n",
       "      <td>REAC:R-HSA-4086398</td>\n",
       "      <td>5.876216e-12</td>\n",
       "      <td>[REAC:R-HSA-3858494]</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>14</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>Transcriptional Regulation by MECP2</td>\n",
       "      <td>REAC:R-HSA-8986944</td>\n",
       "      <td>5.876216e-12</td>\n",
       "      <td>[REAC:R-HSA-212436]</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>14</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TP53 Regulates Metabolic Genes</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>TP53 Regulates Metabolic Genes</td>\n",
       "      <td>REAC:R-HSA-5628897</td>\n",
       "      <td>5.126282e-11</td>\n",
       "      <td>[REAC:R-HSA-3700989]</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>14</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Pre-NOTCH Transcription and Translation</td>\n",
       "      <td>10622</td>\n",
       "      <td>7</td>\n",
       "      <td>Pre-NOTCH Transcription and Translation</td>\n",
       "      <td>REAC:R-HSA-1912408</td>\n",
       "      <td>9.845763e-11</td>\n",
       "      <td>[REAC:R-HSA-1912422]</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>14</td>\n",
       "      <td>0.075269</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MAPK6/MAPK4 signaling</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>MAPK6/MAPK4 signaling</td>\n",
       "      <td>REAC:R-HSA-5687128</td>\n",
       "      <td>4.511748e-09</td>\n",
       "      <td>[REAC:R-HSA-5683057]</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>query_1</td>\n",
       "      <td>14</td>\n",
       "      <td>0.067416</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                description  effective_domain_size  \\\n",
       "6                              Ca2+ pathway                  10622   \n",
       "7       Transcriptional Regulation by MECP2                  10622   \n",
       "11           TP53 Regulates Metabolic Genes                  10622   \n",
       "12  Pre-NOTCH Transcription and Translation                  10622   \n",
       "22                    MAPK6/MAPK4 signaling                  10622   \n",
       "\n",
       "    intersection_size                                     name  \\\n",
       "6                   7                             Ca2+ pathway   \n",
       "7                   7      Transcriptional Regulation by MECP2   \n",
       "11                  7           TP53 Regulates Metabolic Genes   \n",
       "12                  7  Pre-NOTCH Transcription and Translation   \n",
       "22                  6                    MAPK6/MAPK4 signaling   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "6   REAC:R-HSA-4086398  5.876216e-12  [REAC:R-HSA-3858494]   0.500000   \n",
       "7   REAC:R-HSA-8986944  5.876216e-12   [REAC:R-HSA-212436]   0.500000   \n",
       "11  REAC:R-HSA-5628897  5.126282e-11  [REAC:R-HSA-3700989]   0.500000   \n",
       "12  REAC:R-HSA-1912408  9.845763e-11  [REAC:R-HSA-1912422]   0.500000   \n",
       "22  REAC:R-HSA-5687128  4.511748e-09  [REAC:R-HSA-5683057]   0.428571   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "6   query_1          14  0.118644         True   REAC         59  \n",
       "7   query_1          14  0.118644         True   REAC         59  \n",
       "11  query_1          14  0.083333         True   REAC         84  \n",
       "12  query_1          14  0.075269         True   REAC         93  \n",
       "22  query_1          14  0.067416         True   REAC         89  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C373\n",
      "14\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C359\n",
      "28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C368\n",
      "17\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C366\n",
      "18\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C423\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C360\n",
      "25\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>1.091687e-08</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.084507</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>14</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>6.506519e-19</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>14</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>6.506519e-19</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       description  effective_domain_size  intersection_size  \\\n",
       "3  PKMTs methylate histone lysines                  10622                  6   \n",
       "0      Chromatin modifying enzymes                  10622                 14   \n",
       "1           Chromatin organization                  10622                 14   \n",
       "\n",
       "                              name              native       p_value  \\\n",
       "3  PKMTs methylate histone lysines  REAC:R-HSA-3214841  1.091687e-08   \n",
       "0      Chromatin modifying enzymes  REAC:R-HSA-3247509  6.506519e-19   \n",
       "1           Chromatin organization  REAC:R-HSA-4839726  6.506519e-19   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "3  [REAC:R-HSA-3247509]   0.352941  query_1          17  0.084507   \n",
       "0  [REAC:R-HSA-4839726]   0.823529  query_1          17  0.051282   \n",
       "1        [REAC:0000000]   0.823529  query_1          17  0.051282   \n",
       "\n",
       "   significant source  term_size  \n",
       "3         True   REAC         71  \n",
       "0         True   REAC        273  \n",
       "1         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C379\n",
      "11\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Recruitment and ATM-mediated phosphorylation o...</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>Recruitment and ATM-mediated phosphorylation o...</td>\n",
       "      <td>REAC:R-HSA-5693565</td>\n",
       "      <td>2.320935e-12</td>\n",
       "      <td>[REAC:R-HSA-5693606]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNA Double Strand Break Response</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>DNA Double Strand Break Response</td>\n",
       "      <td>REAC:R-HSA-5693606</td>\n",
       "      <td>2.320935e-12</td>\n",
       "      <td>[REAC:R-HSA-5693532]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA Double-Strand Break Repair</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>DNA Double-Strand Break Repair</td>\n",
       "      <td>REAC:R-HSA-5693532</td>\n",
       "      <td>1.657341e-10</td>\n",
       "      <td>[REAC:R-HSA-73894]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.035928</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNA Repair</td>\n",
       "      <td>10622</td>\n",
       "      <td>6</td>\n",
       "      <td>DNA Repair</td>\n",
       "      <td>REAC:R-HSA-73894</td>\n",
       "      <td>7.601190e-09</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.018237</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "0  Recruitment and ATM-mediated phosphorylation o...                  10622   \n",
       "1                   DNA Double Strand Break Response                  10622   \n",
       "2                     DNA Double-Strand Break Repair                  10622   \n",
       "3                                         DNA Repair                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "0                  6  Recruitment and ATM-mediated phosphorylation o...   \n",
       "1                  6                   DNA Double Strand Break Response   \n",
       "2                  6                     DNA Double-Strand Break Repair   \n",
       "3                  6                                         DNA Repair   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "0  REAC:R-HSA-5693565  2.320935e-12  [REAC:R-HSA-5693606]        1.0  query_1   \n",
       "1  REAC:R-HSA-5693606  2.320935e-12  [REAC:R-HSA-5693532]        1.0  query_1   \n",
       "2  REAC:R-HSA-5693532  1.657341e-10    [REAC:R-HSA-73894]        1.0  query_1   \n",
       "3    REAC:R-HSA-73894  7.601190e-09        [REAC:0000000]        1.0  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "0           6  0.077922         True   REAC         77  \n",
       "1           6  0.076923         True   REAC         78  \n",
       "2           6  0.035928         True   REAC        167  \n",
       "3           6  0.018237         True   REAC        329  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C388\n",
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C363\n",
      "20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C406\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C387\n",
      "10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>1.339083e-08</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.070423</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>3.979204e-06</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>0.625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.018315</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>3.979204e-06</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.018315</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       description  effective_domain_size  intersection_size  \\\n",
       "0  PKMTs methylate histone lysines                  10622                  5   \n",
       "1      Chromatin modifying enzymes                  10622                  5   \n",
       "2           Chromatin organization                  10622                  5   \n",
       "\n",
       "                              name              native       p_value  \\\n",
       "0  PKMTs methylate histone lysines  REAC:R-HSA-3214841  1.339083e-08   \n",
       "1      Chromatin modifying enzymes  REAC:R-HSA-3247509  3.979204e-06   \n",
       "2           Chromatin organization  REAC:R-HSA-4839726  3.979204e-06   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-3247509]      0.625  query_1           8  0.070423   \n",
       "1  [REAC:R-HSA-4839726]      0.625  query_1           8  0.018315   \n",
       "2        [REAC:0000000]      0.625  query_1           8  0.018315   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         71  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C374\n",
      "13\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>REAC:R-HSA-3247509</td>\n",
       "      <td>2.237601e-12</td>\n",
       "      <td>[REAC:R-HSA-4839726]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.029304</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Chromatin organization</td>\n",
       "      <td>REAC:R-HSA-4839726</td>\n",
       "      <td>2.237601e-12</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.029304</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "1  Chromatin modifying enzymes                  10622                  8   \n",
       "2       Chromatin organization                  10622                  8   \n",
       "\n",
       "                          name              native       p_value  \\\n",
       "1  Chromatin modifying enzymes  REAC:R-HSA-3247509  2.237601e-12   \n",
       "2       Chromatin organization  REAC:R-HSA-4839726  2.237601e-12   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "1  [REAC:R-HSA-4839726]        1.0  query_1           8  0.029304   \n",
       "2        [REAC:0000000]        1.0  query_1           8  0.029304   \n",
       "\n",
       "   significant source  term_size  \n",
       "1         True   REAC        273  \n",
       "2         True   REAC        273  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C369\n",
      "16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Regulation of PTEN gene transcription</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Regulation of PTEN gene transcription</td>\n",
       "      <td>REAC:R-HSA-8943724</td>\n",
       "      <td>2.852300e-08</td>\n",
       "      <td>[REAC:R-HSA-6807070]</td>\n",
       "      <td>0.625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Potential therapeutics for SARS</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Potential therapeutics for SARS</td>\n",
       "      <td>REAC:R-HSA-9679191</td>\n",
       "      <td>4.495376e-08</td>\n",
       "      <td>[REAC:R-HSA-9679506]</td>\n",
       "      <td>0.625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.063291</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Signaling by TGF-beta Receptor Complex</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Signaling by TGF-beta Receptor Complex</td>\n",
       "      <td>REAC:R-HSA-170834</td>\n",
       "      <td>1.458481e-06</td>\n",
       "      <td>[REAC:R-HSA-9006936]</td>\n",
       "      <td>0.500</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.056338</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HDACs deacetylate histones</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>HDACs deacetylate histones</td>\n",
       "      <td>REAC:R-HSA-3214815</td>\n",
       "      <td>7.753783e-08</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>0.625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.053191</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ERCC6 (CSB) and EHMT2 (G9a) positively regulat...</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>ERCC6 (CSB) and EHMT2 (G9a) positively regulat...</td>\n",
       "      <td>REAC:R-HSA-427389</td>\n",
       "      <td>1.748097e-06</td>\n",
       "      <td>[REAC:R-HSA-5250913]</td>\n",
       "      <td>0.500</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "0               Regulation of PTEN gene transcription                  10622   \n",
       "1                     Potential therapeutics for SARS                  10622   \n",
       "9              Signaling by TGF-beta Receptor Complex                  10622   \n",
       "3                          HDACs deacetylate histones                  10622   \n",
       "10  ERCC6 (CSB) and EHMT2 (G9a) positively regulat...                  10622   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "0                   5              Regulation of PTEN gene transcription   \n",
       "1                   5                    Potential therapeutics for SARS   \n",
       "9                   4             Signaling by TGF-beta Receptor Complex   \n",
       "3                   5                         HDACs deacetylate histones   \n",
       "10                  4  ERCC6 (CSB) and EHMT2 (G9a) positively regulat...   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "0   REAC:R-HSA-8943724  2.852300e-08  [REAC:R-HSA-6807070]      0.625   \n",
       "1   REAC:R-HSA-9679191  4.495376e-08  [REAC:R-HSA-9679506]      0.625   \n",
       "9    REAC:R-HSA-170834  1.458481e-06  [REAC:R-HSA-9006936]      0.500   \n",
       "3   REAC:R-HSA-3214815  7.753783e-08  [REAC:R-HSA-3247509]      0.625   \n",
       "10   REAC:R-HSA-427389  1.748097e-06  [REAC:R-HSA-5250913]      0.500   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "0   query_1           8  0.084746         True   REAC         59  \n",
       "1   query_1           8  0.063291         True   REAC         79  \n",
       "9   query_1           8  0.056338         True   REAC         71  \n",
       "3   query_1           8  0.053191         True   REAC         94  \n",
       "10  query_1           8  0.052632         True   REAC         76  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C418\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C397\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C383\n",
      "10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C410\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>10622</td>\n",
       "      <td>3</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>REAC:R-HSA-3214841</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>[REAC:R-HSA-3247509]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.042254</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       description  effective_domain_size  intersection_size  \\\n",
       "0  PKMTs methylate histone lysines                  10622                  3   \n",
       "\n",
       "                              name              native   p_value  \\\n",
       "0  PKMTs methylate histone lysines  REAC:R-HSA-3214841  0.000003   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-3247509]        1.0  query_1           3  0.042254   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         71  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C376\n",
      "12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C380\n",
      "11\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C405\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C390\n",
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C385\n",
      "10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C389\n",
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C414\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C391\n",
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>REAC:R-HSA-375165</td>\n",
       "      <td>9.343743e-10</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>REAC:R-HSA-6807878</td>\n",
       "      <td>1.039436e-15</td>\n",
       "      <td>[REAC:R-HSA-199977]</td>\n",
       "      <td>1.000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.078431</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>2.478158e-15</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>1.000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.067227</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ER to Golgi Anterograde Transport</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>ER to Golgi Anterograde Transport</td>\n",
       "      <td>REAC:R-HSA-199977</td>\n",
       "      <td>1.629754e-14</td>\n",
       "      <td>[REAC:R-HSA-199991, REAC:R-HSA-948021]</td>\n",
       "      <td>1.000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.051613</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Transport to the Golgi and subsequent modifica...</td>\n",
       "      <td>10622</td>\n",
       "      <td>8</td>\n",
       "      <td>Transport to the Golgi and subsequent modifica...</td>\n",
       "      <td>REAC:R-HSA-948021</td>\n",
       "      <td>5.782790e-14</td>\n",
       "      <td>[REAC:R-HSA-446203]</td>\n",
       "      <td>1.000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.043011</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "9              NCAM signaling for neurite out-growth                  10622   \n",
       "1                COPI-mediated anterograde transport                  10622   \n",
       "2                                 L1CAM interactions                  10622   \n",
       "3                  ER to Golgi Anterograde Transport                  10622   \n",
       "4  Transport to the Golgi and subsequent modifica...                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "9                  5              NCAM signaling for neurite out-growth   \n",
       "1                  8                COPI-mediated anterograde transport   \n",
       "2                  8                                 L1CAM interactions   \n",
       "3                  8                  ER to Golgi Anterograde Transport   \n",
       "4                  8  Transport to the Golgi and subsequent modifica...   \n",
       "\n",
       "               native       p_value                                 parents  \\\n",
       "9   REAC:R-HSA-375165  9.343743e-10                     [REAC:R-HSA-422475]   \n",
       "1  REAC:R-HSA-6807878  1.039436e-15                     [REAC:R-HSA-199977]   \n",
       "2   REAC:R-HSA-373760  2.478158e-15                     [REAC:R-HSA-422475]   \n",
       "3   REAC:R-HSA-199977  1.629754e-14  [REAC:R-HSA-199991, REAC:R-HSA-948021]   \n",
       "4   REAC:R-HSA-948021  5.782790e-14                     [REAC:R-HSA-446203]   \n",
       "\n",
       "   precision    query  query_size    recall  significant source  term_size  \n",
       "9      0.625  query_1           8  0.084746         True   REAC         59  \n",
       "1      1.000  query_1           8  0.078431         True   REAC        102  \n",
       "2      1.000  query_1           8  0.067227         True   REAC        119  \n",
       "3      1.000  query_1           8  0.051613         True   REAC        155  \n",
       "4      1.000  query_1           8  0.043011         True   REAC        186  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C392\n",
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>9.164725e-09</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          description  effective_domain_size  intersection_size  \\\n",
       "1  L1CAM interactions                  10622                  5   \n",
       "\n",
       "                 name             native       p_value              parents  \\\n",
       "1  L1CAM interactions  REAC:R-HSA-373760  9.164725e-09  [REAC:R-HSA-422475]   \n",
       "\n",
       "   precision    query  query_size    recall  significant source  term_size  \n",
       "1   0.833333  query_1           6  0.042017         True   REAC        119  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C407\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Recruitment and ATM-mediated phosphorylation o...</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>Recruitment and ATM-mediated phosphorylation o...</td>\n",
       "      <td>REAC:R-HSA-5693565</td>\n",
       "      <td>2.906706e-10</td>\n",
       "      <td>[REAC:R-HSA-5693606]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.064935</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNA Double Strand Break Response</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>DNA Double Strand Break Response</td>\n",
       "      <td>REAC:R-HSA-5693606</td>\n",
       "      <td>2.906706e-10</td>\n",
       "      <td>[REAC:R-HSA-5693532]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA Double-Strand Break Repair</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>DNA Double-Strand Break Repair</td>\n",
       "      <td>REAC:R-HSA-5693532</td>\n",
       "      <td>9.353149e-09</td>\n",
       "      <td>[REAC:R-HSA-73894]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.029940</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNA Repair</td>\n",
       "      <td>10622</td>\n",
       "      <td>5</td>\n",
       "      <td>DNA Repair</td>\n",
       "      <td>REAC:R-HSA-73894</td>\n",
       "      <td>2.144853e-07</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.015198</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "0  Recruitment and ATM-mediated phosphorylation o...                  10622   \n",
       "1                   DNA Double Strand Break Response                  10622   \n",
       "2                     DNA Double-Strand Break Repair                  10622   \n",
       "3                                         DNA Repair                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "0                  5  Recruitment and ATM-mediated phosphorylation o...   \n",
       "1                  5                   DNA Double Strand Break Response   \n",
       "2                  5                     DNA Double-Strand Break Repair   \n",
       "3                  5                                         DNA Repair   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "0  REAC:R-HSA-5693565  2.906706e-10  [REAC:R-HSA-5693606]        1.0  query_1   \n",
       "1  REAC:R-HSA-5693606  2.906706e-10  [REAC:R-HSA-5693532]        1.0  query_1   \n",
       "2  REAC:R-HSA-5693532  9.353149e-09    [REAC:R-HSA-73894]        1.0  query_1   \n",
       "3    REAC:R-HSA-73894  2.144853e-07        [REAC:0000000]        1.0  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "0           5  0.064935         True   REAC         77  \n",
       "1           5  0.064103         True   REAC         78  \n",
       "2           5  0.029940         True   REAC        167  \n",
       "3           5  0.015198         True   REAC        329  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C412\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C411\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C402\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C432\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>10622</td>\n",
       "      <td>3</td>\n",
       "      <td>NCAM signaling for neurite out-growth</td>\n",
       "      <td>REAC:R-HSA-375165</td>\n",
       "      <td>2.095887e-06</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.050847</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>COPI-mediated anterograde transport</td>\n",
       "      <td>REAC:R-HSA-6807878</td>\n",
       "      <td>8.817990e-08</td>\n",
       "      <td>[REAC:R-HSA-199977]</td>\n",
       "      <td>1.00</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.039216</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>L1CAM interactions</td>\n",
       "      <td>REAC:R-HSA-373760</td>\n",
       "      <td>1.098486e-07</td>\n",
       "      <td>[REAC:R-HSA-422475]</td>\n",
       "      <td>1.00</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.033613</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ER to Golgi Anterograde Transport</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>ER to Golgi Anterograde Transport</td>\n",
       "      <td>REAC:R-HSA-199977</td>\n",
       "      <td>2.399774e-07</td>\n",
       "      <td>[REAC:R-HSA-199991, REAC:R-HSA-948021]</td>\n",
       "      <td>1.00</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.025806</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Transport to the Golgi and subsequent modifica...</td>\n",
       "      <td>10622</td>\n",
       "      <td>4</td>\n",
       "      <td>Transport to the Golgi and subsequent modifica...</td>\n",
       "      <td>REAC:R-HSA-948021</td>\n",
       "      <td>4.007065e-07</td>\n",
       "      <td>[REAC:R-HSA-446203]</td>\n",
       "      <td>1.00</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.021505</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "6              NCAM signaling for neurite out-growth                  10622   \n",
       "1                COPI-mediated anterograde transport                  10622   \n",
       "2                                 L1CAM interactions                  10622   \n",
       "3                  ER to Golgi Anterograde Transport                  10622   \n",
       "4  Transport to the Golgi and subsequent modifica...                  10622   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "6                  3              NCAM signaling for neurite out-growth   \n",
       "1                  4                COPI-mediated anterograde transport   \n",
       "2                  4                                 L1CAM interactions   \n",
       "3                  4                  ER to Golgi Anterograde Transport   \n",
       "4                  4  Transport to the Golgi and subsequent modifica...   \n",
       "\n",
       "               native       p_value                                 parents  \\\n",
       "6   REAC:R-HSA-375165  2.095887e-06                     [REAC:R-HSA-422475]   \n",
       "1  REAC:R-HSA-6807878  8.817990e-08                     [REAC:R-HSA-199977]   \n",
       "2   REAC:R-HSA-373760  1.098486e-07                     [REAC:R-HSA-422475]   \n",
       "3   REAC:R-HSA-199977  2.399774e-07  [REAC:R-HSA-199991, REAC:R-HSA-948021]   \n",
       "4   REAC:R-HSA-948021  4.007065e-07                     [REAC:R-HSA-446203]   \n",
       "\n",
       "   precision    query  query_size    recall  significant source  term_size  \n",
       "6       0.75  query_1           4  0.050847         True   REAC         59  \n",
       "1       1.00  query_1           4  0.039216         True   REAC        102  \n",
       "2       1.00  query_1           4  0.033613         True   REAC        119  \n",
       "3       1.00  query_1           4  0.025806         True   REAC        155  \n",
       "4       1.00  query_1           4  0.021505         True   REAC        186  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C436\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C437\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# gprofiler annotation of clusters\n",
    "\n",
    "# hier_df = hier_df.sort_values('name')\n",
    "\n",
    "system_name_list = []\n",
    "for p in hier_df.index.tolist():\n",
    "    focal_genes=hier_df['CD_MemberList'].loc[p].split(' ')\n",
    "    print(p)\n",
    "    print(len(focal_genes))\n",
    "    if len(focal_genes)>2:\n",
    "        gp_temp = pd.DataFrame(gp.profile(focal_genes,significance_threshold_method='fdr',\n",
    "                                               sources=['REAC']))\n",
    "        if len(gp_temp)>0: # make sure data is not empty\n",
    "            \n",
    "            # make sure terms are specific, and overlap with at least 3 genes\n",
    "            gp_temp = gp_temp[(gp_temp['term_size']<1000)&(gp_temp['term_size']>50)]\n",
    "            gp_temp = gp_temp[gp_temp['intersection_size']>=3]\n",
    "            \n",
    "            gp_temp = gp_temp[gp_temp['p_value']<1E-5] # set a stringent pvalue threshold\n",
    "            \n",
    "            gp_temp = gp_temp.sort_values('recall',ascending=False)\n",
    "            \n",
    "            if len(gp_temp)>1:\n",
    "                system_name_list.append(gp_temp.head(1)['name'].tolist()[0])\n",
    "            else:\n",
    "                system_name_list.append(p)\n",
    "        else:\n",
    "            system_name_list.append(p)\n",
    "            \n",
    "\n",
    "        display(gp_temp.head())\n",
    "        \n",
    "    else:\n",
    "        system_name_list.append(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Validate identified genes and systems_\n",
    "\n",
    "### 16. Load and parse mouse knockout database\n",
    "\n",
    "Requires DDOT for ontology parsing https://github.com/michaelkyu/ddot/blob/master/examples/Tutorial.ipynb\n",
    "\n",
    "Parse the ontology, data from http://www.informatics.jax.org/vocab/mp_ontology\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1001-2000...done.\n",
      "querying 2001-3000...done.\n",
      "querying 3001-4000...done.\n",
      "querying 4001-5000...done.\n",
      "querying 5001-6000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 7001-8000...done.\n",
      "querying 8001-9000...done.\n",
      "querying 9001-10000...done.\n",
      "querying 10001-11000...done.\n",
      "querying 11001-12000...done.\n",
      "querying 12001-13000...done.\n",
      "querying 13001-14000...done.\n",
      "querying 14001-15000...done.\n",
      "querying 15001-16000...done.\n",
      "querying 16001-17000...done.\n",
      "querying 17001-18000...done.\n",
      "querying 18001-19000...done.\n",
      "querying 19001-20000...done.\n",
      "querying 20001-21000...done.\n",
      "querying 21001-22000...done.\n",
      "querying 22001-23000...done.\n",
      "querying 23001-24000...done.\n",
      "querying 24001-24136...done.\n",
      "Finished.\n",
      "35 input query terms found dup hits:\n",
      "\t[('Adam1a', 2), ('Adam1b', 2), ('Ccdc39', 2), ('Chaserr', 2), ('Dnm3os', 2), ('Eldr', 2), ('Fer1l4',\n",
      "12241 input query terms found no hit:\n",
      "\t['+|Cav1', '+|Cm', '+|Del(10Prmt2-Pdxk)4Yey', '+|Del(11Mpo-Chad)18Brd', '+|Del(11Stat5a-Stat5b)1Mam'\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "24179\n",
      "11938\n",
      "11895\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI_Allele_Accession_ID</th>\n",
       "      <th>Allele symbol</th>\n",
       "      <th>involves</th>\n",
       "      <th>MP</th>\n",
       "      <th>PMID</th>\n",
       "      <th>MGI_marker_accession_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>human_ortholog</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000600</td>\n",
       "      <td>12529408</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001716</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001698</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001092</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000961</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           MGI_Allele_Accession_ID Allele symbol               involves  \\\n",
       "gene_name                                                                 \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "\n",
       "                   MP      PMID MGI_marker_accession_ID gene_name  \\\n",
       "gene_name                                                           \n",
       "Rb1        MP:0000600  12529408               MGI:97874       Rb1   \n",
       "Rb1        MP:0001716  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0001698  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0001092  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0000961  16449662               MGI:97874       Rb1   \n",
       "\n",
       "          human_ortholog  \n",
       "gene_name                 \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_df = validation.load_MGI_mouseKO_data()\n",
    "mgi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14470\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0000001</th>\n",
       "      <td>mammalian phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000002</th>\n",
       "      <td>obsolete Morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000003</th>\n",
       "      <td>abnormal adipose tissue morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000011</th>\n",
       "      <td>abnormal adipose tissue morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000005</th>\n",
       "      <td>increased brown adipose tissue amount</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      description\n",
       "MP                                               \n",
       "MP:0000001                    mammalian phenotype\n",
       "MP:0000002                    obsolete Morphology\n",
       "MP:0000003     abnormal adipose tissue morphology\n",
       "MP:0000011     abnormal adipose tissue morphology\n",
       "MP:0000005  increased brown adipose tissue amount"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent</th>\n",
       "      <th>Child</th>\n",
       "      <th>Relation</th>\n",
       "      <th>Namespace</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MP:0005375</td>\n",
       "      <td>MP:0000003</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MP:0001778</td>\n",
       "      <td>MP:0000005</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MP:0001781</td>\n",
       "      <td>MP:0000008</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MP:0005334</td>\n",
       "      <td>MP:0000010</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MP:0000003</td>\n",
       "      <td>MP:0000013</td>\n",
       "      <td>is_a</td>\n",
       "      <td>MPheno.ontology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Parent       Child Relation        Namespace\n",
       "0  MP:0005375  MP:0000003     is_a  MPheno.ontology\n",
       "1  MP:0001778  MP:0000005     is_a  MPheno.ontology\n",
       "2  MP:0001781  MP:0000008     is_a  MPheno.ontology\n",
       "3  MP:0005334  MP:0000010     is_a  MPheno.ontology\n",
       "4  MP:0000003  MP:0000013     is_a  MPheno.ontology"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "0 genes, 13337 terms, 0 gene-term relations, 16846 term-term relations\n",
       "node_attributes: ['description']\n",
       "edge_attributes: ['Relation', 'Namespace']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MPO = validation.load_MPO()\n",
    "MPO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 17. Identify phenotype(s) of interest. Recommend including a negative control.\n",
    "\n",
    "#### here we find MPO terms related to brain or heart\n",
    "\n",
    "Modify as needed for each specific project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "179\n",
      "['MP:0000778', 'MP:0000811', 'MP:0000937', 'MP:0000938', 'MP:0000939', 'MP:0000940', 'MP:0000958', 'MP:0000965', 'MP:0000966', 'MP:0000968']\n"
     ]
    }
   ],
   "source": [
    "# find terms related to brain\n",
    "# ---- modify this part as needed for your project -----\n",
    "MP_focal_brain_list = []\n",
    "for t in MPO.node_attr.index.tolist():\n",
    "    descr_temp = MPO.node_attr.loc[t]['description']\n",
    "    if descr_temp.find('nervous')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "    elif descr_temp.find('neuron')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "    elif descr_temp.find('synapt')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "        \n",
    "print(len(MP_focal_brain_list))\n",
    "print(MP_focal_brain_list[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "283\n",
      "['MP:0000266', 'MP:0000267', 'MP:0000269', 'MP:0000270', 'MP:0000274', 'MP:0000275', 'MP:0000276', 'MP:0000277', 'MP:0000278', 'MP:0000285']\n"
     ]
    }
   ],
   "source": [
    "# find terms related to heart\n",
    "MP_focal_heart_list = []\n",
    "for t in MPO.node_attr.index.tolist():\n",
    "    descr_temp = MPO.node_attr.loc[t]['description']\n",
    "    if descr_temp.find('cardi')>-1:\n",
    "        MP_focal_heart_list.append(t)\n",
    "    elif descr_temp.find('heart')>-1:\n",
    "        MP_focal_heart_list.append(t)\n",
    "        \n",
    "print(len(MP_focal_heart_list))\n",
    "print(MP_focal_heart_list[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 18. Compute the enrichment of selected phenotype(s) in NetColoc subgraph as a whole."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal innate immunity\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 598\n",
      "number overlapping genes = 7\n",
      "0.17603489889324875\n",
      "(-1.2732427290637403, 0.23323969663440758)\n",
      "-0.5200015162146663\n",
      "\n",
      "abnormal nervous system tract morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 209\n",
      "number overlapping genes = 19\n",
      "1.2466819195683438e-11\n",
      "(1.1907083666153984, 2.1601270376591604)\n",
      "1.6754177021372794\n",
      "\n",
      "hippocampal neuron degeneration\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 27\n",
      "number overlapping genes = 2\n",
      "0.0558728244993557\n",
      "(-0.03532781607510116, 2.8534727613539923)\n",
      "1.4090724726394455\n",
      "\n",
      "abnormal motor neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 166\n",
      "number overlapping genes = 8\n",
      "0.008701499385237328\n",
      "(0.24335683286099508, 1.6808804355135014)\n",
      "0.9621186341872483\n",
      "\n",
      "motor neuron degeneration\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 2\n",
      "0.20851174416552776\n",
      "(-0.5100665052925273, 2.3369165828309155)\n",
      "0.9134250387691942\n",
      "\n",
      "decreased motor neuron number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 76\n",
      "number overlapping genes = 4\n",
      "0.042598127064946166\n",
      "(0.0349643258801593, 2.0604961413930325)\n",
      "1.047730233636596\n",
      "\n",
      "abnormal motor neuron innervation pattern\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 1\n",
      "0.8302989440126214\n",
      "(-1.7694684498018842, 2.2039543588300745)\n",
      "0.21724295451409503\n",
      "\n",
      "peripheral nervous system degeneration\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 2\n",
      "0.001917309267094988\n",
      "(0.895619848364519, 3.967728168233494)\n",
      "2.4316740082990065\n",
      "\n",
      "abnormal sensory neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 477\n",
      "number overlapping genes = 8\n",
      "0.6885248657088157\n",
      "(-0.852152664395796, 0.5627686866834254)\n",
      "-0.14469198885618528\n",
      "\n",
      "decreased sensory neuron number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 195\n",
      "number overlapping genes = 4\n",
      "0.8982326692443807\n",
      "(-0.9312137106445342, 1.0612271780737974)\n",
      "0.06500673371463161\n",
      "\n",
      "abnormal sensory neuron innervation pattern\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 93\n",
      "number overlapping genes = 3\n",
      "0.3661417265391369\n",
      "(-0.6226916099162574, 1.6882476479135826)\n",
      "0.5327780189986626\n",
      "\n",
      "abnormal sympathetic neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 1\n",
      "0.16487518105154264\n",
      "(-0.5952704250467558, 3.4905339077295783)\n",
      "1.4476317413414113\n",
      "\n",
      "abnormal cholinergic neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 1\n",
      "0.236523929350149\n",
      "(-0.8028616241082172, 3.251481720203364)\n",
      "1.2243100480475735\n",
      "\n",
      "abnormal enteric nervous system morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 1\n",
      "0.6903003835663676\n",
      "(-1.586689456624333, 2.3964309655248375)\n",
      "0.40487075445025233\n",
      "\n",
      "abnormal enteric neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 1\n",
      "0.46084895600620923\n",
      "(-1.2495725613982311, 2.75711616615099)\n",
      "0.7537718023763795\n",
      "\n",
      "impaired synaptic plasticity\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 41\n",
      "number overlapping genes = 1\n",
      "0.811364414498692\n",
      "(-1.745305775424278, 2.229295861665837)\n",
      "0.24199504312077957\n",
      "\n",
      "abnormal CNS synaptic transmission\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 848\n",
      "number overlapping genes = 31\n",
      "0.0002307329732464978\n",
      "(0.3308362383214899, 1.083602694634012)\n",
      "0.7072194664777509\n",
      "\n",
      "abnormal nervous system electrophysiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 331\n",
      "number overlapping genes = 30\n",
      "1.8565985386892684e-17\n",
      "(1.3091192450609364, 2.0934797257882036)\n",
      "1.70129948542457\n",
      "\n",
      "abnormal autonomic nervous system morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 94\n",
      "number overlapping genes = 4\n",
      "0.10914084946400303\n",
      "(-0.1839706201924607, 1.8309976780671289)\n",
      "0.8235135289373341\n",
      "\n",
      "abnormal somatic nervous system morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 1002\n",
      "number overlapping genes = 39\n",
      "5.366263329234705e-06\n",
      "(0.4498431503865469, 1.1306820941652043)\n",
      "0.7902626222758755\n",
      "\n",
      "abnormal neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 1517\n",
      "number overlapping genes = 59\n",
      "1.4403071047002733e-08\n",
      "(0.5422172981353959, 1.1153376607541134)\n",
      "0.8287774794447547\n",
      "\n",
      "abnormal AMPA-mediated synaptic currents\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 30\n",
      "number overlapping genes = 0\n",
      "0.9074494066040462\n",
      "(-2.962828143929808, 2.6310257034965523)\n",
      "-0.1659012202166279\n",
      "\n",
      "abnormal NMDA-mediated synaptic currents\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 1\n",
      "0.4851706576007908\n",
      "(-1.2888554051825123, 2.714636181951205)\n",
      "0.7128903883843463\n",
      "\n",
      "abnormal excitatory postsynaptic currents\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 137\n",
      "number overlapping genes = 5\n",
      "0.1478899925446971\n",
      "(-0.23545142659203988, 1.5638896369740896)\n",
      "0.6642191051910249\n",
      "\n",
      "abnormal excitatory postsynaptic potential\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 95\n",
      "number overlapping genes = 4\n",
      "0.1139192957963337\n",
      "(-0.19484731053876558, 1.8196553892128957)\n",
      "0.8124040393370651\n",
      "\n",
      "abnormal PNS synaptic transmission\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 39\n",
      "number overlapping genes = 4\n",
      "0.0008459698398705934\n",
      "(0.7310118599011535, 2.811489355020159)\n",
      "1.7712506074606562\n",
      "\n",
      "abnormal synaptic depression\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 114\n",
      "number overlapping genes = 4\n",
      "0.2247444833572393\n",
      "(-0.3819763702615466, 1.6252745473736594)\n",
      "0.6216490885560564\n",
      "\n",
      "increased synaptic depression\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.6784171882009766\n",
      "(-2.225777350669714, 3.4201561551578146)\n",
      "0.5971894022440503\n",
      "\n",
      "decreased synaptic depression\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.6415721829608467\n",
      "(-2.155345951473278, 3.4980594615882956)\n",
      "0.6713567550575088\n",
      "\n",
      "abnormal inhibitory postsynaptic currents\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 81\n",
      "number overlapping genes = 3\n",
      "0.25226691676096036\n",
      "(-0.4817131736034985, 1.8349020259557578)\n",
      "0.6765944261761296\n",
      "\n",
      "abnormal neuron specification\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "0.7135421081868061\n",
      "(-2.291587776280851, 3.3478620694230328)\n",
      "0.5281371465710909\n",
      "\n",
      "abnormal neuron apoptosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 233\n",
      "number overlapping genes = 9\n",
      "0.03413761400480176\n",
      "(0.05460983012501275, 1.4053712741224778)\n",
      "0.7299905521237453\n",
      "\n",
      "increased neuron apoptosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 167\n",
      "number overlapping genes = 7\n",
      "0.03724707842289154\n",
      "(0.04802599235600913, 1.5775689485236704)\n",
      "0.8127974704398397\n",
      "\n",
      "decreased neuron apoptosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 71\n",
      "number overlapping genes = 1\n",
      "0.7514960846279515\n",
      "(-2.2963249922112574, 1.6575118604636432)\n",
      "-0.3194065658738072\n",
      "\n",
      "neuron degeneration\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 403\n",
      "number overlapping genes = 11\n",
      "0.23763580464531042\n",
      "(-0.24218956264848412, 0.9764343841419373)\n",
      "0.3671224107467266\n",
      "\n",
      "loss of hippocampal neurons\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "0.7135421081868061\n",
      "(-2.291587776280851, 3.3478620694230328)\n",
      "0.5281371465710909\n",
      "\n",
      "loss of cortex neurons\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 3\n",
      "0.0003369711508133942\n",
      "(1.026454623309234, 3.502352324574354)\n",
      "2.264403473941794\n",
      "\n",
      "abnormal dopaminergic neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 3\n",
      "0.033625139465023504\n",
      "(0.09870320773974939, 2.4496713475071754)\n",
      "1.2741872776234624\n",
      "\n",
      "loss of dopaminergic neurons\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 1\n",
      "0.236523929350149\n",
      "(-0.8028616241082172, 3.251481720203364)\n",
      "1.2243100480475735\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal GABAergic neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 29\n",
      "number overlapping genes = 0\n",
      "0.926354928883276\n",
      "(-2.929656384441364, 2.665775921620256)\n",
      "-0.131940231410554\n",
      "\n",
      "loss of GABAergic neurons\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.6028952844028312\n",
      "(-2.0795962485420336, 3.582513928271891)\n",
      "0.7514588398649287\n",
      "\n",
      "abnormal glutaminergic neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.12065395286534009\n",
      "(-0.42844041818587897, 3.6885845018403742)\n",
      "1.6300720418272476\n",
      "\n",
      "premature neuronal precursor differentiation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 21\n",
      "number overlapping genes = 2\n",
      "0.02389661210409854\n",
      "(0.22276489486067574, 3.1449666695859406)\n",
      "1.6838657822233083\n",
      "\n",
      "abnormal synaptic transmission\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 1049\n",
      "number overlapping genes = 33\n",
      "0.0033834104889511624\n",
      "(0.18081977703494068, 0.9110859061510717)\n",
      "0.5459528415930062\n",
      "\n",
      "abnormal nervous system development\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 1216\n",
      "number overlapping genes = 73\n",
      "1.8242453796827825e-23\n",
      "(1.0915259995603819, 1.6248831511987412)\n",
      "1.3582045753795615\n",
      "\n",
      "abnormal spinal cord interneuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 32\n",
      "number overlapping genes = 1\n",
      "0.624942636445099\n",
      "(-1.496880963640767, 2.491725799187745)\n",
      "0.49742241777348895\n",
      "\n",
      "abnormal brain interneuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 39\n",
      "number overlapping genes = 0\n",
      "0.7633734607418926\n",
      "(-3.220439998656937, 2.362838842872367)\n",
      "-0.428800577892285\n",
      "\n",
      "abnormal sympathetic nervous system physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 29\n",
      "number overlapping genes = 0\n",
      "0.926354928883276\n",
      "(-2.929656384441364, 2.665775921620256)\n",
      "-0.131940231410554\n",
      "\n",
      "neuronal intranuclear inclusions\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.6415721829608467\n",
      "(-2.155345951473278, 3.4980594615882956)\n",
      "0.6713567550575088\n",
      "\n",
      "abnormal peripheral nervous system regeneration\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 1\n",
      "0.1422973440400102\n",
      "(-0.515309997938143, 3.5847149818171236)\n",
      "1.5347024919394903\n",
      "\n",
      "abnormal proprioceptive neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "0.7470538473053291\n",
      "(-2.3533460236416786, 3.280424499083319)\n",
      "0.4635392377208203\n",
      "\n",
      "abnormal synaptic glutamate release\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.6784171882009766\n",
      "(-2.225777350669714, 3.4201561551578146)\n",
      "0.5971894022440503\n",
      "\n",
      "abnormal miniature excitatory postsynaptic currents\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 119\n",
      "number overlapping genes = 5\n",
      "0.0777740351168504\n",
      "(-0.0903220924357444, 1.714118945868472)\n",
      "0.8118984267163638\n",
      "\n",
      "abnormal synaptic vesicle morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 57\n",
      "number overlapping genes = 2\n",
      "0.3913846089516939\n",
      "(-0.7962588494186797, 2.033920618163313)\n",
      "0.6188308843723167\n",
      "\n",
      "abnormal synaptic vesicle recycling\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 0\n",
      "0.8668918286175664\n",
      "(-2.56819095114319, 3.0485071517042344)\n",
      "0.2401581002805222\n",
      "\n",
      "abnormal synaptic vesicle number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 49\n",
      "number overlapping genes = 2\n",
      "0.2835855736212626\n",
      "(-0.6427915089852256, 2.19577665570739)\n",
      "0.7764925733610823\n",
      "\n",
      "abnormal neuron physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 626\n",
      "number overlapping genes = 28\n",
      "5.042685590887021e-06\n",
      "(0.5254555402686185, 1.3167410619182067)\n",
      "0.9210983010934126\n",
      "\n",
      "abnormal nervous system regeneration\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 1\n",
      "0.38659928277881084\n",
      "(-1.1215617294470674, 2.896524593917333)\n",
      "0.8874814322351328\n",
      "\n",
      "abnormal synaptic plasticity\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 66\n",
      "number overlapping genes = 2\n",
      "0.5169496704753361\n",
      "(-0.9448699946210364, 1.8783601876340117)\n",
      "0.46674509650648766\n",
      "\n",
      "abnormal neuronal precursor proliferation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 93\n",
      "number overlapping genes = 7\n",
      "0.0002922997260040195\n",
      "(0.6598866537691314, 2.2162071944913606)\n",
      "1.438046924130246\n",
      "\n",
      "abnormal neuronal precursor cell number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 67\n",
      "number overlapping genes = 7\n",
      "8.122645367139606e-06\n",
      "(1.0090996689074827, 2.590098390661304)\n",
      "1.7995990297843933\n",
      "\n",
      "increased neuronal precursor cell number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 21\n",
      "number overlapping genes = 1\n",
      "0.36154553940910195\n",
      "(-1.074987158671867, 2.9476491621858765)\n",
      "0.9363310017570048\n",
      "\n",
      "decreased neuronal precursor cell number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 44\n",
      "number overlapping genes = 4\n",
      "0.0019034909511515916\n",
      "(0.6037985405175399, 2.6710452350903444)\n",
      "1.6374218878039422\n",
      "\n",
      "increased motor neuron number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.5622719262373211\n",
      "(-1.9976606113421846, 3.6747197922682)\n",
      "0.8385295904630077\n",
      "\n",
      "abnormal somatic nervous system physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 75\n",
      "number overlapping genes = 3\n",
      "0.20091439850169635\n",
      "(-0.40308309115361163, 1.9170725385494605)\n",
      "0.7569947236979244\n",
      "\n",
      "increased sensory neuron number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 51\n",
      "number overlapping genes = 3\n",
      "0.051510929751894134\n",
      "(-0.007630188407836025, 2.335408016136313)\n",
      "1.1638889138642385\n",
      "\n",
      "abnormal neuronal migration\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 127\n",
      "number overlapping genes = 6\n",
      "0.026340942469388253\n",
      "(0.11021462319888975, 1.7643490523934098)\n",
      "0.9372818377961498\n",
      "\n",
      "abnormal retinal neuronal layer morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 422\n",
      "number overlapping genes = 9\n",
      "0.7546491096562803\n",
      "(-0.5627884952284066, 0.7763063609117151)\n",
      "0.10675893284165427\n",
      "\n",
      "abnormal olfactory sensory neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 0\n",
      "0.9678695347335673\n",
      "(-2.745044903087506, 2.8602427048873)\n",
      "0.05759890089989694\n",
      "\n",
      "abnormal autonomic nervous system physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 1\n",
      "0.902601652375179\n",
      "(-1.8606997011041668, 2.108528885109264)\n",
      "0.1239145920025484\n",
      "\n",
      "abnormal parasympathetic nervous system physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 1\n",
      "0.1422973440400102\n",
      "(-0.515309997938143, 3.5847149818171236)\n",
      "1.5347024919394903\n",
      "\n",
      "increased susceptibility to neuronal excitotoxicity\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 1\n",
      "0.6903003835663676\n",
      "(-1.586689456624333, 2.3964309655248375)\n",
      "0.40487075445025233\n",
      "\n",
      "decreased susceptibility to neuronal excitotoxicity\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 1\n",
      "0.602456873126797\n",
      "(-1.4650670279397116, 2.52561039045942)\n",
      "0.5302716812598542\n",
      "\n",
      "abnormal synaptic bouton morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 0\n",
      "0.990668350775242\n",
      "(-2.784783873831427, 2.818218847647154)\n",
      "0.016717486907863677\n",
      "\n",
      "abnormal neuron number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 510\n",
      "number overlapping genes = 20\n",
      "0.0011543148692976922\n",
      "(0.3033965193998726, 1.225349579238875)\n",
      "0.7643730493193739\n",
      "\n",
      "increased neuron number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 83\n",
      "number overlapping genes = 5\n",
      "0.01021713385602555\n",
      "(0.28273781538942344, 2.1043347385296896)\n",
      "1.1935362769595566\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "decreased neuron number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 428\n",
      "number overlapping genes = 15\n",
      "0.017954468358835216\n",
      "(0.10933717984104752, 1.1634505972892346)\n",
      "0.636393888565141\n",
      "\n",
      "abnormal miniature inhibitory postsynaptic currents\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 50\n",
      "number overlapping genes = 3\n",
      "0.047558532303664185\n",
      "(0.012756214906582963, 2.3572474327943667)\n",
      "1.1850018238504747\n",
      "\n",
      "abnormal synaptonemal complex\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 21\n",
      "number overlapping genes = 1\n",
      "0.36154553940910195\n",
      "(-1.074987158671867, 2.9476491621858765)\n",
      "0.9363310017570048\n",
      "\n",
      "abnormal neuron differentiation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 323\n",
      "number overlapping genes = 21\n",
      "1.7634119262306598e-08\n",
      "(0.8554186479076099, 1.7681520366818702)\n",
      "1.31178534229474\n",
      "\n",
      "abnormal hippocampus neuron morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 115\n",
      "number overlapping genes = 8\n",
      "0.00025821872975417094\n",
      "(0.6281567877401328, 2.081712813719372)\n",
      "1.3549348007297524\n",
      "\n",
      "abnormal neuron proliferation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 124\n",
      "number overlapping genes = 8\n",
      "0.0005735700998532681\n",
      "(0.5487766322154467, 1.9984951476718202)\n",
      "1.2736358899436335\n",
      "\n",
      "abnormal sensory neuron physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 64\n",
      "number overlapping genes = 2\n",
      "0.4889590144778104\n",
      "(-0.9136878256808278, 1.9109136381807301)\n",
      "0.49861290624995114\n",
      "\n",
      "abnormal synaptic dopamine release\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.6415721829608467\n",
      "(-2.155345951473278, 3.4980594615882956)\n",
      "0.6713567550575088\n",
      "\n",
      "increased nervous system tumor incidence\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 68\n",
      "number overlapping genes = 4\n",
      "0.024500304975207332\n",
      "(0.14993677063019595, 2.182042943238907)\n",
      "1.1659898569345515\n",
      "\n",
      "reduced AMPA-mediated synaptic currents\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.6784171882009766\n",
      "(-2.225777350669714, 3.4201561551578146)\n",
      "0.5971894022440503\n",
      "\n",
      "abnormal excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 36\n",
      "number overlapping genes = 2\n",
      "0.13133897445775147\n",
      "(-0.3291982044026409, 2.531303881813344)\n",
      "1.1010528387053515\n",
      "\n",
      "increased excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 2\n",
      "0.0037970318555873607\n",
      "(0.7203028403622165, 3.7414663120269624)\n",
      "2.2308845761945895\n",
      "\n",
      "decreased excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 1\n",
      "0.43629129894425456\n",
      "(-1.2086896621156273, 2.8014713275886094)\n",
      "0.796390832736491\n",
      "\n",
      "decreased dopaminergic neuron number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 1\n",
      "0.46084895600620923\n",
      "(-1.2495725613982311, 2.75711616615099)\n",
      "0.7537718023763795\n",
      "\n",
      "increased susceptibility to dopaminergic neuron neurotoxicity\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 1\n",
      "0.16487518105154264\n",
      "(-0.5952704250467558, 3.4905339077295783)\n",
      "1.4476317413414113\n",
      "\n",
      "decreased susceptibility to dopaminergic neuron neurotoxicity\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.6028952844028312\n",
      "(-2.0795962485420336, 3.582513928271891)\n",
      "0.7514588398649287\n",
      "\n",
      "ectopic neuron\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 63\n",
      "number overlapping genes = 2\n",
      "0.4749599234705568\n",
      "(-0.8977273019982055, 1.9275932570307466)\n",
      "0.5149329775162705\n",
      "\n",
      "impaired neuron differentiation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 2\n",
      "0.0067336994257893015\n",
      "(0.5710193709676845, 3.5564041106104964)\n",
      "2.0637117407890906\n",
      "\n",
      "abnormal neuronal stem cell physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.12065395286534009\n",
      "(-0.42844041818587897, 3.6885845018403742)\n",
      "1.6300720418272476\n",
      "\n",
      "nervous system inclusion bodies\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 63\n",
      "number overlapping genes = 4\n",
      "0.016360421725439316\n",
      "(0.22908011595509392, 2.2661863807715275)\n",
      "1.2476332483633108\n",
      "\n",
      "abnormal pyramidal neuron dendrite morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.6784171882009766\n",
      "(-2.225777350669714, 3.4201561551578146)\n",
      "0.5971894022440503\n",
      "\n",
      "abnormal hippocampal pyramidal neuron dendrite morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.5622719262373211\n",
      "(-1.9976606113421846, 3.6747197922682)\n",
      "0.8385295904630077\n",
      "\n",
      "increased neuronal precursor proliferation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.5622719262373211\n",
      "(-1.9976606113421846, 3.6747197922682)\n",
      "0.8385295904630077\n",
      "\n",
      "decreased neuronal precursor proliferation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "0.7470538473053291\n",
      "(-2.3533460236416786, 3.280424499083319)\n",
      "0.4635392377208203\n",
      "\n",
      "abnormal postsynaptic density morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 1\n",
      "0.2861938705200482\n",
      "(-0.9206985698996828, 3.1187548221817973)\n",
      "1.0990281261410573\n",
      "\n",
      "abnormal synaptic physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 1055\n",
      "number overlapping genes = 33\n",
      "0.0037654612536952275\n",
      "(0.17458299867071736, 0.9047885842889375)\n",
      "0.5396857914798274\n",
      "\n",
      "impaired synaptic physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 41\n",
      "number overlapping genes = 1\n",
      "0.811364414498692\n",
      "(-1.745305775424278, 2.229295861665837)\n",
      "0.24199504312077957\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0002419</th>\n",
       "      <td>1.760349e-01</td>\n",
       "      <td>-0.520002</td>\n",
       "      <td>-1.273243</td>\n",
       "      <td>0.233240</td>\n",
       "      <td>598</td>\n",
       "      <td>abnormal innate immunity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000778</th>\n",
       "      <td>1.246682e-11</td>\n",
       "      <td>1.675418</td>\n",
       "      <td>1.190708</td>\n",
       "      <td>2.160127</td>\n",
       "      <td>209</td>\n",
       "      <td>abnormal nervous system tract morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000811</th>\n",
       "      <td>5.587282e-02</td>\n",
       "      <td>1.409072</td>\n",
       "      <td>-0.035328</td>\n",
       "      <td>2.853473</td>\n",
       "      <td>27</td>\n",
       "      <td>hippocampal neuron degeneration</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000937</th>\n",
       "      <td>8.701499e-03</td>\n",
       "      <td>0.962119</td>\n",
       "      <td>0.243357</td>\n",
       "      <td>1.680880</td>\n",
       "      <td>166</td>\n",
       "      <td>abnormal motor neuron morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000938</th>\n",
       "      <td>2.085117e-01</td>\n",
       "      <td>0.913425</td>\n",
       "      <td>-0.510067</td>\n",
       "      <td>2.336917</td>\n",
       "      <td>43</td>\n",
       "      <td>motor neuron degeneration</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0002419  1.760349e-01 -0.520002        -1.273243         0.233240   \n",
       "MP:0000778  1.246682e-11  1.675418         1.190708         2.160127   \n",
       "MP:0000811  5.587282e-02  1.409072        -0.035328         2.853473   \n",
       "MP:0000937  8.701499e-03  0.962119         0.243357         1.680880   \n",
       "MP:0000938  2.085117e-01  0.913425        -0.510067         2.336917   \n",
       "\n",
       "            num_genes_in_term                            MP_description  \n",
       "MP:0002419                598                  abnormal innate immunity  \n",
       "MP:0000778                209  abnormal nervous system tract morphology  \n",
       "MP:0000811                 27           hippocampal neuron degeneration  \n",
       "MP:0000937                166          abnormal motor neuron morphology  \n",
       "MP:0000938                 43                 motor neuron degeneration  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add a negative control phenotype: abnormal innate immunity: MP:0008171\n",
    "# negative controls are tough here because we're dealing with development... which impacts almost everything\n",
    "MP_focal_list = ['MP:0002419']+MP_focal_brain_list\n",
    "root_KO_brain_df=validation.MPO_enrichment_root(hier_df,MPO,mgi_df,MP_focal_list,G_int,verbose=True)\n",
    "root_KO_brain_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0002419</th>\n",
       "      <td>1.760349e-01</td>\n",
       "      <td>-0.520002</td>\n",
       "      <td>-1.273243</td>\n",
       "      <td>0.233240</td>\n",
       "      <td>598</td>\n",
       "      <td>abnormal innate immunity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000778</th>\n",
       "      <td>1.246682e-11</td>\n",
       "      <td>1.675418</td>\n",
       "      <td>1.190708</td>\n",
       "      <td>2.160127</td>\n",
       "      <td>209</td>\n",
       "      <td>abnormal nervous system tract morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000811</th>\n",
       "      <td>5.587282e-02</td>\n",
       "      <td>1.409072</td>\n",
       "      <td>-0.035328</td>\n",
       "      <td>2.853473</td>\n",
       "      <td>27</td>\n",
       "      <td>hippocampal neuron degeneration</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000937</th>\n",
       "      <td>8.701499e-03</td>\n",
       "      <td>0.962119</td>\n",
       "      <td>0.243357</td>\n",
       "      <td>1.680880</td>\n",
       "      <td>166</td>\n",
       "      <td>abnormal motor neuron morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000938</th>\n",
       "      <td>2.085117e-01</td>\n",
       "      <td>0.913425</td>\n",
       "      <td>-0.510067</td>\n",
       "      <td>2.336917</td>\n",
       "      <td>43</td>\n",
       "      <td>motor neuron degeneration</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0002419  1.760349e-01 -0.520002        -1.273243         0.233240   \n",
       "MP:0000778  1.246682e-11  1.675418         1.190708         2.160127   \n",
       "MP:0000811  5.587282e-02  1.409072        -0.035328         2.853473   \n",
       "MP:0000937  8.701499e-03  0.962119         0.243357         1.680880   \n",
       "MP:0000938  2.085117e-01  0.913425        -0.510067         2.336917   \n",
       "\n",
       "            num_genes_in_term                            MP_description  \n",
       "MP:0002419                598                  abnormal innate immunity  \n",
       "MP:0000778                209  abnormal nervous system tract morphology  \n",
       "MP:0000811                 27           hippocampal neuron degeneration  \n",
       "MP:0000937                166          abnormal motor neuron morphology  \n",
       "MP:0000938                 43                 motor neuron degeneration  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_brain_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal heart morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 1739\n",
      "number overlapping genes = 63\n",
      "1.1114278529864863e-07\n",
      "(0.47656061221016366, 1.0346088973948495)\n",
      "0.7555847548025065\n",
      "\n",
      "abnormal heart development\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 363\n",
      "number overlapping genes = 18\n",
      "4.6016226497267544e-05\n",
      "(0.5254024173284649, 1.4991678147765732)\n",
      "1.012285116052519\n",
      "\n",
      "abnormal heart looping\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 130\n",
      "number overlapping genes = 5\n",
      "0.11756738029707699\n",
      "(-0.18144993576466373, 1.6197015619574069)\n",
      "0.7191258130963716\n",
      "\n",
      "abnormal heart tube morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 84\n",
      "number overlapping genes = 4\n",
      "0.06761475511058576\n",
      "(-0.06823557072156028, 1.952021376907787)\n",
      "0.9418929030931134\n",
      "\n",
      "enlarged heart\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 756\n",
      "number overlapping genes = 29\n",
      "0.000134639760611611\n",
      "(0.3679833458397005, 1.1444016542775026)\n",
      "0.7561925000586016\n",
      "\n",
      "heart right ventricle hypertrophy\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 44\n",
      "number overlapping genes = 1\n",
      "0.8671289868305121\n",
      "(-1.8161184117447284, 2.155110297502671)\n",
      "0.1694959428789713\n",
      "\n",
      "abnormal heart shape\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 34\n",
      "number overlapping genes = 1\n",
      "0.6688690040389004\n",
      "(-1.5576364206915345, 2.427202758421087)\n",
      "0.4347831688647763\n",
      "\n",
      "abnormal myocardial fiber morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 205\n",
      "number overlapping genes = 6\n",
      "0.298067852907112\n",
      "(-0.3844152948977445, 1.2546057205009542)\n",
      "0.4350952128016048\n",
      "\n",
      "abnormal heart valve morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 180\n",
      "number overlapping genes = 10\n",
      "0.000714195291367813\n",
      "(0.4702951591354815, 1.7648068171257258)\n",
      "1.1175509881306036\n",
      "\n",
      "abnormal pericardium morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 232\n",
      "number overlapping genes = 11\n",
      "0.002470819314169057\n",
      "(0.3350547495332826, 1.5660563070214804)\n",
      "0.9505555282773814\n",
      "\n",
      "enlarged pericardium\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 2\n",
      "0.24522552757786475\n",
      "(-0.578616903485393, 2.2638749480433398)\n",
      "0.8426290222789734\n",
      "\n",
      "distended pericardium\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 57\n",
      "number overlapping genes = 2\n",
      "0.3913846089516939\n",
      "(-0.7962588494186797, 2.033920618163313)\n",
      "0.6188308843723167\n",
      "\n",
      "absent myocardial trabeculae\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 1\n",
      "0.43629129894425456\n",
      "(-1.2086896621156273, 2.8014713275886094)\n",
      "0.796390832736491\n",
      "\n",
      "abnormal cardiac stroke volume\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 53\n",
      "number overlapping genes = 0\n",
      "0.6045601140001619\n",
      "(-3.5233392496448865, 2.0506116855254324)\n",
      "-0.7363637820597271\n",
      "\n",
      "abnormal response to cardiac infarction\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 80\n",
      "number overlapping genes = 0\n",
      "0.41804522721829573\n",
      "(-3.9322902659142565, 1.632876579781192)\n",
      "-1.1497068430665323\n",
      "\n",
      "dextrocardia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 38\n",
      "number overlapping genes = 1\n",
      "0.7524424058446197\n",
      "(-1.6691239770137176, 2.309393959849718)\n",
      "0.32013499141800006\n",
      "\n",
      "mesocardia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "0.7135421081868061\n",
      "(-2.291587776280851, 3.3478620694230328)\n",
      "0.5281371465710909\n",
      "\n",
      "abnormal cardiovascular system physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 1858\n",
      "number overlapping genes = 52\n",
      "0.004036779439547747\n",
      "(0.14012055726191647, 0.7402045902660805)\n",
      "0.4401625737639985\n",
      "\n",
      "cardiac hypertrophy\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 200\n",
      "number overlapping genes = 9\n",
      "0.009927695462957863\n",
      "(0.21377517872955454, 1.568915971838056)\n",
      "0.8913455752838053\n",
      "\n",
      "abnormal cardiac output\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 63\n",
      "number overlapping genes = 0\n",
      "0.5219773972771133\n",
      "(-3.694712668996348, 1.875108719458205)\n",
      "-0.9098019747690715\n",
      "\n",
      "abnormal heart rate\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 412\n",
      "number overlapping genes = 12\n",
      "0.1434222990618046\n",
      "(-0.1482335191776789, 1.0213832944257781)\n",
      "0.4365748876240496\n",
      "\n",
      "irregular heartbeat\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 70\n",
      "number overlapping genes = 6\n",
      "0.00024870939897339746\n",
      "(0.7336026322596957, 2.421577274960768)\n",
      "1.5775899536102318\n",
      "\n",
      "pericardial edema\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 86\n",
      "number overlapping genes = 4\n",
      "0.07500315203225826\n",
      "(-0.09246755305407861, 1.9266296555521518)\n",
      "0.9170810512490366\n",
      "\n",
      "heart inflammation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 41\n",
      "number overlapping genes = 0\n",
      "0.7366050513506434\n",
      "(-3.2697086655641785, 2.3118487716394034)\n",
      "-0.47892994696238755\n",
      "\n",
      "myocarditis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.6028952844028312\n",
      "(-2.0795962485420336, 3.582513928271891)\n",
      "0.7514588398649287\n",
      "\n",
      "small heart\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 345\n",
      "number overlapping genes = 12\n",
      "0.037041935502941006\n",
      "(0.03750331239863569, 1.2104179575079723)\n",
      "0.623960634953304\n",
      "\n",
      "abnormal myocardial trabeculae morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 145\n",
      "number overlapping genes = 10\n",
      "4.903473979380597e-05\n",
      "(0.6984049263846094, 2.0019356089418836)\n",
      "1.3501702676632465\n",
      "\n",
      "disorganized myocardium\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 1\n",
      "0.8302989440126214\n",
      "(-1.7694684498018842, 2.2039543588300745)\n",
      "0.21724295451409503\n",
      "\n",
      "heart left ventricle hypertrophy\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 68\n",
      "number overlapping genes = 0\n",
      "0.4873935254337888\n",
      "(-3.770578404425602, 1.7976328514849174)\n",
      "-0.9864727764703423\n",
      "\n",
      "increased heart rate\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 170\n",
      "number overlapping genes = 2\n",
      "0.4794621379612579\n",
      "(-1.9029417403437565, 0.893843288050425)\n",
      "-0.5045492261466658\n",
      "\n",
      "thin myocardium\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 90\n",
      "number overlapping genes = 1\n",
      "0.5777277152420159\n",
      "(-2.5346335560463045, 1.4132728254616336)\n",
      "-0.5606803652923356\n",
      "\n",
      "heart hypoplasia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 102\n",
      "number overlapping genes = 5\n",
      "0.034933341333036265\n",
      "(0.06891358923842728, 1.8798882366910412)\n",
      "0.9744009129647342\n",
      "\n",
      "dilated heart left ventricle\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 111\n",
      "number overlapping genes = 1\n",
      "0.44182937217302065\n",
      "(-2.7456427060973505, 1.1980871938431374)\n",
      "-0.7737777561271066\n",
      "\n",
      "dilated heart right ventricle\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 52\n",
      "number overlapping genes = 0\n",
      "0.614000607297486\n",
      "(-3.5044815880477427, 2.0699694508604556)\n",
      "-0.7172560685936435\n",
      "\n",
      "dilated cardiomyopathy\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 83\n",
      "number overlapping genes = 1\n",
      "0.6349784295648716\n",
      "(-2.453232044281166, 1.4965404574129133)\n",
      "-0.4783457934341264\n",
      "\n",
      "increased heart weight\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 256\n",
      "number overlapping genes = 8\n",
      "0.1641055208677279\n",
      "(-0.20671858491694262, 1.2184782391452875)\n",
      "0.5058798271141725\n",
      "\n",
      "decreased heart weight\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 156\n",
      "number overlapping genes = 2\n",
      "0.5594029374543461\n",
      "(-1.815834672386678, 0.9824363239582588)\n",
      "-0.41669917421420966\n",
      "\n",
      "dystrophic cardiac calcinosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "0.7135421081868061\n",
      "(-2.291587776280851, 3.3478620694230328)\n",
      "0.5281371465710909\n",
      "\n",
      "abnormal cardiovascular development\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 768\n",
      "number overlapping genes = 40\n",
      "1.1773918541892142e-10\n",
      "(0.7746921990608373, 1.4521943889206124)\n",
      "1.1134432939907248\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal cardiac muscle contractility\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 292\n",
      "number overlapping genes = 3\n",
      "0.26800726615733117\n",
      "(-1.7884627072386075, 0.49690132104966556)\n",
      "-0.645780693094471\n",
      "\n",
      "increased myocardial infarct size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 28\n",
      "number overlapping genes = 0\n",
      "0.9459558176920977\n",
      "(-2.895350987058694, 2.70177203105579)\n",
      "-0.09678947800145199\n",
      "\n",
      "decreased myocardial infarct size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 0\n",
      "0.990668350775242\n",
      "(-2.784783873831427, 2.818218847647154)\n",
      "0.016717486907863677\n",
      "\n",
      "abnormal epicardium morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 0\n",
      "0.8388873413196973\n",
      "(-2.5186375062206596, 3.1016590992041126)\n",
      "0.2915107964917265\n",
      "\n",
      "abnormal heart atrium morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 227\n",
      "number overlapping genes = 6\n",
      "0.43083057957662463\n",
      "(-0.4894218804666577, 1.1472529568088585)\n",
      "0.3289155381711004\n",
      "\n",
      "dilated heart atrium\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 58\n",
      "number overlapping genes = 0\n",
      "0.5607708655218802\n",
      "(-3.6126667955470686, 1.9590417461883827)\n",
      "-0.826812524679343\n",
      "\n",
      "cardiac fibrosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 172\n",
      "number overlapping genes = 2\n",
      "0.4690825014429477\n",
      "(-1.9147998873552334, 0.8817927834775274)\n",
      "-0.516503551938853\n",
      "\n",
      "abnormal cardiomyocyte apoptosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 63\n",
      "number overlapping genes = 3\n",
      "0.11366474439225407\n",
      "(-0.22463672643130805, 2.104698880042058)\n",
      "0.9400310768053748\n",
      "\n",
      "increased cardiomyocyte apoptosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 53\n",
      "number overlapping genes = 3\n",
      "0.059985701712958986\n",
      "(-0.04720414216708346, 2.293099956457266)\n",
      "1.1229479071450914\n",
      "\n",
      "decreased cardiomyocyte apoptosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.5622719262373211\n",
      "(-1.9976606113421846, 3.6747197922682)\n",
      "0.8385295904630077\n",
      "\n",
      "decreased cardiac output\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 44\n",
      "number overlapping genes = 0\n",
      "0.69932560394501\n",
      "(-3.3393599289913203, 2.2399080020861044)\n",
      "-0.5497259634526079\n",
      "\n",
      "increased cardiac output\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.6784171882009766\n",
      "(-2.225777350669714, 3.4201561551578146)\n",
      "0.5971894022440503\n",
      "\n",
      "abnormal fetal cardiomyocyte proliferation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 54\n",
      "number overlapping genes = 6\n",
      "1.917730499444551e-05\n",
      "(1.0104663894607087, 2.721982651496473)\n",
      "1.8662245204785908\n",
      "\n",
      "decreased heart right ventricle size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 27\n",
      "number overlapping genes = 1\n",
      "0.5092300065850646\n",
      "(-1.3266588044187113, 2.6738793088755193)\n",
      "0.673610252228404\n",
      "\n",
      "abnormal heart right ventricle morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 182\n",
      "number overlapping genes = 6\n",
      "0.1818360673709758\n",
      "(-0.26175193938178554, 1.3803440523959796)\n",
      "0.559296056507097\n",
      "\n",
      "abnormal heart left ventricle morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 338\n",
      "number overlapping genes = 11\n",
      "0.07637096917621494\n",
      "(-0.05854281434833308, 1.163272398844675)\n",
      "0.5523647922481709\n",
      "\n",
      "abnormal heart right atrium morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 39\n",
      "number overlapping genes = 0\n",
      "0.7633734607418926\n",
      "(-3.220439998656937, 2.362838842872367)\n",
      "-0.428800577892285\n",
      "\n",
      "abnormal heart left atrium morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 1\n",
      "0.3112785525063787\n",
      "(-0.9748061936689836, 3.058426828736527)\n",
      "1.0418103175337716\n",
      "\n",
      "increased heart rate variability\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 1\n",
      "0.3112785525063787\n",
      "(-0.9748061936689836, 3.058426828736527)\n",
      "1.0418103175337716\n",
      "\n",
      "decreased heart rate variability\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 45\n",
      "number overlapping genes = 1\n",
      "0.8850327323342608\n",
      "(-1.8386560843721316, 2.13154996041988)\n",
      "0.1464469380238742\n",
      "\n",
      "abnormal endocardium morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 27\n",
      "number overlapping genes = 2\n",
      "0.0558728244993557\n",
      "(-0.03532781607510116, 2.8534727613539923)\n",
      "1.4090724726394455\n",
      "\n",
      "abnormal cardiac cell glucose uptake\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "0.7470538473053291\n",
      "(-2.3533460236416786, 3.280424499083319)\n",
      "0.4635392377208203\n",
      "\n",
      "abnormal myocardium compact layer morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 73\n",
      "number overlapping genes = 6\n",
      "0.0003659187440027501\n",
      "(0.6892080796859394, 2.373995367973592)\n",
      "1.5316017238297657\n",
      "\n",
      "thin myocardium compact layer\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 5\n",
      "0.00011564159591604275\n",
      "(0.9039975494254912, 2.7737547306585357)\n",
      "1.8388761400420135\n",
      "\n",
      "abnormal cardiac muscle relaxation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 33\n",
      "number overlapping genes = 1\n",
      "0.6470818274825195\n",
      "(-1.5277179981853386, 2.458946548709598)\n",
      "0.4656142752621295\n",
      "\n",
      "abnormal heartbeat\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 464\n",
      "number overlapping genes = 15\n",
      "0.04032880003504268\n",
      "(0.024277306121827658, 1.0769757842146634)\n",
      "0.5506265451682455\n",
      "\n",
      "absent heartbeat\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 23\n",
      "number overlapping genes = 2\n",
      "0.03284122283609013\n",
      "(0.12916690340443182, 3.038160129576032)\n",
      "1.583663516490232\n",
      "\n",
      "cardia bifida\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 2\n",
      "0.02005671266843713\n",
      "(0.27305276984780513, 3.2029320337299367)\n",
      "1.7379924017888708\n",
      "\n",
      "abnormal myocardial fiber physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 144\n",
      "number overlapping genes = 6\n",
      "0.05578951372267483\n",
      "(-0.019897402129705477, 1.6295018527475418)\n",
      "0.8048022253089182\n",
      "\n",
      "failure of heart looping\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 47\n",
      "number overlapping genes = 3\n",
      "0.03683016780263633\n",
      "(0.0765181457534887, 2.425758399783242)\n",
      "1.2511382727683653\n",
      "\n",
      "abnormal direction of heart looping\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 0\n",
      "0.7238121144177112\n",
      "(-3.2934659949079155, 2.287292038717176)\n",
      "-0.5030869780953697\n",
      "\n",
      "abnormal response of heart to induced stress\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 97\n",
      "number overlapping genes = 1\n",
      "0.5270260385623192\n",
      "(-2.6099656668368647, 1.3363460453266065)\n",
      "-0.6368098107551292\n",
      "\n",
      "increased response of heart to induced stress\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 50\n",
      "number overlapping genes = 0\n",
      "0.6336349012557134\n",
      "(-3.465671884609451, 2.1098392264576873)\n",
      "-0.6779163290758818\n",
      "\n",
      "decreased response of heart to induced stress\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 32\n",
      "number overlapping genes = 1\n",
      "0.624942636445099\n",
      "(-1.496880963640767, 2.491725799187745)\n",
      "0.49742241777348895\n",
      "\n",
      "enlarged myocardial fiber\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 63\n",
      "number overlapping genes = 2\n",
      "0.4749599234705568\n",
      "(-0.8977273019982055, 1.9275932570307466)\n",
      "0.5149329775162705\n",
      "\n",
      "small myocardial fiber\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "0.7135421081868061\n",
      "(-2.291587776280851, 3.3478620694230328)\n",
      "0.5281371465710909\n",
      "\n",
      "myocardial fiber degeneration\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 33\n",
      "number overlapping genes = 1\n",
      "0.6470818274825195\n",
      "(-1.5277179981853386, 2.458946548709598)\n",
      "0.4656142752621295\n",
      "\n",
      "decreased myocardial fiber number\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 1\n",
      "0.33641881046090405\n",
      "(-1.0261459450442825, 3.0015133409807007)\n",
      "0.9876836979682091\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal cardinal vein morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 0\n",
      "0.8892032714177291\n",
      "(-2.994938894608421, 2.5974379272024337)\n",
      "-0.19875048370299364\n",
      "\n",
      "abnormal anterior cardinal vein morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.6415721829608467\n",
      "(-2.155345951473278, 3.4980594615882956)\n",
      "0.6713567550575088\n",
      "\n",
      "abnormal heart weight\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 406\n",
      "number overlapping genes = 10\n",
      "0.4269205176403893\n",
      "(-0.37892153811359064, 0.8955149756640064)\n",
      "0.2582967187752079\n",
      "\n",
      "dilated heart\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 229\n",
      "number overlapping genes = 6\n",
      "0.44367923959328825\n",
      "(-0.49845596159501415, 1.1380283978116932)\n",
      "0.3197862181083395\n",
      "\n",
      "decreased cardiac muscle contractility\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 232\n",
      "number overlapping genes = 3\n",
      "0.48306286339127735\n",
      "(-1.5536649992518803, 0.7347440837594408)\n",
      "-0.40946045774621975\n",
      "\n",
      "hemopericardium\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 49\n",
      "number overlapping genes = 1\n",
      "0.9533488172631164\n",
      "(-1.92407208830465, 2.042467246544145)\n",
      "0.059197579119747434\n",
      "\n",
      "abnormal heart ventricle morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 716\n",
      "number overlapping genes = 27\n",
      "0.0003316379119689218\n",
      "(0.3332655065166066, 1.135077660971359)\n",
      "0.7341715837439828\n",
      "\n",
      "pericardial effusion\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 118\n",
      "number overlapping genes = 5\n",
      "0.07463620146812659\n",
      "(-0.08161696558302223, 1.7231545491246956)\n",
      "0.8207687917708366\n",
      "\n",
      "abnormal myocardium layer morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 425\n",
      "number overlapping genes = 19\n",
      "0.0001854338661012072\n",
      "(0.42955506928725734, 1.3765285307329762)\n",
      "0.9030418000101168\n",
      "\n",
      "cardiomyopathy\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 129\n",
      "number overlapping genes = 1\n",
      "0.35684472380239696\n",
      "(-2.897022606959073, 1.0442161180817313)\n",
      "-0.9264032444386707\n",
      "\n",
      "decreased heart rate\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 184\n",
      "number overlapping genes = 9\n",
      "0.004676306673629251\n",
      "(0.30086696327021334, 1.6587185966207252)\n",
      "0.9797927799454693\n",
      "\n",
      "abnormal heart size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 1037\n",
      "number overlapping genes = 36\n",
      "0.0002423319883299945\n",
      "(0.3070377758166717, 1.0107679014719233)\n",
      "0.6589028386442974\n",
      "\n",
      "increased cardiac muscle contractility\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 60\n",
      "number overlapping genes = 1\n",
      "0.883657198582396\n",
      "(-2.1272980299781605, 1.831710837338456)\n",
      "-0.14779359631985223\n",
      "\n",
      "cardiac interstitial fibrosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 65\n",
      "number overlapping genes = 0\n",
      "0.5076770025980152\n",
      "(-3.7257474698469912, 1.8434001962544957)\n",
      "-0.9411736367962478\n",
      "\n",
      "abnormal heart position or orientation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 58\n",
      "number overlapping genes = 2\n",
      "0.4052230357592965\n",
      "(-0.8138961319475919, 2.015401831437586)\n",
      "0.600752849744997\n",
      "\n",
      "myocardium necrosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 1\n",
      "0.43629129894425456\n",
      "(-1.2086896621156273, 2.8014713275886094)\n",
      "0.796390832736491\n",
      "\n",
      "abnormal heart septum morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 376\n",
      "number overlapping genes = 21\n",
      "7.492056941790554e-07\n",
      "(0.6926052726735004, 1.6011695328470175)\n",
      "1.146887402760259\n",
      "\n",
      "abnormal cardiac outflow tract development\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 145\n",
      "number overlapping genes = 11\n",
      "4.89697449392355e-06\n",
      "(0.8314951368032111, 2.0806955905815485)\n",
      "1.4560953636923797\n",
      "\n",
      "congestive heart failure\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 85\n",
      "number overlapping genes = 2\n",
      "0.7744701531715026\n",
      "(-1.2010846375960933, 1.6123952482552986)\n",
      "0.20565530532960263\n",
      "\n",
      "dilated heart ventricle\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 150\n",
      "number overlapping genes = 2\n",
      "0.5979024040125371\n",
      "(-1.7760964398959347, 1.022897182588557)\n",
      "-0.3765996286536888\n",
      "\n",
      "enlarged heart atrium\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 1\n",
      "0.8302989440126214\n",
      "(-1.7694684498018842, 2.2039543588300745)\n",
      "0.21724295451409503\n",
      "\n",
      "enlarged heart right atrium\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.5622719262373211\n",
      "(-1.9976606113421846, 3.6747197922682)\n",
      "0.8385295904630077\n",
      "\n",
      "increased heart ventricle size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 196\n",
      "number overlapping genes = 4\n",
      "0.9064569379185186\n",
      "(-0.9364431005670318, 1.0558926500264514)\n",
      "0.05972477472970983\n",
      "\n",
      "abnormal heart ventricle pressure\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 67\n",
      "number overlapping genes = 0\n",
      "0.49400567021746666\n",
      "(-3.755855165109217, 1.8126589270980964)\n",
      "-0.9715981190055603\n",
      "\n",
      "abnormal fetal cardiomyocyte morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 33\n",
      "number overlapping genes = 2\n",
      "0.10282318271149668\n",
      "(-0.24046381890189994, 2.6276728229590764)\n",
      "1.1936045020285881\n",
      "\n",
      "abnormal cardiac epithelial to mesenchymal transition\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 1\n",
      "0.4851706576007908\n",
      "(-1.2888554051825123, 2.714636181951205)\n",
      "0.7128903883843463\n",
      "\n",
      "ventricular tachycardia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 21\n",
      "number overlapping genes = 1\n",
      "0.36154553940910195\n",
      "(-1.074987158671867, 2.9476491621858765)\n",
      "0.9363310017570048\n",
      "\n",
      "cardiac valve regurgitation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 0\n",
      "0.9874010911641669\n",
      "(-2.823009033512813, 2.7778837350166556)\n",
      "-0.022562649248078603\n",
      "\n",
      "heart hemorrhage\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 59\n",
      "number overlapping genes = 1\n",
      "0.8970953433892608\n",
      "(-2.1104275831496984, 1.8491483367439776)\n",
      "-0.13063962320286038\n",
      "\n",
      "abnormal heart ventricle outflow tract morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 98\n",
      "number overlapping genes = 6\n",
      "0.004320295295052934\n",
      "(0.3799204716029031, 2.046110345413356)\n",
      "1.2130154085081295\n",
      "\n",
      "heart right ventricle hypoplasia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 1\n",
      "0.21213765295609144\n",
      "(-0.7383162229895852, 3.325040830430651)\n",
      "1.2933623037205328\n",
      "\n",
      "abnormal heart and great vessel attachment\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 212\n",
      "number overlapping genes = 13\n",
      "2.49471035218006e-05\n",
      "(0.6581224061217505, 1.8019953950475434)\n",
      "1.230058900584647\n",
      "\n",
      "abnormal heart and great artery attachment\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 210\n",
      "number overlapping genes = 13\n",
      "2.1469618378555573e-05\n",
      "(0.6681723005286573, 1.8123878463122995)\n",
      "1.2402800734204784\n",
      "\n",
      "abnormal heart right ventricle outflow tract morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 56\n",
      "number overlapping genes = 2\n",
      "0.3776032373568248\n",
      "(-0.7783071166450239, 2.0527862138883233)\n",
      "0.6372395486216496\n",
      "\n",
      "abnormal heart left ventricle outflow tract morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 76\n",
      "number overlapping genes = 5\n",
      "0.005717508336908736\n",
      "(0.3745223190007202, 2.2014424856677937)\n",
      "1.287982402334257\n",
      "\n",
      "abnormal heart atrium and ventricle connection\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 86\n",
      "number overlapping genes = 4\n",
      "0.07500315203225826\n",
      "(-0.09246755305407861, 1.9266296555521518)\n",
      "0.9170810512490366\n",
      "\n",
      "heart left ventricle outflow tract stenosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 0\n",
      "0.919455735137897\n",
      "(-2.6605094300127567, 2.9499676424969805)\n",
      "0.14472910624211188\n",
      "\n",
      "heart right ventricle outflow tract stenosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 1\n",
      "0.33641881046090405\n",
      "(-1.0261459450442825, 3.0015133409807007)\n",
      "0.9876836979682091\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal atrium myocardium morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.12065395286534009\n",
      "(-0.42844041818587897, 3.6885845018403742)\n",
      "1.6300720418272476\n",
      "\n",
      "abnormal ventricle myocardium morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 107\n",
      "number overlapping genes = 5\n",
      "0.04527806571782813\n",
      "(0.01942589284001872, 1.8282554089468608)\n",
      "0.9238406508934398\n",
      "\n",
      "myocardium hypoplasia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 86\n",
      "number overlapping genes = 5\n",
      "0.012791132451432695\n",
      "(0.2458205222129366, 2.0654136899664794)\n",
      "1.155617106089708\n",
      "\n",
      "ventricle myocardium hypoplasia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 70\n",
      "number overlapping genes = 4\n",
      "0.028413113485876874\n",
      "(0.119941435187495, 2.1502567101118926)\n",
      "1.1350990726496937\n",
      "\n",
      "myocardial trabeculae hypoplasia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 62\n",
      "number overlapping genes = 4\n",
      "0.014996339606971646\n",
      "(0.24568318016942947, 2.283891262791176)\n",
      "1.2647872214803026\n",
      "\n",
      "abnormal heart electrocardiography waveform feature\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 345\n",
      "number overlapping genes = 9\n",
      "0.3534084639892082\n",
      "(-0.3533686744367437, 0.9889043668352804)\n",
      "0.31776784619926834\n",
      "\n",
      "abnormal heart layer morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 434\n",
      "number overlapping genes = 20\n",
      "7.103107154207121e-05\n",
      "(0.4750123275773309, 1.4000797847418962)\n",
      "0.9375460561596136\n",
      "\n",
      "thin ventricle myocardium compact layer\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 1\n",
      "0.21213765295609144\n",
      "(-0.7383162229895852, 3.325040830430651)\n",
      "1.2933623037205328\n",
      "\n",
      "heart block\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 112\n",
      "number overlapping genes = 6\n",
      "0.01145103211529131\n",
      "(0.24075070135110044, 1.9003079384994686)\n",
      "1.0705293199252846\n",
      "\n",
      "increased heart right ventricle size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 100\n",
      "number overlapping genes = 1\n",
      "0.5070747967656504\n",
      "(-2.6406090773596262, 1.3050880685122181)\n",
      "-0.667760504423704\n",
      "\n",
      "abnormal heart right ventricle size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 123\n",
      "number overlapping genes = 2\n",
      "0.8082280849771922\n",
      "(-1.5751274111070523, 1.2280013724918637)\n",
      "-0.17356301930759432\n",
      "\n",
      "abnormal heart left ventricle size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 152\n",
      "number overlapping genes = 4\n",
      "0.5267913299911449\n",
      "(-0.6765035225142911, 1.3217937860630413)\n",
      "0.32264513177437504\n",
      "\n",
      "increased heart left ventricle size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 119\n",
      "number overlapping genes = 3\n",
      "0.6367044469422826\n",
      "(-0.8738771781164321, 1.4287678816572722)\n",
      "0.27744535177042007\n",
      "\n",
      "decreased heart left ventricle size\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 1\n",
      "0.602456873126797\n",
      "(-1.4650670279397116, 2.52561039045942)\n",
      "0.5302716812598542\n",
      "\n",
      "abnormal cardiac muscle tissue morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 493\n",
      "number overlapping genes = 20\n",
      "0.0006683924010700141\n",
      "(0.3394395957791387, 1.262003161195772)\n",
      "0.8007213784874554\n",
      "\n",
      "cardiac muscle necrosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 32\n",
      "number overlapping genes = 1\n",
      "0.624942636445099\n",
      "(-1.496880963640767, 2.491725799187745)\n",
      "0.49742241777348895\n",
      "\n",
      "sinus bradycardia\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 2\n",
      "0.0037970318555873607\n",
      "(0.7203028403622165, 3.7414663120269624)\n",
      "2.2308845761945895\n",
      "\n",
      "abnormal heart left ventricle pressure\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 48\n",
      "number overlapping genes = 0\n",
      "0.6543403389235298\n",
      "(-3.4253049703515996, 2.1513543256381302)\n",
      "-0.6369753223567347\n",
      "\n",
      "abnormal heart right ventricle pressure\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 0\n",
      "0.919455735137897\n",
      "(-2.6605094300127567, 2.9499676424969805)\n",
      "0.14472910624211188\n",
      "\n",
      "absent heart\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.5622719262373211\n",
      "(-1.9976606113421846, 3.6747197922682)\n",
      "0.8385295904630077\n",
      "\n",
      "abnormal fetal cardiomyocyte physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 60\n",
      "number overlapping genes = 6\n",
      "5.620270885492509e-05\n",
      "(0.8975376435488012, 2.5986311679019862)\n",
      "1.7480844057253937\n",
      "\n",
      "decreased fetal cardiomyocyte proliferation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 17\n",
      "number overlapping genes = 4\n",
      "1.5138827348618648e-06\n",
      "(1.6369737217694915, 3.8889392806648164)\n",
      "2.762956501217154\n",
      "\n",
      "thick heart valve cusps\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 0\n",
      "0.9678695347335673\n",
      "(-2.745044903087506, 2.8602427048873)\n",
      "0.05759890089989694\n",
      "\n",
      "abnormal heart echocardiography feature\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 52\n",
      "number overlapping genes = 2\n",
      "0.3232461857261787\n",
      "(-0.7031167592709047, 2.1319940725823034)\n",
      "0.7144386566556995\n",
      "\n",
      "abnormal heart valve physiology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 27\n",
      "number overlapping genes = 0\n",
      "0.9662907006091268\n",
      "(-2.859831603657139, 2.7391067961273565)\n",
      "-0.060362403764891326\n",
      "\n",
      "increased cardiac stroke volume\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 0\n",
      "0.8096351339015802\n",
      "(-2.466508704864783, 3.157783536979361)\n",
      "0.34563741605728904\n",
      "\n",
      "decreased cardiac stroke volume\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 0\n",
      "0.8221740885419456\n",
      "(-3.1140040249403, 2.473305747542323)\n",
      "-0.3203491386989885\n",
      "\n",
      "delayed heart development\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 0\n",
      "0.8388873413196973\n",
      "(-2.5186375062206596, 3.1016590992041126)\n",
      "0.2915107964917265\n",
      "\n",
      "cardiac edema\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.6028952844028312\n",
      "(-2.0795962485420336, 3.582513928271891)\n",
      "0.7514588398649287\n",
      "\n",
      "abnormal pericardial cavity morphology\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 132\n",
      "number overlapping genes = 5\n",
      "0.12583937898194422\n",
      "(-0.19717412347410945, 1.6034399313157301)\n",
      "0.7031329039208103\n",
      "\n",
      "prenatal lethality prior to heart atrial septation\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 55\n",
      "number overlapping genes = 0\n",
      "0.5863866558919899\n",
      "(-3.560026682529091, 2.012978479810561)\n",
      "-0.7735241013592651\n",
      "\n",
      "abnormal cardiac muscle glycogen level\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 0\n",
      "0.8096351339015802\n",
      "(-2.466508704864783, 3.157783536979361)\n",
      "0.34563741605728904\n",
      "\n",
      "increased cardiac muscle glycogen level\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.6784171882009766\n",
      "(-2.225777350669714, 3.4201561551578146)\n",
      "0.5971894022440503\n",
      "\n",
      "abnormal heart ventricle wall thickness\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 192\n",
      "number overlapping genes = 7\n",
      "0.08691785693244913\n",
      "(-0.09653075451586857, 1.4287643427713648)\n",
      "0.6661167941277482\n",
      "\n",
      "abnormal heart left ventricle weight\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 1\n",
      "0.6903003835663676\n",
      "(-1.586689456624333, 2.3964309655248375)\n",
      "0.40487075445025233\n",
      "\n",
      "abnormal cardiac thrombosis\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 34\n",
      "number overlapping genes = 3\n",
      "0.008327384296153062\n",
      "(0.41201106351042127, 2.7922164415255395)\n",
      "1.6021137525179805\n",
      "\n",
      "abnormal circardian behavior entrainment\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 1\n",
      "0.236523929350149\n",
      "(-0.8028616241082172, 3.251481720203364)\n",
      "1.2243100480475735\n",
      "\n",
      "increased cardiac cell glucose uptake\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.6028952844028312\n",
      "(-2.0795962485420336, 3.582513928271891)\n",
      "0.7514588398649287\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "myocardial fiber disarray\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 1\n",
      "0.236523929350149\n",
      "(-0.8028616241082172, 3.251481720203364)\n",
      "1.2243100480475735\n",
      "\n",
      "abnormal response to myocardial ischemic injury\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 50\n",
      "number overlapping genes = 0\n",
      "0.6336349012557134\n",
      "(-3.465671884609451, 2.1098392264576873)\n",
      "-0.6779163290758818\n",
      "\n",
      "decreased susceptibility to myocardial ischemic injury\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 0\n",
      "0.990668350775242\n",
      "(-2.784783873831427, 2.818218847647154)\n",
      "0.016717486907863677\n",
      "\n",
      "increased susceptibility to myocardial ischemic injury\n",
      "number of genes in root node = 331\n",
      "number of genes in focal MPO term = 29\n",
      "number overlapping genes = 0\n",
      "0.926354928883276\n",
      "(-2.929656384441364, 2.665775921620256)\n",
      "-0.131940231410554\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0000266</th>\n",
       "      <td>1.111428e-07</td>\n",
       "      <td>0.755585</td>\n",
       "      <td>0.476561</td>\n",
       "      <td>1.034609</td>\n",
       "      <td>1739</td>\n",
       "      <td>abnormal heart morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000267</th>\n",
       "      <td>4.601623e-05</td>\n",
       "      <td>1.012285</td>\n",
       "      <td>0.525402</td>\n",
       "      <td>1.499168</td>\n",
       "      <td>363</td>\n",
       "      <td>abnormal heart development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000269</th>\n",
       "      <td>1.175674e-01</td>\n",
       "      <td>0.719126</td>\n",
       "      <td>-0.181450</td>\n",
       "      <td>1.619702</td>\n",
       "      <td>130</td>\n",
       "      <td>abnormal heart looping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000270</th>\n",
       "      <td>6.761476e-02</td>\n",
       "      <td>0.941893</td>\n",
       "      <td>-0.068236</td>\n",
       "      <td>1.952021</td>\n",
       "      <td>84</td>\n",
       "      <td>abnormal heart tube morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000274</th>\n",
       "      <td>1.346398e-04</td>\n",
       "      <td>0.756193</td>\n",
       "      <td>0.367983</td>\n",
       "      <td>1.144402</td>\n",
       "      <td>756</td>\n",
       "      <td>enlarged heart</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0000266  1.111428e-07  0.755585         0.476561         1.034609   \n",
       "MP:0000267  4.601623e-05  1.012285         0.525402         1.499168   \n",
       "MP:0000269  1.175674e-01  0.719126        -0.181450         1.619702   \n",
       "MP:0000270  6.761476e-02  0.941893        -0.068236         1.952021   \n",
       "MP:0000274  1.346398e-04  0.756193         0.367983         1.144402   \n",
       "\n",
       "            num_genes_in_term                  MP_description  \n",
       "MP:0000266               1739       abnormal heart morphology  \n",
       "MP:0000267                363      abnormal heart development  \n",
       "MP:0000269                130          abnormal heart looping  \n",
       "MP:0000270                 84  abnormal heart tube morphology  \n",
       "MP:0000274                756                  enlarged heart  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MP_focal_list = MP_focal_heart_list\n",
    "root_KO_heart_df=validation.MPO_enrichment_root(hier_df,MPO,mgi_df,MP_focal_list,G_int,verbose=True)\n",
    "root_KO_heart_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0000266</th>\n",
       "      <td>1.111428e-07</td>\n",
       "      <td>0.755585</td>\n",
       "      <td>0.476561</td>\n",
       "      <td>1.034609</td>\n",
       "      <td>1739</td>\n",
       "      <td>abnormal heart morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000267</th>\n",
       "      <td>4.601623e-05</td>\n",
       "      <td>1.012285</td>\n",
       "      <td>0.525402</td>\n",
       "      <td>1.499168</td>\n",
       "      <td>363</td>\n",
       "      <td>abnormal heart development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000269</th>\n",
       "      <td>1.175674e-01</td>\n",
       "      <td>0.719126</td>\n",
       "      <td>-0.181450</td>\n",
       "      <td>1.619702</td>\n",
       "      <td>130</td>\n",
       "      <td>abnormal heart looping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000270</th>\n",
       "      <td>6.761476e-02</td>\n",
       "      <td>0.941893</td>\n",
       "      <td>-0.068236</td>\n",
       "      <td>1.952021</td>\n",
       "      <td>84</td>\n",
       "      <td>abnormal heart tube morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000274</th>\n",
       "      <td>1.346398e-04</td>\n",
       "      <td>0.756193</td>\n",
       "      <td>0.367983</td>\n",
       "      <td>1.144402</td>\n",
       "      <td>756</td>\n",
       "      <td>enlarged heart</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0000266  1.111428e-07  0.755585         0.476561         1.034609   \n",
       "MP:0000267  4.601623e-05  1.012285         0.525402         1.499168   \n",
       "MP:0000269  1.175674e-01  0.719126        -0.181450         1.619702   \n",
       "MP:0000270  6.761476e-02  0.941893        -0.068236         1.952021   \n",
       "MP:0000274  1.346398e-04  0.756193         0.367983         1.144402   \n",
       "\n",
       "            num_genes_in_term                  MP_description  \n",
       "MP:0000266               1739       abnormal heart morphology  \n",
       "MP:0000267                363      abnormal heart development  \n",
       "MP:0000269                130          abnormal heart looping  \n",
       "MP:0000270                 84  abnormal heart tube morphology  \n",
       "MP:0000274                756                  enlarged heart  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_heart_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/brinrosenthal/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_with_indexer(indexer, value)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "      <th>MPO_term_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0003861</th>\n",
       "      <td>1.824245e-23</td>\n",
       "      <td>1.358205</td>\n",
       "      <td>1.091526</td>\n",
       "      <td>1.624883</td>\n",
       "      <td>1216</td>\n",
       "      <td>abnormal nervous system development</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0002272</th>\n",
       "      <td>1.856599e-17</td>\n",
       "      <td>1.701299</td>\n",
       "      <td>1.309119</td>\n",
       "      <td>2.093480</td>\n",
       "      <td>331</td>\n",
       "      <td>abnormal nervous system electrophysiology</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000778</th>\n",
       "      <td>1.246682e-11</td>\n",
       "      <td>1.675418</td>\n",
       "      <td>1.190708</td>\n",
       "      <td>2.160127</td>\n",
       "      <td>209</td>\n",
       "      <td>abnormal nervous system tract morphology</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0002925</th>\n",
       "      <td>1.177392e-10</td>\n",
       "      <td>1.113443</td>\n",
       "      <td>0.774692</td>\n",
       "      <td>1.452194</td>\n",
       "      <td>768</td>\n",
       "      <td>abnormal cardiovascular development</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0002882</th>\n",
       "      <td>1.440307e-08</td>\n",
       "      <td>0.828777</td>\n",
       "      <td>0.542217</td>\n",
       "      <td>1.115338</td>\n",
       "      <td>1517</td>\n",
       "      <td>abnormal neuron morphology</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0003861  1.824245e-23  1.358205         1.091526         1.624883   \n",
       "MP:0002272  1.856599e-17  1.701299         1.309119         2.093480   \n",
       "MP:0000778  1.246682e-11  1.675418         1.190708         2.160127   \n",
       "MP:0002925  1.177392e-10  1.113443         0.774692         1.452194   \n",
       "MP:0002882  1.440307e-08  0.828777         0.542217         1.115338   \n",
       "\n",
       "            num_genes_in_term                             MP_description  \\\n",
       "MP:0003861               1216        abnormal nervous system development   \n",
       "MP:0002272                331  abnormal nervous system electrophysiology   \n",
       "MP:0000778                209   abnormal nervous system tract morphology   \n",
       "MP:0002925                768        abnormal cardiovascular development   \n",
       "MP:0002882               1517                 abnormal neuron morphology   \n",
       "\n",
       "           MPO_term_type  \n",
       "MP:0003861         brain  \n",
       "MP:0002272         brain  \n",
       "MP:0000778         brain  \n",
       "MP:0002925         heart  \n",
       "MP:0002882         brain  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# join brain and heart results together\n",
    "root_KO_brain_df['MPO_term_type']='brain'\n",
    "root_KO_heart_df['MPO_term_type']='heart'\n",
    "root_KO_df = root_KO_brain_df.append(root_KO_heart_df)\n",
    "root_KO_df['MPO_term_type'].loc['MP:0002419']='neg_ctrl'\n",
    "root_KO_df = root_KO_df.sort_values('OR_p')\n",
    "root_KO_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "heart       157\n",
       "brain       102\n",
       "neg_ctrl      1\n",
       "Name: MPO_term_type, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_df['MPO_term_type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot top performaing brain and heart terms + negative control term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot top performing brain and heart terms + negative control term (MP:0002419')\n",
    "\n",
    "# plot 5 best and 5 worst (for terms which have at least 150 genes)\n",
    "\n",
    "heart_terms_plot = root_KO_heart_df[root_KO_heart_df['num_genes_in_term']>150]\n",
    "heart_terms_plot = heart_terms_plot.sort_values('OR_p',ascending=True).head(5).index.tolist() #+heart_terms_plot.sort_values('OR_p',ascending=True).tail(5).index.tolist()\n",
    "\n",
    "brain_terms_plot = root_KO_brain_df[root_KO_brain_df['num_genes_in_term']>150]\n",
    "brain_terms_plot = brain_terms_plot.sort_values('OR_p',ascending=True).head(5).index.tolist() #+brain_terms_plot.sort_values('OR_p',ascending=True).tail(5).index.tolist()\n",
    "\n",
    "neg_ctrl_terms_plot=['MP:0002419']\n",
    "\n",
    "terms_plot = brain_terms_plot+heart_terms_plot +neg_ctrl_terms_plot\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAewAAAF/CAYAAACL2PqNAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nOydeVxO2f/A3yG0YJQmJnv0DC1EqYYyUwwaSVkrayPGbiJlyIwlJEZTzC9LKTKZhlRjaRgzg2TsY1dKtCAxtohS9/fHo/vtaVHWMu779XpedO+553zOcu/nnM/5nHOUBEEQkJCQkJCQkKjW1KhqASQkJCQkJCQqRlLYEhISEhIS7wCSwpaQkJCQkHgHkBS2hISEhITEO4CksCUkJCQkJN4BalW1ABISEhJvk8ePH3P27Fm0tLSoWbNmVYsjISFSUFBAdnY2BgYG1K1bt9R9SWFLSEi8V5w9exYXF5eqFkNColw2bdqEiYlJqeuSwpaQkHiv0NLSAuQfxcaNG1exNG+fLVu2ADBw4MAqluTluHsX1qyB7Gxo0gRqvKaJ3dxcuHULunUDe/vXF++LcOPGDVxcXMQ2WhJJYUtISLxXFJnBGzduTNOmTatYmrfPBx98APBO5v3WLVi/HvLzQSZ7vXGrq4OGBhw9CqqqMGoUVNWMSXlTNZLTmYSEhIREtefBA/Dzg0ePQFv7zaRRsya0bAn79kFk5JtJ41WQFLaEhISERLVGEODnn+Uj7DelrIuoUQNatIBdu+DcuTeb1osiKWwJCQmJ9wh9fX309fWrWowX4vRp2L8fdHTeTno1a0KjRrB2LTx8+HbSrAySwpaQkJB4jzA1NcXU1LSqxag0ubkQHAxaWm/XEax+fbh/H7Zte3tpVoSksCUkJCTeI/Lz88nPz69qMSrNiRNyxVmv3ttP+6OP4M8/5fPn1QFJYUtISEi8R2zatIlNmzZVtRiVQhBg505o2LBq0q9VCwoL4ciRqkm/JJLClpCQkJColqSmwrVrcvN0VaGpKe80FBZWnQxFvPMKOzAwECsrq6oW47URFRWFTCbj6dOnVS3Kf56kpCT++uuvqhaDq1evIpPJOHz48FtL08vLCycnp7eW3pskMzOTHTt2VLUYEm+A5GRQUpL/qgp1dbhzB27frjoZinjnFbaExMvi5ubGmTNnqloMiVfEw8ODAwcOVLUYEm+ACxfkm5hUB65dq2oJpJ3OJN5jBEGoahEkXgNSPf43EQT5CPtVnM2Sk6M4enQJOTnXUFf/CFNTL9q0cXypuNLSoEOHl5fldVDtR9hJSUmMGzcOU1NTDAwMsLa2Zu3ataXCBQUFYW5uTqdOnZg1axY5OTniPZlMRmRkJKNHj8bIyIiePXuyatUqhef37dvHkCFDMDY2xsLCgjlz5nDv3j3xvrW1NUuWLKFv37506dKF/fv3M3z4cJYvX463tzfGxsaYmZmxcuVKLl++zPDhwzEyMqJXr17s379fjOf69eu4u7tjYWGBvr4+lpaW+Pr6UlBQUKny8PLywsPDA19fXywsLDAzM2PChAlkZWWJYe7fv4+3t7dYHi4uLhw7dky8HxgYiJOTE9OnTxfLq3PnzmzcuFEhrbCwMMzMzMjLywMgOjoae3t7OnTogJWVFb6+vjx58kShnH/55ReFOKysrAgMDATkJ9H4+fnRvXt3DAwM6NWrFz/99FO5eX1e+IsXLyKTyTh06JDCM7Nnz2bEiBEAnDp1CmdnZ4yNjTExMWHixIlkZmYC8vrMyspi5cqVWFtbA1BYWMiaNWuwsbHByMgIOzs7hfwcPnxYNF3b2tpiaGjIoEGDSE1NJSgoiG7dumFiYsL06dMVyqUkSUlJjBgxgo4dO/L5559zpAyPlr179+Lo6IihoSE2Njb4+vqSm5sr5tHRUfGjc+fOHQwMDPjtt98A+Z7E06ZNw8TEBDMzM8aMGUNSUlK5Ml2/fp2ZM2fSrVs3jIyMym0z//d//4eZmRkmJiZ4enryoJj7rEwmIyYmhhEjRmBkZISNjQ179+7ljz/+oE+fPnTo0AFnZ2fS0tLEZyrTVocPH87atWuxsrKic+fOjBo1ipSUFACGDx/OiRMn2LZtG7LXvVflf5SOHTvSsWPHqhajQh4/hpwcqFPn5Z5PTo7iwIGZ5ORkAgI5OZkcODCT5OSoF45LVRWuXn05OV4n1Vph5+bm4urqSr169di8eTPbt2/H1taWZcuWKZgys7KyOHjwICEhIaxevZoTJ04wbdo0hbiWLl1K//79iYmJoXfv3gQEBIgfyj179jBu3DgsLCzYunUry5Yt49ixY7i6ulJYzNMgPDwcDw8P1q9fL65jXL9+PY0bNyY2NpZhw4YRGBjI2LFjGTFiBFFRUTRt2hRPT09xFPDVV19x9+5dQkJCiIuLY8yYMYSEhLBnz55Kl8uuXbu4e/cu4eHhLF26lGPHjuHv7w/IRxtubm5cuXKFoKAgtmzZgoWFBSNHjuTo0aNiHCdOnKBevXrExMTg5uZGnz59iI2NVUgnNjYWOzs7ateuTWhoKHPmzMHR0ZGYmBi8vb359ddfmT59eqXlDg8PJy4ujhUrVvDbb78xfPhw5s2bV+7c7fPCf/zxx+jr6xMdHS2Gf/z4MXFxcTg6OlJQUCB29GJjYwkLC+PGjRt4eXkB8gMQtLS0cHV1FQ9DWL58OREREcyePZsdO3YwZswY/Pz8CAoKUpBr8eLFLFq0iMjISG7fvs3QoUNJTEwkLCyMefPmERcXV6rjUsSDBw8YNWoUqqqqREZGMnfu3DI7j9OmTWPgwIHs2LGDhQsXEh8fz8SJEwFwdHTk3LlzotIC2LlzJ2pqanz22Wc8evSIYcOGUVhYyMaNGwkPD6dZs2YMGTKE1NTUUjLl5OTg5OREWloagYGBbNmyhZYtWzJq1CiF9+zMmTPs37+fkJAQVq1axfHjx5kyZYpCXIsWLcLFxYVff/2VNm3a4OHhwcqVK1m8eDEbNmzg2rVrLFu2DKh8Wz158iRHjx5lzZo1rF69mrS0NL777jtArtCNjIzo06cP8fHxpfJ2//59MjIyFH43btwos27eF94Vhf30aflz19u3D6zwt2/fdJ4+zS0RZy779k2v8NmS1KwJz+mDvzWqtUk8NzeXESNG4OzsjLq6OgBTpkxh7dq1JCUlYWhoCICysjIrVqygUaNGAMydOxdXV1dSUlLQ1dUFoH///tjb2wPg7u7Opk2bOHnyJF26dGHNmjV8+umnopJv3bo1y5YtY8CAARw4cIDu3bsD0K1bN/H/Rejq6oof0tGjR7Ny5UpsbW3p2bMnAC4uLowfP57s7Gzq16+Pvb09vXr1QufZlj0jR45k3bp1JCUl0bt370qVi5qaGvPnz0dZWRldXV3s7OzEj9WhQ4f4559/OHToEBoaGgBMmjSJ06dPK3Q0AKZOnUrDZ+slHB0dcXJy4sqVK7Rs2ZLLly9z9uxZ5s+fjyAIrF27lqFDhzJy5EgAWrZsSWFhIVOmTCE5OZk2bdpUKHdGRgaqqqro6Oigra3NsGHDaN26tVhHLxp+4MCB+Pn58e2336Kqqsru3bsB6NWrFzk5Ody9exctLS10dHRo1qwZ/v7+3Lp1CwANDQ1q1KiBqqoqGhoaPHz4kA0bNuDn5yeOuJs1a0Z2djZr1qxh7NixolyTJk0SP3g9e/YkPDwcHx8fVFVV0dXVJTg4uNzR7I4dO3j06BG+vr40aNAAPT09vLy8mDp1qhgmKCiIQYMG4ezsDEDz5s3x8fFh0KBBXLhwgc6dO9OyZUtiY2P5+uuvAcXOVUxMDPfu3WP58uUoKysD8O2333L06FF++uknZs+erSBTbGwst2/fJjIykg8//BCAhQsXcubMGdatW8cPP/wghvX390f72d6Q3377LWPGjOHSpUu0bdsWgH79+tGrVy8ABg8ezF9//YW7u7tYXr179+bPP/8EKt9W8/Pz8fX1FdvqsGHD+P777wH5QRa1atWibt26ZZ5wFBYWxsqVK8usi/eVR48eAaBaXSaH3xCFhXkvdP1doForbA0NDZydndmxYwfnz58nLS2NixcvAiiMfJs3by4qa4AOzyYaLl26JH7cW7VqJd5XUlJCXV1d3DwgMTFR4YMJYGBggKqqKomJiaKSbtGiRSkZi8dbdOB48+bNS13Ly8ujbt26uLi4EBcXR0hICGlpaSQmJnLz5s1Km8RBfspO0YcYUMjL+fPnAbCxsVF4Jj8/X+wkADRo0ED8AAJ06tSJVq1a8euvvzJ58mRiYmKQyWTo6+tz69Ytbt26Vep81qIPamJiYqUUtouLC7///jvdu3enffv2dO3alS+++EKh7l4kvJ2dHb6+vuzevZv+/fuzbds2bG1tUVFRQUVFhTFjxrBw4UICAwMxNzfHysqKvn37lplWSkoKeXl5eHp6MmvWLPF6QUEBT548ERU9lK5zLS0thY9fnTp1xGmEkiQlJdGsWTMaNGggXjM2NlYIc+HCBc6cOcO2MrZYSklJoV27djg6OvLzzz8zbdo00tLS+Oeff/j2228BeRvIycmhS5cuCs8+efJEVLbFSUxMpEWLFqKyBvk7YmJiQkJCgnitefPmCs8XyZ2UlCQq7OJlo6KiAii+N3Xq1HnhttqwYUOFtlqvXr1Kb/wxcuRIHBwcFK4VHWH4vhL57FSLUaNGVa0gFaCsLJ/HFoTSI+2+fbdU+HxERJdn5nBF1NV1KvV8cQoK4NmnvEqp1go7OzubwYMHo6GhgY2NDZaWlhgaGpZaxlXyKLIi5Ve7dm3xWvH/F1GRs4ogCArP1S2jxmrVKl2ENcrZP+/Ro0e4uLiQm5tLnz59cHBwEOf1XoSy8lJEYWEhKioqCqbismQtKy+Ojo5s2bKFSZMmsX37doYPH/5cOYrK73nyFF+e1rJlS3bv3s2RI0dISEhg3759hISEsGjRItH6UZyKwterV4/PP/+cmJgYLCws+Pvvv4mIiBCfnzFjBs7Ozuzbt49Dhw7h4+PDxo0b2bx5c6n8F+Vl+fLlovIpjoaGhmhOLlnn5dV3ZSne+QJ5HY4ePbrM84o1NTUBucXohx9+4Pjx4xw6dAiZTEb79u3F55s3b86aNWtKPV9WvZdHYWGhQt2WzHfRe1Y8/yXzAnLlX178lWmrz2tfFVG/fn3qV+UiXomXpk4d+frrJ09eTlmamnpx4MBMBbN4rVoqmJp6vXBcjx7JDwSpaqr1HPb27du5c+cOERERTJgwgR49enD37l1AUdmmpaUpOJkdP34cJSWlMj+8ZSGTyRTmzABOnz5Nbm5upUaOlSU+Pp7z588TFhbG1KlTsbW1RU1NjduvcYGfTCYjNzeX3NxcWrRoIf42btxY4VpVe3t7MjIyiIiIICsrCzs7OwAaNWpEo0aNFJyBANEHoKiMlJWVFerh3r17Yn2BfL7/t99+o2vXrnh4eBAbG4u5uTkxMTFlylOZ8AMHDuTIkSNs3bqVVq1aiabXlJQU5s6di6amJk5OTgQEBBAcHMyFCxfEkV1xRdK6dWuUlZW5du2aQrkdOXKEH3/88ZWVchHt2rXj6tWr/Pvvv+K1kkvL9PT0SE1NVZDj/v37LFmyhDt37gCgra3NJ598wu7du9m+fbuCE5qenh7Xr19HXV1dfL558+YEBARw8ODBUjLJZDKuXr3KzZs3xWuFhYUcP35c4R1KS0tTcMQ8efIkwEsfJPEqbbU45XUIJN5tlJSgbVu549nL0KaNI5aWS1FX1wGUUFfXwdJy6Ut7iUsKuwIaN27M48eP2blzJ5mZmcTHx+Pu7g6gYHLMy8tj6tSpnD9/noSEBBYuXEjfvn1p1qxZpdIZO3Ysf/75J/7+/ly+fJmEhARmzpyJgYEB5ubmrzU/IJ8zzMzM5NixY0yYMIH8/PxyTagvSrdu3dDX12f69OkcOnSItLQ0VqxYQXh4eIWdD21tbbp27cqyZcuwsrISR3MgX7O8efNmwsLCuHr1Knv27MHHx4cePXqIZlBjY2MiIyM5e/YsiYmJeHp6KoyU7ty5w4IFC9i7dy+ZmZkcOHCAc+fOlTIJv0j4Ll268NFHH7F69WoFpdWwYUN27NjB3LlzSUlJITU1la1bt1K/fn2xHNTU1Lhy5QpZWVnUq1ePoUOHEhAQQHR0NOnp6cTGxuLj4yPOd78Oikz606dP5+LFixw5coTFixcrhBk3bhy///47AQEBpKamcuTIEWbMmEF2djZNmzYVwxVZRDIzM+nXr594vV+/fmhoaDB58mROnjzJ5cuXmT17Nrt370ZPT6+UTP369UNTU5MpU6Zw8uRJLl26hLe3N6mpqaLPAsh9Sjw9PUlKSuLQoUPMnz+fXr16KUwBvQiv0laLo6amRkZGhrgCQOK/g0z2aqdltWnjiJPTEdzcMnByOvLSyhrk+4pXNdXaJN67d2/OnTuHn58fOTk56OjoMGjQIPbu3cupU6cYNmwYIB+1GBgYMHLkSARBwNbWVvQGrgw9e/YkICCAoKAg1q1bxwcffECPHj1wd3cv0+T9shgZGTFr1ixCQ0MJDAxEW1sbW1tbmjRpwqlTp15LGjVr1iQ4OBg/Pz/c3d15+PAhrVu3xt/fX3Smeh6Ojo7s37+/1LKhUaNGUbduXUJDQ/Hz80NLS4sBAwYwYcIEMcx3333HvHnzcHJyQkNDgy+//JKHxd62yZMnU1hYyMKFC8nOzkZLS4shQ4Ywfvz4MmWpTHglJSUcHBxYtWqVglldQ0ODdevW8f333zN48GAKCgowMjIiJCRENJG6urri6+tLfHw8Bw8eZNasWWhqahIYGEhWVhba2tq4ubmVK9/LoKqqyoYNG5g/fz5Dhw6lYcOGTJ48WWHevGfPnvzwww+sXr2aNWvWUL9+fbp3746Hh4fCaLJHjx7MmzcPc3Nz0WkL5HO8RSsIxo4dy9OnT5HJZKxduxYDA4NSMqmrq7Np0yZ8fX0ZM2YMhYWFGBoasmHDBoyMjMRwH374IQYGBri4uFCjRg3s7OyYMWPGS5fFq7bVIoYNG8asWbOwtbVl9+7dZc7TS7yb6OmVP4/9trh/X35SWLFXrMpQEqRdByTecebOncvt27dLLY+SeH0EBgbyyy+/KOwp8K6SkZEhrhEvbrF4Xzh79ixAmZ236oYgwLx58q1Bq+oAkCtXYPRoKLFA6I1QUdus1iNsCYnnkZCQQEpKCtHR0QQHB1e1OBIS7wTvgqIuQkkJbG3hxx+rRmHn58u91UsskKkyJIUt8c6ydetW/vjjD8aOHauwvlxCQqJ8ihwHiy8trM506CA3R9+7B29b5GvX5B0GNbW3m255SApb4p1l+fLlVS3Ce8PkyZOZPHlyVYsh8RooWt9f3ddhF1GnDowdC4sWyU/OKrGK941x5w40agTPFstUC6q1l7iEhISEhMTHH0PPnpCR8XbSe/oU7t6VdxSqw4YpRUgKW0JCQkKi2jNgADRv/uaPuSwokJ/M5egoXwdenZAUtoSEhIREtUdVFb7+GjQ135zSfvpU7hXeuzcU29qg2iApbAkJCQmJd4IPPgAvL9DRgdRUuRf36+L+ffnI2sEBnJyqbt3385CcziQkJCTeIywsLKpahFeiQQOYNQt27ICYGLkjmqbmyyvYggLIzJTvW+7pCS+50+5bQVLYEhISEu8RMpmsqkV4ZWrXlo+EjY0hOFg+Mq5dGz78sPJe5Lm5kJ0t/3/37jBwYPVZvlUeksKWkJCQeI8oOiq2vGNt3yVatoT58yE5GfbuhWfnEQFy5auiAkXHADx9Kt+X/PFj+WhcTU3uWGZhIR+hvwtICltCQkLiPWL79u3Au7MOuyKKTvVq2xacnSE9XT7ivngRsrIgL0+utOvWhfbt5eF0dKBZM/kuZu8SksKWkJCQkPhPUL++fA5aXx/69KlqaV4/kpe4hISEhITEO4A0wpaQkJCQkHjGjRtw4AAcPgw5OfITw+rVg86dwcpKfi52VS35khS2hISEhMR7z+XLsGULnD8v9zTX1JTvJQ7y9d5798Jvv8nP6B44UP7v20ZS2BISEhLvEVZWVlUtQrXj6FH5EZ6qqtCiRekRdK1a0LSpfLR9/Tr4+MCYMWBp+XbllBS2hISExHtE69atq1qEasXp07ByJTRuLF8G9jyUlOQjbzU1WLNGfpJYly5vR054TU5ngYGB/6leW1RUFDKZjKdPn1a1KP95kpKS+Ouvv6pajFK87jadn5/P+vXrX1t85ZGRkYFMJiMhIaFaxP+m5ZF4cW7cuMGNGzeqWoxqwd27sGqVfMOVipR1cerWlc9lr1kDN2++OflKInmJS1Qpbm5unDlzpqrFeONER0ezZMmSqhbjlWnSpAnx8fGYmJhUtSgSL0lcXBxxcXFVLUa14NAh+Trtl9nhTEVFbiKPj3/9cpWHpLAlqhRBEKpaBIkXoGbNmmhpaVG7du2qFkVC4pV4+hTi4kBL6+Xj+PBD2L0bnjx5fXI9j0op7KSkJMaNG4epqSkGBgZYW1uzdu3aUuGCgoIwNzenU6dOzJo1i5ycHPGeTCYjMjKS0aNHY2RkRM+ePVm1apXC8/v27WPIkCEYGxtjYWHBnDlzuHfvnnjf2tqaJUuW0LdvX7p06cL+/fsZPnw4y5cvx9vbG2NjY8zMzFi5ciWXL19m+PDhGBkZ0atXL/bv3y/Gc/36ddzd3bGwsEBfXx9LS0t8fX0pKCioVKF5eXnh4eGBr68vFhYWmJmZMWHCBLKyssQw9+/fx9vbWywPFxcXjh07Jt4PDAzEycmJ6dOni+XVuXNnNm7cqJBWWFgYZmZm5OXlAfKRmr29PR06dMDKygpfX1+eFGstMpmMX375RSEOKysrAgMDASgoKMDPz4/u3btjYGBAr169+Omnn8rN6/PCX7x4EZlMxqFDhxSemT17NiNGjADg1KlTODs7Y2xsjImJCRMnTiQzMxOQ12dWVhYrV67E2toagMLCQtasWYONjQ1GRkbY2dkp5Ofw4cPIZDIOHz6Mra0thoaGDBo0iNTUVIKCgujWrRsmJiZMnz5doVxKUlH9lCQvL4+lS5diaWlJx44dGTBgAL///rtCmPPnz+Pq6iq231mzZnH//n2ioqKYM2eOWD9RUVFERUVhbW3N4sWLMTExwdXVFYDk5GQmTJiAubk5xsbGuLm5kZSUJKbh5eWFu7s7CxcupFOnTpibm7No0SKxfRRx+vRphgwZIr6vRWX4qnVW0sRdUFBAaGgoffr0wdDQEBsbG9asWUNhYWG5ZVlRG/7333/5+uuvMTExwczMjGXLljFixAgCAwP5999/MTAwYMuWLQpx/vjjj/To0UPqAEpUmosX5Sd0qaqWfT85OYqIiC6sXduUiIguJCdHlQpTp458q9OzZ9+wsM+oUGHn5ubi6upKvXr12Lx5M9u3b8fW1pZly5YpmDKzsrI4ePAgISEhrF69mhMnTjBt2jSFuJYuXUr//v2JiYmhd+/eBAQEcOTZ5q979uxh3LhxWFhYsHXrVpYtW8axY8dwdXVVePnDw8Px8PBg/fr1mJqaArB+/XoaN25MbGwsw4YNIzAwkLFjxzJixAiioqJo2rQpnp6e4sv81VdfcffuXUJCQoiLi2PMmDGEhISwZ8+eShfcrl27uHv3LuHh4SxdupRjx47h7+8PyEeNbm5uXLlyhaCgILZs2YKFhQUjR47k6NGjYhwnTpygXr16xMTE4ObmRp8+fYiNjVVIJzY2Fjs7O2rXrk1oaChz5szB0dGRmJgYvL29+fXXX5k+fXql5Q4PDycuLo4VK1bw22+/MXz4cObNm8fhw4dfOPzHH3+Mvr4+0dHRYvjHjx8TFxeHo6MjBQUFYkcvNjaWsLAwbty4gZeXFwBbtmxBS0sLV1dX8QO8fPlyIiIimD17Njt27GDMmDH4+fkRFBSkINfixYtZtGgRkZGR3L59m6FDh5KYmEhYWBjz5s0jLi6uVMeliMrWT3FmzpxJfHw8S5cuJSYmBgcHB77++muxvtLT03FxcaFhw4Zs3ryZ//u//+P06dN4eXlha2sr5jk+Ph5bW1sAMjMzSU9PJyoqCi8vLzIzMxk6dCiCIBASEsKmTZuoUaMGLi4uXCt2APDu3bu5fv06mzdvxsfHh9jYWObPn68g78aNGxk3bhw7d+6ke/fueHt7c/ny5Veus5IsWbJEfN+2b9/OhAkTCAoKwtfXt8zwFbXhwsJCxo0bR2pqKmvXriUkJIRTp06J3wkNDQ2sra0V5AfEOlGqjmciSlRLsrPlJu2ySE6O4sCBmeTkZAICOTmZHDgws0ylXbOm3HP8rSBUwO3bt4XVq1cLDx48EK89efJE0NPTE7Zs2SIIgiAEBAQI+vr6QnZ2thgmPj5e0NPTE5KTkwVBEAQ9PT1hwYIF4v3CwkLB2NhYCAoKEgRBEAYOHCiMGzdOIe0zZ84Ienp6wl9//SUIgiB89tlnpcIMGzZM6Nevn/j3gwcPBD09PWH58uXitb179wp6enpCVlaWkJubKwQHBwsZGRkK8XTr1k344YcfBEEQhK1btwp6enpCfn5+mWXi6ekpdOnSRcjLyxOvzZ8/X/j8888FQRCEgwcPCnp6esLt27cVnnNzcxPGjx8vlpmenp7w77//ivePHz8u6OnpCampqYIgCEJKSoqgp6cnnD17VigsLBQ++eQThTIUBEGIi4sT9PT0hEuXLgmCIC/nyMhIhTCWlpZCQECAIAiCsHDhQqFv377CjRs3xPsHDx5UqLviVBR+06ZNQseOHYWHDx8KgiAIMTExQqdOnYRHjx4Jd+/eFWQymbBx40ahoKBAEARBSEtLE06cOFGmbDk5OYKBgYGwa9cuBRnWrl0rmJqaCgUFBcLff/8t6OnpCXv27BHvL1q0SGjfvr0ogyAIgoODg+Dt7V1mnipbP5aWloIgCMKVK1cEPT094fTp0wrhv/vuO7HtLVu2TLC0tBSePHki3j958qSwYsUKoaCgQIiMjBT09PTEe0Vt7MyZM+K1pUuXCp988omQm5srXnv06JHwySefCL6+voIgyNueubm58OjRIzFMRESE0L59e+HevXtCenq6oKenJ2zYsEG8f/fuXUFPT0+IjY0VBOHV6qwo/oMHD+HlZikAACAASURBVAoPHjwQ9PX1hZCQEIVyCQ4OFvT19YX79+8rhK9MGz506JCgp6cnJCUlifdv3bolGBoaiu3kr7/+EmQymZCWliYIgiAcO3ZM+Pjjj4Vr164JJSkqk+K/o0ePCnp6ekJ6enqp8O8D69evF9avX1/VYrwVBgwYUO7v008HCPr6A4TOnUv/mjZtKXz00Uelfk2btiwV1sBggGBpWX46AwYMqLS8Re9LeW2zwmVdGhoaODs7s2PHDs6fP09aWhoXL14EUBj5Nm/eXOH0lw4dOgBw6dIldHV1AWjVqpV4X0lJCXV1dfKfnUCemJjI1KlTFdI2MDBAVVWVxMREunfvDkCLFi1KyVg83rp164rylLyWl5dH3bp1cXFxIS4ujpCQENLS0khMTOTmzZuVNokDNG3aFOViO8cXz8v58+cBsLGxUXgmPz8fHR0d8e8GDRrQsGFD8e9OnTrRqlUrfv31VyZPnkxMTAwymQx9fX1u3brFrVu3Sjn7FFkZEhMTadOmTYVyu7i48Pvvv9O9e3fat29P165d+eKLL8o9uaei8HZ2dvj6+rJ792769+/Ptm3bsLW1RUVFBRUVFcaMGcPChQsJDAzE3NwcKysr+vbtW2ZaKSkp5OXl4enpyaxZs8TrBQUFPHnyRDxlCErXuZaWFqrFbFt16tQpZSYuorL1UzJ8kcm4ePgikpKSaN++vcLcbseOHenYsWOZMhTRsmVL8f+JiYkYGBiI7RVARUUFAwMDEhMTxWsGBgaoFHNp7dixI0+fPuXy5ctivRSPt0GDBoB8JA2vr85SUlLIz88vs03m5+eTkpKi0K5u375dYRvOyspCTU2Ntm3bivc1NTUV6tvS0hJtbW1iYmKYNGkS27Ztw8LCgiZNmpSSMSwsjJUrV5a6/j5Tst2/r9R4jn25sLDsb0d5158X1+ukQoWdnZ3N4MGD0dDQwMbGBktLSwwNDUstealZ4hDSIuVX/ANWlqOKUMGckyAICs8V/5iJmahVOhs1yinBR48e4eLiQm5uLn369MHBwYEOHTrg7Oz8XDlK8jynm8LCQlRUVEqZ7UrKWlZeHB0d2bJlC5MmTWL79u0MHz78uXIUld/z5Cm+PK1ly5bs3r2bI0eOkJCQwL59+wgJCWHRokXY29uXerai8PXq1ePzzz8nJiYGCwsL/v77byIiIsTnZ8yYgbOzM/v27ePQoUP4+PiwceNGNm/eXCr/RXlZvny5wge7CA0NDVJTU4HSdV5efZdFZeunpFwbNmygfv36ZcaprKz8UssAy2oDJSksLFSo35IyFnWci5dByfexOK9SZ5WhMm2yvPA1a9as8JtQo0YNHBwciI2NZezYsezatavUlEARI0eOxMHBQeHajRs3cHFxqZRs/0WaNWtW1SK8NUr6OhTn2DH5+utifVuRiIguz8zhiqir69C3r2KcV67Al1++nU1UKvzKbd++nTt37hAREcGECRPo0aMHd+/eBRSVbVpamoKT2fHjx1FSUirzw1sWMpms1Pzh6dOnyc3NrdTIsbLEx8dz/vx5wsLCmDp1Kra2tqipqXH79u3XloZMJiM3N5fc3FxatGgh/jZu3MiOHTue+6y9vT0ZGRlERESQlZWFnZ0dID+7tlGjRqUco4rm9orKSFlZWaEe7t27J9YXyOf7f/vtN7p27YqHhwexsbGYm5sTExNTpjyVCT9w4ECOHDnC1q1badWqlTiqTElJYe7cuWhqauLk5ERAQADBwcFcuHBBHLUWn3Ns3bo1ysrKXLt2TaHcjhw5wo8//vhCSvl5vGj96D3bg/DmzZsK4Xft2iUqOl1dXS5cuKCgtOPj4+natSv379+vtFxnz54lNzdXvPbo0SPOnj2r8B6VTOfkyZPUqVNHtGRVhlepsyJ0dXVRVlYus03Wrl1bwcoFlWvDH3/8MY8ePSI5OVm8f/fuXa5evarwzIABA0hLSyM0NJQaNWrQo0ePMvNZv359mjZtqvBr3Lhxpcvpv0h6ejrp6elVLUaV07491K4t33a0JKamXtSqpbgwu1YtFUxNFX05Cgrkc9hGRm9S0v9R4RewcePGPH78mJ07d5KZmUl8fDzu7u4ACibHvLw8pk6dyvnz50lISGDhwoX07du30r25sWPH8ueff+Lv78/ly5dJSEhg5syZGBgYYG5u/pLZKzs/IHfmyszM5NixY0yYMIH8/PxyTagvSrdu3dDX12f69OkcOnSItLQ0VqxYQXh4eIWdD21tbbp27cqyZcuwsrJCs9jJ6m5ubmzevJmwsDCuXr3Knj178PHxoUePHqLJ0NjYmMjISM6ePUtiYiKenp4KI7I7d+6wYMEC9u7dS2ZmJgcOHODcuXMYGxuXKU9lwnfp0oWPPvqI1atX4+joKF5v2LAhO3bsYO7cuaSkpJCamsrWrVupX7++WA5qampcuXKFrKws6tWrx9ChQwkICCA6Opr09HRiY2Px8fFBQ0PjtSnsF62fNm3aYG1tzbx58/j9999JT09nw4YNBAQE0LRpU0A+dVDkeZ6cnMzx48dZtGgRnTt3pn79+qg9W+h5+vRpHj58WKZczs7OPHnyBHd3d1FBuru7k5+fz5AhQ8Rw169f59tvvyUlJYU9e/YQEBCAi4uLmEZleJU6K0JdXR0nJydWrlxJdHQ0V69eJSoqih9//BEnJyfU1dVLpVtRGzYzM6Njx454enryzz//cPHiRWbMmEFubq5C565Zs2aYmZmxatUq+vbtS506dSqd9/edvXv3snfv3qoWo8pRVYXPPpOfmV2SNm0csbRcirq6DqCEuroOlpZLadPGUSFcVhZYWMCzWac3ToUm8d69e3Pu3Dn8/PzIyclBR0eHQYMGsXfvXk6dOsWwYcMAaNeuHQYGBowcORJBEBQ8YytDz549CQgIICgoiHXr1vHBBx/Qo0cP3N3dyzRTvixGRkbMmjWL0NBQAgMD0dbWxtbWliZNmnDq1KnXkkbNmjUJDg7Gz88Pd3d3Hj58SOvWrfH39xeXLz0PR0dH9u/fr/AhBfmB83Xr1iU0NBQ/Pz+0tLQYMGAAEyZMEMN89913zJs3DycnJzQ0NPjyyy8VFMTkyZMpLCxk4cKFZGdno6WlxZAhQxg/fnyZslQmvJKSEg4ODqxatUrBrK6hocG6dev4/vvvGTx4MAUFBRgZGRESEiKall1dXfH19SU+Pp6DBw8ya9YsNDU1CQwMJCsrC21tbdzc3MqV72V4mfpZsWIF/v7+zJs3j7t379K0aVPmzJkjTqV8+OGHhISEsHz5chwdHUWzc5H3c7du3ejUqRPOzs58/fXXCr4LRejo6BAeHo6fnx/Ozs7UqFEDU1NTfv75Z4W5dUNDQ+rUqcPAgQNRV1dn2LBhCm2gMrxsnZW0Fnh5eaGhoUFAQAA3b95ER0eHiRMnMnr06DLTrUwbDggIYP78+YwaNYo6derg7OxMSkqKgs8IyN+Tv//+mwEDBrxQ3iUkirCykh/okZ8PJZoXbdo4llLQxXn6VL7+uhKf9NeGklDRhJGERCWYO3cut2/fLrW2XuL14uXlxdWrVxXmnF+W6lhn//77LydPnsTS0lKcA8/Ly8PMzIxvv/2W/v37i2HXrFnDzp07y/RFeB4ZGRnY2Niwd+9e0ULyPhEaGgrIO08SsH07REbKD/14juuHAoWFkJoK/frJT+56XasJK2qb0uEfEq9EQkICKSkpREdHExwcXNXiSFSC6lxntWrVYvr06QwZMgRnZ2fy8/MJDg6mdu3aoqPr8ePHuXLlCsHBwQqrCSQkXoYvvoAHD2DnTtDRke8T/jzy8iA9HWxsYMCAt3s2tqSwJV6JrVu38scffzB27FhxeY5E9aY611n9+vUJCgrC39+fyMhIlJSU6Ny5Mxs2bEBDQwOAv/76iw0bNmBvb1/mygYJiRdBSQmGDgVtbfjlF/m8dIMG8l+RMhYE+a5od+7Idzdzdobevd+usgbJJC4hIfGe8b6bxItO6nrfveXL4skTOHUKduyAtLT/ra8uLJSPvm1twdj4xU72ehEkk7iEhISEhIikqMun6HxrU1O5mTw3Vz66VlGB+vXf/oi6JJLClpCQkHiPuHz5MiDf90CibJSU5Aq6nH2SqgxJYUtISEi8RxSdXCgp7HcP6TxsCQkJCQmJdwBJYUtISEhISLwDSApbQkJCQkLiHUBS2BISEhISEu8AktOZhISExHtEeefRS1R/JIUtISEh8R7RqFGjqhZB4iWRTOISEhIS7xGJiYkkJiZWtRgSL4E0wpaQkJB4jzh06BAAMpmsiiWReFGkEbaEhISEhMQ7gDTClpCQkJCQeMbjx3DpEuTkyA//aNgQdHUrf1b2m0RS2BISEhIS7z23bsGff8Lvv8vPvC5Ow4byk7o++QRUVatGPpAUtoSEhITEe86lS/D99/LjNRs3htq1Fe8/fAgbN8K+ffD11/DsaPa3zhuZww4MDMTKyupNRF0lREVFIZPJePr0aVWL8p8nKSmJv/766z+ZriAIbNu2jdu3b7/RdF4X1tbWrFix4pXiGD58ODNmzHhNEkm8DhwcHHBwcKhqMaoNaWng5wd160Lz5qWVNYCaGrRqBTdvwvLlcnN5VSA5nUlUK9zc3Dhz5sx/Mt2///4bLy8vcnNz32g6EhLPo0GDBjRo0KCqxagWCAKsXw/KylCZImnSBDIz4bff3rxsZSEpbIlqhSAI/9l0qypvEhLFOXv2LGfPnq1qMaoFaWmQmgol95JJTo4iIqILa9c2JSKiC8nJUeK9Jk1gzx65+fxt81IKOykpiXHjxmFqaoqBgQHW1tasXbu2VLigoCDMzc3p1KkTs2bNIqeYHUEmkxEZGcno0aMxMjKiZ8+erFq1SuH5ffv2MWTIEIyNjbGwsGDOnDncu3dPvG9tbc2SJUvo27cvXbp0Yf/+/QwfPpzly5fj7e2NsbExZmZmrFy5ksuXLzN8+HCMjIzo1auXeCYswPXr13F3d8fCwgJ9fX0sLS3x9fWloKCgUuXh5eWFh4cHvr6+WFhYYGZmxoQJE8jKyhLD3L9/H29vb7E8XFxcOHbsmHg/MDAQJycnpk+fLpZX586d2bhxo0JaYWFhmJmZkffMKyI6Ohp7e3s6dOiAlZUVvr6+PCnWkmQyGb/88otCHFZWVgQGBgJQUFCAn58f3bt3x8DAgF69evHTTz+Vm9fnhb948SIymUxc51nE7NmzGTFiBACnTp3C2dkZY2NjTExMmDhxIpmZmYC8PrOysli5ciXW1tYAFBYWsmbNGmxsbDAyMsLOzk4hP4cPH0Ymk3H48GFsbW0xNDRk0KBBpKamEhQURLdu3TAxMWH69OkK5VKcstItq20V1aGlpSX6+vpYWFjwzTff8OjRIzGu9PR0Jk+ejImJCV26dGHy5MncuHGDw4cPM3r0aABsbGzE8i+JTCYjJiaGESNGYGRkhI2NDXv37uWPP/6gT58+dOjQAWdnZ9LS0sRnrl+/zsyZM+nWrRtGRkYVtq1vvvmGqKgorKys+OWXX7CyssLY2Jjx48dz48YNBXlu377NlClTxHdp4cKFClNDJ0+eZNSoUZiYmGBqasq0adO4fv16mXkDSE5OZsKECZibm2NsbIybmxtJSUkKYUJDQ7G2tsbIyIjRo0cr1MvEiRNxcXFRCH/9+nXatWvH0aNHy01X4n8cO3ZMoX28z+zfD7VqgZLS/64lJ0dx4MBMcnIyAYGcnEwOHJgpKu06deSe5FVgCHxxp7Pc3FxcXV0xNzdn8+bN1KxZky1btrBs2TLMzc0xNDQEICsri4MHDxISEsLDhw+ZM2cO06ZNY926dWJcS5cuxdvbm7lz5xIVFUVAQACmpqZ06dKFPXv2MHnyZL766isWL17M9evXWbBgAa6urvzyyy/UqCHva4SHh7Nq1SoaNWpE69atWbt2LevXr2f8+PHExsYSHR1NYGAg0dHReHp60qpVKxYvXoynpycJCQkoKSnx1VdfoampSUhICOrq6vzxxx8sWrSIDh060Lt370qVy65du7CzsyM8PJyMjAw8PDzw9/dn8eLFCIKAm5sbtWvXJigoiPr167Nz505GjhxJaGgopqamAJw4cUL8YOfn51OzZk1iY2MZPny4mE5sbCx2dnbUrl2b0NBQli1bhoeHB927d+fSpUvMmzeP9PR0Vq5cWSm5w8PDiYuLY8WKFWhra/Pnn38yb948dHV1MTMze+Hw+vr6REdHY2FhAcDjx4+Ji4vD29ubgoICxo0bx5AhQ/D19eX+/fvMnTsXLy8vNm7cyJYtW+jXrx92dna4ubkBsHz5cnbu3Im3tzdt27blxIkT+Pj4cPv2bb766itRrsWLF7No0SLq1KnDxIkTGTp0KJ988glhYWFcvHiRmTNnYmxszLBhw0rlqax0i/JavG25u7tz48YNVq5ciaamJidOnOCbb75BV1eXL7/8kgcPHuDi4oKuri7r169HWVmZBQsW8NVXXxEZGYm/vz/Tpk3jl19+QVdXt9w6WbRoEfPnz2fBggUsWrQIDw8PWrZsyeLFi1FSUmLq1KksW7aMgIAAcnJycHJyonHjxgQGBqKmpkZYWBijRo0iIiJCfB9Ltq1//vmHf//9l9DQUL7//nvq1KnDd999h6urKzExMSgrKwNy/w1PT088PDw4fPgws2fPpm3btgwZMoTTp08zfPhwHBwcmD17Ng8ePMDX1xcXFxdiY2NRV1dXyFdmZiZDhw7F1NSUkJAQAH744QdcXFyIiYnho48+YtOmTXz//fd4e3tjYmJCXFwcAQEBNGnSBICBAwcyfvx40tPTadasGQDbtm2jadOmmJiYlCrL+/fvc//+fYVrJTslEu8+AwcOfKnnrlyBp0/hn3/+dy0r6ziFhYpu4k+f5rJv33QuXpQPTh4/hhkzSo/MK8OWLVteSlZ4iRF2bm4uI0aM4LvvvkNXV5eWLVsyZcoUAIWesrKyMitWrKB9+/aYmpoyd+5cDhw4QEpKihimf//+2Nvb06pVK9zd3VFTU+PkyZMArFmzhk8//ZRp06bRunVrunbtyrJlyzh79iwHDhwQ4+jWrRvdu3dHX18fFRUVAHR1dZk4cSLNmjUTRzW2trb07NmTNm3a4OLiwr///kt2djaPHz/G3t6eBQsW0K5dO5o1a8bIkSP58MMPS/X8n4eamhrz589HV1eX7t27Y2dnx4kTJwD5zkL//PMPP/zwAx07dqR169ZMmjSJrl27sn79eoV4pk6dSrNmzWjdujWOjo6cPn2aK1euAHD58mXOnj2Lg4MDgiCwdu1ahg4dysiRI2nZsiU9e/bE29ubPXv2kJycXCm5MzIyUFVVRUdHBx0dHYYNG8b69evLVSgVhR84cCC7d+8WR527d+8GoFevXuTk5HD37l20tLTQ0dFBX18ff39/3N3dAdDQ0KBGjRqoqqqioaHBw4cP2bBhA56enlhbW9OsWTPs7e0ZO3YsISEhFBYWinJNmjSJjh070q5dO3r27ElOTg4+Pj7o6uryxRdfIJPJyq3PkukWUbJtde3alcWLF9OhQweaNm1Kv379MDAwEOPdsWMH9+7d4/vvv8fQ0JCPP/4YHx8fPvnkE/Ly8sR5Qw0NDdTU1Mqtk379+tGrVy9atGjB4MGDefjwIe7u7nTs2FHsRBZtLRkbG8vt27cJCAjA2NgYPT09Fi5cSOvWrRU6x6DYtgDy8/NZvHgxJiYmGBoa4ufnR0pKCgcPHhSfsbGxYeTIkTRr1oyBAwcik8lEc2pISAi6urrMnz+ftm3b0qlTJwICAsjOziYmJqZUvn766Sfq1Kkjfhfat2+Pv78/tWvXJjw8HIDg4GBcXFwYNGgQrVq1Yvz48fTo0UOMw8rKCi0tLYX4Y2JicHR0RKn4MOkZYWFh2NjYKPxKjtAl3l8EQXF0DZRS1mVdV1KSP/u2eeERtoaGBs7OzuzYsYPz58+TlpbGxYsXARQ+oM2bN1fYZL5Dhw4AXLp0Sfy4t2rVSryvpKSEuro6+fn5gHy/26lTpyqkbWBggKqqKomJiXTv3h2AFi1alJKxeLx169YV5Sl5LS8vj7p16+Li4kJcXBwhISGkpaWRmJjIzZs3K20SB2jatKk4KgEU8nL+/HlA/vErTn5+Pjo6OuLfDRo0oGHDhuLfnTp1olWrVvz6669MnjyZmJgYZDIZ+vr63Lp1i1u3bpUaVRSN1hMTE2nTpk2Fcru4uPD777/TvXt32rdvT9euXfniiy/KPSCgovB2dnb4+vqye/du+vfvz7Zt27C1tUVFRQUVFRXGjBnDwoULCQwMxNzcHCsrq3JPD0pJSSEvLw9PT09mzZolXi8oKODJkyfcunVLvFayzrW0tFAttmCyTp064jRCZSnZtpycnPjjjz/Ytm0bV69eJTk5mYyMDJo2bQrIO6zNmzdXqMOWLVsyc+bMF0q3eF6KOqHFZalTp47Ce9KiRQs+/PBD8b6SkhImJiYkJCSI10q2raK4jYyMxL9bt25NgwYNSEpK4tNPPxXlL06DBg14/PixmLaFhYWCotTW1qZFixZl7lWdmJiIgYGB+P4VyWBgYEBiYiJ37twhMzMTY2Njhec6d+7MuXPnAKhZsyYODg5ER0czadIkjh8/TlpaGv379y+VHsDIkSNLeUTfuHFDUtr/MV521Lp0KWRkKC7Tiojo8swcroi6ug59+8rTuXoVnJzg889fKtmX5oUVdnZ2NoMHD0ZDQwMbGxssLS0xNDQstYyrZoltYYqUX+1iPvO1y/Cfr8gxRxAEheeKv/xF1KpVOltFJvSSPHr0CBcXF3Jzc+nTpw8ODg7iPOGLUFZeiigsLERFRYXo6OjnylpWXhwdHdmyZQuTJk1i+/btCubxsigqv+fJU3wOsmXLluzevZsjR46QkJDAvn37CAkJYdGiRdjb25d6tqLw9erV4/PPPycmJgYLCwv+/vtvIiIixOdnzJiBs7Mz+/bt49ChQ/j4+LBx40Y2b95cKv9FeVm+fDlt27YtJYuGhgapqalA6Tovr75fhOLyFBYWMnbsWC5dukS/fv2wtbXFwMCAOXPmiGGKd9hehbLiKWv0+DwKCwsrfE9KvqMgf0+Ll11Z5VjRO1oy7crKWlSHFcU/YMAA1qxZw/Hjx4mJieGTTz4RTeYlqV+/PvXr16+0LBLvFxYWEBKiqLBNTb04cGAmT5/+bzVHrVoqmJp6AfKRtSBAu3ZvW9qXMIlv376dO3fuEBERwYQJE+jRowd3794FFF+0tLQ0BSez48ePo6SkVOaHtyxkMlkpJ5LTp0+Tm5tbqZFjZYmPj+f8+fOEhYUxdepUbG1tUVNTe61rZWUyGbm5ueTm5tKiRQvxt3HjRnbs2PHcZ+3t7cnIyCAiIoKsrCzs7OwA+RF5jRo1KuU8cuTIEQCxjJSVlRXq4d69e2J9Aaxfv57ffvuNrl274uHhQWxsLObm5mWaNCsbfuDAgRw5coStW7fSqlUrOnbsCMhHzHPnzkVTUxMnJycCAgIIDg7mwoULohWiuGJq3bo1ysrKXLt2TaHcjhw5wo8//vhalHIRFSnECxcuEB8fz4oVK5gxYwb9+vWjWbNmpKWlie1eV1eXtLQ0BcfI1NRUTE1NSU5OfmGlWxlkMhlXr17l5s2b4rXCwkKOHz9e4buWk5MjdnhAbv3KyclBX1+/0mkfO3ZM4b3PysoiLS2tzLSLzOnFl7U9evSIs2fP0rZtW+rVq4eOjg6nTp1SeO706dMKf7do0QJTU1Pi4uLYu3cvAwYMqJS8EnIGDx7M4MGDq1qMakHnznKns2cGKwDatHHE0nIp6uo6gBLq6jpYWi6lTRtHAO7cgTZt4Jlh7a3ywl+8xo0b8/jxY3bu3ElmZibx8fHiHGRxk2NeXh5Tp07l/PnzJCQksHDhQvr27Ss6ilTE2LFj+fPPP/H39+fy5cskJCQwc+ZMDAwMMDc3f1Gxn5sfkM8FZmZmcuzYMSZMmEB+fv4Lm1DLo1u3bujr6zN9+nQOHTpEWloaK1asIDw8vMLOh7a2tjh/b2VlhaampnjPzc2NzZs3ExYWxtWrV9mzZw8+Pj706NFDNKsaGxsTGRnJ2bNnSUxMxNPTU2E0eufOHRYsWMDevXvJzMzkwIEDnDt3rpRZ8kXCd+nShY8++ojVq1fj6OgoXm/YsCE7duxg7ty5pKSkkJqaytatW6lfv75YDmpqaly5coWsrCzq1avH0KFDCQgIIDo6mvT0dGJjY/Hx8RHnnV8XxdMti0aNGlGrVi127dpFeno6Z86cYdq0aWRnZ4vmaTs7Oxo2bIiHh4fYCZk9ezYtWrRAV1dXnLe+cOECDx48eC1y9+vXD01NTaZMmcLJkye5dOkS3t7epKamMnLkyAqf9/T05MyZM5w6dYqZM2diZGRUprNhWbi6unL58mXmzp3LpUuXOHHiBFOmTOHDDz/E1ta2VHhnZ2eePHmCu7u7WD7u7u7k5+czZMgQQN6mf/rpJ3HaITQ0lF27dpWKa8CAAURGRvL06VOFOW6JilFVVVWYLnqfUVWFXr3kZvHihp02bRxxcjqCm1sGTk5HRGWdny9X2Pb2pee+3wYvbBLv3bs3586dw8/Pj5ycHHR0dBg0aBB79+7l1KlTohduu3btMDAwYOTIkQiCgK2tLV5eXpVOp2fPngQEBBAUFMS6dev44IMP6NGjB+7u7mWavF8WIyMjZs2aRWhoKIGBgWhra2Nra0uTJk1K9fRflpo1axIcHIyfnx/u7u48fPiQ1q1b4+/vLy5XeR6Ojo7s379fQfkBjBo1irp16xIaGoqfnx9aWloMGDCACRMmiGG+++475s2bh5OTExoaGnz55Zc8fPhQvD958mQKCwtZuHAh2dnZaGlpMWTIEMaPH1+mLJUJr6SkhIODA6tWrVIwq2toaLBu3Tq+//57Bg8eTEFBAUZGRoSEhIhmS1dXV3x9fYmPpR5EwQAAIABJREFUj+fgwYPMmjULTU1NAgMDycrKQltbGzc3t3Lle1lKplsSbW1tlixZQmBgID///DNaWlp8+umnjBo1it27d4vTHsHBwfj6+uLs7Ezt2rXp1q0bXl5eKCkp8fHHH2Ntbc3XX3/N0KFDFczpL4u6ujqbNm3C19eXMWPGUFhYiKGhIRs2bFCYny4Pe3t7vvrqK548ecJnn33GN998U+mOkJGREcHBwfj7+zNgwABUVFSwtLQkICCAevXqlQqvo6NDeHg4fn5+ODs7U6NGDUxNTfn5559FXw4nJyfu3bvHihUruHPnDl26dMHBwYHjx48rxNW7d2/mz59P3759X8j8LgH/PHOJLrJ8ve/07w/XrsHx4/Kdzso75OPJE7liHzoUni2+eOsoCdJuDhJvgLlz53L79u1Sa+slqgdRUVHMmjWLc+fOvdYO8Kuyb98+dHV1RUc+AG9vb9LS0ggLCxOvZWVl8dlnnxEVFcXHH3/8QmlkZGSI69ubVoVds4oJDQ0F5B1+CTn5+fDzz/DHH/K/GzWSr7cWBPk+4nfuyLcsdXaG7t3f3Oi6orZZfd5Uif8ECQkJpKSkEB0dTXBwcFWLI/GOERsbS3JyMvPmzUNLS4ujR48SGxvLt99+C8g9vE+dOkVkZCQmJiYvrKwlJMpCWRmGDZOfyJWQAH/9Bdevy0fb2towcCB06gTPFm1UGZLClnitbN26lT/++IOxY8eKS8wkJCqLt7c3S5YsYeLEidy/f5/mzZvzzTffiNNBd+7cwcvLi2bNmlV6cyAJicqioQF9+8p/1RFJYUu8VpYvX17VIkhUAkdHx1I+EdWBDz74gCVLlpR7v127duLmShLvPoIABQXyX61aUKNG1ThzvStICltCQkJC4o0jCHDjBqSnQ3IyJCbKnbiK709Vty60bi1f49y8ufxIS2kZ/f+QFLbE/7N39nE9Xv0Df4fQpLnzvJKHrC/rSZtKm4zKjeaputkIkc3TjK1u4R42m20esqxiKRIyZkRp2I2foeUeGWW69Wh9q1EeFktZ0fX749J1961QilTn/Xr18nVd1znX55zr+n4/5/M553w+AoGgEfG0o7wVF8Ovv8L+/bKiBnnOuHVrMDCQreqy12Zny8q8NGzoq6/CoEGy8m7s1rdQ2AKBQNCIqK2IfI+ipERewPXtt/Dnn7KCNjJ6uNLV1oa//U3+Azkxx88/w4kT0LUrTJ4sK+7GilDYAoFA0IgojSD5JBeFXr8OYWEQHw+dOkGZeE/VolkzeOEF2dq+cQOWLoWRI+VFYY1x+71Q2AKBQNCIKE2k8qQU9pkzEBwsK9nacmNraclK//nnISoK4uJgzhx5MNCYqL3YjgKBQCBo1Bw7Bl99JSvWF16o/TnnZs2gWzfIy4Nly0Ctrt36n3WEwhYIBAJBjfnpJ9i4UV5I9qRDlXfoIC9WW7FCDivaWBAKWyAQCAQ1IjkZQkJkq7pFi6dzz9KUmKtXy+FDGwNCYQsEAoHgsSkshPXroU0beR/106R9e9k9vnOnZrathopYdCYQCASNiNpO+rFnj5wcw8ioVqutMgYGcPQoWFuDmVndyPC0EBa2QCAQCB6LtDT44QdZadYVTZrI2bU2bJBTYDZkhMIWCASCRkRsbCyxsbG1UtcPP8gZrB6UQ/ppoacHN2/C/VTfDRahsAUCgaARkZycTHJyco3ruXEDTp+WV2w/Czz/vBz+tCHPZQuFLRAIBIJqExsr77Nu8oxokeefh4wM+a+h8ox0taC2CAgIYMCAAXUtRq0RERGBSqXi7t27dS1Ko+Pnn39GpVKRUYNfwKysLFQqVa25YAXPDidOPH7I0SeBlpbsmm/I2VeFwhYIBAJBtSgogNzcJx8gpbq0bg0XL9a1FE8Osa1LIBAIGhG1ka3r8mXZFf6spbvU1YVLl+RMYc+Kq742aYBNatgkJyczffp0rK2tMTMzw8HBgZCQkArXBQUF0a9fP15++WUWLlxIfn6+ck6lUrFz506mTJmChYUFgwcPZu3atRrljx07xptvvomVlRV2dnYsWrSImzdvKucdHBxYvnw5w4cPx8bGhuPHjzNx4kRWr17N4sWLsbKywtbWlsDAQNLT05k4cSIWFhYMGTKE48ePK/VcvnwZLy8v7OzsMDU1xd7enhUrVnCvbFb7h7BgwQLmzZvHihUrsLOzw9bWllmzZpGTk6Ncc+vWLRYvXqz0h7u7O3Fxccr5yqYRvvvuO1Qq1UPbe+/ePcLCwhg2bBjm5uY4OjoSHBxMSUkJ8D938IEDBxgzZgwWFhY4Ozvz7bffPrA9ERERODg4sH//fhwcHLCwsGDKlCnk5OTwxRdfYGNjg62tLcuWLUMqs7rmcZ/X559/jre3N3369MHe3p6vv/5ao16A48ePM2LECMzMzBg2bBhHjhxRzj2qDypj7969jBo1CktLSwYMGMCKFSv4q8x+nBs3bvDBBx/Qt29fbG1t8fX1ZdKkSQQEBHDjxg3MzMzYtWuXRp3r1q3DycmpguyCiri7u9c4J3Z2tqwUa0pqagTbt9sQEmLI9u02pKZG1Ki+Zs3klJxXr9ZctmcRobDrEYWFhXh6etK6dWt27NhBdHQ0zs7O+Pr6cv78eeW6nJwcfvrpJ0JDQ1m/fj2//PIL77//vkZdK1euZPTo0URGRjJ06FD8/f05deoUAIcOHWL69OnY2dmxe/dufH19iYuLw9PTU+OHODw8nHnz5rFp0yYl88+mTZvo1KkTUVFRTJgwgYCAAKZNm8akSZOIiIjA0NCQ+fPnKz+sM2bMIC8vj9DQUA4ePMjbb79NaGgohw4dqnK/HDhwgLy8PMLDw1m5ciVxcXGsWbMGAEmSeOedd/jtt98ICgpi165d2NnZ4eHhoaQZrCrl27t8+XKlfdHR0cyaNYugoCBWrFihUW758uXMmDGDPXv2YGlpyccff0xmZuYD75Obm8s333zDunXrCA0N5fz584wcORJJkvj2229599132bp1Kz/++CNQs+f1zTffoKury+7du/H29iYoKIigoKAK7V60aBHR0dEYGxvj5eXFn3/+qbStKn1QSlhYGIsWLcLV1ZXIyEgWL17Mvn378Pb2BqCkpITp06dz6dIlQkJCCA0NJT4+Xnk39fX1cXBwYO/evRr1RkZG4uLigtazZvI1UK5dq/lWrtTUCE6c8CE/PxuQyM/P5sQJnxorbS0tOf92Q0S4xOsRhYWFTJo0ifHjx6OrqwvAnDlzCAkJITk5GXNzc0B2efn5+dGuXTsAlixZgqenJ2lpaRgbGwMwevRoRo0aBYCXlxfbtm3j7Nmz2NjYEBwczMCBAxUl36NHD3x9fXFzc+PEiRO8/vrrAPTv31/5XIqxsTHvvvsuAFOmTCEwMBBnZ2cGDx4MyKP7mTNncvXqVfT09Bg1ahRDhgzB4H7kBQ8PDzZs2EBycjJDhw6tUr+0atWKTz75BG1tbYyNjRkxYgQxMTEAnDx5knPnznHy5En07wcfnj17NgkJCRqKqyqUbW9+fj7bt2/H29sbFxcXALp27crNmzf58ssvmT17tlJu8uTJODo6AuDj40NERATx8fF06dKl0vsUFxfz4Ycf0qtXLwD69etHQkICCxYsoEmTJnTv3p21a9eSnJzMoEGDavS8unfvzscff4yWlhbGxsakpaWxdetWpk+frlwzf/58bG1tAZg1axaHDh0iJSUFExOTKvcByIOnkJAQ3nrrLTw8PADo1q0bJSUlzJkzh9TUVK5du0ZCQgLR0dG8+OKLAKxZs4ZBgwYp9bi5uTF9+nQyMzPp0qULZ86cQa1W4+rqWqEvb926xa1btzSOXblypdJ+bywcO3YMoMK7UB2Kih7uco6O/scj68jJOUNJSZHGsbt3Czl2zJuLF795ZPnhw3c98Fxx8SOL10uEwq5H6OvrM378eL7//nsSExNRq9VcvL/CoqwlZWRkpChrAEtLSwBSUlIUhd29e3flvJaWFrq6uhTff8uTkpKYO3euxr3NzMx47rnnSEpKUr7oXbt2rSBj2Xpb3g8sbFQmZmHpsaKiIlq2bIm7uzsHDx4kNDQUtVpNUlISubm5VXaJAxgaGmrMy5VtS2JiIoCiMEspLi5WBglVpWx709LSKC4upm/fvhrXWFtbU1xcTFpamvIMyvZJ69atlfs/jPL92KVLF5qU+YVs0aIFRUXyj11Nntcrr7yiYZX26dOH4OBgrl+/rhzr1q2b8llPTw+Av/76q1p9AHD9+nWuXbtW6fWl7cjJyaFVq1aKsgZo27atRn/Y29vTsWNHIiMjmT17Nnv27MHOzo7OnTtXaN/mzZsJDAyscLwxc+nSJaBmCrs2HBnllfWjjleHhupoEQq7HnH16lXGjh2Lvr4+jo6O2NvbY25uXmH+tWk5X1Wp8mvevLlyrOznUh41/ydJkka5lpVE+m/WrOIr1eQBQ/GCggLc3d0pLCxk2LBhuLi4YGlpyfjx4x8qR3kqa0spJSUl6OjoVHChlpW1MjdqZQq1svaWp7QPa9rX5fvxQX34KFmq+7xKB35l36Hy71Np3Q+7Lzz8uTzo+qZNmz6yb5o0aYKLiwtRUVFMmzaNAwcO8Mknn1R6rYeHh2L9l3LlypUaz+E2dlq2fPgc9sOs31K2b7e57w7XRFfXoErlH0YlP0MNAjGHXY+Ijo7mjz/+YPv27cyaNQsnJyfy8vIAzR9QtVqtscjszJkzaGlpaVgtD0OlUlWY301ISKCwsJCePXvWQktkYmJiSExMZPPmzcydOxdnZ2datWqlYd3VFJVKRWFhIYWFhXTt2lX527p1K99//z0gTyHcvn27Qh8+DGNjY7S1tTUWrwGcOnWK5s2ba3gVnjQ1eV6//vqrxv9/+eUXOnfurEwfPIzq9kG7du1o165dpdcD9OzZk169elFQUEBqaqpyPi8vr8JecDc3N9RqNWFhYTRp0gQnJ6dKZdTT08PQ0FDjr1OnTo9sm+DhdOwI1XCCVYq19QKaNdPRONasmQ7W1gtqVK8kwd/+VqMqnlmEwq5HdOrUiTt37rB//36ys7OJiYnBy8sLQHGPln6eO3cuiYmJxMbGsmzZMoYPH/7AOdPyTJs2jaNHj7JmzRrS09OJjY3Fx8cHMzMz+vXrV6vtAYiKiiI7O5u4uDhmzZpFcXGxRntqQv/+/TE1NcXb25uTJ0+iVqvx8/MjPDxcUWZWVlbk5+ezfv16srKyiIqKqtQiL4uuri7jxo0jMDCQvXv3kpGRQUREBOvWrWPcuHHKGoOnQU2e19mzZ1mzZg2XLl1i165dbNu2jbfffrtK932cPnjnnXfYsWMHmzdvJiMjg0OHDvHZZ5/h5ORE9+7dsbW1pU+fPsyfP59z585x8eJF/vnPf1JYWKjhCenSpQu2trasXbuW4cOH0+JpJWEWANC5c823TfXs6Yq9/Up0dQ0ALXR1DbC3X0nPnhXXIlSVoiI5tnkVxpv1kgbqOGiYDB06lAsXLrBq1Sry8/MxMDBgzJgxHDlyhPj4eCZMmABA7969MTMzw8PDA0mScHZ2ZsGCqo9aBw8ejL+/P0FBQWzYsIE2bdrg5OSEl5dXpS7vx8XCwoKFCxcSFhZGQEAAHTt2xNnZmc6dOxMfH18r92jatCkbN25k1apVeHl5cfv2bXr06MGaNWtwcHAAwMbGhg8++IDw8HDWrVuHtbU1Pj4+LFy48KF1L1iwAH19ffz9/cnNzcXAwIB3332XKVOm1IrsVaUmz2vQoEFkZGQwatQoOnTowPz586vlLq5uH0yePJmWLVsSFhbGqlWraN++PW5ubsyaNUu5xt/fn08++YTJkyfTokULxo8fT1paWoX9w66urvznP//Bzc2tyvIK4LlaiHbSubPsEq/pfueePV1rpKDLk58PPXo03DlsLUlsXBQIGiUTJ06kY8eO+Pr61rUoCjdu3ODs2bPY29src+BFRUXY2try0UcfMXr0aOXa4OBg9u/f/0hvSHmysrJwdHTkyJEjGBoa1qr8jYnFi+H2bTlT1rOCWg3/+Ac4O9e1JI/Ho95N4RIXCATPDM2aNcPb25vVq1eTkZFBamoqH330Ec2bN1cWV545c4bdu3ezceNGJk+eXLcCN2IcHeGPP+paiv8hSbLF//LLdS3Jk0MobIFA8Mygp6dHUFAQ8fHxjB49mrFjx3Lt2jW2bNmiLIT78ccf+eSTTxgyZIgSS0BQdQ4fPszhw4drXI+1tRw85VnZ83ztGpiZQUNeUyjmsAWCRsrWrVvrWoRK6devHzt27HjgeW9vbyUymqD6ZGVl1Uo9rVrBgAFy1q5qhjR4IhQUwJAhdS3Fk0VY2AKBQCB4LAYPlt3QtbSp47G5dk0eNPTuXbdyPGmEwhYIBALBY9G5M4wZIycDqavly8XF8urwd95puAFTShEKWyAQCASPjZOTvJWqrjJkZWeDqytUEnm3wSEUtkAgEDQi9PT0lJjwtUGzZrJ1e+8elMuz8sT5/Xfo3h2GDXu6960rGrgDQSAQCARlqSyrWU154QXw8oKVK+WgJfdz3DxRLl+Gtm3h/fehXEydBouwsAUCgUBQY1QqmDdPtrJv3Hhy95EkyMyEDh3Axweef/7J3etZQyhsgUAgaEQcPHiQgwcPPpG6e/WCDz+EFi0gIwPu3q3d+gsK4NIlsLCQlXVDTfLxIIRLXCAQCBoRV65ceaL1d+sGn34KkZGwf7/sHm/btmbxve/dk13gzZrBrFlga9tw44U/DKGwBQKBQFCrtGwJb74phwnduRNSUmRl27Fj9eabb9+W91g3aQKvviqvBm9sVnVZhMIWCAQCwRPhxRfhX/+St14dOyb/3b0rz0O3aCFHS2vWTLaWS0rkPdV//ilb1Fpa8vz0+PFgY9O45qofhFDYAoFAIHhiaGmBoSG4u8uZtH7/Xf5LSZHno+/ckRW1tja0aydb0j16yCvPO3SQ45ULZITCFggEgkZE27Zt6+zeLVrI+6a7d4fXXqvduiUJsrJkK/7iRdmd3qwZ6OvLMc9ffhl0dGr3nk8bobAFAoGgETFixIi6FqHWOX8e9u6FtDTZUm/TRp5HLymBK1dgwwb5uIODnCv7WcrhXR2EwhYIBAJBvUSS4MAB2LFDVtJdu1ZcPa6jIy9UKyqCH36As2fB21t2t9c3xD5sgUAgaETs27ePffv21bUYtcKhQ7B9O3TpIru+H7bVq3lzWaHfugUrVkBe3tOTs7YQCvsZJyAggAEDBtS1GLVGREQEKpWKu7UdUaEe8d1336FSqZT/Ozg44Ofnp/z/008/5ZVXXsHKyor4+HiOHTuGo6MjZmZmfPbZZ3UhssKNGzf47rvvlP8vWLCAcePGVbn87du3CQ8PV/7f0N7v+sD169e5fv16XYtRY9LTYds2WVlXZ6tYx47ySvTg4LrLMPa4CJe4QFDH7Nq1ixYtWgDw66+/Eh4ezpIlSxg4cCDt27fHzc0NIyMjNm/eTKtWrepU1i+++IIrV64wZswYAD788EPu3btX5fIhISFERUUxYcIEADw9PXF3d38isgoaNocPy/PUzZtXv2znzpCYKG83MzSsfdmeFMLCFgjqGH19fUUR//nnnwAMGDAAAwMDmjdvzq1bt7CyssLQ0JC/1XHUCKmcSdK6dWvatGnz2OVbtWqFvr5+rcgmaDzk5cF//vP489BaWvIK8h9/rFWxnjhCYdcxycnJTJ8+HWtra8zMzHBwcCAkJKTCdUFBQfTr14+XX36ZhQsXkp+fr5xTqVTs3LmTKVOmYGFhweDBg1m7dq1G+WPHjvHmm29iZWWFnZ0dixYt4ubNm8p5BwcHli9fzvDhw7GxseH48eNMnDiR1atXs3jxYqysrLC1tSUwMJD09HQmTpyIhYUFQ4YM4fjx40o9ly9fxsvLCzs7O0xNTbG3t2fFihVVtsIWLFjAvHnzWLFiBXZ2dtja2jJr1ixycnKUa27dusXixYuV/nB3dycuLk45X5mbtTI3dPn23rt3j7CwMIYNG4a5uTmOjo4EBwdTUlICQFZWFiqVigMHDjBmzBgsLCxwdnbm22+/fWibDh06xIgRIzA3N2fChAlcvnxZ43ypSzwiIoLJkycD4OTkxMSJE1GpVFy5coW1a9eiUqnIysoCZKu8VM6hQ4cSHBysTDOUyhkUFET//v0ZOHAgN27coKioiJUrV2Jvb0+fPn1wc3Pj8OHDihwRERE4ODiwZ88eBg8eTJ8+fXjrrbc4c+aM8mz27dvHqVOnFFnKu8QPHz7MmDFj6NOnD+bm5ri4uHDs2DHluQQFBZGdnY1KpeLnn3+u8KwuX76Mj48P/fv3x8LCotJnO3HiREJCQhgwYACvvPIKkydPJi0t7aHPQNCwOH1admc/aI92amoE27fbEBJiyPbtNqSmRlS4pmNHOH5c3gdeXxAKuw4pLCzE09OT1q1bs2PHDqKjo3F2dsbX15fz588r1+Xk5PDTTz8RGhrK+vXr+eWXX3j//fc16lq5ciWjR48mMjKSoUOH4u/vz6lTpwBZYUyfPh07Ozt2796Nr68vcXFxeHp6KsoIIDw8nHnz5rFp0yasra0B2LRpE506dVLcmAEBAUybNo1JkyYRERGBoaEh8+fPVyynGTNmkJeXR2hoKAcPHuTtt98mNDSUQ4cOVblfDhw4QF5eHuHh4axcuZK4uDjWrFkDyBbaO++8w2+//UZQUBC7du3Czs4ODw8PTp8+Xa3+L9/e5cuXK+2Ljo5m1qxZBAUFsWLFCo1yy5cvZ8aMGezZswdLS0s+/vhjMjMzK73HL7/8wnvvvYeTkxNRUVGMGDGi0gEZgLOzs9LO7777joCAAGJiYmjfvj2enp7ExMTQuXNnvvnmG1auXMmsWbPYv38/8+bN45tvvmHp0qUa9e3evZvQ0FD8/f3R19fHx8eHmJgYVq5cSWRkJC4uLnzwwQdERUUpZXJzc9mxYwerVq0iPDyc4uJi5s+fT0lJCR9++CFDhgzByspKkaUsv/76K++99x7Dhg1j37597Ny5k3bt2uHj40NRURGenp54eHjQqVMnYmJisLKy0iifn5/PuHHjUKvVBAQEsGvXLrp168bkyZM1vg9nz57l9OnTBAcHs379etRqNR9//PHDH7ZAoVOnTnTq1KmuxagRmZmyO7wyUlMjOHHCh/z8bEAiPz+bEyd8KihtbW05oloZu+WZR8xh1yGFhYVMmjSJ8ePHo6urC8CcOXMICQkhOTkZc3NzALS1tfHz86Ndu3YALFmyBE9PT9LS0jA2NgZg9OjRjBo1CgAvLy+2bdvG2bNnsbGxITg4mIEDBypKvkePHvj6+uLm5saJEyd4/fXXAejfv7/yuRRjY2PeffddAKZMmUJgYCDOzs4MHjwYAHd3d2bOnMnVq1fR09Nj1KhRDBkyBAMDAwA8PDzYsGEDycnJDB06tEr90qpVKz755BO0tbUxNjZmxIgRxMTEAHDy5EnOnTvHyZMnFVfq7NmzSUhI0BhoVIWy7c3Pz2f79u14e3vj4uICQNeuXbl58yZffvkls2fPVspNnjwZR0dHAHx8fIiIiCA+Pp4uXbpUuEd4eDiWlpbMnTsXgO7du5OUlMS2bdsqXNuyZUuevx9/UV9fX3E1N2nShOeee4727dsDsrdl+vTpyn7aLl26UFRUhJeXF15eXkp948aNw8TEBICMjAwOHDjArl27lPeqa9eupKWlsXHjRkaOHAlAcXExS5YswdTUFIBp06YxZ84crl27RocOHWjevDna2tqKLGVp2rQpixYt0piTnjx5Mp6enuTk5NClSxd0dHRo2rRppeWjoqK4fv06O3fupMN9X+eyZcs4f/48GzZs4KuvvlJkXLFihTI9MGHCBL788ssK9YHsjbl165bGsSed/OJZp6rfw7rgH//4R5Wuy86WM3dVttgsJ+cMJSVFGsfu3i3k2DFvLl78RuP4nTswdeqDlX95du3aVbULnxBCYdch+vr6jB8/nu+//57ExETUajUXL14E0LB8jYyMFGUNYGlpCUBKSoqisLt3766c19LSQldXl+LiYgCSkpIUhVGKmZkZzz33HElJSYrS6tq1awUZy9bb8v5bbWRkVOFYUVERLVu2xN3dnYMHDxIaGoparSYpKYnc3NxqLUwyNDREu8w3sWxbEhMTARSFWUpxcbEySKgqZdublpZGcXExffv21bjG2tqa4uJi0tLSlGdQtk9at26t3L8ykpOT6devn8YxKyurShV2Vbhx4wY5OTl89dVXBAYGKsdLSkooKSnht99+U5Rh2faV9tukSZM06qtM7uq0ryy9e/dGT0+P4OBg0tPTycjIqPR9fhBJSUl07dpVUdYgv8t9+/YlNjZWOfa3v/1NYy6/devWD5Rv8+bNGv0kaBg0afLgFd7llfWjjjepR35mobDrkKtXrzJ27Fj09fVxdHTE3t4ec3PzCvOvTctN1JQqv+Zllkc2r2SpZPkFPpWdL1uuZSXDzGbNKr4iTR7whhcUFODu7k5hYSHDhg3DxcUFS0tLxo8f/1A5ylNZW0opKSlBR0eHvXv3PlBWrUo2Y1b2g15Ze8tT2oc16evy5yrr06pSqvjmz59P//79K5zv2LEj165dA0CnTBzGUhm2bNmC3iPCPD3OuwTw888/M3XqVAYOHEjfvn0ZOXIkBQUFiofmcSkpKXlk/z8IDw8PxWNSypUrVxr1yvSICNk17OrqWseSVKSqFmx0NEREQBnbQWH7dpv77nBNdHUNGD78f/XfuyfHNPf3l5OQ1AeEwq5DoqOj+eOPP/jhhx+UH6GkpCRA8wdSrVaTn5+vuM3PnDmDlpYWL774YpXuo1KpOH36NFOnTlWOJSQkUFhYSM+ePWurOcTExJCYmMjx48fp2LEjAHl5ebW651OlUlFYWEhhYSG9evVSji9btowOHTowbdo0tLW1uX37NpIkKcpbrVY/tF5jY2O0tbVPO+9BAAAgAElEQVSJi4tTXMYAp06donnz5hgZGZH3GJEWevfuzdmzZzWOlZ2PrS5t27albdu2qNVqDQv66NGj7Nmzh88//7zScqWu8dzcXI32BQUFkZeXx4IFC6p0/8oGQ6WUTkmUtWi3bt0K/O99flh5lUpFREQEubm5ipVdUlLCmTNnFPmri56e3iMHKI2N8lME9REbG9i1Sw49Wt5+sLZewIkTPty9W6gca9ZMB2trzXc8N1fOq11flDWIRWd1SqdOnbhz5w779+8nOzubmJgYZQ6yqOh/7puioiLmzp1LYmIisbGxLFu2jOHDh1c6Z1oZ06ZN4+jRo6xZs4b09HRiY2Px8fHBzMysgru2pu0BeS4yOzubuLg4Zs2aRXFxsUZ7akL//v0xNTXF29ubkydPolar8fPzIzw8XBl8WFlZkZ+fz/r168nKyiIqKqpSi7wsurq6jBs3jsDAQPbu3UtGRgYRERGsW7eOcePGKYOl6uLp6UlycjIrVqzg0qVLREZGsmPHjseqC2SFN23aNLZt28aWLVtQq9UcPXqUDz/8UGlHZfTs2RMHBweWLl3K4cOHyczMZMuWLfj7+2NYjY2orVq1Iicnh8zMzArBbzp37kxKSgpxcXFkZWWxe/duZd659Pm3atWKmzdvkp6ezl9//aVRfuTIkbRt25Y5c+Zw9uxZUlJSWLx4MZcuXcLDw6PKMgoaPh06gLk5VGYL9Ozpir39SnR1DQAtdHUNsLdfSc+e//MoSJI8f+3g8PRkrg2EhV2HDB06lAsXLrBq1Sry8/MxMDBgzJgxHDlyhPj4eCW4RO/evTEzM8PDwwNJknB2dq6yRQQwePBg/P39CQoKYsOGDbRp0wYnJye8vLxq5J4tj4WFBQsXLiQsLIyAgAA6duyIs7MznTt3Jj4+vlbu0bRpUzZu3MiqVavw8vLi9u3b9OjRgzVr1uBw/9tnY2PDBx98QHh4OOvWrcPa2hofHx8WLlz40LoXLFiAvr4+/v7+5ObmYmBgwLvvvsuUKVMeW97evXsTEhKirLp+8cUXmTFjBqtXr37sOidPnkzLli3ZvHkzK1euRF9fnxEjRlTYOVAePz8/1qxZw9KlS8nLy8PQ0JBFixZVa8pizJgxHD9+HGdnZ8V6LqV0cdqMGTMAeZDw+eefM2/ePOLj4zExMcHZ2ZnIyEhGjhzJypUrNcrr6uqybds2VqxYwdtvv01JSQnm5uZs2bIFCwuLKssoaBwMHSqHGNXXr7i9q2dPVw0FXZ5r12R3+v0lQPUGLakqk1MCgUDQQMjKysLR0ZEjR45Uy7vQUAgLCwNQ9vzXVyRJTvpx4AB061b1xWN5ebJ1vWSJnHP7WeJR76awsAUCgaAR0VAGKVpaMHasnIXr8GE53Ohzzz34+pISyMmRI5z5+Dx7yroqCIUtEAgEjQgnJ6e6FqHWaNoUJk2SE4BERsoLyXR1ZTd56davv/6Cq1dlhW1hAePGQX2NGyMUtkAgEAjqLVpa8uKxAQPkhB4HD0JysrxtC6B1axg+HF57rX7mwC6LUNgCgUDQiNi5cycAY8eOrWNJapdmzWQL2sJCtqzv3pUt8PoUGOVRCIUtEAgEjYiCgoK6FuGJo6VVvRzZ9YUGNPYQCAQCgaDhIhS2QCAQCAT1AKGwBQKBQCCoB4g5bIFAIGhElM3GVi8pLJT3aoGcF7OquTEbAEJhCwQCQSOifM77Z547dyA+Hs6dg5QUOYB4aRIZSYL27cHEBPr0kZeIVyObW31DKGyBQCAQPHvcvi1vqj50SLaodXTkTdVGRpoKu7AQzpyBmBg51NnQoTB4sHx9A0MobIFAIGhEbNu2DeDZzQkuSXDhAmzYADdvymHJWrSo/FotLVlJl8YkvXMH9uyBY8dg2jRQqZ6e3E8BsehMIBAIGhHFxcUUFxfXtRiVI0kQFQUrV8oRT7p2fbCyroyWLeUy9+7B55/LFnoDym8lLGyBQCAQ1D2SJFvHe/fKbu+apP5t00a2urdtk4OIOzvXnpx1iLCwBQKBQFD3nDolK+yaKutSmjeX69qxQ1601gAQClsgEAgEdcsff8CmTXKOzNpQ1qVoa8uryDdsgD//rL166wihsAUCgaARYWJigomJSV2Locn27XK2jocltH5cWreGggKIiKj9up8yYg5bIBAIGhGvvvpqXYugydWrcPq0nNT6SfHCC3D8OLi4gJ7ek7vPE6ZRWNgBAQEMGDCgrsWoNSIiIlCpVNy9e7dO5YiNjUWlUpGVlQXAxIkT+ec//1mnMtUmT6uf7969i0qlIuIpWgAN6Ttx48YNvvvuu7oWQ/C4xMbK27OeZB7Mpk3lxWc///zk7vEUEBa2oNYICAigadOmdS2GoJHxxRdfcOXKFcaMGVPXotQLwsLCAJg8eXKdyqFw4gS0a/fk7/O3v8nBVQYPfvL3ekIIhS2oNdq0aVPXIggaIVID2mfb6Lh9Ww41amRU5SIRqaksP32a3/PzeUFXlwXW1rj27Pnogrq6kJUFxcX1Nll2g3CJJycnM336dKytrTEzM8PBwYGQkJAK1wUFBdGvXz9efvllFi5cSH5+vnJOpVKxc+dOpkyZgoWFBYMHD2bt2rUa5Y8dO8abb76JlZUVdnZ2LFq0iJs3byrnHRwcWL58OcOHD8fGxobjx48zceJEVq9ezeLFi7GyssLW1pbAwEDS09OZOHEiFhYWDBkyhOPHjyv1XL58GS8vL+zs7DA1NcXe3p4VK1Zw7969KvfJ/v37GT16NBYWFjg4OLB+/Xrlh+1R9UdERODg4MAXX3xB37598fT0BCAuLo4xY8ZgYWHB6NGjSU5O1rhneZf42bNnmTx5Mn379sXa2pr333+fy5cvAxAZGYmZmRl//PGHRh2jR4/m008/BR79XAsLC/nwww957bXXMDc3Z9SoURw8eFCjvm3btjFs2DCln8u6TlUqVQVX6oABAwgICKi0Tx+338pz5coVZs6ciZWVFa+//jr79++vcM2ZM2dwd3fHwsKC119/nUWLFil9tW7dOl599VUNV/3du3exs7MjNDQUgFu3brF48WLlfXd3dycuLq5SeQDy8vL49NNPGThwIObm5ri6unL48GHlfEREBAMGDOC7775jwIABWFlZMXPmTK5cuaJc4+DgQFhYGO+99x6WlpbY29uzc+dOzp07h6urq/LenD9/XilTVFTEypUrsbe3p0+fPri5uVW4r4ODA3v27GHw4MH06dOHt956izNnzgCwYMEC9u3bx6lTpzSmZwT1hMuXZVd4aajRRxCRmorPiRNk5+cjAdn5+ficOEFEauqjCzdpIu/1zsmpmcx1iVTPKSgokF577TXJ29tbSk1NlS5duiStWrVKMjExkRISEiRJkiR/f3/JxMREmjBhgnThwgXp1KlT0t///ndp6tSpSj0mJibSK6+8Iu3du1dKT0+XfH19JRMTE+nnn3+WJEmS/v3vf0sqlUry8/OT0tLSpJiYGGnIkCGSq6urdO/ePUmSJGnQoEGSqamp9OOPP0q//vqrVFBQIE2YMEEyNTWVAgMDJbVarcji6Ogo/fvf/5ZSUlIkT09PqV+/flJJSYkkSZI0cuRIacqUKVJiYqKkVqulsLAwycTERDpw4IAkSZK0e/duycTERCouLq60Tw4cOCCpVCpp/fr10qVLl6QffvhB6tOnj7Rp06Zq1T9z5kwpIyNDSkpKktRqtWRubi4tWrRISk1Nlfbv3y9ZW1tLJiYmUmZmpiRJkjRhwgTJ29tbkiRJio+Pl0xNTaVFixZJycnJ0pkzZ6SxY8dKgwYNkv7880+psLBQevnll6Vt27YpcqekpEgmJiZK3z3quX722WeSm5ub9Ouvv0pqtVpavXq11Lt3bykjI0OSJEnasGGDZGFhIe3YsUPKyMiQvv32W+mll16SDh06pDzznTt3avSdvb295O/vX2k/P06/lae4uFh64403pLFjx0rnz5+Xzpw5Iw0fPlwyMTGRdu/eLUmSJCUmJkrm5uZSYGCglJ6eLp09e1Zyd3eX3njjDamoqEj6/fffpV69ekk//vijUu/Ro0ell156Sbp69apUUlIijR07VpowYYJ09uxZKS0tTQoICJBeeukl6dSpU8p3wt7eXpIkSbp7967k4uIiDR06VPrpp5+ktLQ06csvv5RUKpV0+PBhpW2mpqaSs7OzdPr0aSkhIUFydXWVhg0bJhUVFSnvv6WlpfTtt99KarVa+vDDD6WXXnpJeuONN6SffvpJ+u9//yuNHj1acnFxUeSeO3euNGLECCk2Nlb67bffpK1bt0pmZmZSZGSkxn3Hjh0rnT17Vjp//rzk6uoqOTo6Svfu3ZNu3bolvffee9Kbb74p5ebmSnfv3tXo75s3b0qZmZkaf6dPn9Z4bxsbmzZtUn4L6hq3IUMkN1NTye2VV6r0183QUHrhhRcq/HUzNKxaHaamktuwYXXd7AeSmZn50Hez3rvECwsLmTRpEuPHj0dXVxeAOXPmEBISQnJyMubm5gBoa2vj5+dHu/tzJUuWLMHT05O0tDSMjY0B2bobNWoUAF5eXmzbto2zZ89iY2NDcHAwAwcO5P333wegR48e+Pr64ubmxokTJ5QMOP3796+QDcfY2Jh3330XgClTphAYGIizszOD78+luLu7M3PmTK5evYqenh6jRo1iyJAhGBgYAODh4cGGDRtITk5m6NChj+yTTZs2MWTIEKZNmwZAt27dKCgooKSkhDt37lS5/lmzZmF031W1evVq9PX1+eijj2jWrBnGxsb8/vvvrFy5slIZQkNDMTY25pNPPkHr/ujZ398fJycnIiMjcXd3x9nZmaioKMaPHw/IVreJiQmmpqbcuHHjkc81KyuLVq1aYWhoyPPPP8/cuXPp27cvbdq0QZIkNm3ahLu7O2+++SYARkZG3LlzR5GnOjxuv5Xn5MmTpKSkcODAAXr06AHA559/zj/+8Q/lmo0bN2JnZ6e8MwBfffUVr732GseOHcPJyQk7OzuioqKUdy0qKgp7e3vatWtHbGws586d4+TJk+jr6wMwe/ZsEhIS2LRpE9bW1hoyxcTEcOHCBSIiIjA1NQXggw8+ICkpia+//hpHR0dADmn5xRdfYGFhAcCqVasYNmwYP/30EwMHDgTkFchjx44FYMKECXz33XdMnDhRWZns6urKihUrAMjIyODAgQPs2rVL+Z527dqVtLQ0Nm7cyMiRI5X7LlmyRJFt2rRpzJkzh2vXrtGhQweaN2+OtrY27du3r9DfmzdvJjAwsNJnIah/FJWUVOt4pdTjKZR6r7D19fUZP34833//PYmJiajVai5evAhASZmHaGRkpChrAEtLSwBSUlIUhV02T6yWlha6urpKzN2kpCTmzp2rcW8zMzOee+45kpKSlB/Orl27VpCxbL0t7+duLfuDXnqsqKiIli1b4u7uzsGDBwkNDUWtVpOUlERubm6VXeKVKfbRo0crn6taf7du3TTq7NWrF83KBDWwsrJ6oAxJSUnY2dlpKMeOHTvStWtXkpKSAPnH+6233iIzMxNDQ0Oio6Px8PAAqvZc33nnHWbMmMGrr76KhYUFr732GiNHjkRPT48bN25w9epV+vTpoyHXpEmTHtl/lVGd51K238qTnJxMq1atFGUN8nukXWZO7b///S8ZGRkV+leSJNLS0nBycsLV1ZXFixcrA7EjR47g6+sLQGJiIoCiaEspLi5WBhtlSUpK4rnnnlMUYinW1tZ89dVXyv91dHQUZQ3yoPX5558nOTlZUdhl33Wd+9mSyr/rpd+pUjnLP5PK4lyXrbd169YPvK48Hh4euLi4aBy7cuXKs5v44ilQ/jnXJbtWr4Yvv6zyli6b7dvJLjOVWYqBri67hg9/dAWZmbBwYXXFfGao9wr76tWrjB07Fn19fRwdHbG3t8fc3LzClpXyq5dLf2Sbl8md2rySPKrSI0ZjkiRplGtZSTL1ZpVE7mnygC0MBQUFuLu7U1hYyLBhw3BxccHS0lKxQquC9kMWVFSn/sraUpbK2vUoSkpKlP6ysrKiR48e7Nu3j1deeYXc3FxGjBgBVO25WllZcezYMX766SdOnjzJvn372LBhA19//bVisVWXB23hqs1+K/9OaWlpaTyzkpISnJ2dNSzsUp5//nkABg8ezNKlSzl06BD37t1DR0dHUZolJSXo6Oiwd+/eCuWr88xKSko05KpsB8C9e/c03uXqvOul/bBlyxb0HrE39nG+mwB6enqPrLuxUd7DUqe0b18ti3eBtTU+J05QWOZ7qtOsGQuq0iZJkrd2VeKJqS/Ue4UdHR3NH3/8wQ8//KB8qUstuLJfaLVaTX5+vuJePXPmDFpaWrz44otVuo9KpeL06dNMnTpVOZaQkEBhYSE9q7JCsYrExMSQmJjI8ePH6dixIyAvCLp+/XqV6+jRo4fGwh4APz8/EhISGDdu3GPV37t3b3bt2kVRUZHSz+XvURaVSkVcXBySJClWdk5ODmq1WrGiAVxcXNi/fz9Xrlzh9ddfp23btkDVnqufnx82NjY4Ojri6OhISUkJb7zxBvv378fOzo4OHTpw/vx5/v73vyv38/HxoVmzZnz++edoa2trLDy8efMmeXl5lbanNp4LwEsvvURBQQFJSUmo7qf+S01NpaCgQLnGxMSE1NRUjIyMlL67cuUKS5YsYfbs2bRp04YWLVrg7OzMv//9b+7cucOIESMU5apSqSgsLKSwsJBevXop9S5btowOHTooUyWlqFQqCgoKuHDhgob1dfr0aY3vR35+PpcuXVKs3ZSUFPLz8x/bYiuNtpWbm6sxwAoKCiIvL48FCxZUqZ7HmeJozJR6Jh42sH9qtGsnr9i+e7dKIUlLV4M/1irxv/6Sg6bU4wFcvV8l3qlTJ+7cucP+/fvJzs4mJiYGLy8vQHYxl1JUVMTcuXNJTEwkNjaWZcuWMXz4cLpU0RUzbdo0jh49ypo1a0hPTyc2NhYfHx/MzMzo169frbYH5DnJ7Oxs4uLimDVrFsXFxRrteRjTp0/n4MGDhIWFoVarOXjwIFu2bMHJyemx6x83bhx//fUX//rXv0hLS+P//u//+Prrrx94vaenJ+np6SxZsoSUlBR++eUX5syZQ4cOHXAukzln9OjRpKSkEB0djaurq0Y/POq5ZmVl8dFHH3Hy5Emys7M5ePAg2dnZiit5+vTphIeHs2fPHjIzM/n222/5/vvvlbUDVlZW7Ny5k19//ZWkpCTmz5//QAu0Np4LgK2tLZaWlixYsIBz585x/vx5Fi5cqGGFTp06leTkZD766CNSU1NJSEhg7ty5pKamKtM3AG5ubsTExPCf//xHo+/69++Pqakp3t7enDx5ErVajZ+fH+Hh4ZUOLkuvnzdvHidPniQ9PR0/Pz+OHTvG22+/rXHt/PnzOX/+PPHx8fj4+GBhYYGtrW2V21+Wnj174uDgwNKlSzl8+DCZmZls2bIFf39/DA0Nq1xPq1atyMnJITMzs86DCdUHtm3bpuTErnOaNAFTU7hxo8pFXHv25NS4cWS98w6nxo2rmrIGOV65uXmVV6Q/i9R7C3vo0KFcuHCBVatWkZ+fj4GBAWPGjOHIkSPEx8czYcIEQLYQzczM8PDwQJIknJ2dqzyCB9kF6e/vT1BQEBs2bKBNmzY4OTnh5eX1WK7hB2FhYcHChQsJCwsjICCAjh074uzsTOfOnYmvYsaZQYMGsWzZMjZs2ICvry+dO3dm7ty5yrzd49TfsWNHNm/ezGeffYaLiwsGBgbMnDmTTz755IHt2LhxI2vWrMHNzQ0dHR3s7e3x9/dX5iABOnTowGuvvcb58+c1FutV5bkuXbqUlStXMm/ePPLy8jAwMMDLy0tRXu7u7ty5c4fAwEByc3Pp2rUry5cvZ9CgQQB8/PHHLF26lHHjxqGvr8/UqVO5ffv2E3suILuH169fz7Jly/D09ERHR4eZM2eiVqsr9N1XX32Fq6srOjo62NjY4OvrS6tWrTSu69KlC02bNqV3797K8aZNm7Jx40ZWrVqFl5cXt2/fpkePHqxZswYHB4cKMpW9/oMPPqCgoIBevXrx9ddfV7h+1KhRzJgxg7/++otBgwbxr3/964Eu76rg5+fHmjVrWLp0KXl5eRgaGrJo0aJqTQGNGTOG48eP4+zszNatWyusWxA84zg6wi+/PNl7SJJsYd//7tdXtKSqTAQJBIJGTUREBAsXLuTChQu1OkCtC7KysnB0dOTIkSPVsuQbCs9cpLN798DHR/5cZjBfq+TlQatWsGzZM21hP+rdrPcucYFAIBDUY5o2BXd3OQlIdbZnVZV792R3+IQJz7Syrgr1e6gsEAgEgmcfSYJbt+D33+V/S7fkNWsmp9Q0NIR+/eDMmdrP2pWVJbvdy0wb1VeEwhYIBI/E1dVVY2GboP7y1Ob479yBc+fk1JkpKVC6I6N0FlaS/pelq/RYejpcuSIvDnvE9sgqkZ0tDwDKBCaqzwiFLRAIBI2IJ66wL1+Ws2IdPiwv9NLVleem//a3h7uki4tli/vkSVnBm5hAjx6PLlcZkiRb1i+8AF5eshXfABAKWyAQCBoRpXv+n6ttJVZYCBERcOiQPC/dsSNUEvDmgWhrywr2jTdk13hSEvz2m3ysT5+qK93bt+UEHzY24OHx5Bay1QFCYQsEAkEjYufOnUAtrxK/eBGCg+XFXV26yAr7cWnZEl59Va4nPl5W2pcvwyuvQNeulVvbkgQ3b/5vNfh770HfvvV+kVl5hMIWCAQCweNRUgK7d8O+faCvLyvU2kBLS86R3bmzrKwTE+HECfjvf2VXeanlfu/e/+bAjYxg3DiwtKyd+e9nEKGwBQKBQFB97t2DLVvg//4PunWrmVX9ILS1ZUXcpYu8iC05WZ7n/vvfQUdHnh9/4QX5r3XrBmdRl0cobIFAIBBUj5ISCA+Ho0ehe3fZyn2SaGnJCtrCQl5MlpIC3t7ysUaECJwiEAgEguqxf7+8CrxbtyevrMuipSXv2U5Ph/Xrn0yglWcYYWELBAJBI6Jv3741qyAjQ563NjJ6usq6FC0t2UX+yy8QGwv9+z99GeoIobAFAoGgEWFmZvb4hYuLYcMGeSV2Xabn1NKSF6Rt3SpHMLuflrehI1ziAoFA0Ii4efMmN2/efLzCBw9CZqacx7qu0dGRV4dv3vy/SGkNHKGwBQKBoBGxZ88e9uzZU/2CBQUQFSWvyH5W6NwZEhLg0qW6luSpIBS2QCAQCB5NXBzcvVu96GVPGi0tWZ6jR+takqeCUNgCgUAgeDglJfLKcH39upakIh07yovPbt2qa0meOEJhCwQCgeDhpKfL8bmfxbjcTZvKA4rTp+takieOUNgCAgICGDBgQF2LUWtERESgUqm4e/dupecXLFjAuHHjnrJUMjdu3OC7776rk3tXl59//hmVSkVGRsZj15GVlYVKpSI2NrYWJRM8ddLT61qCh9O6tRx3vIEjtnUJBE+RL774gitXrjBmzJi6FkXQSLGzs6t+of/+V97K9ayiqwtpaf/Lsd1AEQpbIHiKSI1k+4ng2UWlUlWvgCRBauqz6Q4vpXlzOb3nH388m/PstYRwiTcCkpOTmT59OtbW1piZmeHg4EBISEiF64KCgujXrx8vv/wyCxcuJD8/XzmnUqnYuXMnU6ZMwcLCgsGDB7N27VqN8seOHePNN9/EysoKOzs7Fi1apLHf08HBgeXLlzN8+HBsbGw4fvw4EydOZPXq1SxevBgrKytsbW0JDAwkPT2diRMnYmFhwZAhQzh+/LhSz+XLl/Hy8sLOzg5TU1Ps7e1ZsWIF9+7dq3Kf3Lt3j1WrVmFnZ0efPn2YMWMGubm5yvlbt26xePFipT/c3d2Ji4tTzpeUlBAcHMyQIUMwMzPj5ZdfZurUqfz2228affbVV1/h4ODAq6++ynvvvce+ffs4deoUKpWKrKysCnJFRETg4ODA/v37cXBwwMLCgilTppCTk8MXX3yBjY0Ntra2LFu2TEP5P27ff/7553h7e9OnTx/s7e35+uuvKwwqjh8/zogRIzAzM2PYsGEcOXJEox/DwsIYNmwY5ubmODo6EhwcTMlDQkbu3buXUaNGYWlpyYABA1ixYgV//fWXcv7GjRt88MEH9O3bF1tbW3x9fZk0aRIBAQHcuHEDMzMzdu3apVHnunXrcHJyEgOiKnDt2jWuXbtW9QK3bkF+PrRoUSv3j0hNxWb7dgxDQrDZvp2I1NRaqReQM3s1YITCbuAUFhbi6elJ69at2bFjB9HR0Tg7O+Pr68v58+eV63Jycvjpp58IDQ1l/fr1/PLLL7z//vsada1cuZLRo0cTGRnJ0KFD8ff359SpUwAcOnSI6dOnY2dnx+7du/H19SUuLg5PT0+NH+/w8HDmzZvHpk2bsLa2BmDTpk106tSJqKgoJkyYQEBAANOmTWPSpElERERgaGjI/PnzlR/jGTNmkJeXR2hoKAcPHuTtt98mNDSUQ4cOVblf4uPj+eOPP9i2bRvr168nISGBFStWALIV/M477/Dbb78RFBTErl27sLOzw8PDg9P3F7Zs2bKFkJAQ5s+fzw8//MDatWtRq9V88cUXGvfZtm0bfn5+rFu3js8//5whQ4ZgZWVFTEwMnTt3rlS23NxcvvnmG9atW0doaCjnz59n5MiRSJLEt99+y7vvvsvWrVv58ccfa9z333zzDbq6uuzevRtvb2+CgoIICgrSkCc8PJxFixYRHR2NsbExXl5e/PnnnwAsX75ceV7R0dHMmjWLoKAgpS/LExYWxqJFi3B1dSUyMpLFixezb98+vL29AXkgNH36dC5dukRISAihoaHEx8cr75m+vj4ODg7s3btXo97IyEhcXFzQasDu0NoiOjqa6Ojoqhe4fbvWMnFFpKbic+IE2fn5SEB2fj4+J07UntIuKKidep5RhEu8gVNYWMikSZMYP348urq6ADzCixcAACAASURBVMyZM4eQkBCSk5MxNzcHQFtbGz8/P9rdj2C0ZMkSPD09SUtLw9jYGIDRo0czatQoALy8vNi2bRtnz57FxsaG4OBgBg4cqCj5Hj164Ovri5ubGydOnOD1118HoH///srnUoyNjXn33XcBmDJlCoGBgTg7OzN48GAA3N3dmTlzJlevXkVPT49Ro0YxZMgQDAwMAPDw8GDDhg0kJyczdOjQKvVL27Zt+fTTT2natCk9evTA2dmZEydOAHDy5EnOnTvHyZMn0b/vXps9ezYJCQmKsjMyMmL58uU4ODgAYGBggLOzM1FRURr3GTFiBJaWlsr/mzdvjra2Nu3bt3+gbMXFxXz44Yf06tULgH79+pGQkMCCBQto0qQJ3bt3Z+3atSQnJzNo0KAa9X337t35+OOP0dLSwtjYmLS0NLZu3cr06dOVa+bPn4+trS0As2bN4tChQ6SkpGBiYsL27dvx9vbGxcUFgK5du3Lz5k2+/PJLZs+erXEvSZIICQnhrbfewsPDA4Bu3bpRUlLCnDlzSE1N5dq1ayQkJBAdHc2LL74IwJo1axg0aJBSj5ubG9OnTyczM5MuXbpw5swZ1Go1rq6uFfry1q1b3Cq33efKlSsP7HtBJdy9W6VIYv+owiDgTE4OReW8L4V37+J97BjfXLz40LK7hg9/eOWSJIdObcAIhd3A0dfXZ/z48Xz//fckJiaiVqu5eP+LUdb6MjIyUpQ1oCiZlJQURWF3795dOa+lpYWuri7F978gSUlJzJ07V+PeZmZmPPfccyQlJSmKomslCe7L1tvyfuJ5IyOjCseKiopo2bIl7u7uHDx4kNDQUNRqNUlJSeTm5lbLJd6lSxealrEann/+ee7cuQNAYmIiAI6OjhpliouLlUGCg4MD586dw8/Pj4yMDC5dukRqaipty8U0rqy9VaF8n3Tp0oUmZRIttGjRgqKiIqBmff/KK69oWKV9+vQhODiY69evK8e6deumfNbT0wPgr7/+Ii0tjeLi4grJJKytrSkuLiYtLU3jnbp+/TrXrl2r9PrSduTk5NCqVStFWYM8uCrbH/b29nTs2JHIyEhmz57Nnj17sLOzq9RjsXnzZgIDAyscF1QDLa1aW8hVXlk/6ni1qYtkJE8RobAbOFevXmXs2LHo6+vj6OiIvb095ubmFbZxNS3n8ipVfs3LRDVqXkmEo0fNGUqSpFGuVPmWpVmziq9hkwd88QoKCnB3d6ewsJBhw4bh4uKCpaUl48ePf6gcVam/tC0lJSXo6OhUcLuWlTUoKIh169bh4uKCnZ0dnp6eHDp0iMjISI3rdR4zX2/5PnlQfzyMx+n70kFc2feh/LtRWvfD7guVvy+Pur5p06aPfKeaNGmCi4sLUVFRTJs2jQMHDvDJJ59Ueq2Hh4di/Zdy5coV3N3dqySbAGjWrEoW9iMtYMBm+3ayy6yNKcVAV7dK5R+KlpYsawOmYbdOQHR0NH/88Qc//PCD8gOalJQEaP7oqtVq8vPzFbf5mTNn0NLS0rB0HoZKpeL06dNMnTpVOZaQkEBhYSE9e/asreYQExNDYmIix48fp2PHjgDk5eVpWIQ1RaVSUVhYSGFhoeKWBli2bBkdOnRg2rRpBAcHM3PmTGbOnKmcDwkJeaSyeRJzrDXp+19//VXj/7/88gudO3dWpgIehrGxMdra2sTFxSlTKwCnTp2iefPmGBkZkZeXpxxv164d7dq1Iy4uTmPqonR+umfPnujp6VFQUEBqaqoie15eXoW94G5ubgQFBREWFkaTJk1wcnKqVEY9PT3FKyB4TNq0kQOT1MKWqQXW1vicOEFhmRgJOs2aseC+l6VGaGnB88/XvJ5nmIbtPxDQqVMn7ty5w/79+8nOziYmJgYvLy8AxaVa+nnu3LkkJiYSGxvLsmXLGD58OF26dKnSfaZNm8bRo0dZs2YN6enpxMbG4uPjg5mZGf369avV9gBERUWRnZ1NXFwcs2bNori4WKM9NaF///6Ympri7e3NyZMnUavV+Pn5ER4eriiRzp07ExsbS0pKCunp6fj5+fHvf//7kTK0atWKnJwcMjMzHxjYpbrUpO/Pnj3LmjVruHTpErt27WLbtm28/fbbVbqvrq4u48aNIzAwkL1795KRkUFERATr1q1j3LhxyuCvLO+88w47duxg8+bNZGRkcOjQIT777DOcnJzo3r07tra29OnTh/nz53Pu3DkuXrzIP//5TwoLCzUGO126dMHW1pa1a9cyfPhwWtTSCubGwIABA6oXKElHB9q3l7dN1RDXnj1ZaW+Pga4uWsiW9Up7e1xrOqiXJHlQ8YCFnA0FYWE3cIYOHcqFCxdYtWoV+fn5GBgYMGbMGI4cOUJ8fDwTJkwAoHfv3piZmeHh8f/snXlYVdXawH+EmrOGOaI4do7GIGgYqKSihp5ABRVDcEBzngq/FMrhalZOhSJ5DSTAIb2pODRoKCXqRUW9ICoGMh7gpphDZqIgrO+PI/tyBARHRNbvec6j7L322u9699r7Xetd71prDEIINBoN3t7e5b5P//798fPzY926daxfv56GDRvSr18/vLy8SnR5PyoWFhb4+PgQEhLCmjVraNq0KRqNhubNm3P6Ca10ZGhoSFBQECtWrMDLy4u///6bdu3asWrVKiXIbPny5SxevJhhw4ZRp04dOnfuzKJFi1i4cCFarVZvDL4ow4cP59ChQ2g0GjZu3IilpeVjy/s4uu/Tpw/p6ekMHjyYJk2aMHfu3IdyF3t7e2NkZISfnx/Z2dkYGxszbdo0PD09S0w/duxYatasSUhICCtWrKBx48YMHTqUqVOnKmn8/PxYvHgxY8eO5eWXX2bkyJEkJydT/b79l11cXDh27BhDhw4tt7wSXVDiQ6NSwalTULv2Y9/fpUOHxzfQ93P7tm7+dQmNxBcJAyEnLkokVZJRo0bRtGlTVq5cWdGiKFy9epWYmBjs7OyUIZzc3FzefPNNFi5cyJAhQ5S0AQEB/PTTTyXGGjyIzMxM+vbtS0REBC1btnyi8lcGCqPkC71V5eLQIQgOhkcMonzqZGeDhQVMnlzRkjwWZdVN2cOWSCTPDdWqVWP27NmMGDGCkSNHkpeXR1BQEDVq1FDcuKdOnSItLY2goCB8fHwqWOLKx759+wCdt6PcqFQ6t/PzuvTnrVvQtWtFS/HUkWPYEonkuaF+/fqsW7eO06dPM2TIEFxdXfnjjz/YsGGDEgh38OBBFi9ejIODg7IugOQp06wZmJrCEwzufGLcvq1z1RdZ7+BFRfawJZIqysaNGytahBKxsbFh69atpZ6fPXu2sjKa5Bni4AC+vlBkbv1zQXY2ODnp1hN/wZE9bIlEIpGUzeuv66Z4lTCPusIoXIWtZ8+KluSZIA22RCKRSMqmWjXw8ND1aJ/UymSPS1YWaDS6aWdVAOkSl0gkkirE/UvuPhRduoCtrW6KV0VH2F+9Ck2awKBBFSvHM0T2sCUSiaQK0apVq3IviFQMAwMYOVK31WZFusbv3oU//4SJE5/Ytp+VAWmwJRKJpAqRkZFBRkbGo2fQoIFuvvPly7oI7WdNfj5oteDiAvc2JqoqSIMtkUgkVYiIiAgiIiIeLxNzc5g0Cf7732drtPPzIS1NF7FehVzhhcgxbIlEIpE8PN276/79+mtd0NfTXhY0NxcyMnRBZiNGPJ8LuDxlpMGWSCQSyaPRvTvUq6cz2teugbHxk9+TWoj/ud9HjYL+/auksQbpEpdIJBLJ42BuDp99BtbWOnf1X389ubzv3IHUVGjaFJYsgbffrrLGGmQPWyKRSCSPS/36ujHtbt1g0yZIT9ctF9qo0cP3uIXQRYBfv66LAB85Evr1080Dr+JIDUgkEkkVYsCAAU8nYwMD3Tztzp3h/Hn4+Wc4c0Z3rmZNneu8du3iPWQhdD3pv/7SbeIhBLRqBe++C5aWumuLcvkyxMbCH3/o0tepo5uPbWUFr7zydMr2nCANtkQikVQhHmpbzUfB0BDMzHS/7GxITISkJEhI0AWN3W+wCwp0htbUFDp2hHbtdNt4Fk2Xn6+7/uefIS5Od6xGDV3vPT8f8vJ0PXtra+jbF1577YV0nUuDLZFIJFWIlJQUANq1a/f0b9akie5XuNb37dtw48b/1gCvVk3X665Xr/Q8bt3SBbXFxup6061alexmz8/XpTl2THe/MWNeuA1BpMGWSCSSKsShQ4eAZ2Sw76dmzeIu7gdx6xasXKkLZmvT5sG9ZkNDaN5c12M/ckQ3Dj5z5gtltGWUuEQikUieP+7ehXXrdMbaxKT8Lu6XXtIZ97NnYcMGXU/+BUEa7HKyZs0a3nrrrYoW44kRFhaGWq3m7t27JZ739vbGzc3tGUul4+rVq2zbtq1C7l2IEIKdO3dy5cqVCpWjIimrjpSH48ePo1arSU9Pf4KSSaoE587pXNyPsu65gYHOyB8+DPeGAF4EpMGWPHd8/vnn7Nmzp0JlOHbsGN7e3uTk5FSoHBJJlWXfPt10sUcNHnvpJZ37/ZdfnqxcFYg02JLnDvEcuLCeBxkkkirLf/8Lv/2mm8f9ODRpAkeP6sazXwCkwb5HYmIikyZNwtraGjMzM+zt7QkMDCyWbt26ddjY2NClSxd8fHy4WWSLObVazXfffYenpycWFhb079+fr776Su/6yMhIRowYgZWVFba2tsybN48/i1Qme3t7li5diqOjI926dePQoUOMGjWKL774gvnz52NlZcWbb76Jv78/KSkpjBo1CgsLCxwcHJRgEoDff/8dLy8vbG1tMTU1xc7OjmXLlpGfn19uneTn57NixQpsbW2xtLRk8uTJZGdnK+dv3LjB/PnzFX24u7tz8uRJ5XxBQQEBAQE4ODhgZmZGly5dGD9+PGlpaXo6W716Nfb29nTv3p0ZM2bw/fffEx0djVqtJjMzs1S5evXqhZmZGQ4ODnz77bd6aSIiInBxccHc3Jy+ffuybNkyvd6yWq3m22+/5d1338Xc3BxHR0cOHDgA6Ny4np6egG7v4DVr1pTo2k1PT0etVnP8+HFAN4wwd+5cfH19efPNN+nSpQv/+Mc/uHTpElOmTKFz58706dOHXbt2larzNWvWMGrUKDZu3EjPnj2xtLTk/fff5/Lly3h7e2NlZUXPnj0JCAjQu27Xrl0MHjyYzp0789Zbb7Fs2TLu3LlTqp4TEhKwt7fnn//8J++99x4WFhb069ePf/3rX8Vk2rVrF/3798fc3BwXFxdiYmKUc7dv32b16tX069cPc3NzNBrNA4cz8vPzCQkJYeDAgcqzCQgIoKCgQEmj1WqZMGECVlZW2NnZERwcTP/+/QkLC+O3335DrVZz9OhRvXw//vhjRo8eXep9Jf/D0dERR0fHihbjwZw4oeshl9K7DktKotuWLbQMDKTbli2EJSWVnI+hoS4ILTb2KQr7DBEScevWLdGjRw8xe/ZskZSUJFJTU8WKFSuESqUScXFxQggh/Pz8hEqlEh4eHuLcuXMiOjpavP3222L8+PFKPiqVSnTt2lXs2rVLpKSkiJUrVwqVSiWOHz8uhBAiPDxcqNVq4evrK5KTk8WRI0eEg4ODcHFxEfn5+UIIIfr06SNMTU3FwYMHxdmzZ8WtW7eEh4eHMDU1Ff7+/kKr1Sqy9O3bV4SHh4sLFy6IcePGCRsbG1FQUCCEEGLQoEHC09NTxMfHC61WK0JCQoRKpRJ79+4VQgixY8cOoVKpRF5eXok6mTt3rlCpVMLHx0ckJyeLY8eOCVtbW+Hl5SWEEKKgoEC4uroKDw8PERMTI5KTk8WaNWvE66+/LqKjo4UQQgQHB4s33nhDREREiMzMTBEVFSX69esnJk6cqKcza2trERsbK2JiYsSNGzfEjBkzxIgRI0R2dra4e/duMdlCQkKEvb29OHXqlMjMzBQbN24UKpVKHDt2TAghxMGDB4WZmZnYvHmzSE9PF1FRUcLR0VF4enrq3dfS0lJs2rRJJCcni5UrVwq1Wi1OnDgh7ty5I3766SehUqnE6dOnxc2bN8WxY8eESqUSaWlpSh5paWl69507d64wNTUVCxcuFGlpaeJf//qXUKlUonv37mL79u0iOTlZzJ07V5iZmYmrV6+WqHc/Pz9hamoqpk+fLpKSkkRERITo1KmT6NatmwgICNCrV7/99puiZ1NTUxESEiJSU1NFeHi46NGjh5g2bVqpei4oKBB9+vQRZmZmYs2aNSI5OVkEBweLjh07it27d+vVEVdXV3H69GmRmJgoRowYIezs7JR6NmXKFNGzZ0+xf/9+kZqaqsgSGhoqhBDF9LZkyRLRpUsXERYWJtLS0sT27duFlZWV+Oyzz4QQunexT58+YuzYsSI+Pl6cOHFCODo6CrVaLXbs2CGEEMLZ2VnMmTNHKVtOTo7o0qWL2LlzZ4k6vZ+MjAyhUqlERkZGudJLKoCvvxZi6lQhFi4s9tvh7i7am5iIFi1aKL/2JiZih7t7ienF+PFC3Ks7zztl1U05rQvIyclh9OjRjBw5krr3dpyZOXMmgYGBJCYmYm5uDkD16tXx9fXl1VdfBWDBggWMGzeO5ORk2t/bl3XIkCEMHjwYAC8vLzZv3kxMTAzdunUjICCA3r178/777wO6aRUrV65k6NChHD58mF69egHQs2dP5f+FtG/fnmnTpgHg6emJv78/Go2G/v37A+Du7s6UKVO4fPky9evXZ/DgwTg4OGBsbAzAmDFjWL9+PYmJieVe6ahRo0Z88sknGBoa0q5dOzQaDYcPHwbg6NGjxMbGcvToUYyMjACYPn06cXFxBAcHY21tjYmJCUuXLsXe3h4AY2NjNBpNsfFpJycnOnfurPxdo0YNqlevTuPGjUuUKzMzk9q1a2NsbEzTpk3x8PCgXbt2yjNYt24dw4cPZ+TIkQCYmJjw6aefMnz4cM6fP0+nTp2UZ+Xu7g7A7NmziY6OZsOGDfj5+dGgQQMAjIyMqFOnTrn0BVCnTh3mz5+PoaEhrVu3ZuXKldjY2DB06FBA9+x27txJWloar5SyKlN+fj6ffPIJDRs2pH379qjVal5++WUmTJgAwKRJkwgICODChQuoVCoCAwN59913GTNmDABt2rShoKCAmTNnkpSURIcOHUrUM4CtrS3Tp08HdPUxLi6O0NBQBhXZunDx4sWo1WoAxo8fz/Tp08nOzubmzZtERESwZs0a+vXrB8DYsWPJyMjg66+/ZtSoUXr3unnzJlu2bGH27Nk4OzsD0Lp1a/7880++/PJLpk+fTnh4OH/88Qfbt29X6tXKlSv15Bk2bBgrVqxg4cKF1K5dm/DwcAAcHByK6fLGjRvcuHFD79jFixdL1HtVISEhAUB5ps+SYcOGlS9hZibk5ED16sVOnbp0idwiHhmAnLt3mR0Zybe//VY8r9xcnXv9Pi9caWzfvr18MlYA0mCj+yiPHDmSH3/8kfj4eLRaLb/de/BFXXUmJiaKsQaUj9+FCxcUY9G2bVvlvIGBAXXr1iUvLw/QvSizZs3Su7eZmRm1a9cmISFBMdKtW7cuJmPRfGvem8doYmJS7Fhubi41a9bE3d2dffv28c0336DVaklISCA7O/uhXOKtWrXC0NBQ+btBgwbcvrf3bXx8PKBzGRclLy9PaSTY29sTGxuLr68v6enppKamkpSURKP7xqVKKu+DcHd358CBA/Tq1YvXX3+dHj168M477yjP5vz585w5c4adO3cWuzY5OVkx2NbW1nrnLC0tiYyMfChZ7sfExERPZzVr1iz1OZXGK6+8QsOGDfWuaVUkUrZoHleuXOGPP/7gjTfe0MujsGwJCQmKwS5JzyXpYP/+/XrHita9+vXrA3Dnzh3lw1/SvTdt2sQff/yhdzw5OZm8vLwS0+fl5ZGcnEx8fDytW7dWjDXoDEu9IgtrODk5sWzZMsLDwxkyZAg7d+5Eo9FQq1atYuULDQ3F39+/2PGqTOFwQkUY7HLz0kulTse631iXdRwhdK7xFwBpsIHLly/j6uqKkZERffv2xc7ODnNz82LTuAzve+iFxq9GkYn5NUqYpC/KCGASQuhdV7OEhQWqlbDw/UulLKp/69Yt3N3dycnJYeDAgTg7O9O5c2elx1leSsq/sCwFBQXUqlWrxPHYQlnXrVvH2rVrcXZ2xtbWlnHjxrF//352796tl76kD+2DaNOmDeHh4URHRxMVFUVkZCTffPMNn332GYMHD6agoABPT88SW/NFGwvV72u95+fnF3vGhRiUMJZW0nSnh3lOpfEk8ih8TkXrVUl6vl8HBQUFxXTwoHpQEoWN3JLehbJkLU3/RalXrx5vv/02u3fvxtbWlmPHjrFly5YS044ZM0bpzRdy8eJFxbMiebaUu/e6cyf88EOJU7q6bdlCVpHYoUKM69Zle0lj82lpMHYs9O79ULI+j8igM+CHH37g2rVrbNmyhalTp9KvXz+uX78O6H+YtFqtXpDZqVOnMDAw4LXXXivXfdRqNSdOnNA7FhcXR05OjtILehIcOXKE+Ph4QkNDmTVrFhqNhjp16jzROcVqtZqcnBxycnJo3bq18tu4cSM//vgjAAEBAUyZMoVFixYxYsQILCwsSEtLK7MBU5JxLEpwcDA///wzPXr04MMPP2TPnj3Y2NgoDQGVSkVqaqqeXDdu3GDp0qVcu3ZNyedM4cYE94iJieH1118vUYZCw1b0+T8Pc4tfffVVXn31Vb1gP4Do6GiAMuvV/Tr4z3/+o+igLAp7aPff+8SJEzRp0kQZViikffv2VK9evURZa9SogYmJCR07diQ9PV3vOaWkpPDXfVs2Dhs2jOjoaHbs2EHbtm2xtLQsUcb69evTsmVLvd9TX0tb8vhYW/9v+dL78La2ptZ9jdpa1arhfZ+3CNAtV2pgoNuQ5AVAGmx0i+Hfvn2bn376iaysLI4cOYKXlxeg77rMzc1l1qxZxMfHExUVxZIlS3B0dNRzVz6IiRMn8uuvv7Jq1SpSUlKIiopizpw5mJmZYWNj80TLA7Bnzx6ysrI4efIkU6dOJS8v74Gu2IehZ8+emJqaMnv2bI4ePYpWq8XX15dNmzYpRqJ58+ZERUVx4cIFUlJS8PX1JTw8vEwZ6tSpw6VLl8jIyCixF3vt2jU++eQTIiIiyMrK4vDhw5w7dw4rKytAN8Z74MAB/Pz8SE1NJTo6mv/7v//j8uXLtGzZUsln48aN7Nmzh9TUVJYtW8Zvv/2mRIcXjlufP3+ev/76i9dee406derw9ddfk56eTnR0NKtXr34iunxcJkyYwNatWwkNDSU9PZ39+/fz6aef0q9fPz13dkns3buXjRs3kpaWxvr169m/f78yVl4W7du3p2/fvnzyySccOHCAtLQ0QkJC2LZtG++9916x9HXr1sXNzQ1/f3927dpFeno6YWFhrF27Fjc3N+rWrYujoyONGjXiww8/5LfffiM2NpYPP/wQ0G9EdevWjRYtWvD111/j4uLyENqSVAqMjXUbeBRpuBXi0qEDy+3sMK5bFwN0Pevldna4lNQ4vXxZZ/xfkF28pEsc3XZz586dY8WKFdy8eRNjY2OGDx9OREQEp0+fxsPDA4BOnTphZmbGmDFjEEKg0Wjw9vYu93369++Pn58f69atY/369TRs2JB+/frh5eVVohv0UbGwsMDHx4eQkBDWrFlD06ZN0Wg0NG/enNOnTz+RexgaGhIUFMSKFSvw8vLi77//pl27dqxatUoJMlu+fDmLFy9m2LBh1KlTh86dO7No0SIWLlyIVqvVG9styvDhwzl06BAajYaNGzcW6z3NmDGDgoIClixZwuXLl2ncuDEjRoxgypQpgE7Pq1ev5uuvvyYgIID69evTq1cvPvzwQ72P/rvvvsvGjRtJSEhApVKxfv16pXfZsWNH7O3t+eCDD3j33XeZN28eK1eu5IsvvuCdd96hXbt2+Pj4MHbs2Ceiz8dh7Nix1KxZk5CQEFasWEHjxo0ZOnQoU6dOLfPawYMH8+uvv7J8+XJat27NF198QZ8+fcp97y+//JJVq1bxj3/8g+vXr9O2bVsWLVqkBNndj7e3N0ZGRvj5+ZGdnY2xsTHTpk1TGko1atRg/fr1LF68GFdXVxo0aMDkyZM5e/asnvvewMAAZ2dnvvrqKyXIU/ICYWAAGg2sWqUztvd5vFw6dCjZQBdFCF3g2r2AyBcBA1GWf1IieQFRq9UsWbKE4cOHV7QoFYa9vT1OTk588MEHFS2KQmZmJsnJyXqzJC5dusRbb73F5s2b9QLWFixYwJUrV4qtdVCee/Tt25eIiAg9j0tVoXDdh/uHLJ478vLgyy9123M+7PKkQoBWC926waRJJe/u9RxSVt2sHKWQSCRVgjt37jBlyhTWr19PRkYG8fHxzJ8/nzZt2iizMqKioti4cSO7du16LjwclY0GDRo8/8YadFO6pk2DFi10+2iXt29ZaKxVKhg3rtIY6/Lw4pREIpFUetq3b8+XX37JDz/8gKOjI56entSuXZvg4GDFJb5jxw6+/PJLJk6cWGxamqRszp49y9mzZytajPJRty7MnaszvmlpcOVK6Ya7oACysyE1Fd54Az74AF5++ZmK+7SRY9iSKknhHOKqzC/P6aYIAwYMeODiPl988cUzlObFozBK38zMrIIlKSd168Ls2brtMvftg4QE3Zj2yy/r5lfn58OdOzqD3bkzvP02dOr0QvWsC5EGWyKRSCTPN9WqgaWl7vff/8KpU7oI8JwcqFNHt8mHtTWUsjrii4I02BKJRCKpPLRooftVQV48n4FEIpFIJC8g0mBLJBKJRFIJkC5xiUQiqUK4urpWtAiSR0QabIlEIqlC1K5du6JFkDwi0iUukUgkVYjY2FhiY2MrWgzJIyANtkQikVQhpMGuvEiDLZFIJBJJJUAabIlEIpFIKgHSYEskEolEUgmQBlsikUgkkkqAnNYlkUgkVQh3d/eKFuHJIgSkp0NKkhcW+QAAIABJREFUCvz9N9SqBcbGuh2+DA0rWronijTYEolEUoUo3Ka00iOEbhOQ77/XGWz43+5dBgZgZATvvAN2drq9tV8ApMGWSCSSKsSJEycAKvde4gUFsGWLbrvNRo2gdWudkS7KzZsQHAyxsTBliq7nXcl5KmPYa9as4a233noaWVcIYWFhqNVq7t69W+J5b29v3NzcnrFUOq5evcq2bdue+X23bduGWq1+5vctD09LJ3///TebNm164vk+CU6ePKnscwyQlZXFjz/++MTyf57f6Sdd1hedc+fOce7cuYoW4/HYtUtnrNu0gYYNixtr0O2j3a4dnDkDgYG6nnclRwadVXI+//xz9uzZU9FiPFc8LZ0EBgbyzTffPPF8H5f8/Hzc3d3RarXKsQ8//JDDhw9XoFTPjqpUVglw6ZLODd66ddlj1AYGYGICJ09CZW+kIF3ilR4hREWL8NzxtHTyvOq6JLmeV1mfBlWprBLg3/+Gl16CavrmKywpiaUnTvDfmzdpUbcu3tbWuHTooDPa9evreuQWFhUk9JPhkXrYiYmJTJo0CWtra8zMzLC3tycwMLBYunXr1mFjY0OXLl3w8fHh5s2byjm1Ws13332Hp6cnFhYW9O/fn6+++krv+sjISEaMGIGVlRW2trbMmzePP//8Uzlvb2/P0qVLcXR0pFu3bhw6dIhRo0bxxRdfMH/+fKysrHjzzTfx9/cnJSWFUaNGYWFhgYODA4cOHVLy+f333/Hy8sLW1hZTU1Ps7OxYtmwZ+Q/hQsnPz2fFihXY2tpiaWnJ5MmTyc7OVs7fuHGD+fPnK/pwd3fXc2EWFBQQEBCAg4MDZmZmdOnShfHjx5OWlqans9WrV2Nvb0/37t2ZMWMG33//PdHR0ajVajIzM0uUbdu2bTg5OWFhYYGlpSVubm7ExcXp6TEgIIAZM2ZgZWVF7969WbJkid4QwP79+3FycsLc3BwPDw9+//33MnUSEhKCvb09FhYWeHp64u/vj729PQCZmZmo1WrWrVtHz5496d27N1evXiU3N5fly5djZ2eHpaUlQ4cO5cCBA+Uuj7e39xPTSdG65erqyrp168jKykKtVnP8+HEATp06hbu7OxYWFvTq1Yt58+Zx7dq1UvM5cOAAarVarzcMMHbsWHx8fEqUtbDu2NnZYWpqiq2tLR999BG3bt0CwNTUFAAfHx9GjRrFqFGj+M9//sPOnTuVYYuy8gDIyMhgxowZvPHGG3Tr1o0ZM2Zw8eJFPVmCgoLo1asXFhYWeHh4kJKSopxTq9Xs3r2b0aNHY2FhQd++fYmIiOCXX35h4MCBdO7cmZEjR+qVvaz3Ys2aNYwaNYrAwEDeeustunbtytixY0lOTgYosaySF5i7d2H/fmjSRO9wWFIScw4fJuvmTQSQdfMmcw4fJiwpSZegUSM4fx4uX372Mj9JxENy69Yt0aNHDzF79myRlJQkUlNTxYoVK4RKpRJxcXFCCCH8/PyESqUSHh4e4ty5cyI6Olq8/fbbYvz48Uo+KpVKdO3aVezatUukpKSIlStXCpVKJY4fPy6EECI8PFyo1Wrh6+srkpOTxZEjR4SDg4NwcXER+fn5Qggh+vTpI0xNTcXBgwfF2bNnxa1bt4SHh4cwNTUV/v7+QqvVKrL07dtXhIeHiwsXLohx48YJGxsbUVBQIIQQYtCgQcLT01PEx8cLrVYrQkJChEqlEnv37hVCCLFjxw6hUqlEXl5eiTqZO3euUKlUwsfHRyQnJ4tjx44JW1tb4eXlJYQQoqCgQLi6ugoPDw8RExMjkpOTxZo1a8Trr78uoqOjhRBCBAcHizfeeENERESIzMxMERUVJfr16ycmTpyopzNra2sRGxsrYmJixI0bN8SMGTPEiBEjRHZ2trh7924x2cLDw4WpqanYuXOnyMzMFDExMWLo0KHinXfeUdL06dNHmJmZidDQUJGamipCQkKEWq0WYWFhQgghTp06JdRqtVi1apVISUkRW7duFWZmZkKlUpVaTzZt2iTMzc3Fd999J1JSUsTatWtFx44dRZ8+fYQQQmRkZAiVSiX69esnEhISxOnTp4UQQsyaNUs4OTmJqKgokZaWJjZu3CjMzMzE7t27y1WeJ6mTonXrr7/+Ep9++ql46623RHZ2trhz546Ij48X5ubmwt/fX6SkpIiYmBjh7u4u3nnnHZGbm1tqPj169BCrV69W7vXf//5XdOzYUZw4caJEXU6ePFkMGTJExMbGioyMDLF7925hamoq1q9fL4QQ4uLFi0KlUomQkBBx7do1ce3aNTFs2DAxa9YskZ2dXa48bty4Iezs7MTYsWNFXFycOH/+vBg5cqQYPHiwEOJ/7/SkSZPE+fPnxenTp4WDg4MYMWKEIqdKpRLdunUT+/btE2lpaWLixInCyspKODs7i5iYGBEbGyt69eolZsyYIYQo33vh5+cnTE1NxYQJE8T58+fFiRMnRJ8+fYSHh4cQQpRY1qL8+eefIiMjQ+934sQJoVKpREZGRmnV94UmODhYBAcHP/P7Dh069PF/gweLoaamYmjXrnq/Ni1bihYtWhT7tWnZ8n/pTE3F0IEDn4gcT4vC72JpdfOhXeI5OTmMHj2akSNHUrduXQBmzpxJYGAgiYmJmJubA7qpA76+vrz66qsALFiwgHHjxpGcnEz79u0BGDJkCIMHDwbAy8uLzZs3ExMTQ7du3QgICKB37968//77ALRr146VK1cydOhQDh8+TK9evQDo2bOn8v9C2rdvz7Rp0wCUnp1Go6F///6Abh7ilClTuHz5MvXr12fw4ME4ODhgbGwMwJgxY1i/fj2JiYkMGDCgXHpp1KgRn3zyCYaGhrRr1w6NRqOMqx09epTY2FiOHj2KkZERANOnTycuLo7g4GCsra0xMTFh6dKlSg/U2NgYjUZTbCzWycmJzp07K3/XqFGD6tWr07hx4xLlatiwIUuWLGHIkCFKvq6ursyfPx8hBAb3gjV69OjB6NGjAWjTpg3bt28nJiYGZ2dnNm3aROfOnZk1axYAbdu2JSEhgc2bN5eqj6CgINzd3Rk+fDgAU6ZMIT4+vliwi5ubGyqVCoD09HT27t3L9u3blXrUunVrkpOTCQoKYtCgQWWWp169ek9MJ/fXrVq1amFoaKjkGxQUhK2trVLXAFavXk2PHj2IjIykX79+JeYzePBgdu/ezYwZMzAwMGDXrl20atWKN954o0R5e/TowRtvvEHHjh0BaNmyJd9++y2JiYmAru4B1KtXj4YNGwJQrVo1atasqchaVh4//vgjf/75J19++SWvvPIKAJ9++infffed4hkzNDRk+fLl1K9fH4B3332XlStX6sk6aNAgHBwcAN2+ywcPHsTLywtLS0sABgwYwK+//gqU770AyMvLY9myZYpcHh4efPnll8qzvL+sRQkNDcXf379EvVZVxo4dW9EiPDqlDH/kFhSU73glHz55aINtZGTEyJEj+fHHH4mPj0er1fLbb78BOrduISYmJoqxBhQjc+HCBcVgt23bVjlvYGBA3bp1ycvLAyAhIUExEIWYmZlRu3ZtEhISlA9g69ati8lYNN+aNWsq8tx/LDc3l5o1a+Lu7s6+ffv45ptv0Gq1JCQkkJ2d/VAu8VatWmFYJACiQYMG3L59G4D4+HgA+vbtq3dNXl6e0kiwt7cnNjYWX19f0tPTSU1NJSkpSfkYF1JSeR+EtbU1DRs2xN/fn9TUVNLT00lISAB0bvxq98aBiuoM0HsWiYmJ2NjY6J23srIq1WBfu3aNrKwsrKys9I537dq1mMEuWp5CPRU2HAoplONhyvMgyptHWbo+f/486enpxcophCA5OVkx2PfnM2zYMNavX8/JkyextrZm165duLi4lHofNzc3fvnlF3bu3El6ejpJSUlkZmbSsmXLMsta3jwSExMxMTFRjCLoGm5z5sxR/jYyMlKMNUD9+vXJy8sjPz9fqftF61Gte9Noipb/5ZdfVp5ned4LgFdeeUVPrnr16unViQcxZswYnJ2d9Y5dvHjxxVs8pBKwffv2x8/k9m2YOhVattSNY9+j25YtZBUZci3EuG5dtjs66v5IT4d//EMXrFZJeWiDffnyZVxdXTEyMqJv377Y2dlhbm5ebMqH4X3Re4XGr0aNGsqxov8vRJTRAhJC6F1XaHyLUtJH+6WXSh6uv3XrFu7u7uTk5DBw4ECcnZ2VsbaHoaT8C8tSUFBArVq12LVrV6myrlu3jrVr1+Ls7IytrS3jxo1j//797N69Wy99rYecS7hnzx68vb1xcnLCysoKNzc3EhISWLx4sV66sp7F/c/lQYax8FxZzxL0y1OYfsOGDXqGoSjlLc+DKG8eJdWtohQUFKDRaPR62IU0aNCg1Hzatm1L165d2b17N9WrV0er1RYzKkXvMXHiRC5cuMCgQYPQaDSYmZkxb9688ha3XHmUZzGN0t6hos+5pHwMSppyQ/neCyi5bpaX+vXrl1qXqipRUVEAdO/evYIleQRq1gRLS914dNOmymFva2vmHD5MTpG4m1rVquFdONf85k145RWdoa/EPLTB/uGHH7h27Ro///yz8iIV9k6KvrharZabN28qbvNTp05hYGDAa6+9Vq77qNVqTpw4wfjx45VjcXFx5OTk0KFDh4cVu1SOHDlCfHw8hw4doum9CnD9+nWuXLnyxO6hVqvJyckhJydHcUkCLFmyhCZNmjBx4kQCAgKYMmUKU6ZMUc4HBgaWafRK+xgWzcPFxYUlS5YoxwqDuMpjUAE6depETEyM3rEzZ86Umr5evXoYGxtz+vRpxT0K6AV1lUShazw7O1txiYOuMXP9+nW8vb3LVZ6npZP781WpVCQlJWFiYqKcu3jxIgsWLGD69OmKe7okhg0bxvLly6lbty49evRQ6t79nD9/niNHjvDtt9/StWtXQNcD1Wq1NG/evNTyFj1Wnjzat2/P1q1b+fPPP5XGRmpqKq6urmzZsqXUcjwO5XkvykNZz1uiT+EwSKU02AD9+8N//qN3yOWeTSgxShzgyhXw8Kj0S5U+dJR4s2bNuH37Nj/99BNZWVkcOXIELy8vQOdiLiQ3N5dZs2YRHx9PVFQUS5YswdHRkVatWpXrPhMnTuTXX39l1apVpKSkEBUVxZw5czAzMyvmnn0cmjVrBuh6XVlZWZw8eZKpU6eSl5enV57HoWfPnpiamjJ79myOHj2KVqvF19eXTZs2KY2P5s2bExUVxYULF0hJScHX15fw8PAyZahTpw6XLl0iIyOjxIVdmjdvTmxsLGfPnkWr1RISEqIs/lHe8o0bN47ExESWLVtGamoqu3fvZuvWrQ+8ZsKECXz77beKCzYkJIS9e/c+8JoOHTpgb2/PokWLOHDgABkZGWzYsAE/Pz/FdVue8jwtndSpU4c///yTlJQU7ty5w/jx40lMTGThwoUkJSURFxfHrFmzSEpKUoZ9SmPAgAHk5uayZcsWhg4dWmq6V199lWrVqrF3714yMjI4c+YM77//PpcvX1bcwoaGhtSsWZOkpCSloVmnTh0yMzPJysoqVx5OTk688sorfPjhh5w/f574+Hg+/vhjWrduXWZZHpXyvBfloWhZJVUAtRo6dID7nrdLhw5Eu7mROWEC0W5u/zPWV6/qpnV161YBwj5ZHtpgDxgwgPfee48VK1ag0Wj47LPPGDZsGNbW1pw+fVpJ16lTJ8zMzBgzZgwzZ86kZ8+eD+W27N+/P35+fkRGRjJo0CDmzJmDjY0NwcHB5RqnLC8WFhb4+PiwefNmBg4ciI+PD9bW1jg6OuqV53EwNDQkKCiIzp074+XlhaOjI5GRkaxatUoJMlu+fDm5ubkMGzYMDw8PEhMTWbRoEdevXy82Bagow4cP5+7du2g0Gs6ePVvs/Pz582ncuDGjRo1i+PDh/Prrryxfvhwou8dbSKdOnQgMDOT48eMMGjSI0NBQJk+e/MBr3NzcmDx5Mr6+vjg6OnL48GGcnZ3LdL36+vqi0WhYtGgRGo2GLVu2MG/ePDw8PMpdnqelE41GQ/PmzRk0aBARERFYWFgQFBREUlISLi4uTJgwgSZNmhAaGkqdOnUeWM7atWszcOBAatasWWwMtyhNmzZl6dKlHDp0CI1Gw6xZs2jatCljx47lzJkzStzIxIkT2bJlC56engDKlCuNRgNQZh61atUiKCgIgJEjR+Lp6Unz5s355z//+dR6sOV5L8pD0bJeunTpqcgqeY546SWYPl03VSszU7dMaUkIAdnZkJcHs2dDvXrPVs6ngIEor19UInkIIiMjad++vV5g1Pz589FqtYSGhlagZM8P48ePp3379nz00UcVLUqVIjMzU5kj/jCBey8KISEhQCWPFge4cQNCQnTucUND3WYf1arpliC9dk1nqNu0gYkToUWLipa2XJRVN+VKZ5Knwp49e0hKSmLRokU0btyYEydOsGfPHhYuXFjRolU4ERERJCQkcPz4cakPyTPnhdmtq359mDlTt1TpkSO65UdzcuDll6F7d+jTp+RNQSox0mBLngrz589n6dKlTJs2jRs3bmBiYsJHH330wOlLVYXAwEBSUlJYuHCh3nRDieRZ8MJNaWvaFIYO1f1ecKTBljwVGjZsyNKlSytajOeSsgL2JBKJpCTkbl0SiURShYiMjCQyMrKixZA8AtJgSyQSSRUiNTWV1NTUihZD8ghIgy2RSCQSSSVAGmyJRCKRSCoB0mBLJBKJRFIJkFHiEolEUoWoXbt2RYsgeUSkwZZIJJIqhKura0WLIHlEpEtcIpFIJJJKgDTYEolEUoU4cOCAsp2spHIhXeISiURShcjMzKxoESSPiOxhSyQSiURSCZAGWyKRSCSSSoB0iUskEomkcnLnDsTFwYED8N//QkEBNGgAvXrBm29Cw4YVLeETRRpsiUQiqULUr1+/okV4Mhw9Chs3wq1bur2x69XT7X195w5s3Qr/+hf07QuurvCC7AH+2C7xNWvW8NZbbz0JWZ4LwsLCUKvV3L17t8Tz3t7euLm5PWOpdFy9epVt27Y98/tu27YNtVr9zO9bHipKJ08Se3t7fH19H/n6vLw8goOD9Y798ssvJCUlPa5oCo8r49PkSZf1RcfFxaXy70sfHg5r1+qMdJs2YGQENWroDHPdutC6NRgb69J99RXk5VW0xE8EOYZdifj888/Zs2dPRYvxXCF1Art27dLbe1yr1TJlyhSuXLlSgVI9G6pSWSX3OHsWNm2CVq2gVq3S0xka6oz5f/4DO3Y8M/GeJtJgVyKEEBUtwnOH1ElxqpJOqlJZnxT79u1j3759FS3GoyEE7NqlG5uuUaPs9AYGYGIC+/fDX389ffmeMmUa7MTERCZNmoS1tTVmZmbY29sTGBhYLN26deuwsbGhS5cu+Pj4cPPmTeWcWq3mu+++w9PTEwsLC/r3789XX32ld31kZCQjRozAysoKW1tb5s2bx59//qmct7e3Z+nSpTg6OtKtWzcOHTrEqFGj+OKLL5g/fz5WVla8+eab+Pv7k5KSwqhRo7CwsMDBwYFDhw4p+fz+++94eXlha2uLqakpdnZ2LFu2jPz8/HIrLT8/nxUrVmBra4ulpSWTJ08mOztbOX/jxg3mz5+v6MPd3Z2TJ08q5wsKCggICMDBwQEzMzO6dOnC+PHjSUtL09PZ6tWrsbe3p3v37syYMYPvv/+e6Oho1Gp1qXMpt23bhpOTExYWFlhaWuLm5kZcXJyeHgMCApgxYwZWVlb07t2bJUuW6A0B7N+/HycnJ8zNzfHw8OD3338vUychISHY29tjYWGBp6cn/v7+2NvbA7p5n2q1mnXr1tGzZ0969+7N1atXyc3NZfny5djZ2WFpacnQoUOLLejwoPJ4e3s/UCc5OTl07dqVjRs36h0PDQ3lzTffJDc3F4CIiAhcXFwwNzenb9++LFu2jJycHCV9ZGQkLi4udO7cGRsbG+bOncv169eV80FBQfTr1095P/z9/SkoKFDOnzp1Cnd3dywsLOjVqxfz5s3j2rVrZeq0kJMnTzJ69Gi6dOmCmZkZAwcOZOfOnYBuCGfevHmArs6EhYXx9ttvAzB69Gi8vb3LzKOQn376iSFDhmBhYYG9vT1ff/21nkG8cuUKM2fOVN61ovUmLCwMe3t7fvrpJ716cOnSJT7//HO6deumXFM0z7J086D6mpmZWWJZJQ/m4sWLXLx4saLFeDSysiApCV55Re9wWFIS3bZsoWVgIN22bCGs6BBJtWq6YLTo6Gcs7FNAPIBbt26JHj16iNmzZ4ukpCSRmpoqVqxYIVQqlYiLixNCCOHn5ydUKpXw8PAQ586dE9HR0eLtt98W48ePV/JRqVSia9euYteuXSIlJUWsXLlSqFQqcfz4cSGEEOHh4UKtVgtfX1+RnJwsjhw5IhwcHISLi4vIz88XQgjRp08fYWpqKg4ePCjOnj0rbt26JTw8PISpqanw9/cXWq1WkaVv374iPDxcXLhwQYwbN07Y2NiIgoICIYQQgwYNEp6eniI+Pl5otVoREhIiVCqV2Lt3rxBCiB07dgiVSiXy8vJK1MncuXOFSqUSPj4+Ijk5WRw7dkzY2toKLy8vIYQQBQUFwtXVVXh4eIiYmBiRnJws1qxZI15//XURHR0thBAiODhYvPHGGyIiIkJkZmaKqKgo0a9fPzFx4kQ9nVlbW4vY2FgRExMjbty4IWbMmCFGjBghsrOzxd27d4vJFh4eLkxNTcXOnTtFZmamiImJEUOHDhXvvPOOkqZPnz7CzMxMhIaGitTUVBESEiLUarUICwsTQghx6tQpoVarxapVq0RKSorYunWrMDMzEyqVqtR6smnTJmFubi6+++47kZKSItauXSs6duwo+vTpI4QQIiMjQ6hUKtGvXz+RkJAgTp8+LYQQYtasWcLJyUlERUWJtLQ0sXHjRmFmZiZ2795drvKURycff/yxGDZsmN4xFxcX8cknnwghhDh48KAwMzMTmzdvFunp6SIqKko4OjoKT09PIYQQV65cEaampmLTpk0iMzNTnDx5Utjb24u5c+cKIYTYv3+/sLa2FkeOHBFZWVnixx9/FKampmLHjh1CCCHi4+OFubm58Pf3FykpKSImJka4u7uLd955R+Tm5irP5MsvvyxRtxcvXhQWFhZi6dKlIi0tTVy4cEHMnTtXmJqaikuXLomcnBzxzTffCJVKJbKzs8Xff/8tYmJihEqlEj///LO4ceNGmXkIIcTevXuFWq0WX3/9tUhNTRU///yzsLS0FMHBwYqMnTp1EiEhIUKr1Ypt27YJlUoltm7dKoTQvTempqbC3d1dnD9/Xpw4cUJ07dpVdOvWTXz66aciJSVFhIaGCpVKJX755ZeH0k1p9fXu3bvFyloeCutjRkZGudK/aAQHByvPtdLx009CeHoKsXCh8tvh7i7am5iIFi1aKL/2JiZih7v7/9J98IEQCxZUrOzloKy6+cAo8ZycHEaPHs3IkSOpW7cuADNnziQwMJDExETMzc0BqF69Or6+vrz66qsALFiwgHHjxpGcnEz79u0BGDJkCIMHDwbAy8uLzZs3ExMTQ7du3QgICKB37968//77ALRr146VK1cydOhQDh8+TK9evQDo2bOn8v9C2rdvz7Rp0wCUnp1Go6F///4AuLu7M2XKFC5fvkz9+vUZPHgwDg4OGBsbAzBmzBjWr19PYmIiAwYMKFcjp1GjRnzyyScYGhrSrl07NBoNhw8fBuDo0aPExsZy9OhRjIyMAJg+fTpxcXEEBwdjbW2NiYkJS5cuVXqgxsbGaDSaYmOxTk5OdO7cWfm7Ro0aVK9encaNG5coV8OGDVmyZAlDhgxR8nV1dWX+/PkIITAwMACgR48ejB49GoA2bdqwfft2YmJicHZ2ZtOmTXTu3JlZs2YB0LZtWxISEti8eXOp+ggKCsLd3Z3hw4cDMGXKFOLj4zl37pxeOjc3N1QqFQDp6ens3buX7du3K/WodevWJCcnExQUxKBBg8osT7169crUiYuLC25ubqSlpdGmTRtSUlI4e/YsixcvBnSeoeHDhzNy5EgATExM+PTTTxk+fDjnz58HdEFdzZo1w9jYGGNjY9atW6d4ZDIzM6lRowbNmzenRYsWtGjRgiZNmtC8eXNFN7a2tkodBVi9ejU9evQgMjKSfv36lapXgNzcXKZPn8748eN56SWdQ2zy5Mns3LmTlJQUbGxslHezUAev3Ot9NGjQgHr16nH9+vUH5tGkSROCg4NxcHBg4sSJgK5e3Lp1S89T0LdvX8aMGQNAq1at2LBhA2fPnmXEiBGKnj7++GM6duwIgI2NDXFxcXh7e/PSSy/Rtm1bvvrqKxITE+nTp0+5dfOg+np/We/nxo0b3LhxQ+9Ype1dVnKGDRv2+JlcugTXr+umct3j1KVL5BappwA5d+8yOzKSb3/7TXegoED3u++b9Chs3779sfN4VB5osI2MjBg5ciQ//vgj8fHxaLVafrungKIvsomJiWKsAcXIXLhwQTHYbdu2Vc4bGBhQt25d8u5F7iUkJCgGohAzMzNq165NQkKCYqRbt25dTMai+dasWVOR5/5jubm51KxZE3d3d/bt28c333yDVqslISGB7Ozsh3KJt2rVCkNDQ+XvBg0acPv2bQDi4+MB3cetKHl5eUojwd7entjYWHx9fUlPTyc1NZWkpCQaNWqkd01J5X0Q1tbWNGzYEH9/f1JTU0lPTychIQHQufGrVdM97qI6A/SeRWJiIjY2NnrnraysSjXY165dIysrCysrK73jXbt2LWawi5anUE+FH+JC8opEc5a3PA+iS5cutG3blu+//54ZM2awe/du1Go1pqamAJw/f54zZ84Ucw8DJCcn4+joiKOjI1OnTqVx48Z0796d3r17K427QYMGsWPHDgYOHEiHDh3o3r07AwYMUJ71+fPnSU9PL6YfIQTJycllGuxWrVrrdBdtAAAgAElEQVTh4uLChg0buHDhQqnv4OPmUVKDtbChVEibNm30/i5a7wu5/31s1aqV0kgAePnll5WhiPLq5kH1tSxCQ0Px9/cvV1pJJeBep6Mo9xvrEo8LUeK1lY0HfvEuX76Mq6srRkZG9O3bFzs7O8zNzYtN4ypqvADF+NUoEhRQo4QAAVFGwIgQQu+6QuOrV4ASPtpFPxBFuXXrFu7u7uTk5DBw4ECcnZ3p3Lmz0rsqLyXlX1iWgoICatWqxa5du0qVdd26daxduxZnZ2dsbW0ZN24c+/fvZ/fu3Xrpaz0oArIE9uzZg7e3N05OTlhZWeHm5kZCQoLSmyykrGdx/3N5kGEsPFfWswT98hSm37BhQ6nzQstbnrJwcXFh+/btTJ8+nR9++IFRo0Yp5woKCvD09Cyx9V/YgPriiy+YNm0ahw4d4ujRo8ydO5ewsDACAwMxMjJi9+7dxMTE8O9//5t///vfbNy4kQ8++IBJkyZRUFCARqPR60UW0qBBgzJlT0pKws3Njddff50ePXrw9ttvY2Rk9FC9lfLkUb0c81QfVO8Lub+ulPYuAuXWzaN8OwoZM2YMzs7OescuXryIu7t7ua5/Ebm/Y/CseCI908OH4ZtvdNO27tFtyxayisRMFWJcty7bHR11f1y+DK+9Bve8uJWVBxrsH374gWvXrvHzzz8rL01hD6foC6PVarl586bimjt16hQGBga89tpr5RJCrVZz4sQJxo8frxyLi4sjJyeHDh06PFyJHsCRI0eIj4/n0KFDNG3aFIDr168/0SkharWanJwccnJyFNcgwJIlS2jSpAkTJ04kICCAKVOmMGXKFOV8YGBgmR8hgzJaiIGBgbi4uLBkyRLlWGEQV3k/cJ06dSImJkbv2JkzZ0pNX69ePYyNjTl9+jQODg7K8aKBbiVR6BrPzs5WXOKga8xcv34db2/vcpWnLJ0ADB48mFWrVrFlyxYuXbqEk5OTnhypqal6vf8zZ86wdu1aPvroIxISEti3bx8ff/wx7dq1Y+zYsfzwww/Mnj2b7OxsoqKi+Pvvv/Hw8KBr167MnDmTBQsWEBYWxqRJk1CpVCQlJWFiYqLIevHiRRYsWMD06dNpWMZKTFu3buWVV14hNDRUOfbLL7/o6eB+7tdJefJo165dsefs6+urDOU8DR5XN1D2869fv/6Ls1DIE6Jo/a90WFrqpmvdvasLJgO8ra2Zc/gwOUUCZ2tVq4a3tfX/rrt1S7eISiXngVHizZo14/bt2/z0009kZWVx5MgRvLy8ABS3VuH/Z82aRXx8PFFRUSxZsgRHR0datWpVLiEmTpzIr7/+yqpVq0hJSSEqKoo5c+ZgZmZWzD37ODRr1gzQ9dyysrI4efIkU6dOJS8vT688j0PPnj0xNTVl9uzZHD16FK1Wi6+vL5s2bVIaH82bNycqKooLFy6QkpKCr68v4eHhZcpQp04dLl26REZGRokLuzRv3pzY2FjOnj2LVqslJCSETZs2AZS7fOPGjSMxMZFly5aRmprK7t272bp16wOvmTBhAt9++y07d+4kPT2dkJAQ9u7d+8BrOnTogL29PYsWLeLAgQNkZGSwYcMG/Pz8aNmyZbnLU5ZOAJo2bUqPHj1YuXIlb731ll4PY9KkSRw4cAA/Pz9SU1OJjo7m//7v/7h8+TItW7akfv36bNmyhRUrVpCWlsaFCxf44YcfaNWqFa+++qoS6b5r1y4yMzM5efIkx48fV9y848ePJzExkYULF5KUlERcXByzZs0iKSlJGS56EM2aNSM7O5uDBw+SlZVFeHg4//jHP4rpAHSNpL///lv5OyEhgWvXrpUrj0mTJrFv3z5CQkLQarXs27ePDRs2lOmyfxweVzdAsbJKXnDq1YOePaFIHIJLhw4st7PDuG5dDND1rJfb2eFS2Nm7eVO3XGmnThUj8xPkgQZ7wIABvPfee6xYsQKNRsNnn33GsGHDsLa25vTp00q6Tp06YWZmxpgxY5g5cyY9e/Z8KLdl//798fPzIzIykkGDBjFnzhxsbGwIDg4u1zhlebGwsMDHx4fNmzczcOBAfHx8sLa2xtHRUa88j4OhoSFBQUF07twZLy8vHB0diYyMZNWqVUqQ2fLly8nNzWXYsGF4eHiQmJjIokWLuH79OlqtttS8hw8fzt27d9FoNJw9e7bY+fnz59O4cWNGjRrF8OHD+fXXX1m+fDlQdo+3kE6dOhEYGMjx48cZNGgQoaGhTJ48+YHXuLm5MXnyZHx9fXF0dOTw4cM4OzuX6Wb19fVFo9GwaNEiNBoNW7ZsYd68eXh4eJS7PGXppBAXFxf+/vvvYis89e/fn9WrV3Pw4EGcnJx4//336dKlCwEBAYqXaM2aNRw/fhxnZ2feffddDAwMCAoKwtDQkBEjRjBz5kzWrl3LwIEDef/99+nevTvz588HdHUuKCiIpKQkXFxcmDBhAk2aNCE0NFQxNg9i9OjRaDQa5s6di6OjI//85z/x8vJSvBqgayR26dKFkSNHsnXrVho1asSIESNYsWIFPj4+5cqjT58+LFmyhO+++w6NRsMXX3zBrFmznqrr+HF1AxQrq6Rsvv/+e77//vuKFuPRcXLSLZhSxDPq0qED0W5uZE6YQLSb2/+M9e3bkJ0NY8cqPfLKjIEor69UIimFyMhI2rdvr/SMQWdstVqtnhtWInkeyMzMpG/fvkREROjV2apCSEgIAGPHjq1QOR6LjAxYuVLXe27WrPha4QUF8McfOlf4e+/peuWVgLLqZuVvckgqnD179pCUlMSiRYto3LgxJ06cYM+ePSxcuLCiRZNIJC8irVrBggWwdy9ERurWCjc01EWC372r+9fCQtcbf4JxUBWNNNiSx2b+/PksXbqUadOmcePGDUxMTPjoo48q/wYDEonk+aVRI/DwAGdn3bzsy5d1xrphQzAzgyZNKlrCJ4402JLHpmHDhnqbT0gkEskzo04dsLWtaCmeCdJgSyQSSRWicLaMpPIhDbZEIpFUIcq7BLPk+UNurymRSCQSSSVAGmyJRCKpQoSFhREWFlbRYkgeAekSl0gkkirE/buXSSoPsoctkUgkEkklQBpsiUQikUgqAdIlLpFIJM8Jt2/f5sqVK+Tl5WFoaIiRkVG511WXvPhIgy2RSCQVyOXLlzly5AjR0dFcunRJ2UNcCEFBQQGNGjWia9eu2NnZYWxsXK4tZR9EVVw//UVBGmyJRCKpAK5fv87WrVs5fvw4L730EkZGRrRq1UrPIAshuH37NhEREfz888+YmpoyevRomjZt+sj3fZpbpkqeLtJgSyQSyTMmJiaGwMBAcnNzadWqldKrvh8DAwNq1apFy5YtEUKQnJzMxx9/jLu7O717937s3rakciENtkQikTxDDh8+TGBgIE2bNqXJQ2xQYWBgQLNmzbhz5w7ffPMN169fZ8iQIQ9ttL/77jsAXF1dH+o6ScUjo8QlEonkGREXF0dgYCAtWrR45GCyl19+mdatW7Njx47/b+/Ow6Kq/j+Av1kGUCYEA7QUEUkQ2RxkAAVEUVKJLyJhhiCIKyZCiaAikSWJG5qCPiJCWJAbGtpXSoNiSZLFytxCyZJ9KUFFdub+/uDL/TkOywAjiH5ez8PzwLnnnvO5l3nmM+feM+fixx9/7PH+dXV1qKur61XfZGBRwiaEkH7w6NEjREdHQ01NDQoKCn1qS1ZWFhoaGkhISEBZWZmEIiTPO0rYhBDSD06fPo26ujq88sorEmlPXl4ecnJyiIuLA8MwEmmTPN8oYZMei4iIwLRp0wY6DIk5c+YMdHV10dLS0uH2jRs3wtXVtV9jevz4MeLj4/vUxkDELQlP/z9KSkpw/vz5AY6qb2pqapCRkYHXX39dou2qq6sjPz8fhYWFEm2XPJ9o0hkh3di8eTNaW1v7tc/o6GicO3cO7u7uvW5jIOKWBHt7e1hbW0NWtu3tKSAgAGPGjMFbb701wJH1Xk5ODhiGgYyMjETblZKSAofDQVpaGjw9PcXaR0tLS6IxkP5DCZuQbkjqEmZPSOIS50DELQkKCgpC93hfhMu9V65cwbBhw8SqW1BQgNzcXNTW1oLL5YLP5+ONN97otL6qqip++eUXeHh4iDVj3MbGRuy4yfOFLokTEbdv38aqVavA5/NhYGAAW1tbREdHi9Q7dOgQLCwsYGJigk2bNqG2tpbdpquri5MnT8LLywtGRkaws7PDgQMHhPZPT0/HwoULwePxMGXKFAQHB+PBgwfsdltbW2zfvh0ODg4wMzNDRkYGFi9ejPDwcHz44Yfg8XgwNzdHZGQk7t69i8WLF8PIyAizZ89GRkYG205ZWRnWrVuHKVOmQF9fH9bW1tixY4fYo88nLy1nZ2dDV1cX6enpcHBwgLGxMZycnJCSksLWX7x4MXbs2IGgoCCYmprC0tISAQEBQucnJSUFCxYswKRJk2BoaIj58+cjPT0dQNsth0OHDqGkpAS6urrIzs4G0Pam7+bmBiMjI9jY2CA4OBjV1dU9ijs7Oxv29vYwNDTEggUL8Ndff+HQoUOwsrKCqakp/P390djYCKDt0rStrS2Sk5Nha2sLIyMjeHl5oaKiAmFhYTAzM4O5uTlCQ0PZpNrR7ZJTp05BV1dX6P96+PBhrF27FjweD9OnT0doaCh7CfzJS+KLFy/GL7/8gq+//hq6urpISUmBrq6uyCXgJUuWYNOmTWL9P/ubQCDAX3/9BS6X223dgoICZGZmsq+V2tpaZGZmoqCgoNN95OTk8PDhQ3oK10uARthESH19PZYuXQoLCwscP34cMjIySExMxO7du2FhYQFDQ0MAQEVFBS5duoTY2Fg8fvwYwcHBeP/993HkyBG2rZ07d+LDDz9ESEgIzpw5g/3794PP58PMzAzff/891q5dC29vb4SFhaGsrAxbt27F0qVLcerUKXYhifj4eBw4cACqqqoYN24coqOj8fnnn2P16tU4d+4ckpKSEBERgaSkJGzYsAFaWloICwvDhg0bkJWVBSkpKXh7e+PVV19FbGwsuFwufvjhB2zbtg3GxsaYM2dOr87Trl27sHnzZgwfPhzh4eEIDAxERkYG+6b85Zdfssdy7do1BAUFQVNTEz4+Prh+/TrWrl2LgIAA7NmzB7W1tdizZw8CAwORmZmJpUuX4tGjR7hw4QISExMxbNgw3Lp1C15eXli1ahVCQ0Px4MED7N69G4sXL8bXX38NDocjVtxhYWHYtm0b5OXlsWbNGrz77ruYOnUqjh49ij/++AOBgYHg8XjspfjKykp89dVXOHjwIGpra+Ht7Q1HR0fMmzcPJ06cQGZmJj799FNYWlpixowZYp+/iIgIBAQEwN/fH+np6QgLC4O+vj7mz58vUm/FihUYNWoUNm/eDBUVFaipqSEpKQm+vr4A2j6QZWdn48svvxTpp6NEVl5e3mVsLi4uYh+HOFpaWlBQUCDWzPCKigoIBAKR/dPT0/HHH390ul9DQwMWLVqEIUOGdNvH2LFjAQB///13t3UTExO7rUP6D42wiZD6+np4eHhgy5Yt0NbWxtixY9k3xtu3b7P1OBwO9u7di4kTJ4LP5yMkJASZmZn4888/2TpOTk6YN28etLS0sG7dOigqKuLXX38FABw+fBjTp0/H+++/j3HjxsHS0hK7d+/G9evXkZmZybZhZWUFGxsb6Ovrs29G2traWLNmDTQ0NODl5QWg7b6nnZ0d3njjDbi5ueH+/fuoqqpCQ0MD5s2bh61bt0JPTw8aGhrw9PSEurq60PH0lK+vL6ZMmQJdXV34+vri8ePHuHPnDru9/Zi1tLTg6OgIKysr9thlZGQQHByMpUuXQkNDA3p6eliyZAlqampQUVEBRUVFDBkyBDIyMlBTU4OcnBxiYmIwZcoUrFmzBlpaWpg0aRL27duHgoICdmQuDh8fH0yaNAl6enqws7NDbW0tPv30U2hra+Ott96Crq6u0Hlpbm7G5s2bMWHCBJiamsLCwgLy8vLYuHEjtLS04OHhAWVl5R6fS0tLS3h4eGDs2LHw9PTE+PHj2fPzJGVlZcjKykJBQQFqamqQlZXFvHnzcPbsWXZUn5SUBA0NDZiamorsf/ToUcycOVPox83NrUex9lVPLuk/nay7K+9NP1JSUrRC2iBFI2wiZPjw4Vi0aBHOnz+PmzdvorCwkP1k/+SbxpgxY6Cqqsr+bWxsDAC4c+cOtLW1AQhPbpGSkgKXy0VzczMAID8/H35+fkJ9GxgYYOjQocjPz2fvs2lqaorE+GS77aOWMWPGiJQ1NTVBQUEBbm5u+O677xAbG4vCwkLk5+ejsrKyTxOynoyh/V5x+7E9vR0AuFwuKioqAAB6enpQUlLC4cOHcffuXdy7d6/Dc/ykW7du4d69e+DxeELl7ctVirs+9NPnTk1NDUOHDmXL5OXl0dTU1OU+Ty+l2dE+PYkDgNBrozsuLi44cuQI8vLywOfzkZSUBGdn5w7renp6iozay8vLu0zakh5V1tTUYN26dUKv0c4cO3ZM6NZJOy6XCwcHh073KyoqQnBwsFgTyuLi4gC0XSUigwslbCKkqqoK77zzDoYPH46ZM2fC2toahoaGIvcln57t2p785OTk2LInf2/X3SiAYRih/Tq6jNg+e/hJna3FXFdXBzc3N9TX12Pu3LmYP38+jI2NsWjRoi7j6E53x9bR9nbZ2dlYtmwZpk+fDlNTUzg6OqKurg5r1qzpdB+BQAB7e/sO64g7mQkQPXednbfe7tPRyK2jRNyb10Y7LS0tTJ48GWfPngWHw0FhYaFIUm6npKQEJSUlsdp9VpSUlMDhcNDS0tLha/dJfD4fmZmZQl8xlJWVBZ/P73SfJ5/qRV5slLCJkP/+97+orq7GhQsX2DfV/Px8AMJvqIWFhewsVqBtQpSUlBTGjx8vVj+6urrIzc3FsmXL2LLff/8d9fX1Xc6I7amffvoJN2/eREZGBvuEo5qaGvz7778S66OnPv/8c/D5fERGRrJl7fdf28/x04lPR0cHBQUFGDNmDLutvLwcISEh8PHxgbKycj9F3zUOh4PHjx+DYRg2zr5+R7ijDwEuLi7YuXMnuFwuLC0t+/T0qmdNWloa48aNQ2lpKVRUVLqs2/7a78ks8cbGRgwbNmzQfiuAiI/uYRMhI0eORENDA5KTk1FSUoKffvoJ69atAwChy55NTU3w8/PDzZs3kZWVhdDQUDg4OEBDQ0OsflauXIkff/wRn332Ge7evYusrCwEBgbCwMAAFhYWEj0eADh37hxKSkqQl5eH9957D83NzT2+jCspr732Gu7cuYO8vDwUFxfj9OnT2LdvH4D/P8eKiop48OAB7t69i8bGRixbtgy3b9/GRx99hIKCAvz+++/w8/NDQUEBewviecDj8VBbW4uoqCgUFxezEwP7QlFREcXFxSgpKWHL5syZg6amJhw7dgxvv/12X8N+5kxNTcWexf3GG2/A1dUVK1asgKura7cfYP/991/w+Xyx70vr6OhAR0dHrLrk+UIJmwiZM2cOli9fjl27dsHe3h7btm2Di4sL+Hw+rl69ytbT09ODgYEBPD094evrCysrK3zyySdi92NnZ4f9+/cjPT0djo6OCAwMhIWFBT7//PNuLxv2hJGRETZt2oSEhATMnTsXmzZtAp/Ph4ODg9Dx9CdfX1/weDx4e3vDyckJp06dwrZt26CgoMDGZG9vj9deew2Ojo5ITU2FkZERYmJiUFBQAGdnZ6xYsQLq6uo4evRorx8i8SyYmZnhgw8+QHx8POzt7XH27FkEBgb2qU13d3fcvXsX9vb27DyAoUOHYu7cuVBQUMDMmTMlEfozZWZmBmlp6U5X0+sthmHQ3Nzco+9WT506FVOnTpVoHKR/SDEvwqoEhJCXzrJly6CtrY2goKAe7VdcXIyZM2ciNTUVo0ePfkbRiUpISEBqaqpYk8/EVVZWhvHjx8Pf359mfr8Auntt0gibEDKopKam4uDBg8jOzu7T0q39zcnJCUpKSkKLA/VFQ0MDBAIBPD09e5Ss4+Li2JniZHChhE0IGVSio6MRFxeHjz76SKKj1WdNUVER3t7eqK6uRn19fZ/aam5uRklJCTw8PKCmpiahCMnzjmaJE0IGlePHjw90CL02YcIErFmzBgcOHMDw4cN79ZWz+vp6lJaWwtXVFVZWVs8gSvK8ohE2IYT0Iz6fj8DAQLS2tqKwsFDsBXwEAgFKS0tRU1OD9957D/b29nTf+iVDI2xCCOlnEydOxKeffoozZ84gLS0NDMNAWVkZXC5XaFEigUCAx48fo7q6GgzDYPLkyXB1daVFUl5SlLAJIWQAcLlceHh4wNHREdnZ2cjJyUFhYaHI8rSjRo2ClZUVLC0toa6u3ud+9fX1+9wGGRiUsAkhZAApKytj9uzZmD17NlpaWlBdXY2WlhbIyMhAWVm5y2Vue6OrZU7J840SNiHkpdJ+z7i7x2wOtNbWVlRWVkq83fbFWyS5QBGRjPbXZGfzGug/Rgh5qVRVVQFAvz9mkxBxVVVVdfikQlrpjBDyUmloaMD169ehpqYm8tS5l0H740UTEhLYtfafZy9TvK2traiqqoKBgUHHTyqUVJCEEDIYKCgowNTUdKDDGHAjR47s16VZ++plibejkXU7+h42IYQQMghQwiaEEEIGAUrYhBBCyCAgs2XLli0DHQQhhJD+Iy8vD3Nzc8jLyw90KGKheNvQLHFCCCFkEKBL4oQQQsggQAmbEEIIGQQoYRNCyEtAIBBg//79sLa2hrGxMZYuXYp79+4NdFhiiYqKgqur60CH0aXa2lps27YNtra24PF4cHZ2RmpqqkT7oIRNCCEvgQMHDuDYsWMIDQ3FiRMnICMjg2XLlqGxsXGgQ+tSQkIC9u7dO9BhdGvTpk1IS0tDaGgokpKS8Oabb8LHxwc///yzxPqghE0IIS+4pqYmxMbGwsfHBzY2NpgwYQL27t2Lf/75B99+++1Ah9ehiooKeHt7Y/fu3dDS0hrocLpUVVWFixcvIigoCFOnToWmpia8vb1hZmaGxMREifVDCZsQQl5wt27dQl1dHSwsLNgyLpeLiRMnIi8vbwAj69yNGzegqKiIc+fOwdjYeKDD6dKQIUMQHR0tsuStlJQUHjx4ILF+aC1xQgh5wVVUVAAARowYIVSurq6OsrKygQipW7a2trC1tR3oMMTC5XIxbdo0obLffvsNly9fRnBwsMT6oRE2IYS84Orr6wEAcnJyQuVycnJoamoaiJBeaH/++Sd8fHxgbGyMhQsXSqxdStiEEPKCa39U49PJuampCUOHDh2IkF5Yubm5WLRoEdTU1BAVFQUOhyOxtilhE0LIC+61114DAFRWVgqVV1ZWilwmJ7137tw5eHl5QV9fH19++SWUlZUl2j4lbEIIecFNmDABXC4XOTk5bFltbS1u3rwJMzOzAYzsxfHNN98gMDAQc+fORVRUFLhcrsT7oElnhBDygpOTk4O7uzv27t0LVVVVjB49GuHh4RgxYgTefPPNgQ5v0CsvL8eHH34Ic3NzBAQEoKamht3G4XAkNtKmhE0IIS8BX19ftLa2IiQkBPX19Zg8eTKOHDkiMhGN9NzFixdRX1+Py5cvw9raWmibiYkJjh07JpF+6GldhBBCyCBA97AJIYSQQYASNiGEEDIIUMImhBBCBgFK2IQQQsggQAmbEEIIGQQoYRNCCCGDACVsQgjpAVtbW6xfv/6ZtV9aWopp06aJLCP622+/wc/PD1ZWVjAwMICtrS0CAgJw7do1kTY2btwIXV1dkR8ej4e33noLkZGRaG1tZet/9dVX8Pb2fmbH1JUbN25g2bJlMDMzw5QpU7Bp0yb8888/QnWuXbvW4fHMmTOHrVNVVYXly5ezx/jTTz8JtVFZWQkTExPcvn1b7NgEAgFOnz6NxYsXw8LCAjweD46OjoiJiWEfqNJu48aNIk/skjRaOIUQQp4TDMMgKCgI7u7uUFdXZ8tjY2Oxc+dO8Pl8+Pv7Y+TIkSgtLUViYiIWLlyIjRs3wsPDQ6it4cOHIyIiQqispqYGycnJiIiIQGNjI/z9/QEA7777Lo4fP45Tp05hwYIFz/5A/+e3336Du7s7tLS0EBISAgUFBcTFxWHhwoVISkrCK6+8AgD4448/IC0tjbi4OMjIyLD7DxkyhP19586dqK6uxr59+5CdnQ1fX19kZGSwS4RGRkZi9uzZ0NHRESu2xsZGrF69Gnl5eXjnnXfg4eEBOTk55OXlYf/+/fjhhx8QHR3dvw9PYQghhIhtxowZjL+//zNp++LFi8zkyZOZuro6tiwtLY3R0dFhwsPDReoLBALmo48+YnR1dZmsrCy2fMOGDYy1tXWn/bzzzjuMhYWFUFlycjJjbm4u1HdXTp8+zejo6IhVtzOrVq1izMzMmPv377Nl9fX1jI2NDbNnzx627OOPP2bmzJnTZVv29vZMQkIC24aOjg7z+++/MwzDMAUFBcykSZOY0tJSsWP7+OOPGX19fSYvL09k248//sjo6Ogwn332GVvW3TmXBLokTgghfdDY2IiDBw9izpw5MDQ0hJ2dHaKiooQuOQNATEwMZs6cCSMjIyxatAjp6enQ1dVFdnY2W+fQoUOws7MTGjlGRkZi7Nix+OCDD0T6lpKSQlBQEEaOHInIyEixY+7owRS2trZobm7GiRMnxG6nr/766y+YmJhARUWFLVNQUICRkRHS0tLYslu3bmHChAldtjV69GhcvnwZjx49wg8//AAZGRn2KWXh4eFwdXVl/+5OdXU1Tp48CRcXF0yePFlk+/Tp07Fq1SpoamqK1Z6k0CVxQgjpJYZh4O3tjV9//RXe3t7Q09NDTk4O9u3bh7///hthYWEA2pJuZGQkvLy8YGFhgczMTPj6+gq19eeff+L69etC5ffv38e1a9fg5eUFKSmpDmOQk5PDrFmzEB8fj+rqaqHk19LSwrtLTyAAAAhWSURBVP4uEAhw//59nDt3DpcuXYKXl5dQO/Ly8pgxYwa++eYbLFmypMO+BAIBBAIB+/vTfcjIyHQaZ0dUVFRQWloqUl5UVISioiIAbec4Pz8fCgoKWLBgAW7duoVhw4bB2dkZvr6+7POm33//faxatQqmpqbgcDgIDAyEqqoqrly5gry8PPZ/IY6srCw0NzdjxowZndZZt26d2O1JCiVsQgjppYyMDGRlZWHHjh1wcnICANjY2GDIkCGIiIiAp6cnxowZg+joaLi6umLDhg1sncbGRpw8eZJt6/LlywAAIyMjtqykpAQMw2D06NFdxqGpqQmGYVBWVsYm7IqKCujr64vUHTVqFPz8/LB8+XKRbYaGhjh//jwePnwIJSUlke1BQUH4+uuvhcqe7CMsLAzOzs5dxvqkt99+G8HBwdi6dStWrlzJ3qcuKChgPwgUFhbi8ePHuHv3Lvz9/TFixAhkZWUhJiYGpaWlCA8PBwDo6ekhJSUFRUVFUFdXZ+9/79q1C6tWrQLDMPD19UV+fj4sLS0RGBgIBQWFDuMqKysDgG7Pe3+jhE0IIb2Uk5MDaWlp2NvbC5U7OTkhIiICOTk5uH//PhoaGoRmNAPAf/7zH6GEXVxcjKFDhwqNkJn/PZupfRTZGVlZWaH6APDqq68iKioKQNuzr48cOYJr164hJCQE06dP77CdUaNGQSAQoLS0tMOE7ePjAzc3NwBAWloaIiMjkZiYyG4fPXo0BAIBmpubhfbjcDiQlha9A7tgwQLU1tZi//79iI+Ph5SUFOzs7LBw4UKcPn0aAKCqqoojR45g/PjxGDlyJADA3Nwc8vLy2LdvH7y9vTF+/HgAbVcbtLW12fYvXryI8vJyuLu7IyAgANLS0jh48CBCQkIQERGBgICALs9n+1WE5wXdwyaEkF568OABlJSURB5RqaamBgB4+PAh7t+/D6Bt1vaTXn31VaG/a2trRWYct4/wiouLu4yjsLAQAITu0crKysLQ0BCGhoaYMmUKDh06BG1tbfj4+ODKlSsdttPef21tbYfbR48ezbY5atQoAGD/NjQ0hIqKCnJzc2FkZCT0k5ub22nsXl5eyMnJQXJyMi5duoSIiAg8evSIfYa0oqIirK2t2WTdrv1y9a1btzpst6WlBXv27IGfnx+kpaWRmpqKJUuWQFtbG25ubvj22287jen1118H0HaFozOVlZVobGzsdPuzQAmbEEJ6adiwYXj48CGampqEytu/Q62iosImmqe/W9yeyNupqKjg4cOHQmXDhw8Hj8fDhQsXOh3ttba24vvvv4eJiYnIh4IncTgcbN++HdLS0tiwYUOHyaampoaNpbf09fWRmJgo9NPRpXmg7fvV3333HTgcDrS1tdkPMTdu3MDEiRMBAPn5+YiPjxf53nNDQ0OXsZ48eRIcDgfz5s1DTU0NWlpa2A8BSkpKIv+PJ1lYWIDD4QhNfHva+vXrMWvWLKF7+M8aJWxCCOklMzMzCAQCJCcnC5WfPXsWADB58mRMmDABr7zyCi5evChU58KFC0J/jxo1Ck1NTSKJ3M/PD/fu3cOuXbs6jGH79u0oLi7G6tWru41XU1MTK1euRFFREQ4fPiyyvby8HNLS0hgxYkS3bXWGy+UKjboNDQ07nJUOALm5uVi/fj2qq6vZskuXLuHOnTuYNWsWgLZR7tatW5GSkiK07/nz5zF06FAYGBiItFtXV4cDBw5g/fr1kJaWhoqKCmRkZFBVVQWg7QOVqqpqp8egpKQEFxcXJCYm4urVqyLbU1JSkJOTAycnJ/byeX+ge9iEENJL06ZNg7m5ObZs2YKKigro6ekhNzcXMTExcHBwYL+KtHz5cuzbtw9DhgyBubk5cnNzcfz4cQBg7+1aWVkBAK5cuQI7Ozu2j/bVv3bs2IGbN2/C2dkZI0aMQHl5OU6fPo0rV64gMDBQ7FW2VqxYgaSkJERHR8PJyQkaGhrstitXrsDExKTTBPskZ2fnHk0w64ijoyOioqLg6+uL5cuXo7y8HDt37oSJiQkcHR0BANbW1uDxePjkk09QXV0NTU1NpKWl4dixYwgICOhwhB0bG4tx48bBxsYGQNvtAWtra+zfvx9Lly5FbGws+4GgM/7+/rh+/To8PDzw7rvvwsLCAq2trfj5559x4sQJ8Pl8rF27tk/H32PP9FvehBDygnl64ZS6ujpm+/btzLRp0xh9fX3mzTffZKKiopiWlha2jkAgYA4cOMDWcXNzY+Li4hgdHR3m+vXrbD0XFxcmKCiow36vXr3KrFu3jpk+fTpjYGDA2NjYMAEBAczVq1dF6na3iEdqaiqjo6PDrF69mi2rr69nTE1NmS+++KJH56Ovbty4wbi7uzOTJk1irK2tmdDQUObRo0dCdWpqaphPPvmEmTFjBmNgYMA4ODgwJ0+e7LC9f/75h+HxeCLnpbS0lHF3d2dMTEwYPz8/kT46UldXx0RHRzPz589nTE1NGR6Px8yfP5+Ji4tjGhsbher2x8IpUgzzxLRCQgghEtXS0oKkpCRMnTqVncwEAPHx8di2bRuys7PZryClpKQgICAAmZmZYo1yJenMmTMIDw9HSkqK0MIt5PlB97AJIeQZkpWVRVxcHFavXo3k5GT8/PPPOHr0KPbu3QtnZ2c2WQPArFmzoKenhy+++KJfY2xtbUVMTAzWrl1Lyfo5RiNsQgh5xoqKirBnzx5kZ2fj4cOHeP311+Hk5ISVK1eKTFoqKirCokWLkJiY2KfJXz2RkJCAtLQ0REdH90t/pHcoYRNCCCGDAF0SJ4QQQgYBStiEEELIIEAJmxBCCBkEKGETQgghgwAlbEIIIWQQoIRNCCGEDAL/B3Fsjk9y8kemAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 216x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(3,6))\n",
    "\n",
    "plt.errorbar(root_KO_df.loc[terms_plot]['log_OR'],np.arange(len(terms_plot)),\n",
    "            xerr=[np.subtract(root_KO_df.loc[terms_plot]['log_OR'],root_KO_df.loc[terms_plot]['log_OR_CI_lower']),\n",
    "                   np.subtract(root_KO_df.loc[terms_plot]['log_OR_CI_upper'],root_KO_df.loc[terms_plot]['log_OR'])],color='k',fmt='o')\n",
    "\n",
    "color_temp = root_KO_df.loc[terms_plot]['MPO_term_type'].map({'brain':'blue','heart':'red','neg_ctrl':'black'})\n",
    "plt.scatter(root_KO_df.loc[terms_plot]['log_OR'],np.arange(len(terms_plot)),s=root_KO_df.loc[terms_plot]['num_genes_in_term'],alpha=.5,\n",
    "           c=color_temp.tolist())\n",
    "\n",
    "plt.yticks(np.arange(len(terms_plot)),root_KO_df.loc[terms_plot]['MP_description'])\n",
    "plt.xticks([0,1,2])\n",
    "plt.xlabel('log(OR) +- 95% CI')\n",
    "\n",
    "plt.plot([0,0],[-.5,len(terms_plot)-.5],'--',color='gray')\n",
    "plt.ylim([-.5,len(terms_plot)-.5])\n",
    "\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# plt.savefig('ASD_CHD_brain_heart_MPO_root_enrichment_210813.png',dpi=300,bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 19. Compute enrichment of phenotype(s) in NetColoc subsystems."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'validation' from '/Users/brinrosenthal/Documents/GitHub/NetColoc/netcoloc/validation.py'>"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import imp\n",
    "import sys\n",
    "sys.path.append('/Users/brinrosenthal/Documents/GitHub/NetColoc/netcoloc')\n",
    "import validation\n",
    "imp.reload(validation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['M', 'P', ':', '0', '0', '1', '1', '3', '9', '5']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list('MP:0011395')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "abnormal nervous system development\n",
      "Genes and Terms to keep: 191\n",
      "abnormal nervous system electrophysiology\n",
      "Genes and Terms to keep: 34\n",
      "abnormal nervous system tract morphology\n",
      "Genes and Terms to keep: 62\n",
      "abnormal cardiovascular development\n",
      "Genes and Terms to keep: 139\n",
      "abnormal neuron morphology\n",
      "Genes and Terms to keep: 262\n",
      "abnormal neuron differentiation\n",
      "Genes and Terms to keep: 16\n",
      "abnormal heart morphology\n",
      "Genes and Terms to keep: 369\n",
      "abnormal heart septum morphology\n",
      "Genes and Terms to keep: 41\n",
      "decreased fetal cardiomyocyte proliferation\n",
      "abnormal cardiac outflow tract development\n",
      "Genes and Terms to keep: 16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abnormal nervous system development:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system development:log_OR</th>\n",
       "      <th>abnormal nervous system development:num_genes</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal nervous system electrophysiology:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system electrophysiology:log_OR</th>\n",
       "      <th>abnormal nervous system electrophysiology:num_genes</th>\n",
       "      <th>abnormal nervous system electrophysiology:gene_ids</th>\n",
       "      <th>abnormal nervous system tract morphology:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system tract morphology:log_OR</th>\n",
       "      <th>...</th>\n",
       "      <th>abnormal heart septum morphology:num_genes</th>\n",
       "      <th>abnormal heart septum morphology:gene_ids</th>\n",
       "      <th>decreased fetal cardiomyocyte proliferation:-log(OR_p)</th>\n",
       "      <th>decreased fetal cardiomyocyte proliferation:log_OR</th>\n",
       "      <th>decreased fetal cardiomyocyte proliferation:num_genes</th>\n",
       "      <th>decreased fetal cardiomyocyte proliferation:gene_ids</th>\n",
       "      <th>abnormal cardiac outflow tract development:-log(OR_p)</th>\n",
       "      <th>abnormal cardiac outflow tract development:log_OR</th>\n",
       "      <th>abnormal cardiac outflow tract development:num_genes</th>\n",
       "      <th>abnormal cardiac outflow tract development:gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C350</th>\n",
       "      <td>22.738917</td>\n",
       "      <td>1.358205</td>\n",
       "      <td>73</td>\n",
       "      <td>ADNP ARID1A CACNA1A CACNA2D2 CCND1 CCND2 CHD7 ...</td>\n",
       "      <td>16.731282</td>\n",
       "      <td>1.701299</td>\n",
       "      <td>30</td>\n",
       "      <td>ANK3 CACNA1A CACNA2D2 CHD7 CHD8 ELAVL3 EPS8L2 ...</td>\n",
       "      <td>10.904244</td>\n",
       "      <td>1.675418</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>ARID1A CCND1 CHD7 HAND1 HDAC2 HOXA1 KDM6A KMT2...</td>\n",
       "      <td>5.819908</td>\n",
       "      <td>2.762957</td>\n",
       "      <td>4</td>\n",
       "      <td>CHD7 KMT2D SMARCD3 SPTBN1</td>\n",
       "      <td>5.310072</td>\n",
       "      <td>1.456095</td>\n",
       "      <td>11</td>\n",
       "      <td>ARID1A CHD7 HAND1 HOXA1 KMT2D PITX2 RARA RARB ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C353</th>\n",
       "      <td>2.092072</td>\n",
       "      <td>0.970252</td>\n",
       "      <td>9</td>\n",
       "      <td>CACNA1A CACNA2D2 CNTN2 KCNA1 NRCAM SCN8A SPTAN...</td>\n",
       "      <td>34.739152</td>\n",
       "      <td>3.541826</td>\n",
       "      <td>21</td>\n",
       "      <td>ANK3 CACNA1A CACNA2D2 FGF14 GLDN KCNA1 KCNQ3 K...</td>\n",
       "      <td>2.073248</td>\n",
       "      <td>1.575458</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>SCN8A SPTBN1</td>\n",
       "      <td>2.413564</td>\n",
       "      <td>3.004830</td>\n",
       "      <td>1</td>\n",
       "      <td>SPTBN1</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C386</th>\n",
       "      <td>3.111527</td>\n",
       "      <td>2.172537</td>\n",
       "      <td>4</td>\n",
       "      <td>ADNP NFIA SMARCA4 ZBTB14</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>1.432669</td>\n",
       "      <td>2.204263</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>SMARCA4 SMARCD3 ZBTB14</td>\n",
       "      <td>4.990873</td>\n",
       "      <td>4.780465</td>\n",
       "      <td>1</td>\n",
       "      <td>SMARCD3</td>\n",
       "      <td>1.828144</td>\n",
       "      <td>2.575753</td>\n",
       "      <td>1</td>\n",
       "      <td>SMARCA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C356</th>\n",
       "      <td>1.862682</td>\n",
       "      <td>1.028190</td>\n",
       "      <td>7</td>\n",
       "      <td>ADNP KATNB1 NFIA SPAST SPG21 VPS26A ZBTB14</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>1.345495</td>\n",
       "      <td>1.460157</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>SMYD1 ZBTB14</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C358</th>\n",
       "      <td>1.040115</td>\n",
       "      <td>0.819527</td>\n",
       "      <td>5</td>\n",
       "      <td>ADNP CHD8 KATNB1 NFIA ZBTB14</td>\n",
       "      <td>0.174659</td>\n",
       "      <td>0.434783</td>\n",
       "      <td>1</td>\n",
       "      <td>CHD8</td>\n",
       "      <td>0.426552</td>\n",
       "      <td>0.903564</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>SHOC2 SMYD1 ZBTB14</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      abnormal nervous system development:-log(OR_p)  \\\n",
       "C350                                       22.738917   \n",
       "C353                                        2.092072   \n",
       "C386                                        3.111527   \n",
       "C356                                        1.862682   \n",
       "C358                                        1.040115   \n",
       "\n",
       "      abnormal nervous system development:log_OR  \\\n",
       "C350                                    1.358205   \n",
       "C353                                    0.970252   \n",
       "C386                                    2.172537   \n",
       "C356                                    1.028190   \n",
       "C358                                    0.819527   \n",
       "\n",
       "      abnormal nervous system development:num_genes  \\\n",
       "C350                                             73   \n",
       "C353                                              9   \n",
       "C386                                              4   \n",
       "C356                                              7   \n",
       "C358                                              5   \n",
       "\n",
       "           abnormal nervous system development:gene_ids  \\\n",
       "C350  ADNP ARID1A CACNA1A CACNA2D2 CCND1 CCND2 CHD7 ...   \n",
       "C353  CACNA1A CACNA2D2 CNTN2 KCNA1 NRCAM SCN8A SPTAN...   \n",
       "C386                           ADNP NFIA SMARCA4 ZBTB14   \n",
       "C356         ADNP KATNB1 NFIA SPAST SPG21 VPS26A ZBTB14   \n",
       "C358                       ADNP CHD8 KATNB1 NFIA ZBTB14   \n",
       "\n",
       "      abnormal nervous system electrophysiology:-log(OR_p)  \\\n",
       "C350                                          16.731282      \n",
       "C353                                          34.739152      \n",
       "C386                                          -0.000000      \n",
       "C356                                          -0.000000      \n",
       "C358                                           0.174659      \n",
       "\n",
       "      abnormal nervous system electrophysiology:log_OR  \\\n",
       "C350                                          1.701299   \n",
       "C353                                          3.541826   \n",
       "C386                                          0.000000   \n",
       "C356                                          0.000000   \n",
       "C358                                          0.434783   \n",
       "\n",
       "      abnormal nervous system electrophysiology:num_genes  \\\n",
       "C350                                                 30     \n",
       "C353                                                 21     \n",
       "C386                                                  0     \n",
       "C356                                                  0     \n",
       "C358                                                  1     \n",
       "\n",
       "     abnormal nervous system electrophysiology:gene_ids  \\\n",
       "C350  ANK3 CACNA1A CACNA2D2 CHD7 CHD8 ELAVL3 EPS8L2 ...   \n",
       "C353  ANK3 CACNA1A CACNA2D2 FGF14 GLDN KCNA1 KCNQ3 K...   \n",
       "C386                                                      \n",
       "C356                                                      \n",
       "C358                                               CHD8   \n",
       "\n",
       "      abnormal nervous system tract morphology:-log(OR_p)  \\\n",
       "C350                                          10.904244     \n",
       "C353                                           2.073248     \n",
       "C386                                           1.432669     \n",
       "C356                                           1.345495     \n",
       "C358                                           0.426552     \n",
       "\n",
       "      abnormal nervous system tract morphology:log_OR  ...  \\\n",
       "C350                                         1.675418  ...   \n",
       "C353                                         1.575458  ...   \n",
       "C386                                         2.204263  ...   \n",
       "C356                                         1.460157  ...   \n",
       "C358                                         0.903564  ...   \n",
       "\n",
       "      abnormal heart septum morphology:num_genes  \\\n",
       "C350                                          21   \n",
       "C353                                           2   \n",
       "C386                                           3   \n",
       "C356                                           2   \n",
       "C358                                           3   \n",
       "\n",
       "              abnormal heart septum morphology:gene_ids  \\\n",
       "C350  ARID1A CCND1 CHD7 HAND1 HDAC2 HOXA1 KDM6A KMT2...   \n",
       "C353                                       SCN8A SPTBN1   \n",
       "C386                             SMARCA4 SMARCD3 ZBTB14   \n",
       "C356                                       SMYD1 ZBTB14   \n",
       "C358                                 SHOC2 SMYD1 ZBTB14   \n",
       "\n",
       "      decreased fetal cardiomyocyte proliferation:-log(OR_p)  \\\n",
       "C350                                           5.819908        \n",
       "C353                                           2.413564        \n",
       "C386                                           4.990873        \n",
       "C356                                          -0.000000        \n",
       "C358                                          -0.000000        \n",
       "\n",
       "      decreased fetal cardiomyocyte proliferation:log_OR  \\\n",
       "C350                                           2.762957    \n",
       "C353                                           3.004830    \n",
       "C386                                           4.780465    \n",
       "C356                                           0.000000    \n",
       "C358                                           0.000000    \n",
       "\n",
       "      decreased fetal cardiomyocyte proliferation:num_genes  \\\n",
       "C350                                                  4       \n",
       "C353                                                  1       \n",
       "C386                                                  1       \n",
       "C356                                                  0       \n",
       "C358                                                  0       \n",
       "\n",
       "     decreased fetal cardiomyocyte proliferation:gene_ids  \\\n",
       "C350                          CHD7 KMT2D SMARCD3 SPTBN1     \n",
       "C353                                             SPTBN1     \n",
       "C386                                            SMARCD3     \n",
       "C356                                                        \n",
       "C358                                                        \n",
       "\n",
       "      abnormal cardiac outflow tract development:-log(OR_p)  \\\n",
       "C350                                           5.310072       \n",
       "C353                                          -0.000000       \n",
       "C386                                           1.828144       \n",
       "C356                                          -0.000000       \n",
       "C358                                          -0.000000       \n",
       "\n",
       "      abnormal cardiac outflow tract development:log_OR  \\\n",
       "C350                                           1.456095   \n",
       "C353                                           0.000000   \n",
       "C386                                           2.575753   \n",
       "C356                                           0.000000   \n",
       "C358                                           0.000000   \n",
       "\n",
       "      abnormal cardiac outflow tract development:num_genes  \\\n",
       "C350                                                 11      \n",
       "C353                                                  0      \n",
       "C386                                                  1      \n",
       "C356                                                  0      \n",
       "C358                                                  0      \n",
       "\n",
       "     abnormal cardiac outflow tract development:gene_ids  \n",
       "C350  ARID1A CHD7 HAND1 HOXA1 KMT2D PITX2 RARA RARB ...   \n",
       "C353                                                      \n",
       "C386                                            SMARCA4   \n",
       "C356                                                      \n",
       "C358                                                      \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MP_focal_top = root_KO_df.head(10).index.tolist() # record the top 10 overall\n",
    "MP_full_results_df = validation.MPO_enrichment_full(hier_df,MPO,mgi_df,MP_focal_top,G_int)\n",
    "MP_full_results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 20. Annotate NetColoc systems map with mouse knockout data, seed genes, and enriched GO terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abnormal neuron morphology:-log(OR_p)</th>\n",
       "      <th>abnormal neuron morphology:log_OR</th>\n",
       "      <th>abnormal neuron morphology:num_genes</th>\n",
       "      <th>abnormal neuron morphology:gene_ids</th>\n",
       "      <th>abnormal nervous system development:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system development:log_OR</th>\n",
       "      <th>abnormal nervous system development:num_genes</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal cardiovascular development:-log(OR_p)</th>\n",
       "      <th>abnormal cardiovascular development:log_OR</th>\n",
       "      <th>...</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:-log(OR_p)</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:log_OR</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:num_genes</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:gene_ids</th>\n",
       "      <th>gprofiler_name</th>\n",
       "      <th>num_d1_seeds</th>\n",
       "      <th>num_d2_seeds</th>\n",
       "      <th>frac_d1_seeds</th>\n",
       "      <th>frac_d2_seeds</th>\n",
       "      <th>frac_no_seeds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C773</th>\n",
       "      <td>24.785255</td>\n",
       "      <td>1.013477</td>\n",
       "      <td>142</td>\n",
       "      <td>AFDN AFF1 ANK2 ANK3 ARID1B ATM ATP1A3 ATP1B2 A...</td>\n",
       "      <td>23.117843</td>\n",
       "      <td>1.048385</td>\n",
       "      <td>120</td>\n",
       "      <td>AFDN AGO2 ANKRD17 APC ARHGEF7 ARID1A ARID1B BC...</td>\n",
       "      <td>22.269918</td>\n",
       "      <td>1.189028</td>\n",
       "      <td>...</td>\n",
       "      <td>15.458468</td>\n",
       "      <td>1.191350</td>\n",
       "      <td>57</td>\n",
       "      <td>AKAP13 AKAP6 ANKRD17 ARID1A ARID2 BRAF BRD2 CA...</td>\n",
       "      <td>ASD-CHD systems map</td>\n",
       "      <td>47</td>\n",
       "      <td>29</td>\n",
       "      <td>0.061039</td>\n",
       "      <td>0.037662</td>\n",
       "      <td>0.901299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C774</th>\n",
       "      <td>4.160306</td>\n",
       "      <td>0.689573</td>\n",
       "      <td>40</td>\n",
       "      <td>ARID1B BCL11A BRD7 BRPF1 CHD7 CLOCK CTNNB1 DNM...</td>\n",
       "      <td>18.839552</td>\n",
       "      <td>1.375076</td>\n",
       "      <td>57</td>\n",
       "      <td>ARID1A ARID1B BCL11A BRD1 BRD2 BRPF1 CHD7 CREB...</td>\n",
       "      <td>21.744214</td>\n",
       "      <td>1.620159</td>\n",
       "      <td>...</td>\n",
       "      <td>8.043151</td>\n",
       "      <td>1.281307</td>\n",
       "      <td>23</td>\n",
       "      <td>ARID1A BRD2 CHD7 CREBBP EP300 EPAS1 EZH2 FOXO1...</td>\n",
       "      <td>PKMTs methylate histone lysines</td>\n",
       "      <td>20</td>\n",
       "      <td>11</td>\n",
       "      <td>0.073260</td>\n",
       "      <td>0.040293</td>\n",
       "      <td>0.886447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C775</th>\n",
       "      <td>2.537936</td>\n",
       "      <td>0.571114</td>\n",
       "      <td>32</td>\n",
       "      <td>AFF1 ATM ATRX ATXN1 ATXN2 BRAF CDC14A CLEC16A ...</td>\n",
       "      <td>3.696631</td>\n",
       "      <td>0.733920</td>\n",
       "      <td>30</td>\n",
       "      <td>AGO2 ANKRD17 APC ARHGEF7 BRAF C2CD3 CDK13 CHD8...</td>\n",
       "      <td>3.498324</td>\n",
       "      <td>0.833387</td>\n",
       "      <td>...</td>\n",
       "      <td>2.170585</td>\n",
       "      <td>0.782679</td>\n",
       "      <td>13</td>\n",
       "      <td>AKAP13 ANKRD17 ARID2 BRAF CDK13 DICER1 FNIP1 N...</td>\n",
       "      <td>Chromatin modifying enzymes</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>0.045643</td>\n",
       "      <td>0.037344</td>\n",
       "      <td>0.917012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C777</th>\n",
       "      <td>23.604394</td>\n",
       "      <td>1.719462</td>\n",
       "      <td>53</td>\n",
       "      <td>ANK2 ANK3 ATP1A3 ATP1B2 CACNA1A CACNA1D CACNA1...</td>\n",
       "      <td>2.438693</td>\n",
       "      <td>0.698978</td>\n",
       "      <td>20</td>\n",
       "      <td>CACNA1A CNTN1 CNTN2 DSCAM FGFR2 ISL1 LGI1 MAPT...</td>\n",
       "      <td>2.337151</td>\n",
       "      <td>0.798715</td>\n",
       "      <td>...</td>\n",
       "      <td>6.070921</td>\n",
       "      <td>1.353035</td>\n",
       "      <td>15</td>\n",
       "      <td>AKAP6 CACNA1C CACNA1H CASQ2 FGFR2 FKBP1B ISL1 ...</td>\n",
       "      <td>Ion homeostasis</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.890909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C779</th>\n",
       "      <td>6.449627</td>\n",
       "      <td>1.184946</td>\n",
       "      <td>24</td>\n",
       "      <td>AFDN ATXN1 BCL11A CTNNB1 FOXG1 FOXN4 GLI2 HIPK...</td>\n",
       "      <td>16.898562</td>\n",
       "      <td>1.815873</td>\n",
       "      <td>32</td>\n",
       "      <td>AFDN BCL11A CREBBP CTBP1 CTBP2 CTNNB1 EP300 FO...</td>\n",
       "      <td>21.136696</td>\n",
       "      <td>2.137514</td>\n",
       "      <td>...</td>\n",
       "      <td>7.498114</td>\n",
       "      <td>1.654940</td>\n",
       "      <td>13</td>\n",
       "      <td>CREBBP EP300 EPAS1 FOXO1 GATA4 HDAC9 MEF2C MYC...</td>\n",
       "      <td>Transcriptional regulation by RUNX3</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>0.064220</td>\n",
       "      <td>0.055046</td>\n",
       "      <td>0.880734</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      abnormal neuron morphology:-log(OR_p)  \\\n",
       "C773                              24.785255   \n",
       "C774                               4.160306   \n",
       "C775                               2.537936   \n",
       "C777                              23.604394   \n",
       "C779                               6.449627   \n",
       "\n",
       "      abnormal neuron morphology:log_OR  abnormal neuron morphology:num_genes  \\\n",
       "C773                           1.013477                                   142   \n",
       "C774                           0.689573                                    40   \n",
       "C775                           0.571114                                    32   \n",
       "C777                           1.719462                                    53   \n",
       "C779                           1.184946                                    24   \n",
       "\n",
       "                    abnormal neuron morphology:gene_ids  \\\n",
       "C773  AFDN AFF1 ANK2 ANK3 ARID1B ATM ATP1A3 ATP1B2 A...   \n",
       "C774  ARID1B BCL11A BRD7 BRPF1 CHD7 CLOCK CTNNB1 DNM...   \n",
       "C775  AFF1 ATM ATRX ATXN1 ATXN2 BRAF CDC14A CLEC16A ...   \n",
       "C777  ANK2 ANK3 ATP1A3 ATP1B2 CACNA1A CACNA1D CACNA1...   \n",
       "C779  AFDN ATXN1 BCL11A CTNNB1 FOXG1 FOXN4 GLI2 HIPK...   \n",
       "\n",
       "      abnormal nervous system development:-log(OR_p)  \\\n",
       "C773                                       23.117843   \n",
       "C774                                       18.839552   \n",
       "C775                                        3.696631   \n",
       "C777                                        2.438693   \n",
       "C779                                       16.898562   \n",
       "\n",
       "      abnormal nervous system development:log_OR  \\\n",
       "C773                                    1.048385   \n",
       "C774                                    1.375076   \n",
       "C775                                    0.733920   \n",
       "C777                                    0.698978   \n",
       "C779                                    1.815873   \n",
       "\n",
       "      abnormal nervous system development:num_genes  \\\n",
       "C773                                            120   \n",
       "C774                                             57   \n",
       "C775                                             30   \n",
       "C777                                             20   \n",
       "C779                                             32   \n",
       "\n",
       "           abnormal nervous system development:gene_ids  \\\n",
       "C773  AFDN AGO2 ANKRD17 APC ARHGEF7 ARID1A ARID1B BC...   \n",
       "C774  ARID1A ARID1B BCL11A BRD1 BRD2 BRPF1 CHD7 CREB...   \n",
       "C775  AGO2 ANKRD17 APC ARHGEF7 BRAF C2CD3 CDK13 CHD8...   \n",
       "C777  CACNA1A CNTN1 CNTN2 DSCAM FGFR2 ISL1 LGI1 MAPT...   \n",
       "C779  AFDN BCL11A CREBBP CTBP1 CTBP2 CTNNB1 EP300 FO...   \n",
       "\n",
       "      abnormal cardiovascular development:-log(OR_p)  \\\n",
       "C773                                       22.269918   \n",
       "C774                                       21.744214   \n",
       "C775                                        3.498324   \n",
       "C777                                        2.337151   \n",
       "C779                                       21.136696   \n",
       "\n",
       "      abnormal cardiovascular development:log_OR  ...  \\\n",
       "C773                                    1.189028  ...   \n",
       "C774                                    1.620159  ...   \n",
       "C775                                    0.833387  ...   \n",
       "C777                                    0.798715  ...   \n",
       "C779                                    2.137514  ...   \n",
       "\n",
       "      abnormal cardiac muscle tissue morphology:-log(OR_p)  \\\n",
       "C773                                          15.458468      \n",
       "C774                                           8.043151      \n",
       "C775                                           2.170585      \n",
       "C777                                           6.070921      \n",
       "C779                                           7.498114      \n",
       "\n",
       "     abnormal cardiac muscle tissue morphology:log_OR  \\\n",
       "C773                                         1.191350   \n",
       "C774                                         1.281307   \n",
       "C775                                         0.782679   \n",
       "C777                                         1.353035   \n",
       "C779                                         1.654940   \n",
       "\n",
       "      abnormal cardiac muscle tissue morphology:num_genes  \\\n",
       "C773                                                 57     \n",
       "C774                                                 23     \n",
       "C775                                                 13     \n",
       "C777                                                 15     \n",
       "C779                                                 13     \n",
       "\n",
       "      abnormal cardiac muscle tissue morphology:gene_ids  \\\n",
       "C773  AKAP13 AKAP6 ANKRD17 ARID1A ARID2 BRAF BRD2 CA...    \n",
       "C774  ARID1A BRD2 CHD7 CREBBP EP300 EPAS1 EZH2 FOXO1...    \n",
       "C775  AKAP13 ANKRD17 ARID2 BRAF CDK13 DICER1 FNIP1 N...    \n",
       "C777  AKAP6 CACNA1C CACNA1H CASQ2 FGFR2 FKBP1B ISL1 ...    \n",
       "C779  CREBBP EP300 EPAS1 FOXO1 GATA4 HDAC9 MEF2C MYC...    \n",
       "\n",
       "                           gprofiler_name num_d1_seeds  num_d2_seeds  \\\n",
       "C773                  ASD-CHD systems map           47            29   \n",
       "C774      PKMTs methylate histone lysines           20            11   \n",
       "C775          Chromatin modifying enzymes           11             9   \n",
       "C777                      Ion homeostasis           12             6   \n",
       "C779  Transcriptional regulation by RUNX3            7             6   \n",
       "\n",
       "      frac_d1_seeds  frac_d2_seeds frac_no_seeds  \n",
       "C773       0.061039       0.037662      0.901299  \n",
       "C774       0.073260       0.040293      0.886447  \n",
       "C775       0.045643       0.037344      0.917012  \n",
       "C777       0.072727       0.036364      0.890909  \n",
       "C779       0.064220       0.055046      0.880734  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add the best gprofiler annotation\n",
    "MP_full_results_df['gprofiler_name']=pd.Series(system_name_list,index=hier_df.index.tolist())\n",
    "# don't annotate the root node\n",
    "root_node = hier_df['CD_MemberList_Size'].sort_values(ascending=False).head(1).index.tolist()[0]\n",
    "MP_full_results_df['gprofiler_name'].loc[root_node]=d1_name+'-'+d2_name+' systems map'\n",
    "\n",
    "# also add the frac_seeds/num_seeds data here\n",
    "MP_full_results_df=MP_full_results_df.join(hier_df[['num_d1_seeds','num_d2_seeds','frac_d1_seeds','frac_d2_seeds','frac_no_seeds']],\n",
    "                                          how='left')\n",
    "\n",
    "MP_full_results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 21. Export NetColoc systems map to NDEx, with default style. \n",
    "\n",
    "Default style maps fraction of seed genes from seed set 1 (red) and seed set 2 (blue) to node pie charts. Fraction of genes in each system which are in neither seed set, but which are implicated by the network are indicated with white. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert G_hier to nice cx network\n",
    "node_id_to_node_name = nx.get_node_attributes(G_hier, 'name')\n",
    "for node_id in list(G_hier.nodes):\n",
    "    del G_hier.nodes[node_id]['name']\n",
    "\n",
    "G_hier_cx = ndex2.create_nice_cx_from_networkx(G_hier)\n",
    "\n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    node['n'] = node_id_to_node_name[node_id]\n",
    "\n",
    "G_hier_cx.set_name(d1_name+'_'+d2_name+'_systems_map') \n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    data = MP_full_results_df.loc[node['r']]\n",
    "    for row, value in data.items():\n",
    "        if (row.find('gene_ids')>-1) or (row=='gprofiler_name'):\n",
    "            data_type = \"string\"\n",
    "            value=str(value)\n",
    "        else:\n",
    "            data_type = \"double\"\n",
    "            value = str(value) # nice cx can only accept strings as values...\n",
    "            if value=='inf': # check if inf, set to -1 if so\n",
    "                value='-1'\n",
    "            \n",
    "        G_hier_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "# some CDAPs properties were corrupted on networkx conversion. Fix them here\n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    for i in np.arange(len(G_hier_cx.nodeAttributes[node_id])):\n",
    "        dict_temp = G_hier_cx.nodeAttributes[node_id][i]\n",
    "        if dict_temp['n'] in ['CD_MemberList_Size','CD_MemberList_LogSize','HiDeF_persistence']:\n",
    "            G_hier_cx.set_node_attribute(node_id, dict_temp['n'], dict_temp['v'], type='double',overwrite=True)\n",
    "            \n",
    "# this is required so we can easily make subgraphs from systems\n",
    "G_hier_cx.set_network_attribute('__CD_OriginalNetwork',\n",
    "                                       values='0', type='long')\n",
    "            \n",
    "          \n",
    "# use apply_style_from_network-- this should overwrite existing style\n",
    "netcoloc_template = ndex2.create_nice_cx_from_server('ndexbio.org',\n",
    "            uuid='36041bac-d2e3-11eb-b666-0ac135e8bacf')\n",
    "G_hier_cx.apply_style_from_network(netcoloc_template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NDEx server (probably ndexbio.org): ndexbio.org\n",
      "NDEx user name: sbrosenthal@health.ucsd.edu\n",
      "NDEx password: \n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "#Upload to NDEx\n",
    "G_hier_cx.set_name(d1_name+'_'+d2_name+'_systems_map') \n",
    "SERVER = input('NDEx server (probably ndexbio.org): ')\n",
    "USERNAME = input('NDEx user name: ')\n",
    "PASSWORD = getpass.getpass('NDEx password: ')\n",
    "network_uuid_hier = G_hier_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 22. Apply another template style to NetColoc systems map for mouse knockout view, and export to NDEx. \n",
    "\n",
    "Select the property to be mapped to system node colors (should be one of the mouse knockout phenotypes we identified above). In this style, the log odds ratio is mapped to the system node color. Systems which are not significantly enriched for the phenotype are white (p<0.05).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------ modify this based on your project. Should be a system identified above -----\n",
    "# set the property we should map to system node colors\n",
    "mouse_KO_mapping_property = 'abnormal neuron morphology'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating CX\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "# apply a template style \n",
    "G_hier_cx.set_name(d1_name+'_'+d2_name+'_systems_map_mouse_KO:'+mouse_KO_mapping_property) \n",
    "# use apply_style_from_network-- this should overwrite existing style\n",
    "netcoloc_template = ndex2.create_nice_cx_from_server('ndexbio.org',\n",
    "            uuid='4958993c-df46-11eb-b666-0ac135e8bacf')\n",
    "raw_cx_st = json.dumps(netcoloc_template.to_cx())\n",
    "\n",
    "# replace the default template values with mouse_KO_mapping_property\n",
    "updated_raw_cx = re.sub('COL=abnormal heart development:log_OR', 'COL='+mouse_KO_mapping_property+':log_OR', raw_cx_st)\n",
    "updated_raw_cx = re.sub('COL=abnormal heart development:-log', \n",
    "                        'COL='+mouse_KO_mapping_property+':-log', updated_raw_cx)\n",
    "updated_raw_cx=json.loads(updated_raw_cx)\n",
    "netcoloc_template_updated = ndex2.create_nice_cx_from_raw_cx(updated_raw_cx)\n",
    "G_hier_cx.apply_style_from_network(netcoloc_template_updated)\n",
    "\n",
    "network_uuid_hier_mouse_KO = G_hier_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 23. Add genes associated with mouse knockout phenotypes to NetColoc subgraph, export to NDEx. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d1_seeds</th>\n",
       "      <th>d2_seeds</th>\n",
       "      <th>z_d1</th>\n",
       "      <th>z_d2</th>\n",
       "      <th>z_both</th>\n",
       "      <th>d1_name</th>\n",
       "      <th>d2_name</th>\n",
       "      <th>sum_seeds</th>\n",
       "      <th>abnormal neuron morphology:gene_ids</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal cardiovascular development:gene_ids</th>\n",
       "      <th>abnormal heart morphology:gene_ids</th>\n",
       "      <th>abnormal heart development:gene_ids</th>\n",
       "      <th>abnormal myocardial trabeculae morphology:gene_ids</th>\n",
       "      <th>abnormal neuron number:gene_ids</th>\n",
       "      <th>abnormal neuron physiology:gene_ids</th>\n",
       "      <th>abnormal heart layer morphology:gene_ids</th>\n",
       "      <th>abnormal cardiac muscle tissue morphology:gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PTEN</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>17.775818</td>\n",
       "      <td>18.270646</td>\n",
       "      <td>324.775666</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNB1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15.369891</td>\n",
       "      <td>16.212593</td>\n",
       "      <td>249.185790</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POGZ</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.518539</td>\n",
       "      <td>13.035299</td>\n",
       "      <td>163.182897</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2C</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.964197</td>\n",
       "      <td>12.579514</td>\n",
       "      <td>163.083297</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NSD1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.263667</td>\n",
       "      <td>12.748115</td>\n",
       "      <td>156.338630</td>\n",
       "      <td>ASD</td>\n",
       "      <td>CHD</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        d1_seeds  d2_seeds       z_d1       z_d2      z_both d1_name d2_name  \\\n",
       "PTEN           1         1  17.775818  18.270646  324.775666     ASD     CHD   \n",
       "CTNNB1         1         1  15.369891  16.212593  249.185790     ASD     CHD   \n",
       "POGZ           1         1  12.518539  13.035299  163.182897     ASD     CHD   \n",
       "KMT2C          1         1  12.964197  12.579514  163.083297     ASD     CHD   \n",
       "NSD1           1         1  12.263667  12.748115  156.338630     ASD     CHD   \n",
       "\n",
       "        sum_seeds  abnormal neuron morphology:gene_ids  \\\n",
       "PTEN            2                                    1   \n",
       "CTNNB1          2                                    1   \n",
       "POGZ            2                                    0   \n",
       "KMT2C           2                                    0   \n",
       "NSD1            2                                    0   \n",
       "\n",
       "        abnormal nervous system development:gene_ids  \\\n",
       "PTEN                                               1   \n",
       "CTNNB1                                             1   \n",
       "POGZ                                               0   \n",
       "KMT2C                                              0   \n",
       "NSD1                                               0   \n",
       "\n",
       "        abnormal cardiovascular development:gene_ids  \\\n",
       "PTEN                                               1   \n",
       "CTNNB1                                             1   \n",
       "POGZ                                               0   \n",
       "KMT2C                                              0   \n",
       "NSD1                                               0   \n",
       "\n",
       "        abnormal heart morphology:gene_ids  \\\n",
       "PTEN                                     1   \n",
       "CTNNB1                                   1   \n",
       "POGZ                                     0   \n",
       "KMT2C                                    0   \n",
       "NSD1                                     0   \n",
       "\n",
       "        abnormal heart development:gene_ids  \\\n",
       "PTEN                                      1   \n",
       "CTNNB1                                    0   \n",
       "POGZ                                      0   \n",
       "KMT2C                                     0   \n",
       "NSD1                                      0   \n",
       "\n",
       "        abnormal myocardial trabeculae morphology:gene_ids  \\\n",
       "PTEN                                                    0    \n",
       "CTNNB1                                                  0    \n",
       "POGZ                                                    0    \n",
       "KMT2C                                                   0    \n",
       "NSD1                                                    0    \n",
       "\n",
       "        abnormal neuron number:gene_ids  abnormal neuron physiology:gene_ids  \\\n",
       "PTEN                                  1                                    1   \n",
       "CTNNB1                                0                                    1   \n",
       "POGZ                                  0                                    0   \n",
       "KMT2C                                 0                                    0   \n",
       "NSD1                                  0                                    0   \n",
       "\n",
       "        abnormal heart layer morphology:gene_ids  \\\n",
       "PTEN                                           1   \n",
       "CTNNB1                                         1   \n",
       "POGZ                                           0   \n",
       "KMT2C                                          0   \n",
       "NSD1                                           0   \n",
       "\n",
       "        abnormal cardiac muscle tissue morphology:gene_ids  \n",
       "PTEN                                                    1   \n",
       "CTNNB1                                                  0   \n",
       "POGZ                                                    0   \n",
       "KMT2C                                                   0   \n",
       "NSD1                                                    0   "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add fields to node_df for genes in each mouse_KO phenotype of interest\n",
    "MP_genes_columns = [c for c in MP_full_results_df.columns.tolist() if c.find(':gene_ids')>-1]\n",
    "\n",
    "# look up overlapping genes in the root node, add them to node_df\n",
    "for MP in MP_genes_columns:\n",
    "    focal_genes = MP_full_results_df.loc[root_node][MP].split(' ')\n",
    "    node_df[MP]=0\n",
    "    node_df[MP].loc[focal_genes]=1\n",
    "node_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "773\n",
      "26351\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "print(len(G_overlap.nodes()))\n",
    "print(len(G_overlap.edges()))\n",
    "G_overlap_cx = ndex2.create_nice_cx_from_networkx(G_overlap)\n",
    "G_overlap_cx.set_name(d1_name+'_'+d2_name+'_NetColoc_subgraph') \n",
    "for node_id, node in G_overlap_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_overlap_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "# apply a template style (834b6ad4-d2ea-11eb-b666-0ac135e8bacf)\n",
    "G_overlap_cx.apply_template('ndexbio.org','834b6ad4-d2ea-11eb-b666-0ac135e8bacf')\n",
    "\n",
    "network_uuid_NetColoc = G_overlap_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Upload cosine-similarity transformed NetColoc subgraph to NDEx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "773\n",
      "7862\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "#Annotate network\n",
    "print(len(G_cosSim.nodes()))\n",
    "print(len(G_cosSim.edges()))\n",
    "G_cosSim_cx = ndex2.create_nice_cx_from_networkx(G_cosSim)\n",
    "G_cosSim_cx.set_name(d1_name+'_'+d2_name+'_NetColoc_subgraph_CosSim95') \n",
    "for node_id, node in G_cosSim_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_cosSim_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "        \n",
    "        \n",
    "\n",
    "# apply a template style (834b6ad4-d2ea-11eb-b666-0ac135e8bacf)\n",
    "G_cosSim_cx.apply_template('ndexbio.org','2cbed84b-e5c3-11eb-b666-0ac135e8bacf')\n",
    "\n",
    "network_uuid_NetColoc_CosSim = G_cosSim_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add 4 networks from above to network set\n",
    "\n",
    "### ----------- is there a way to check if network set exists...? -----------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'http://public.ndexbio.org/v2/networkset/8d3cb8f5-e5cf-11eb-b666-0ac135e8bacf'"
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# need a way to check if network set exists...\n",
    "\n",
    "# networkSetURL=ndex2.client.Ndex2(host=SERVER,username=USERNAME,password=PASSWORD).create_networkset(d1_name+'-'+d2_name+' network set','network set for '+d1_name+'-'+d2_name+' NetColoc subgraph and systems map')\n",
    "# networkSetURL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'http://public.ndexbio.org/v2/networkset/8d3cb8f5-e5cf-11eb-b666-0ac135e8bacf/members'"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parse out UUID from URL strings\n",
    "networkSetUUID = networkSetURL.split('/')[-1]\n",
    "networkSetUUID\n",
    "\n",
    "networkURLs = [network_uuid_NetColoc,network_uuid_NetColoc_CosSim,network_uuid_hier,network_uuid_hier_mouse_KO]\n",
    "networkUUIDs = [n.split('/')[-1] for n in networkURLs]\n",
    "\n",
    "ndex2.client.Ndex2(host=SERVER,username=USERNAME,password=PASSWORD).add_networks_to_networkset(networkSetUUID,\n",
    "                                                                                              networkUUIDs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TODO: ADD PARAMETER LIST TO NDEX METADATA SOMEWHERE\n",
    "\n",
    "Parameters:\n",
    "- background interactome UUID\n",
    "- clustering algorithm + params\n",
    "- d1 name, d2 name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
