{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introducing Quantitative NetColoc (QNetColoc)\n",
    "\n",
    "**Overview:** This notebook demonstrates the implementation of quantitative network colocalization which allows input seed genes to be assigned individual scores prior to propagation. \n",
    "\n",
    "\n",
    "**Background:** NetColoc evaluates the extent to which two gene sets are related in network space, i.e. the extent to which they are colocalized in a molecular interaction network. In the original implementation, the genes within each gene set are all assigned an equal starting weight. However, when generating such gene sets, we may have varying confidence about the membership of each gene, or evidence of the varying strength of each gene's impact. QNetColoc, therefore, extends the functionality of NetColoc by allowing the user to weight the inputs accordingly. The key differences in the pipeline are as follows:\n",
    "\n",
    "1. Gene sets need to be supplied with associated scores. These scores will be normalized as part of the QNetColoc procedure.\n",
    "2. When assessing the expected heat at each gene within the network, maintenance of degree-score correlations is assessed as part of the degree-matched randomization. \n",
    "\n",
    "The outputs of the pipeline remain unchanged and can be analyzed as previously. Additional benchmarking and optimization of QNetColoc can be found in the related manusript: \n",
    ">Wright SN, Yang J, Ideker T. Common and rare genetic variants show network convergence for a majority of human traits. *EMBO Reports* (2026)\n",
    "\n",
    "The intention of this notebook is to demonstrate the usage of QNetColoc. The choice between QNetColoc and the original binary implementation of NetColoc (BNetColoc) for a given study depends on several factors including:\n",
    "* Availability of quantitative scores or measurements for input genes\n",
    "* Confidence in the scores scaling with the biological relevance of genes\n",
    "* Desire to include lower confidence genes with appropriate weighting\n",
    "\n",
    "Note: it is recommended to use scores directly related to biological measurements. Scores related to annotations (e.g. citations) that are correlated with node degree in biological networks may lead to biased results. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:36.753749Z",
     "start_time": "2026-02-16T22:47:35.408874Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import ndex2\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:36.769649Z",
     "start_time": "2026-02-16T22:47:36.755225Z"
    }
   },
   "outputs": [],
   "source": [
    "from netcoloc.network_colocalization import calculate_expected_overlap, get_p_from_permutation_results\n",
    "from netcoloc.netprop_zscore import calculate_heat_zscores, calculate_scored_heat_zscores\n",
    "from netcoloc.netcoloc_utils import Seeds\n",
    "from netcoloc.netprop import get_normalized_adjacency_matrix, get_individual_heats_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Load scored gene sets\n",
    "\n",
    "Here we will examine genes associated with liver fat measurement. Taking gene associations from the GWAS Catalog (Accession: GCST90016673), we partitioned them into two non-overlapping sets A and B.\n",
    "\n",
    "Genes are identified by NCBI gene IDs and scores represent association p-values (which we transform to -log10 transformed values). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.001501Z",
     "start_time": "2026-02-16T22:47:36.770477Z"
    }
   },
   "outputs": [],
   "source": [
    "seeds_a = Seeds('/cellar/users/snwright/Git/NetColoc/example_notebooks/EFO_0010821_GCST90016673_A.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.010974Z",
     "start_time": "2026-02-16T22:47:37.010966Z"
    }
   },
   "outputs": [],
   "source": [
    "seeds_b = Seeds('/cellar/users/snwright/Git/NetColoc/example_notebooks/EFO_0010821_GCST90016673_B.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Source a molecular interaction network and pre-calculate the heat matrix\n",
    "\n",
    "A benchmarking analysis demonstrates that the runtime required scales with the number of edges (w’) and the number of nodes (w’’). The diffusion parameter, which controls the rate of propagation through the network, may be set in this step. In practice, we have found that results are not dependent on the choice of this parameter, and recommend using the default value of 0.5.\n",
    "\n",
    "Here we use the PCNet 2.1 network as an example. Note that due to the network size it may take several minutes to construct the required matrices. Saving the results is recommended for repeated runs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.012200Z",
     "start_time": "2026-02-16T22:47:37.012176Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes: 18839\n",
      "\n",
      "Number of edges: 1749784\n"
     ]
    }
   ],
   "source": [
    "interactome_uuid='e9c574f4-e87a-11ee-9621-005056ae23aa' # for PCNet\n",
    "ndex_server='public.ndexbio.org'\n",
    "ndex_user=None\n",
    "ndex_password=None\n",
    "G_int = ndex2.create_nice_cx_from_server(\n",
    "            ndex_server, \n",
    "            username=ndex_user, \n",
    "            password=ndex_password, \n",
    "            uuid=interactome_uuid\n",
    "        ).to_networkx()\n",
    "node_labels = {key:int(node['GeneID']) for key, node in G_int.nodes(data=True)}\n",
    "G = nx.relabel_nodes(G_int, node_labels)\n",
    "nodes = list(G.nodes)\n",
    "\n",
    "# print out the numbers of nodes and edges in the interatome for diagnostic purposes:\n",
    "print('Number of nodes:', len(G.nodes))\n",
    "print('\\nNumber of edges:', len(G.edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.012672Z",
     "start_time": "2026-02-16T22:47:37.012665Z"
    }
   },
   "outputs": [],
   "source": [
    "node_names = {v:k for k,v in node_labels.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.013043Z",
     "start_time": "2026-02-16T22:47:37.013036Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "calculating w_prime\n",
      "\n",
      "calculating w_double_prime\n"
     ]
    }
   ],
   "source": [
    "# pre-calculate matrices used for network propagation. this step takes a few minutes, more for denser interactomes\n",
    "print('\\ncalculating w_prime')\n",
    "w_prime = get_normalized_adjacency_matrix(G, conserve_heat=True)\n",
    "\n",
    "print('\\ncalculating w_double_prime')\n",
    "w_double_prime = get_individual_heats_matrix(w_prime, .5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.013423Z",
     "start_time": "2026-02-16T22:47:37.013416Z"
    }
   },
   "outputs": [],
   "source": [
    "G_degree = dict(G.degree())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Implement QNetColoc\n",
    "\n",
    "### Prepare the gene scores\n",
    "\n",
    "For quantitative NetColoc scores must be positive, and in increasing order of importance. Therefore, we perform a negative log10 transform of the association p-values. Other transformation options include log10 and log2 transforms.\n",
    "\n",
    "Next, we filter the gene sets to only genes present in the selected network. \n",
    "\n",
    "And, finally, to give a constant input heat to network propagation, we normalize all scores to have a sum of one. Other normalization options include max, minmax and log. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0/5 seeds not found in network\n"
     ]
    }
   ],
   "source": [
    "seeds_a.transform_scores('neglog10')\n",
    "seeds_a.filter_seeds_by_network(nodes)\n",
    "seeds_a.normalize_scores(method='sum')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0/4 seeds not found in network\n"
     ]
    }
   ],
   "source": [
    "seeds_b.transform_scores('neglog10')\n",
    "seeds_b.filter_seeds_by_network(nodes)\n",
    "seeds_b.normalize_scores(method='sum')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform Quantitative network propagation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.013782Z",
     "start_time": "2026-02-16T22:47:37.013775Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [02:09<00:00,  7.71it/s]\n"
     ]
    }
   ],
   "source": [
    "z_AQ, F_QA, F_rand_QA = calculate_scored_heat_zscores(w_double_prime, nodes, G_degree, seeds_a.scores, \n",
    "                                                        num_reps=1000, minimum_bin_size=20, verbose=False,\n",
    "                                                      normalize_heat=None, random_seed=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.014335Z",
     "start_time": "2026-02-16T22:47:37.014327Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [02:10<00:00,  7.65it/s]\n"
     ]
    }
   ],
   "source": [
    "z_BQ, F_QB, F_rand_QB = calculate_scored_heat_zscores(w_double_prime, nodes, G_degree, seeds_b.scores, \n",
    "                                                        num_reps=1000, minimum_bin_size=20, verbose=False,\n",
    "                                                      normalize_heat=None, random_seed=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.014975Z",
     "start_time": "2026-02-16T22:47:37.014967Z"
    }
   },
   "outputs": [],
   "source": [
    "z_Qcoloc = pd.DataFrame({'z_A':z_AQ, 'z_B':z_BQ})\n",
    "z_Qcoloc = z_Qcoloc.assign(z_AB = z_Qcoloc.z_A * z_Qcoloc.z_B)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assess the significance of network colocalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.015570Z",
     "start_time": "2026-02-16T22:47:37.015562Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overlap seed genes: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:12<00:00, 81.23it/s]\n"
     ]
    }
   ],
   "source": [
    "observed_Qsz, permuted_Qsz = calculate_expected_overlap(z_AQ, z_BQ, z_score_threshold=3, z1_threshold=1,\n",
    "                                                    z2_threshold=1, num_reps=1000, plot=False, \n",
    "                                                    overlap_control=\"bin\", seed1=seeds_a.genes, seed2=seeds_b.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.016179Z",
     "start_time": "2026-02-16T22:47:37.016170Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of the colocalized network: 289 genes\n",
      "COLOC Score: 2.24\n",
      "Significance of network colocalization: 9.35e-54\n"
     ]
    }
   ],
   "source": [
    "print(f'Size of the colocalized network: {observed_Qsz} genes')\n",
    "print(f'COLOC Score: {observed_Qsz/np.mean(permuted_Qsz):.2f}')\n",
    "p = get_p_from_permutation_results(observed_Qsz, permuted_Qsz)\n",
    "print(f'Significance of network colocalization: {p:.2e}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.017803Z",
     "start_time": "2026-02-16T22:47:37.017794Z"
    },
    "run_control": {
     "marked": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top Genes:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MTARC1     46.145030\n",
       "GPAM       30.866713\n",
       "TM6SF2     30.155082\n",
       "PNPLA3     24.112447\n",
       "PPP1R3B    19.526084\n",
       "Name: z_AB, dtype: float64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Top Genes:')\n",
    "top = z_Qcoloc.sort_values(by='z_AB', ascending=False).iloc[0:5, 2]\n",
    "top.index = top.index.map(node_names)\n",
    "top"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: QNetColoc prioritizes TM6SF2 as the most colocalized gene, consistent with the TM6SF2 having one of the strongest association signals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4: Implement non-quantitative NetColoc for comparison\n",
    "\n",
    "Here, the inputs are two lists of seed genes, each of which is assigned the same weight. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.018650Z",
     "start_time": "2026-02-16T22:47:37.018642Z"
    }
   },
   "outputs": [],
   "source": [
    "gene_set_a = seeds_a.genes\n",
    "gene_set_b = seeds_b.genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Implement network propagation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.019304Z",
     "start_time": "2026-02-16T22:47:37.019297Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calculating SetA z-scores: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:11<00:00, 83.44it/s]\n"
     ]
    }
   ],
   "source": [
    "print('\\nCalculating SetA z-scores: ')\n",
    "z_A, Fnew_A, Fnew_rand_A = calculate_heat_zscores(w_double_prime, nodes, \n",
    "                                                                    dict(G.degree), \n",
    "                                                                    gene_set_a, num_reps=1000,\n",
    "                                                                    minimum_bin_size=100)\n",
    "\n",
    "z_A = pd.DataFrame({'z':z_A})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.020099Z",
     "start_time": "2026-02-16T22:47:37.020092Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calculating SetA z-scores: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:11<00:00, 85.01it/s]\n"
     ]
    }
   ],
   "source": [
    "print('\\nCalculating SetA z-scores: ')\n",
    "z_B, Fnew_B, Fnew_rand_B = calculate_heat_zscores(w_double_prime, nodes, \n",
    "                                                                    dict(G.degree), \n",
    "                                                                    gene_set_b, num_reps=1000,\n",
    "                                                                    minimum_bin_size=100)\n",
    "\n",
    "z_B = pd.DataFrame({'z':z_B})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.020764Z",
     "start_time": "2026-02-16T22:47:37.020757Z"
    }
   },
   "outputs": [],
   "source": [
    "z_coloc = z_A.join(z_B, lsuffix='_A', rsuffix='_B')\n",
    "z_coloc = z_coloc.assign(z_AB = z_coloc.z_A * z_coloc.z_B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.021390Z",
     "start_time": "2026-02-16T22:47:37.021383Z"
    },
    "run_control": {
     "marked": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top Genes:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MTARC1    72.302288\n",
       "SUGP1     58.058361\n",
       "GPAM      54.911657\n",
       "TM6SF2    48.997318\n",
       "TMC4      42.418630\n",
       "Name: z_AB, dtype: float64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Top Genes:')\n",
    "top = z_coloc.sort_values(by='z_AB', ascending=False).iloc[0:5, 2]\n",
    "top.index = top.index.map(node_names)\n",
    "top"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assess the significance of network colocalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.022192Z",
     "start_time": "2026-02-16T22:47:37.022185Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overlap seed genes: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:12<00:00, 81.01it/s]\n"
     ]
    }
   ],
   "source": [
    "observed_sz, permuted_sz = calculate_expected_overlap(z_A, z_B, z_score_threshold=3, z1_threshold=1,\n",
    "                                                    z2_threshold=1, num_reps=1000, plot=False, \n",
    "                                                    overlap_control=\"bin\", seed1=gene_set_a, seed2=gene_set_b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.022807Z",
     "start_time": "2026-02-16T22:47:37.022800Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of the colocalized network: 438 genes\n",
      "COLOC Score: 3.13\n",
      "Significance of network colocalization: 1.09e-158\n"
     ]
    }
   ],
   "source": [
    "print(f'Size of the colocalized network: {observed_sz} genes')\n",
    "print(f'COLOC Score: {observed_sz/np.mean(permuted_sz):.2f}')\n",
    "p = get_p_from_permutation_results(observed_sz, permuted_sz)\n",
    "print(f'Significance of network colocalization: {p:.2e}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 5: Compare the colocalized networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.023521Z",
     "start_time": "2026-02-16T22:47:37.023513Z"
    }
   },
   "outputs": [],
   "source": [
    "BNetColoc_genes = list(z_coloc[(z_coloc.z_A>1) & (z_coloc.z_B>1) & (z_coloc.z_AB>3)].index.map(node_names).values)\n",
    "QNetColoc_genes = list(z_Qcoloc[(z_Qcoloc.z_A>1) & (z_Qcoloc.z_B>1) & (z_Qcoloc.z_AB>3)].index.map(node_names).values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.024301Z",
     "start_time": "2026-02-16T22:47:37.024294Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of BNetColoc colocalized network: 438\n",
      "Size of QNetColoc colocalized network: 289\n",
      "Intersection of BNetColoc & QNetColoc: 245\n"
     ]
    }
   ],
   "source": [
    "print(f'Size of BNetColoc colocalized network: {len(BNetColoc_genes)}')\n",
    "print(f'Size of QNetColoc colocalized network: {len(QNetColoc_genes)}')\n",
    "print(f'Intersection of BNetColoc & QNetColoc: {len(set(BNetColoc_genes).intersection(set(QNetColoc_genes)))}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to the strength of network colocalization, we can examine simiarlities and differences in the functional annotations of the colocalized network genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.024954Z",
     "start_time": "2026-02-16T22:47:37.024947Z"
    }
   },
   "outputs": [],
   "source": [
    "from gprofiler import GProfiler\n",
    "gp = GProfiler(\"MyToolName/0.1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functional enrichments of the BNetColoc colocalized network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.025648Z",
     "start_time": "2026-02-16T22:47:37.025641Z"
    }
   },
   "outputs": [],
   "source": [
    "gp_bnet = pd.DataFrame(gp.profile(BNetColoc_genes,significance_threshold_method='fdr',\n",
    "                                               sources=['REAC', 'HP']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:15:17.724553Z",
     "start_time": "2026-02-16T22:15:17.722314Z"
    }
   },
   "source": [
    "Human Phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.026429Z",
     "start_time": "2026-02-16T22:47:37.026422Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Abnormal circulating lipid concentration</td>\n",
       "      <td>1.750664e-26</td>\n",
       "      <td>54</td>\n",
       "      <td>135</td>\n",
       "      <td>335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Abnormal circulating cholesterol concentration</td>\n",
       "      <td>2.967069e-22</td>\n",
       "      <td>37</td>\n",
       "      <td>135</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Abnormality of lipoprotein cholesterol concent...</td>\n",
       "      <td>1.971128e-17</td>\n",
       "      <td>26</td>\n",
       "      <td>135</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Hyperlipidemia</td>\n",
       "      <td>4.316553e-17</td>\n",
       "      <td>32</td>\n",
       "      <td>135</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Abnormal circulating organic compound concentr...</td>\n",
       "      <td>9.876500e-17</td>\n",
       "      <td>64</td>\n",
       "      <td>135</td>\n",
       "      <td>764</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 name       p_value  \\\n",
       "6            Abnormal circulating lipid concentration  1.750664e-26   \n",
       "11     Abnormal circulating cholesterol concentration  2.967069e-22   \n",
       "19  Abnormality of lipoprotein cholesterol concent...  1.971128e-17   \n",
       "20                                     Hyperlipidemia  4.316553e-17   \n",
       "21  Abnormal circulating organic compound concentr...  9.876500e-17   \n",
       "\n",
       "    intersection_size  query_size  term_size  \n",
       "6                  54         135        335  \n",
       "11                 37         135        164  \n",
       "19                 26         135         93  \n",
       "20                 32         135        163  \n",
       "21                 64         135        764  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp_bnet[gp_bnet.source=='HP'].loc[:, ['name', 'p_value', 'intersection_size', 'query_size', 'term_size' ]].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.027075Z",
     "start_time": "2026-02-16T22:47:37.027068Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>3.435771e-125</td>\n",
       "      <td>177</td>\n",
       "      <td>314</td>\n",
       "      <td>747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>4.916662e-93</td>\n",
       "      <td>226</td>\n",
       "      <td>314</td>\n",
       "      <td>2100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glycerophospholipid biosynthesis</td>\n",
       "      <td>5.504335e-51</td>\n",
       "      <td>56</td>\n",
       "      <td>314</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Phospholipid metabolism</td>\n",
       "      <td>7.097422e-42</td>\n",
       "      <td>60</td>\n",
       "      <td>314</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Plasma lipoprotein assembly, remodeling, and c...</td>\n",
       "      <td>1.179889e-33</td>\n",
       "      <td>35</td>\n",
       "      <td>314</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name        p_value  \\\n",
       "0                               Metabolism of lipids  3.435771e-125   \n",
       "1                                         Metabolism   4.916662e-93   \n",
       "2                   Glycerophospholipid biosynthesis   5.504335e-51   \n",
       "3                            Phospholipid metabolism   7.097422e-42   \n",
       "4  Plasma lipoprotein assembly, remodeling, and c...   1.179889e-33   \n",
       "\n",
       "   intersection_size  query_size  term_size  \n",
       "0                177         314        747  \n",
       "1                226         314       2100  \n",
       "2                 56         314        128  \n",
       "3                 60         314        211  \n",
       "4                 35         314         71  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp_bnet[gp_bnet.source=='REAC'].loc[:, ['name', 'p_value', 'intersection_size', 'query_size', 'term_size' ]].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functional enrichments of the QNetColoc colocalized network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.027817Z",
     "start_time": "2026-02-16T22:47:37.027809Z"
    }
   },
   "outputs": [],
   "source": [
    "gp_qnet = pd.DataFrame(gp.profile(QNetColoc_genes,significance_threshold_method='fdr',\n",
    "                                               sources=['REAC', 'HP']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Human Phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.028464Z",
     "start_time": "2026-02-16T22:47:37.028457Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Abnormal circulating lipid concentration</td>\n",
       "      <td>3.936809e-17</td>\n",
       "      <td>36</td>\n",
       "      <td>88</td>\n",
       "      <td>335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Hyperlipidemia</td>\n",
       "      <td>1.103466e-14</td>\n",
       "      <td>25</td>\n",
       "      <td>88</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Abnormal circulating cholesterol concentration</td>\n",
       "      <td>1.117265e-12</td>\n",
       "      <td>23</td>\n",
       "      <td>88</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Hyperlipoproteinemia</td>\n",
       "      <td>1.117265e-12</td>\n",
       "      <td>14</td>\n",
       "      <td>88</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Abnormal LDL cholesterol concentration</td>\n",
       "      <td>3.238256e-12</td>\n",
       "      <td>14</td>\n",
       "      <td>88</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              name       p_value  \\\n",
       "10        Abnormal circulating lipid concentration  3.936809e-17   \n",
       "12                                  Hyperlipidemia  1.103466e-14   \n",
       "14  Abnormal circulating cholesterol concentration  1.117265e-12   \n",
       "15                            Hyperlipoproteinemia  1.117265e-12   \n",
       "16          Abnormal LDL cholesterol concentration  3.238256e-12   \n",
       "\n",
       "    intersection_size  query_size  term_size  \n",
       "10                 36          88        335  \n",
       "12                 25          88        163  \n",
       "14                 23          88        164  \n",
       "15                 14          88         40  \n",
       "16                 14          88         44  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp_qnet[gp_qnet.source=='HP'].loc[:, ['name', 'p_value', 'intersection_size', 'query_size', 'term_size' ]].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reactome Pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.029226Z",
     "start_time": "2026-02-16T22:47:37.029219Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>2.208704e-70</td>\n",
       "      <td>109</td>\n",
       "      <td>207</td>\n",
       "      <td>747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>2.255332e-46</td>\n",
       "      <td>135</td>\n",
       "      <td>207</td>\n",
       "      <td>2100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glycerophospholipid biosynthesis</td>\n",
       "      <td>1.519596e-33</td>\n",
       "      <td>38</td>\n",
       "      <td>207</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Phospholipid metabolism</td>\n",
       "      <td>2.328508e-28</td>\n",
       "      <td>41</td>\n",
       "      <td>207</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Triglyceride metabolism</td>\n",
       "      <td>9.405670e-20</td>\n",
       "      <td>18</td>\n",
       "      <td>207</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               name       p_value  intersection_size  \\\n",
       "0              Metabolism of lipids  2.208704e-70                109   \n",
       "1                        Metabolism  2.255332e-46                135   \n",
       "2  Glycerophospholipid biosynthesis  1.519596e-33                 38   \n",
       "3           Phospholipid metabolism  2.328508e-28                 41   \n",
       "4           Triglyceride metabolism  9.405670e-20                 18   \n",
       "\n",
       "   query_size  term_size  \n",
       "0         207        747  \n",
       "1         207       2100  \n",
       "2         207        128  \n",
       "3         207        211  \n",
       "4         207         38  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp_qnet[gp_qnet.source=='REAC'].loc[:, ['name', 'p_value', 'intersection_size', 'query_size', 'term_size' ]].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genes unique to each colocalized network\n",
    "\n",
    "The genes only present in the BNetColoc network appear to be more involved with kidney function, while the genes only present in the QNetColoc network appear to be more relevant to lipid metabolism. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.029972Z",
     "start_time": "2026-02-16T22:47:37.029963Z"
    }
   },
   "outputs": [],
   "source": [
    "only_B = set(BNetColoc_genes).difference(set(QNetColoc_genes))\n",
    "only_Q = set(QNetColoc_genes).difference(set(BNetColoc_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Genes unique to BNetColoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.030646Z",
     "start_time": "2026-02-16T22:47:37.030638Z"
    }
   },
   "outputs": [],
   "source": [
    "gp_onlyB = pd.DataFrame(gp.profile(list(only_B),significance_threshold_method='fdr',\n",
    "                                               sources=['REAC', 'HP']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.031423Z",
     "start_time": "2026-02-16T22:47:37.031416Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Abnormal circulating lipid concentration</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>19</td>\n",
       "      <td>56</td>\n",
       "      <td>335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Abnormal circulating cholesterol concentration</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>14</td>\n",
       "      <td>56</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Abnormal circulating organic compound concentr...</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>27</td>\n",
       "      <td>56</td>\n",
       "      <td>764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Abnormal circulating metabolite concentration</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>36</td>\n",
       "      <td>56</td>\n",
       "      <td>1445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Abnormality of lipoprotein cholesterol concent...</td>\n",
       "      <td>0.000283</td>\n",
       "      <td>9</td>\n",
       "      <td>56</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 name   p_value  \\\n",
       "15           Abnormal circulating lipid concentration  0.000002   \n",
       "16     Abnormal circulating cholesterol concentration  0.000002   \n",
       "19  Abnormal circulating organic compound concentr...  0.000003   \n",
       "22      Abnormal circulating metabolite concentration  0.000014   \n",
       "27  Abnormality of lipoprotein cholesterol concent...  0.000283   \n",
       "\n",
       "    intersection_size  query_size  term_size  \n",
       "15                 19          56        335  \n",
       "16                 14          56        164  \n",
       "19                 27          56        764  \n",
       "22                 36          56       1445  \n",
       "27                  9          56         93  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp_onlyB[gp_onlyB.source=='HP'].loc[:, ['name', 'p_value', 'intersection_size', 'query_size', 'term_size' ]].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.032182Z",
     "start_time": "2026-02-16T22:47:37.032174Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>1.839645e-47</td>\n",
       "      <td>70</td>\n",
       "      <td>123</td>\n",
       "      <td>747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>1.648651e-42</td>\n",
       "      <td>95</td>\n",
       "      <td>123</td>\n",
       "      <td>2100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cholesterol biosynthesis</td>\n",
       "      <td>2.024961e-17</td>\n",
       "      <td>13</td>\n",
       "      <td>123</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glycerophospholipid biosynthesis</td>\n",
       "      <td>1.180368e-14</td>\n",
       "      <td>19</td>\n",
       "      <td>123</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Retinoid metabolism and transport</td>\n",
       "      <td>5.125799e-14</td>\n",
       "      <td>13</td>\n",
       "      <td>123</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                name       p_value  intersection_size  \\\n",
       "0               Metabolism of lipids  1.839645e-47                 70   \n",
       "1                         Metabolism  1.648651e-42                 95   \n",
       "2           Cholesterol biosynthesis  2.024961e-17                 13   \n",
       "3   Glycerophospholipid biosynthesis  1.180368e-14                 19   \n",
       "4  Retinoid metabolism and transport  5.125799e-14                 13   \n",
       "\n",
       "   query_size  term_size  \n",
       "0         123        747  \n",
       "1         123       2100  \n",
       "2         123         26  \n",
       "3         123        128  \n",
       "4         123         44  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp_onlyB[gp_onlyB.source=='REAC'].loc[:, ['name', 'p_value', 'intersection_size', 'query_size', 'term_size' ]].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Genes unique to QNetColoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.032889Z",
     "start_time": "2026-02-16T22:47:37.032882Z"
    }
   },
   "outputs": [],
   "source": [
    "gp_onlyQ = pd.DataFrame(gp.profile(list(only_Q),significance_threshold_method='fdr',\n",
    "                                               sources=['REAC', 'HP']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.033465Z",
     "start_time": "2026-02-16T22:47:37.033456Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [name, p_value, intersection_size, query_size, term_size]\n",
       "Index: []"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp_onlyQ[gp_onlyQ.source=='HP'].loc[:, ['name', 'p_value', 'intersection_size', 'query_size', 'term_size' ]].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-16T22:47:37.034502Z",
     "start_time": "2026-02-16T22:47:37.034494Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Nuclear Envelope Breakdown</td>\n",
       "      <td>0.004759</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Depolymerization of the Nuclear Lamina</td>\n",
       "      <td>0.009267</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mitotic Prophase</td>\n",
       "      <td>0.035511</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     name   p_value  intersection_size  \\\n",
       "0              Nuclear Envelope Breakdown  0.004759                  3   \n",
       "1  Depolymerization of the Nuclear Lamina  0.009267                  2   \n",
       "2                        Mitotic Prophase  0.035511                  3   \n",
       "\n",
       "   query_size  term_size  \n",
       "0          16         47  \n",
       "1          16         13  \n",
       "2          16        134  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp_onlyQ[gp_onlyQ.source=='REAC'].loc[:, ['name', 'p_value', 'intersection_size', 'query_size', 'term_size' ]].head(5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:netcoloc-dev]",
   "language": "python",
   "name": "conda-env-netcoloc-dev-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
